
<!doctype html>
<html lang="en" class="no-js">
  <head>
    
      <meta charset="utf-8">
      <meta name="viewport" content="width=device-width,initial-scale=1">
      
      
      
      
      
      
      <link rel="icon" href="../../assets/images/favicon.png">
      <meta name="generator" content="mkdocs-1.6.1, mkdocs-material-9.6.14">
    
    
      
        <title>GWADataLoader - magenpy</title>
      
    
    
      <link rel="stylesheet" href="../../assets/stylesheets/main.342714a4.min.css">
      
        
        <link rel="stylesheet" href="../../assets/stylesheets/palette.06af60db.min.css">
      
      


    
    
      
    
    
      
        
        
        <link rel="preconnect" href="https://fonts.gstatic.com" crossorigin>
        <link rel="stylesheet" href="https://fonts.googleapis.com/css?family=Roboto:300,300i,400,400i,700,700i%7CRoboto+Mono:400,400i,700,700i&display=fallback">
        <style>:root{--md-text-font:"Roboto";--md-code-font:"Roboto Mono"}</style>
      
    
    
      <link rel="stylesheet" href="../../assets/_mkdocstrings.css">
    
    <script>__md_scope=new URL("../..",location),__md_hash=e=>[...e].reduce(((e,_)=>(e<<5)-e+_.charCodeAt(0)),0),__md_get=(e,_=localStorage,t=__md_scope)=>JSON.parse(_.getItem(t.pathname+"."+e)),__md_set=(e,_,t=localStorage,a=__md_scope)=>{try{t.setItem(a.pathname+"."+e,JSON.stringify(_))}catch(e){}}</script>
    
      

    
    
    
  </head>
  
  
    
    
      
    
    
    
    
    <body dir="ltr" data-md-color-scheme="default" data-md-color-primary="indigo" data-md-color-accent="indigo">
  
    
    <input class="md-toggle" data-md-toggle="drawer" type="checkbox" id="__drawer" autocomplete="off">
    <input class="md-toggle" data-md-toggle="search" type="checkbox" id="__search" autocomplete="off">
    <label class="md-overlay" for="__drawer"></label>
    <div data-md-component="skip">
      
        
        <a href="#magenpy.GWADataLoader.GWADataLoader" class="md-skip">
          Skip to content
        </a>
      
    </div>
    <div data-md-component="announce">
      
    </div>
    
    
      

  

<header class="md-header md-header--shadow" data-md-component="header">
  <nav class="md-header__inner md-grid" aria-label="Header">
    <a href="../.." title="magenpy" class="md-header__button md-logo" aria-label="magenpy" data-md-component="logo">
      
  
  <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 8a3 3 0 0 0 3-3 3 3 0 0 0-3-3 3 3 0 0 0-3 3 3 3 0 0 0 3 3m0 3.54C9.64 9.35 6.5 8 3 8v11c3.5 0 6.64 1.35 9 3.54 2.36-2.19 5.5-3.54 9-3.54V8c-3.5 0-6.64 1.35-9 3.54"/></svg>

    </a>
    <label class="md-header__button md-icon" for="__drawer">
      
      <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M3 6h18v2H3zm0 5h18v2H3zm0 5h18v2H3z"/></svg>
    </label>
    <div class="md-header__title" data-md-component="header-title">
      <div class="md-header__ellipsis">
        <div class="md-header__topic">
          <span class="md-ellipsis">
            magenpy
          </span>
        </div>
        <div class="md-header__topic" data-md-component="header-topic">
          <span class="md-ellipsis">
            
              GWADataLoader
            
          </span>
        </div>
      </div>
    </div>
    
      
        <form class="md-header__option" data-md-component="palette">
  
    
    
    
    <input class="md-option" data-md-color-media="(prefers-color-scheme)" data-md-color-scheme="default" data-md-color-primary="indigo" data-md-color-accent="indigo"  aria-label="Switch to light mode"  type="radio" name="__palette" id="__palette_0">
    
      <label class="md-header__button md-icon" title="Switch to light mode" for="__palette_1" hidden>
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M3.9 12c0-1.71 1.39-3.1 3.1-3.1h4V7H7a5 5 0 0 0-5 5 5 5 0 0 0 5 5h4v-1.9H7c-1.71 0-3.1-1.39-3.1-3.1M8 13h8v-2H8zm9-6h-4v1.9h4c1.71 0 3.1 1.39 3.1 3.1s-1.39 3.1-3.1 3.1h-4V17h4a5 5 0 0 0 5-5 5 5 0 0 0-5-5"/></svg>
      </label>
    
  
    
    
    
    <input class="md-option" data-md-color-media="(prefers-color-scheme: light)" data-md-color-scheme="default" data-md-color-primary="indigo" data-md-color-accent="indigo"  aria-label="Switch to dark mode"  type="radio" name="__palette" id="__palette_1">
    
      <label class="md-header__button md-icon" title="Switch to dark mode" for="__palette_2" hidden>
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M17 7H7a5 5 0 0 0-5 5 5 5 0 0 0 5 5h10a5 5 0 0 0 5-5 5 5 0 0 0-5-5m0 8a3 3 0 0 1-3-3 3 3 0 0 1 3-3 3 3 0 0 1 3 3 3 3 0 0 1-3 3"/></svg>
      </label>
    
  
    
    
    
    <input class="md-option" data-md-color-media="(prefers-color-scheme: dark)" data-md-color-scheme="slate" data-md-color-primary="black" data-md-color-accent="indigo"  aria-label="Switch to system preference"  type="radio" name="__palette" id="__palette_2">
    
      <label class="md-header__button md-icon" title="Switch to system preference" for="__palette_0" hidden>
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M17 7H7a5 5 0 0 0-5 5 5 5 0 0 0 5 5h10a5 5 0 0 0 5-5 5 5 0 0 0-5-5M7 15a3 3 0 0 1-3-3 3 3 0 0 1 3-3 3 3 0 0 1 3 3 3 3 0 0 1-3 3"/></svg>
      </label>
    
  
</form>
      
    
    
      <script>var palette=__md_get("__palette");if(palette&&palette.color){if("(prefers-color-scheme)"===palette.color.media){var media=matchMedia("(prefers-color-scheme: light)"),input=document.querySelector(media.matches?"[data-md-color-media='(prefers-color-scheme: light)']":"[data-md-color-media='(prefers-color-scheme: dark)']");palette.color.media=input.getAttribute("data-md-color-media"),palette.color.scheme=input.getAttribute("data-md-color-scheme"),palette.color.primary=input.getAttribute("data-md-color-primary"),palette.color.accent=input.getAttribute("data-md-color-accent")}for(var[key,value]of Object.entries(palette.color))document.body.setAttribute("data-md-color-"+key,value)}</script>
    
    
    
      
      
        <label class="md-header__button md-icon" for="__search">
          
          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.52 6.52 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5"/></svg>
        </label>
        <div class="md-search" data-md-component="search" role="dialog">
  <label class="md-search__overlay" for="__search"></label>
  <div class="md-search__inner" role="search">
    <form class="md-search__form" name="search">
      <input type="text" class="md-search__input" name="query" aria-label="Search" placeholder="Search" autocapitalize="off" autocorrect="off" autocomplete="off" spellcheck="false" data-md-component="search-query" required>
      <label class="md-search__icon md-icon" for="__search">
        
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.52 6.52 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5"/></svg>
        
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11z"/></svg>
      </label>
      <nav class="md-search__options" aria-label="Search">
        
          <a href="javascript:void(0)" class="md-search__icon md-icon" title="Share" aria-label="Share" data-clipboard data-clipboard-text="" data-md-component="search-share" tabindex="-1">
            
            <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M18 16.08c-.76 0-1.44.3-1.96.77L8.91 12.7c.05-.23.09-.46.09-.7s-.04-.47-.09-.7l7.05-4.11c.54.5 1.25.81 2.04.81a3 3 0 0 0 3-3 3 3 0 0 0-3-3 3 3 0 0 0-3 3c0 .24.04.47.09.7L8.04 9.81C7.5 9.31 6.79 9 6 9a3 3 0 0 0-3 3 3 3 0 0 0 3 3c.79 0 1.5-.31 2.04-.81l7.12 4.15c-.05.21-.08.43-.08.66 0 1.61 1.31 2.91 2.92 2.91s2.92-1.3 2.92-2.91A2.92 2.92 0 0 0 18 16.08"/></svg>
          </a>
        
        <button type="reset" class="md-search__icon md-icon" title="Clear" aria-label="Clear" tabindex="-1">
          
          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M19 6.41 17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"/></svg>
        </button>
      </nav>
      
        <div class="md-search__suggest" data-md-component="search-suggest"></div>
      
    </form>
    <div class="md-search__output">
      <div class="md-search__scrollwrap" tabindex="0" data-md-scrollfix>
        <div class="md-search-result" data-md-component="search-result">
          <div class="md-search-result__meta">
            Initializing search
          </div>
          <ol class="md-search-result__list" role="presentation"></ol>
        </div>
      </div>
    </div>
  </div>
</div>
      
    
    
      <div class="md-header__source">
        <a href="https://github.com/shz9/magenpy" title="Go to repository" class="md-source" data-md-component="source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 496 512"><!--! Font Awesome Free 6.7.2 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2024 Fonticons, Inc.--><path d="M165.9 397.4c0 2-2.3 3.6-5.2 3.6-3.3.3-5.6-1.3-5.6-3.6 0-2 2.3-3.6 5.2-3.6 3-.3 5.6 1.3 5.6 3.6m-31.1-4.5c-.7 2 1.3 4.3 4.3 4.9 2.6 1 5.6 0 6.2-2s-1.3-4.3-4.3-5.2c-2.6-.7-5.5.3-6.2 2.3m44.2-1.7c-2.9.7-4.9 2.6-4.6 4.9.3 2 2.9 3.3 5.9 2.6 2.9-.7 4.9-2.6 4.6-4.6-.3-1.9-3-3.2-5.9-2.9M244.8 8C106.1 8 0 113.3 0 252c0 110.9 69.8 205.8 169.5 239.2 12.8 2.3 17.3-5.6 17.3-12.1 0-6.2-.3-40.4-.3-61.4 0 0-70 15-84.7-29.8 0 0-11.4-29.1-27.8-36.6 0 0-22.9-15.7 1.6-15.4 0 0 24.9 2 38.6 25.8 21.9 38.6 58.6 27.5 72.9 20.9 2.3-16 8.8-27.1 16-33.7-55.9-6.2-112.3-14.3-112.3-110.5 0-27.5 7.6-41.3 23.6-58.9-2.6-6.5-11.1-33.3 2.6-67.9 20.9-6.5 69 27 69 27 20-5.6 41.5-8.5 62.8-8.5s42.8 2.9 62.8 8.5c0 0 48.1-33.6 69-27 13.7 34.7 5.2 61.4 2.6 67.9 16 17.7 25.8 31.5 25.8 58.9 0 96.5-58.9 104.2-114.8 110.5 9.2 7.9 17 22.9 17 46.4 0 33.7-.3 75.4-.3 83.6 0 6.5 4.6 14.4 17.3 12.1C428.2 457.8 496 362.9 496 252 496 113.3 383.5 8 244.8 8M97.2 352.9c-1.3 1-1 3.3.7 5.2 1.6 1.6 3.9 2.3 5.2 1 1.3-1 1-3.3-.7-5.2-1.6-1.6-3.9-2.3-5.2-1m-10.8-8.1c-.7 1.3.3 2.9 2.3 3.9 1.6 1 3.6.7 4.3-.7.7-1.3-.3-2.9-2.3-3.9-2-.6-3.6-.3-4.3.7m32.4 35.6c-1.6 1.3-1 4.3 1.3 6.2 2.3 2.3 5.2 2.6 6.5 1 1.3-1.3.7-4.3-1.3-6.2-2.2-2.3-5.2-2.6-6.5-1m-11.4-14.7c-1.6 1-1.6 3.6 0 5.9s4.3 3.3 5.6 2.3c1.6-1.3 1.6-3.9 0-6.2-1.4-2.3-4-3.3-5.6-2"/></svg>
  </div>
  <div class="md-source__repository">
    magenpy
  </div>
</a>
      </div>
    
  </nav>
  
</header>
    
    <div class="md-container" data-md-component="container">
      
      
        
          
        
      
      <main class="md-main" data-md-component="main">
        <div class="md-main__inner md-grid">
          
            
              
              <div class="md-sidebar md-sidebar--primary" data-md-component="sidebar" data-md-type="navigation" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    



<nav class="md-nav md-nav--primary" aria-label="Navigation" data-md-level="0">
  <label class="md-nav__title" for="__drawer">
    <a href="../.." title="magenpy" class="md-nav__button md-logo" aria-label="magenpy" data-md-component="logo">
      
  
  <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 8a3 3 0 0 0 3-3 3 3 0 0 0-3-3 3 3 0 0 0-3 3 3 3 0 0 0 3 3m0 3.54C9.64 9.35 6.5 8 3 8v11c3.5 0 6.64 1.35 9 3.54 2.36-2.19 5.5-3.54 9-3.54V8c-3.5 0-6.64 1.35-9 3.54"/></svg>

    </a>
    magenpy
  </label>
  
    <div class="md-nav__source">
      <a href="https://github.com/shz9/magenpy" title="Go to repository" class="md-source" data-md-component="source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 496 512"><!--! Font Awesome Free 6.7.2 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2024 Fonticons, Inc.--><path d="M165.9 397.4c0 2-2.3 3.6-5.2 3.6-3.3.3-5.6-1.3-5.6-3.6 0-2 2.3-3.6 5.2-3.6 3-.3 5.6 1.3 5.6 3.6m-31.1-4.5c-.7 2 1.3 4.3 4.3 4.9 2.6 1 5.6 0 6.2-2s-1.3-4.3-4.3-5.2c-2.6-.7-5.5.3-6.2 2.3m44.2-1.7c-2.9.7-4.9 2.6-4.6 4.9.3 2 2.9 3.3 5.9 2.6 2.9-.7 4.9-2.6 4.6-4.6-.3-1.9-3-3.2-5.9-2.9M244.8 8C106.1 8 0 113.3 0 252c0 110.9 69.8 205.8 169.5 239.2 12.8 2.3 17.3-5.6 17.3-12.1 0-6.2-.3-40.4-.3-61.4 0 0-70 15-84.7-29.8 0 0-11.4-29.1-27.8-36.6 0 0-22.9-15.7 1.6-15.4 0 0 24.9 2 38.6 25.8 21.9 38.6 58.6 27.5 72.9 20.9 2.3-16 8.8-27.1 16-33.7-55.9-6.2-112.3-14.3-112.3-110.5 0-27.5 7.6-41.3 23.6-58.9-2.6-6.5-11.1-33.3 2.6-67.9 20.9-6.5 69 27 69 27 20-5.6 41.5-8.5 62.8-8.5s42.8 2.9 62.8 8.5c0 0 48.1-33.6 69-27 13.7 34.7 5.2 61.4 2.6 67.9 16 17.7 25.8 31.5 25.8 58.9 0 96.5-58.9 104.2-114.8 110.5 9.2 7.9 17 22.9 17 46.4 0 33.7-.3 75.4-.3 83.6 0 6.5 4.6 14.4 17.3 12.1C428.2 457.8 496 362.9 496 252 496 113.3 383.5 8 244.8 8M97.2 352.9c-1.3 1-1 3.3.7 5.2 1.6 1.6 3.9 2.3 5.2 1 1.3-1 1-3.3-.7-5.2-1.6-1.6-3.9-2.3-5.2-1m-10.8-8.1c-.7 1.3.3 2.9 2.3 3.9 1.6 1 3.6.7 4.3-.7.7-1.3-.3-2.9-2.3-3.9-2-.6-3.6-.3-4.3.7m32.4 35.6c-1.6 1.3-1 4.3 1.3 6.2 2.3 2.3 5.2 2.6 6.5 1 1.3-1.3.7-4.3-1.3-6.2-2.2-2.3-5.2-2.6-6.5-1m-11.4-14.7c-1.6 1-1.6 3.6 0 5.9s4.3 3.3 5.6 2.3c1.6-1.3 1.6-3.9 0-6.2-1.4-2.3-4-3.3-5.6-2"/></svg>
  </div>
  <div class="md-source__repository">
    magenpy
  </div>
</a>
    </div>
  
  <ul class="md-nav__list" data-md-scrollfix>
    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="../.." class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Home
    
  </span>
  

      </a>
    </li>
  

    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="../../installation/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Installation
    
  </span>
  

      </a>
    </li>
  

    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="../../getting_started/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Getting Started
    
  </span>
  

      </a>
    </li>
  

    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="../../features/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Features and Configurations
    
  </span>
  

      </a>
    </li>
  

    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="../../tutorials/overview/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Tutorials
    
  </span>
  

      </a>
    </li>
  

    
      
      
  
  
  
  
    
    
      
        
      
        
      
        
      
    
    
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
          
        
        <input class="md-nav__toggle md-toggle md-toggle--indeterminate" type="checkbox" id="__nav_6" >
        
          
          <label class="md-nav__link" for="__nav_6" id="__nav_6_label" tabindex="0">
            
  
  
  <span class="md-ellipsis">
    Command Line Scripts
    
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_6_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_6">
            <span class="md-nav__icon md-icon"></span>
            Command Line Scripts
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../commandline/overview/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Overview
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../commandline/magenpy_ld/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    magenpy_ld
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../commandline/magenpy_simulate/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    magenpy_simulate
    
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="https://github.com/shz9/magenpy/issues" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Report issues/bugs
    
  </span>
  

      </a>
    </li>
  

    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="../../faq/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    FAQ
    
  </span>
  

      </a>
    </li>
  

    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="../../citation/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Citation
    
  </span>
  

      </a>
    </li>
  

    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="../overview/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    API Reference
    
  </span>
  

      </a>
    </li>
  

    
  </ul>
</nav>
                  </div>
                </div>
              </div>
            
            
              
              <div class="md-sidebar md-sidebar--secondary" data-md-component="sidebar" data-md-type="toc" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    

<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#magenpy.GWADataLoader.GWADataLoader" class="md-nav__link">
    <span class="md-ellipsis">
      GWADataLoader
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#magenpy.GWADataLoader.GWADataLoader.chromosomes" class="md-nav__link">
    <span class="md-ellipsis">
      chromosomes
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#magenpy.GWADataLoader.GWADataLoader.m" class="md-nav__link">
    <span class="md-ellipsis">
      m
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#magenpy.GWADataLoader.GWADataLoader.n" class="md-nav__link">
    <span class="md-ellipsis">
      n
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#magenpy.GWADataLoader.GWADataLoader.n_annotations" class="md-nav__link">
    <span class="md-ellipsis">
      n_annotations
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#magenpy.GWADataLoader.GWADataLoader.n_snps" class="md-nav__link">
    <span class="md-ellipsis">
      n_snps
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#magenpy.GWADataLoader.GWADataLoader.sample_size" class="md-nav__link">
    <span class="md-ellipsis">
      sample_size
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#magenpy.GWADataLoader.GWADataLoader.samples" class="md-nav__link">
    <span class="md-ellipsis">
      samples
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#magenpy.GWADataLoader.GWADataLoader.shapes" class="md-nav__link">
    <span class="md-ellipsis">
      shapes
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#magenpy.GWADataLoader.GWADataLoader.snps" class="md-nav__link">
    <span class="md-ellipsis">
      snps
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#magenpy.GWADataLoader.GWADataLoader.__init__" class="md-nav__link">
    <span class="md-ellipsis">
      __init__
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#magenpy.GWADataLoader.GWADataLoader.align_with" class="md-nav__link">
    <span class="md-ellipsis">
      align_with
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#magenpy.GWADataLoader.GWADataLoader.cleanup" class="md-nav__link">
    <span class="md-ellipsis">
      cleanup
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#magenpy.GWADataLoader.GWADataLoader.compute_ld" class="md-nav__link">
    <span class="md-ellipsis">
      compute_ld
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#magenpy.GWADataLoader.GWADataLoader.filter_samples" class="md-nav__link">
    <span class="md-ellipsis">
      filter_samples
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#magenpy.GWADataLoader.GWADataLoader.filter_snps" class="md-nav__link">
    <span class="md-ellipsis">
      filter_snps
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#magenpy.GWADataLoader.GWADataLoader.get_ld_matrices" class="md-nav__link">
    <span class="md-ellipsis">
      get_ld_matrices
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#magenpy.GWADataLoader.GWADataLoader.harmonize_data" class="md-nav__link">
    <span class="md-ellipsis">
      harmonize_data
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#magenpy.GWADataLoader.GWADataLoader.load_ld" class="md-nav__link">
    <span class="md-ellipsis">
      load_ld
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#magenpy.GWADataLoader.GWADataLoader.perform_gwas" class="md-nav__link">
    <span class="md-ellipsis">
      perform_gwas
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#magenpy.GWADataLoader.GWADataLoader.predict" class="md-nav__link">
    <span class="md-ellipsis">
      predict
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#magenpy.GWADataLoader.GWADataLoader.read_annotations" class="md-nav__link">
    <span class="md-ellipsis">
      read_annotations
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#magenpy.GWADataLoader.GWADataLoader.read_covariates" class="md-nav__link">
    <span class="md-ellipsis">
      read_covariates
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#magenpy.GWADataLoader.GWADataLoader.read_genotypes" class="md-nav__link">
    <span class="md-ellipsis">
      read_genotypes
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#magenpy.GWADataLoader.GWADataLoader.read_ld" class="md-nav__link">
    <span class="md-ellipsis">
      read_ld
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#magenpy.GWADataLoader.GWADataLoader.read_phenotype" class="md-nav__link">
    <span class="md-ellipsis">
      read_phenotype
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#magenpy.GWADataLoader.GWADataLoader.read_summary_statistics" class="md-nav__link">
    <span class="md-ellipsis">
      read_summary_statistics
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#magenpy.GWADataLoader.GWADataLoader.release_ld" class="md-nav__link">
    <span class="md-ellipsis">
      release_ld
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#magenpy.GWADataLoader.GWADataLoader.score" class="md-nav__link">
    <span class="md-ellipsis">
      score
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#magenpy.GWADataLoader.GWADataLoader.set_phenotype" class="md-nav__link">
    <span class="md-ellipsis">
      set_phenotype
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#magenpy.GWADataLoader.GWADataLoader.split_by_chromosome" class="md-nav__link">
    <span class="md-ellipsis">
      split_by_chromosome
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#magenpy.GWADataLoader.GWADataLoader.split_by_samples" class="md-nav__link">
    <span class="md-ellipsis">
      split_by_samples
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#magenpy.GWADataLoader.GWADataLoader.sync_sample_tables" class="md-nav__link">
    <span class="md-ellipsis">
      sync_sample_tables
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#magenpy.GWADataLoader.GWADataLoader.to_individual_table" class="md-nav__link">
    <span class="md-ellipsis">
      to_individual_table
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#magenpy.GWADataLoader.GWADataLoader.to_phenotype_table" class="md-nav__link">
    <span class="md-ellipsis">
      to_phenotype_table
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#magenpy.GWADataLoader.GWADataLoader.to_snp_table" class="md-nav__link">
    <span class="md-ellipsis">
      to_snp_table
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#magenpy.GWADataLoader.GWADataLoader.to_summary_statistics_table" class="md-nav__link">
    <span class="md-ellipsis">
      to_summary_statistics_table
    </span>
  </a>
  
</li>
      
    </ul>
  
</nav>
                  </div>
                </div>
              </div>
            
          
          
            <div class="md-content" data-md-component="content">
              <article class="md-content__inner md-typeset">
                
                  


  
    <a href="https://github.com/shz9/magenpy/edit/master/docs/api/GWADataLoader.md" title="Edit this page" class="md-content__button md-icon">
      
      <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M10 20H6V4h7v5h5v3.1l2-2V8l-6-6H6c-1.1 0-2 .9-2 2v16c0 1.1.9 2 2 2h4zm10.2-7c.1 0 .3.1.4.2l1.3 1.3c.2.2.2.6 0 .8l-1 1-2.1-2.1 1-1c.1-.1.2-.2.4-.2m0 3.9L14.1 23H12v-2.1l6.1-6.1z"/></svg>
    </a>
  
  
    
      
    
    <a href="https://github.com/shz9/magenpy/raw/master/docs/api/GWADataLoader.md" title="View source of this page" class="md-content__button md-icon">
      
      <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M17 18c.56 0 1 .44 1 1s-.44 1-1 1-1-.44-1-1 .44-1 1-1m0-3c-2.73 0-5.06 1.66-6 4 .94 2.34 3.27 4 6 4s5.06-1.66 6-4c-.94-2.34-3.27-4-6-4m0 6.5a2.5 2.5 0 0 1-2.5-2.5 2.5 2.5 0 0 1 2.5-2.5 2.5 2.5 0 0 1 2.5 2.5 2.5 2.5 0 0 1-2.5 2.5M9.27 20H6V4h7v5h5v4.07c.7.08 1.36.25 2 .49V8l-6-6H6a2 2 0 0 0-2 2v16a2 2 0 0 0 2 2h4.5a8.2 8.2 0 0 1-1.23-2"/></svg>
    </a>
  


  <h1>GWADataLoader</h1>

<div class="doc doc-object doc-class">



<a id="magenpy.GWADataLoader.GWADataLoader"></a>
    <div class="doc doc-contents first">
            <p class="doc doc-class-bases">
              Bases: <code><span title="object">object</span></code></p>


        <p>A class to load and manage multiple data sources for genetic association studies.
This class is designed to handle genotype matrices, summary statistics, LD matrices,
and annotation matrices. It also provides functionalities to filter samples and/or SNPs,
harmonize data sources, and compute LD matrices. This is all done in order to facilitate
downstream statistical genetics analyses that require multiple data sources to be aligned
and harmonized. The use cases include:</p>
<ul>
<li>Summary statistics-based PRS computation</li>
<li>Summary statistics-based heritability estimation.</li>
<li>Complex trait simulation.</li>
<li>Performing Genome-wide association tests.</li>
</ul>


<p><span class="doc-section-title">Attributes:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td><code><span title="magenpy.GWADataLoader.GWADataLoader.genotype">genotype</span></code></td>
            <td>
                  <code><span title="magenpy.GenotypeMatrix.Union">Union</span>[<span title="typing.Dict">Dict</span>[<span title="int">int</span>, <a class="autorefs autorefs-internal" title="&lt;code&gt;GenotypeMatrix&lt;/code&gt; (&lt;code&gt;magenpy.GenotypeMatrix.GenotypeMatrix&lt;/code&gt;)" href="../GenotypeMatrix/#magenpy.GenotypeMatrix.GenotypeMatrix">GenotypeMatrix</a>], None]</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>A dictionary of <code>GenotypeMatrix</code> objects, where the key is the chromosome number.</p>
              </div>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td><code><span title="magenpy.GWADataLoader.GWADataLoader.sample_table">sample_table</span></code></td>
            <td>
                  <code><span title="magenpy.GenotypeMatrix.Union">Union</span>[<a class="autorefs autorefs-internal" title="SampleTable (&lt;code&gt;magenpy.SampleTable.SampleTable&lt;/code&gt;)" href="../SampleTable/#magenpy.SampleTable.SampleTable">SampleTable</a>, None]</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>A <code>SampleTable</code> object containing the sample information.</p>
              </div>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td><code><span title="magenpy.GWADataLoader.GWADataLoader.phenotype_likelihood">phenotype_likelihood</span></code></td>
            <td>
                  <code><span title="str">str</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>The likelihood of the phenotype (e.g. <code>gaussian</code>, <code>binomial</code>).</p>
              </div>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td><code><span title="magenpy.GWADataLoader.GWADataLoader.ld">ld</span></code></td>
            <td>
                  <code><span title="magenpy.GenotypeMatrix.Union">Union</span>[<span title="typing.Dict">Dict</span>[<span title="int">int</span>, <a class="autorefs autorefs-internal" title="LDMatrix (&lt;code&gt;magenpy.LDMatrix.LDMatrix&lt;/code&gt;)" href="../LDMatrix/#magenpy.LDMatrix.LDMatrix">LDMatrix</a>], None]</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>A dictionary of <code>LDMatrix</code> objects, where the key is the chromosome number.</p>
              </div>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td><code><span title="magenpy.GWADataLoader.GWADataLoader.sumstats_table">sumstats_table</span></code></td>
            <td>
                  <code><span title="magenpy.GenotypeMatrix.Union">Union</span>[<span title="typing.Dict">Dict</span>[<span title="int">int</span>, <a class="autorefs autorefs-internal" title="SumstatsTable (&lt;code&gt;magenpy.SumstatsTable.SumstatsTable&lt;/code&gt;)" href="../SumstatsTable/#magenpy.SumstatsTable.SumstatsTable">SumstatsTable</a>], None]</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>A dictionary of <code>SumstatsTable</code> objects, where the key is the chromosome number.</p>
              </div>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td><code><span title="magenpy.GWADataLoader.GWADataLoader.annotation">annotation</span></code></td>
            <td>
                  <code><span title="magenpy.GenotypeMatrix.Union">Union</span>[<span title="typing.Dict">Dict</span>[<span title="int">int</span>, <a class="autorefs autorefs-internal" title="AnnotationMatrix (&lt;code&gt;magenpy.AnnotationMatrix.AnnotationMatrix&lt;/code&gt;)" href="../AnnotationMatrix/#magenpy.AnnotationMatrix.AnnotationMatrix">AnnotationMatrix</a>], None]</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>A dictionary of <code>AnnotationMatrix</code> objects, where the key is the chromosome number.</p>
              </div>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td><code><span title="magenpy.GWADataLoader.GWADataLoader.backend">backend</span></code></td>
            <td>
            </td>
            <td>
              <div class="doc-md-description">
                <p>The backend software used for the computation. Currently, supports <code>xarray</code> and <code>plink</code>.</p>
              </div>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td><code><span title="magenpy.GWADataLoader.GWADataLoader.temp_dir">temp_dir</span></code></td>
            <td>
            </td>
            <td>
              <div class="doc-md-description">
                <p>The temporary directory where we store intermediate files (if necessary).</p>
              </div>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td><code><span title="magenpy.GWADataLoader.GWADataLoader.output_dir">output_dir</span></code></td>
            <td>
            </td>
            <td>
              <div class="doc-md-description">
                <p>The output directory where we store the results of the computation.</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>







              <details class="quote">
                <summary>Source code in <code>magenpy/GWADataLoader.py</code></summary>
                <div class="language-python highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal"><a href="#__codelineno-0-24">  24</a></span>
<span class="normal"><a href="#__codelineno-0-25">  25</a></span>
<span class="normal"><a href="#__codelineno-0-26">  26</a></span>
<span class="normal"><a href="#__codelineno-0-27">  27</a></span>
<span class="normal"><a href="#__codelineno-0-28">  28</a></span>
<span class="normal"><a href="#__codelineno-0-29">  29</a></span>
<span class="normal"><a href="#__codelineno-0-30">  30</a></span>
<span class="normal"><a href="#__codelineno-0-31">  31</a></span>
<span class="normal"><a href="#__codelineno-0-32">  32</a></span>
<span class="normal"><a href="#__codelineno-0-33">  33</a></span>
<span class="normal"><a href="#__codelineno-0-34">  34</a></span>
<span class="normal"><a href="#__codelineno-0-35">  35</a></span>
<span class="normal"><a href="#__codelineno-0-36">  36</a></span>
<span class="normal"><a href="#__codelineno-0-37">  37</a></span>
<span class="normal"><a href="#__codelineno-0-38">  38</a></span>
<span class="normal"><a href="#__codelineno-0-39">  39</a></span>
<span class="normal"><a href="#__codelineno-0-40">  40</a></span>
<span class="normal"><a href="#__codelineno-0-41">  41</a></span>
<span class="normal"><a href="#__codelineno-0-42">  42</a></span>
<span class="normal"><a href="#__codelineno-0-43">  43</a></span>
<span class="normal"><a href="#__codelineno-0-44">  44</a></span>
<span class="normal"><a href="#__codelineno-0-45">  45</a></span>
<span class="normal"><a href="#__codelineno-0-46">  46</a></span>
<span class="normal"><a href="#__codelineno-0-47">  47</a></span>
<span class="normal"><a href="#__codelineno-0-48">  48</a></span>
<span class="normal"><a href="#__codelineno-0-49">  49</a></span>
<span class="normal"><a href="#__codelineno-0-50">  50</a></span>
<span class="normal"><a href="#__codelineno-0-51">  51</a></span>
<span class="normal"><a href="#__codelineno-0-52">  52</a></span>
<span class="normal"><a href="#__codelineno-0-53">  53</a></span>
<span class="normal"><a href="#__codelineno-0-54">  54</a></span>
<span class="normal"><a href="#__codelineno-0-55">  55</a></span>
<span class="normal"><a href="#__codelineno-0-56">  56</a></span>
<span class="normal"><a href="#__codelineno-0-57">  57</a></span>
<span class="normal"><a href="#__codelineno-0-58">  58</a></span>
<span class="normal"><a href="#__codelineno-0-59">  59</a></span>
<span class="normal"><a href="#__codelineno-0-60">  60</a></span>
<span class="normal"><a href="#__codelineno-0-61">  61</a></span>
<span class="normal"><a href="#__codelineno-0-62">  62</a></span>
<span class="normal"><a href="#__codelineno-0-63">  63</a></span>
<span class="normal"><a href="#__codelineno-0-64">  64</a></span>
<span class="normal"><a href="#__codelineno-0-65">  65</a></span>
<span class="normal"><a href="#__codelineno-0-66">  66</a></span>
<span class="normal"><a href="#__codelineno-0-67">  67</a></span>
<span class="normal"><a href="#__codelineno-0-68">  68</a></span>
<span class="normal"><a href="#__codelineno-0-69">  69</a></span>
<span class="normal"><a href="#__codelineno-0-70">  70</a></span>
<span class="normal"><a href="#__codelineno-0-71">  71</a></span>
<span class="normal"><a href="#__codelineno-0-72">  72</a></span>
<span class="normal"><a href="#__codelineno-0-73">  73</a></span>
<span class="normal"><a href="#__codelineno-0-74">  74</a></span>
<span class="normal"><a href="#__codelineno-0-75">  75</a></span>
<span class="normal"><a href="#__codelineno-0-76">  76</a></span>
<span class="normal"><a href="#__codelineno-0-77">  77</a></span>
<span class="normal"><a href="#__codelineno-0-78">  78</a></span>
<span class="normal"><a href="#__codelineno-0-79">  79</a></span>
<span class="normal"><a href="#__codelineno-0-80">  80</a></span>
<span class="normal"><a href="#__codelineno-0-81">  81</a></span>
<span class="normal"><a href="#__codelineno-0-82">  82</a></span>
<span class="normal"><a href="#__codelineno-0-83">  83</a></span>
<span class="normal"><a href="#__codelineno-0-84">  84</a></span>
<span class="normal"><a href="#__codelineno-0-85">  85</a></span>
<span class="normal"><a href="#__codelineno-0-86">  86</a></span>
<span class="normal"><a href="#__codelineno-0-87">  87</a></span>
<span class="normal"><a href="#__codelineno-0-88">  88</a></span>
<span class="normal"><a href="#__codelineno-0-89">  89</a></span>
<span class="normal"><a href="#__codelineno-0-90">  90</a></span>
<span class="normal"><a href="#__codelineno-0-91">  91</a></span>
<span class="normal"><a href="#__codelineno-0-92">  92</a></span>
<span class="normal"><a href="#__codelineno-0-93">  93</a></span>
<span class="normal"><a href="#__codelineno-0-94">  94</a></span>
<span class="normal"><a href="#__codelineno-0-95">  95</a></span>
<span class="normal"><a href="#__codelineno-0-96">  96</a></span>
<span class="normal"><a href="#__codelineno-0-97">  97</a></span>
<span class="normal"><a href="#__codelineno-0-98">  98</a></span>
<span class="normal"><a href="#__codelineno-0-99">  99</a></span>
<span class="normal"><a href="#__codelineno-0-100"> 100</a></span>
<span class="normal"><a href="#__codelineno-0-101"> 101</a></span>
<span class="normal"><a href="#__codelineno-0-102"> 102</a></span>
<span class="normal"><a href="#__codelineno-0-103"> 103</a></span>
<span class="normal"><a href="#__codelineno-0-104"> 104</a></span>
<span class="normal"><a href="#__codelineno-0-105"> 105</a></span>
<span class="normal"><a href="#__codelineno-0-106"> 106</a></span>
<span class="normal"><a href="#__codelineno-0-107"> 107</a></span>
<span class="normal"><a href="#__codelineno-0-108"> 108</a></span>
<span class="normal"><a href="#__codelineno-0-109"> 109</a></span>
<span class="normal"><a href="#__codelineno-0-110"> 110</a></span>
<span class="normal"><a href="#__codelineno-0-111"> 111</a></span>
<span class="normal"><a href="#__codelineno-0-112"> 112</a></span>
<span class="normal"><a href="#__codelineno-0-113"> 113</a></span>
<span class="normal"><a href="#__codelineno-0-114"> 114</a></span>
<span class="normal"><a href="#__codelineno-0-115"> 115</a></span>
<span class="normal"><a href="#__codelineno-0-116"> 116</a></span>
<span class="normal"><a href="#__codelineno-0-117"> 117</a></span>
<span class="normal"><a href="#__codelineno-0-118"> 118</a></span>
<span class="normal"><a href="#__codelineno-0-119"> 119</a></span>
<span class="normal"><a href="#__codelineno-0-120"> 120</a></span>
<span class="normal"><a href="#__codelineno-0-121"> 121</a></span>
<span class="normal"><a href="#__codelineno-0-122"> 122</a></span>
<span class="normal"><a href="#__codelineno-0-123"> 123</a></span>
<span class="normal"><a href="#__codelineno-0-124"> 124</a></span>
<span class="normal"><a href="#__codelineno-0-125"> 125</a></span>
<span class="normal"><a href="#__codelineno-0-126"> 126</a></span>
<span class="normal"><a href="#__codelineno-0-127"> 127</a></span>
<span class="normal"><a href="#__codelineno-0-128"> 128</a></span>
<span class="normal"><a href="#__codelineno-0-129"> 129</a></span>
<span class="normal"><a href="#__codelineno-0-130"> 130</a></span>
<span class="normal"><a href="#__codelineno-0-131"> 131</a></span>
<span class="normal"><a href="#__codelineno-0-132"> 132</a></span>
<span class="normal"><a href="#__codelineno-0-133"> 133</a></span>
<span class="normal"><a href="#__codelineno-0-134"> 134</a></span>
<span class="normal"><a href="#__codelineno-0-135"> 135</a></span>
<span class="normal"><a href="#__codelineno-0-136"> 136</a></span>
<span class="normal"><a href="#__codelineno-0-137"> 137</a></span>
<span class="normal"><a href="#__codelineno-0-138"> 138</a></span>
<span class="normal"><a href="#__codelineno-0-139"> 139</a></span>
<span class="normal"><a href="#__codelineno-0-140"> 140</a></span>
<span class="normal"><a href="#__codelineno-0-141"> 141</a></span>
<span class="normal"><a href="#__codelineno-0-142"> 142</a></span>
<span class="normal"><a href="#__codelineno-0-143"> 143</a></span>
<span class="normal"><a href="#__codelineno-0-144"> 144</a></span>
<span class="normal"><a href="#__codelineno-0-145"> 145</a></span>
<span class="normal"><a href="#__codelineno-0-146"> 146</a></span>
<span class="normal"><a href="#__codelineno-0-147"> 147</a></span>
<span class="normal"><a href="#__codelineno-0-148"> 148</a></span>
<span class="normal"><a href="#__codelineno-0-149"> 149</a></span>
<span class="normal"><a href="#__codelineno-0-150"> 150</a></span>
<span class="normal"><a href="#__codelineno-0-151"> 151</a></span>
<span class="normal"><a href="#__codelineno-0-152"> 152</a></span>
<span class="normal"><a href="#__codelineno-0-153"> 153</a></span>
<span class="normal"><a href="#__codelineno-0-154"> 154</a></span>
<span class="normal"><a href="#__codelineno-0-155"> 155</a></span>
<span class="normal"><a href="#__codelineno-0-156"> 156</a></span>
<span class="normal"><a href="#__codelineno-0-157"> 157</a></span>
<span class="normal"><a href="#__codelineno-0-158"> 158</a></span>
<span class="normal"><a href="#__codelineno-0-159"> 159</a></span>
<span class="normal"><a href="#__codelineno-0-160"> 160</a></span>
<span class="normal"><a href="#__codelineno-0-161"> 161</a></span>
<span class="normal"><a href="#__codelineno-0-162"> 162</a></span>
<span class="normal"><a href="#__codelineno-0-163"> 163</a></span>
<span class="normal"><a href="#__codelineno-0-164"> 164</a></span>
<span class="normal"><a href="#__codelineno-0-165"> 165</a></span>
<span class="normal"><a href="#__codelineno-0-166"> 166</a></span>
<span class="normal"><a href="#__codelineno-0-167"> 167</a></span>
<span class="normal"><a href="#__codelineno-0-168"> 168</a></span>
<span class="normal"><a href="#__codelineno-0-169"> 169</a></span>
<span class="normal"><a href="#__codelineno-0-170"> 170</a></span>
<span class="normal"><a href="#__codelineno-0-171"> 171</a></span>
<span class="normal"><a href="#__codelineno-0-172"> 172</a></span>
<span class="normal"><a href="#__codelineno-0-173"> 173</a></span>
<span class="normal"><a href="#__codelineno-0-174"> 174</a></span>
<span class="normal"><a href="#__codelineno-0-175"> 175</a></span>
<span class="normal"><a href="#__codelineno-0-176"> 176</a></span>
<span class="normal"><a href="#__codelineno-0-177"> 177</a></span>
<span class="normal"><a href="#__codelineno-0-178"> 178</a></span>
<span class="normal"><a href="#__codelineno-0-179"> 179</a></span>
<span class="normal"><a href="#__codelineno-0-180"> 180</a></span>
<span class="normal"><a href="#__codelineno-0-181"> 181</a></span>
<span class="normal"><a href="#__codelineno-0-182"> 182</a></span>
<span class="normal"><a href="#__codelineno-0-183"> 183</a></span>
<span class="normal"><a href="#__codelineno-0-184"> 184</a></span>
<span class="normal"><a href="#__codelineno-0-185"> 185</a></span>
<span class="normal"><a href="#__codelineno-0-186"> 186</a></span>
<span class="normal"><a href="#__codelineno-0-187"> 187</a></span>
<span class="normal"><a href="#__codelineno-0-188"> 188</a></span>
<span class="normal"><a href="#__codelineno-0-189"> 189</a></span>
<span class="normal"><a href="#__codelineno-0-190"> 190</a></span>
<span class="normal"><a href="#__codelineno-0-191"> 191</a></span>
<span class="normal"><a href="#__codelineno-0-192"> 192</a></span>
<span class="normal"><a href="#__codelineno-0-193"> 193</a></span>
<span class="normal"><a href="#__codelineno-0-194"> 194</a></span>
<span class="normal"><a href="#__codelineno-0-195"> 195</a></span>
<span class="normal"><a href="#__codelineno-0-196"> 196</a></span>
<span class="normal"><a href="#__codelineno-0-197"> 197</a></span>
<span class="normal"><a href="#__codelineno-0-198"> 198</a></span>
<span class="normal"><a href="#__codelineno-0-199"> 199</a></span>
<span class="normal"><a href="#__codelineno-0-200"> 200</a></span>
<span class="normal"><a href="#__codelineno-0-201"> 201</a></span>
<span class="normal"><a href="#__codelineno-0-202"> 202</a></span>
<span class="normal"><a href="#__codelineno-0-203"> 203</a></span>
<span class="normal"><a href="#__codelineno-0-204"> 204</a></span>
<span class="normal"><a href="#__codelineno-0-205"> 205</a></span>
<span class="normal"><a href="#__codelineno-0-206"> 206</a></span>
<span class="normal"><a href="#__codelineno-0-207"> 207</a></span>
<span class="normal"><a href="#__codelineno-0-208"> 208</a></span>
<span class="normal"><a href="#__codelineno-0-209"> 209</a></span>
<span class="normal"><a href="#__codelineno-0-210"> 210</a></span>
<span class="normal"><a href="#__codelineno-0-211"> 211</a></span>
<span class="normal"><a href="#__codelineno-0-212"> 212</a></span>
<span class="normal"><a href="#__codelineno-0-213"> 213</a></span>
<span class="normal"><a href="#__codelineno-0-214"> 214</a></span>
<span class="normal"><a href="#__codelineno-0-215"> 215</a></span>
<span class="normal"><a href="#__codelineno-0-216"> 216</a></span>
<span class="normal"><a href="#__codelineno-0-217"> 217</a></span>
<span class="normal"><a href="#__codelineno-0-218"> 218</a></span>
<span class="normal"><a href="#__codelineno-0-219"> 219</a></span>
<span class="normal"><a href="#__codelineno-0-220"> 220</a></span>
<span class="normal"><a href="#__codelineno-0-221"> 221</a></span>
<span class="normal"><a href="#__codelineno-0-222"> 222</a></span>
<span class="normal"><a href="#__codelineno-0-223"> 223</a></span>
<span class="normal"><a href="#__codelineno-0-224"> 224</a></span>
<span class="normal"><a href="#__codelineno-0-225"> 225</a></span>
<span class="normal"><a href="#__codelineno-0-226"> 226</a></span>
<span class="normal"><a href="#__codelineno-0-227"> 227</a></span>
<span class="normal"><a href="#__codelineno-0-228"> 228</a></span>
<span class="normal"><a href="#__codelineno-0-229"> 229</a></span>
<span class="normal"><a href="#__codelineno-0-230"> 230</a></span>
<span class="normal"><a href="#__codelineno-0-231"> 231</a></span>
<span class="normal"><a href="#__codelineno-0-232"> 232</a></span>
<span class="normal"><a href="#__codelineno-0-233"> 233</a></span>
<span class="normal"><a href="#__codelineno-0-234"> 234</a></span>
<span class="normal"><a href="#__codelineno-0-235"> 235</a></span>
<span class="normal"><a href="#__codelineno-0-236"> 236</a></span>
<span class="normal"><a href="#__codelineno-0-237"> 237</a></span>
<span class="normal"><a href="#__codelineno-0-238"> 238</a></span>
<span class="normal"><a href="#__codelineno-0-239"> 239</a></span>
<span class="normal"><a href="#__codelineno-0-240"> 240</a></span>
<span class="normal"><a href="#__codelineno-0-241"> 241</a></span>
<span class="normal"><a href="#__codelineno-0-242"> 242</a></span>
<span class="normal"><a href="#__codelineno-0-243"> 243</a></span>
<span class="normal"><a href="#__codelineno-0-244"> 244</a></span>
<span class="normal"><a href="#__codelineno-0-245"> 245</a></span>
<span class="normal"><a href="#__codelineno-0-246"> 246</a></span>
<span class="normal"><a href="#__codelineno-0-247"> 247</a></span>
<span class="normal"><a href="#__codelineno-0-248"> 248</a></span>
<span class="normal"><a href="#__codelineno-0-249"> 249</a></span>
<span class="normal"><a href="#__codelineno-0-250"> 250</a></span>
<span class="normal"><a href="#__codelineno-0-251"> 251</a></span>
<span class="normal"><a href="#__codelineno-0-252"> 252</a></span>
<span class="normal"><a href="#__codelineno-0-253"> 253</a></span>
<span class="normal"><a href="#__codelineno-0-254"> 254</a></span>
<span class="normal"><a href="#__codelineno-0-255"> 255</a></span>
<span class="normal"><a href="#__codelineno-0-256"> 256</a></span>
<span class="normal"><a href="#__codelineno-0-257"> 257</a></span>
<span class="normal"><a href="#__codelineno-0-258"> 258</a></span>
<span class="normal"><a href="#__codelineno-0-259"> 259</a></span>
<span class="normal"><a href="#__codelineno-0-260"> 260</a></span>
<span class="normal"><a href="#__codelineno-0-261"> 261</a></span>
<span class="normal"><a href="#__codelineno-0-262"> 262</a></span>
<span class="normal"><a href="#__codelineno-0-263"> 263</a></span>
<span class="normal"><a href="#__codelineno-0-264"> 264</a></span>
<span class="normal"><a href="#__codelineno-0-265"> 265</a></span>
<span class="normal"><a href="#__codelineno-0-266"> 266</a></span>
<span class="normal"><a href="#__codelineno-0-267"> 267</a></span>
<span class="normal"><a href="#__codelineno-0-268"> 268</a></span>
<span class="normal"><a href="#__codelineno-0-269"> 269</a></span>
<span class="normal"><a href="#__codelineno-0-270"> 270</a></span>
<span class="normal"><a href="#__codelineno-0-271"> 271</a></span>
<span class="normal"><a href="#__codelineno-0-272"> 272</a></span>
<span class="normal"><a href="#__codelineno-0-273"> 273</a></span>
<span class="normal"><a href="#__codelineno-0-274"> 274</a></span>
<span class="normal"><a href="#__codelineno-0-275"> 275</a></span>
<span class="normal"><a href="#__codelineno-0-276"> 276</a></span>
<span class="normal"><a href="#__codelineno-0-277"> 277</a></span>
<span class="normal"><a href="#__codelineno-0-278"> 278</a></span>
<span class="normal"><a href="#__codelineno-0-279"> 279</a></span>
<span class="normal"><a href="#__codelineno-0-280"> 280</a></span>
<span class="normal"><a href="#__codelineno-0-281"> 281</a></span>
<span class="normal"><a href="#__codelineno-0-282"> 282</a></span>
<span class="normal"><a href="#__codelineno-0-283"> 283</a></span>
<span class="normal"><a href="#__codelineno-0-284"> 284</a></span>
<span class="normal"><a href="#__codelineno-0-285"> 285</a></span>
<span class="normal"><a href="#__codelineno-0-286"> 286</a></span>
<span class="normal"><a href="#__codelineno-0-287"> 287</a></span>
<span class="normal"><a href="#__codelineno-0-288"> 288</a></span>
<span class="normal"><a href="#__codelineno-0-289"> 289</a></span>
<span class="normal"><a href="#__codelineno-0-290"> 290</a></span>
<span class="normal"><a href="#__codelineno-0-291"> 291</a></span>
<span class="normal"><a href="#__codelineno-0-292"> 292</a></span>
<span class="normal"><a href="#__codelineno-0-293"> 293</a></span>
<span class="normal"><a href="#__codelineno-0-294"> 294</a></span>
<span class="normal"><a href="#__codelineno-0-295"> 295</a></span>
<span class="normal"><a href="#__codelineno-0-296"> 296</a></span>
<span class="normal"><a href="#__codelineno-0-297"> 297</a></span>
<span class="normal"><a href="#__codelineno-0-298"> 298</a></span>
<span class="normal"><a href="#__codelineno-0-299"> 299</a></span>
<span class="normal"><a href="#__codelineno-0-300"> 300</a></span>
<span class="normal"><a href="#__codelineno-0-301"> 301</a></span>
<span class="normal"><a href="#__codelineno-0-302"> 302</a></span>
<span class="normal"><a href="#__codelineno-0-303"> 303</a></span>
<span class="normal"><a href="#__codelineno-0-304"> 304</a></span>
<span class="normal"><a href="#__codelineno-0-305"> 305</a></span>
<span class="normal"><a href="#__codelineno-0-306"> 306</a></span>
<span class="normal"><a href="#__codelineno-0-307"> 307</a></span>
<span class="normal"><a href="#__codelineno-0-308"> 308</a></span>
<span class="normal"><a href="#__codelineno-0-309"> 309</a></span>
<span class="normal"><a href="#__codelineno-0-310"> 310</a></span>
<span class="normal"><a href="#__codelineno-0-311"> 311</a></span>
<span class="normal"><a href="#__codelineno-0-312"> 312</a></span>
<span class="normal"><a href="#__codelineno-0-313"> 313</a></span>
<span class="normal"><a href="#__codelineno-0-314"> 314</a></span>
<span class="normal"><a href="#__codelineno-0-315"> 315</a></span>
<span class="normal"><a href="#__codelineno-0-316"> 316</a></span>
<span class="normal"><a href="#__codelineno-0-317"> 317</a></span>
<span class="normal"><a href="#__codelineno-0-318"> 318</a></span>
<span class="normal"><a href="#__codelineno-0-319"> 319</a></span>
<span class="normal"><a href="#__codelineno-0-320"> 320</a></span>
<span class="normal"><a href="#__codelineno-0-321"> 321</a></span>
<span class="normal"><a href="#__codelineno-0-322"> 322</a></span>
<span class="normal"><a href="#__codelineno-0-323"> 323</a></span>
<span class="normal"><a href="#__codelineno-0-324"> 324</a></span>
<span class="normal"><a href="#__codelineno-0-325"> 325</a></span>
<span class="normal"><a href="#__codelineno-0-326"> 326</a></span>
<span class="normal"><a href="#__codelineno-0-327"> 327</a></span>
<span class="normal"><a href="#__codelineno-0-328"> 328</a></span>
<span class="normal"><a href="#__codelineno-0-329"> 329</a></span>
<span class="normal"><a href="#__codelineno-0-330"> 330</a></span>
<span class="normal"><a href="#__codelineno-0-331"> 331</a></span>
<span class="normal"><a href="#__codelineno-0-332"> 332</a></span>
<span class="normal"><a href="#__codelineno-0-333"> 333</a></span>
<span class="normal"><a href="#__codelineno-0-334"> 334</a></span>
<span class="normal"><a href="#__codelineno-0-335"> 335</a></span>
<span class="normal"><a href="#__codelineno-0-336"> 336</a></span>
<span class="normal"><a href="#__codelineno-0-337"> 337</a></span>
<span class="normal"><a href="#__codelineno-0-338"> 338</a></span>
<span class="normal"><a href="#__codelineno-0-339"> 339</a></span>
<span class="normal"><a href="#__codelineno-0-340"> 340</a></span>
<span class="normal"><a href="#__codelineno-0-341"> 341</a></span>
<span class="normal"><a href="#__codelineno-0-342"> 342</a></span>
<span class="normal"><a href="#__codelineno-0-343"> 343</a></span>
<span class="normal"><a href="#__codelineno-0-344"> 344</a></span>
<span class="normal"><a href="#__codelineno-0-345"> 345</a></span>
<span class="normal"><a href="#__codelineno-0-346"> 346</a></span>
<span class="normal"><a href="#__codelineno-0-347"> 347</a></span>
<span class="normal"><a href="#__codelineno-0-348"> 348</a></span>
<span class="normal"><a href="#__codelineno-0-349"> 349</a></span>
<span class="normal"><a href="#__codelineno-0-350"> 350</a></span>
<span class="normal"><a href="#__codelineno-0-351"> 351</a></span>
<span class="normal"><a href="#__codelineno-0-352"> 352</a></span>
<span class="normal"><a href="#__codelineno-0-353"> 353</a></span>
<span class="normal"><a href="#__codelineno-0-354"> 354</a></span>
<span class="normal"><a href="#__codelineno-0-355"> 355</a></span>
<span class="normal"><a href="#__codelineno-0-356"> 356</a></span>
<span class="normal"><a href="#__codelineno-0-357"> 357</a></span>
<span class="normal"><a href="#__codelineno-0-358"> 358</a></span>
<span class="normal"><a href="#__codelineno-0-359"> 359</a></span>
<span class="normal"><a href="#__codelineno-0-360"> 360</a></span>
<span class="normal"><a href="#__codelineno-0-361"> 361</a></span>
<span class="normal"><a href="#__codelineno-0-362"> 362</a></span>
<span class="normal"><a href="#__codelineno-0-363"> 363</a></span>
<span class="normal"><a href="#__codelineno-0-364"> 364</a></span>
<span class="normal"><a href="#__codelineno-0-365"> 365</a></span>
<span class="normal"><a href="#__codelineno-0-366"> 366</a></span>
<span class="normal"><a href="#__codelineno-0-367"> 367</a></span>
<span class="normal"><a href="#__codelineno-0-368"> 368</a></span>
<span class="normal"><a href="#__codelineno-0-369"> 369</a></span>
<span class="normal"><a href="#__codelineno-0-370"> 370</a></span>
<span class="normal"><a href="#__codelineno-0-371"> 371</a></span>
<span class="normal"><a href="#__codelineno-0-372"> 372</a></span>
<span class="normal"><a href="#__codelineno-0-373"> 373</a></span>
<span class="normal"><a href="#__codelineno-0-374"> 374</a></span>
<span class="normal"><a href="#__codelineno-0-375"> 375</a></span>
<span class="normal"><a href="#__codelineno-0-376"> 376</a></span>
<span class="normal"><a href="#__codelineno-0-377"> 377</a></span>
<span class="normal"><a href="#__codelineno-0-378"> 378</a></span>
<span class="normal"><a href="#__codelineno-0-379"> 379</a></span>
<span class="normal"><a href="#__codelineno-0-380"> 380</a></span>
<span class="normal"><a href="#__codelineno-0-381"> 381</a></span>
<span class="normal"><a href="#__codelineno-0-382"> 382</a></span>
<span class="normal"><a href="#__codelineno-0-383"> 383</a></span>
<span class="normal"><a href="#__codelineno-0-384"> 384</a></span>
<span class="normal"><a href="#__codelineno-0-385"> 385</a></span>
<span class="normal"><a href="#__codelineno-0-386"> 386</a></span>
<span class="normal"><a href="#__codelineno-0-387"> 387</a></span>
<span class="normal"><a href="#__codelineno-0-388"> 388</a></span>
<span class="normal"><a href="#__codelineno-0-389"> 389</a></span>
<span class="normal"><a href="#__codelineno-0-390"> 390</a></span>
<span class="normal"><a href="#__codelineno-0-391"> 391</a></span>
<span class="normal"><a href="#__codelineno-0-392"> 392</a></span>
<span class="normal"><a href="#__codelineno-0-393"> 393</a></span>
<span class="normal"><a href="#__codelineno-0-394"> 394</a></span>
<span class="normal"><a href="#__codelineno-0-395"> 395</a></span>
<span class="normal"><a href="#__codelineno-0-396"> 396</a></span>
<span class="normal"><a href="#__codelineno-0-397"> 397</a></span>
<span class="normal"><a href="#__codelineno-0-398"> 398</a></span>
<span class="normal"><a href="#__codelineno-0-399"> 399</a></span>
<span class="normal"><a href="#__codelineno-0-400"> 400</a></span>
<span class="normal"><a href="#__codelineno-0-401"> 401</a></span>
<span class="normal"><a href="#__codelineno-0-402"> 402</a></span>
<span class="normal"><a href="#__codelineno-0-403"> 403</a></span>
<span class="normal"><a href="#__codelineno-0-404"> 404</a></span>
<span class="normal"><a href="#__codelineno-0-405"> 405</a></span>
<span class="normal"><a href="#__codelineno-0-406"> 406</a></span>
<span class="normal"><a href="#__codelineno-0-407"> 407</a></span>
<span class="normal"><a href="#__codelineno-0-408"> 408</a></span>
<span class="normal"><a href="#__codelineno-0-409"> 409</a></span>
<span class="normal"><a href="#__codelineno-0-410"> 410</a></span>
<span class="normal"><a href="#__codelineno-0-411"> 411</a></span>
<span class="normal"><a href="#__codelineno-0-412"> 412</a></span>
<span class="normal"><a href="#__codelineno-0-413"> 413</a></span>
<span class="normal"><a href="#__codelineno-0-414"> 414</a></span>
<span class="normal"><a href="#__codelineno-0-415"> 415</a></span>
<span class="normal"><a href="#__codelineno-0-416"> 416</a></span>
<span class="normal"><a href="#__codelineno-0-417"> 417</a></span>
<span class="normal"><a href="#__codelineno-0-418"> 418</a></span>
<span class="normal"><a href="#__codelineno-0-419"> 419</a></span>
<span class="normal"><a href="#__codelineno-0-420"> 420</a></span>
<span class="normal"><a href="#__codelineno-0-421"> 421</a></span>
<span class="normal"><a href="#__codelineno-0-422"> 422</a></span>
<span class="normal"><a href="#__codelineno-0-423"> 423</a></span>
<span class="normal"><a href="#__codelineno-0-424"> 424</a></span>
<span class="normal"><a href="#__codelineno-0-425"> 425</a></span>
<span class="normal"><a href="#__codelineno-0-426"> 426</a></span>
<span class="normal"><a href="#__codelineno-0-427"> 427</a></span>
<span class="normal"><a href="#__codelineno-0-428"> 428</a></span>
<span class="normal"><a href="#__codelineno-0-429"> 429</a></span>
<span class="normal"><a href="#__codelineno-0-430"> 430</a></span>
<span class="normal"><a href="#__codelineno-0-431"> 431</a></span>
<span class="normal"><a href="#__codelineno-0-432"> 432</a></span>
<span class="normal"><a href="#__codelineno-0-433"> 433</a></span>
<span class="normal"><a href="#__codelineno-0-434"> 434</a></span>
<span class="normal"><a href="#__codelineno-0-435"> 435</a></span>
<span class="normal"><a href="#__codelineno-0-436"> 436</a></span>
<span class="normal"><a href="#__codelineno-0-437"> 437</a></span>
<span class="normal"><a href="#__codelineno-0-438"> 438</a></span>
<span class="normal"><a href="#__codelineno-0-439"> 439</a></span>
<span class="normal"><a href="#__codelineno-0-440"> 440</a></span>
<span class="normal"><a href="#__codelineno-0-441"> 441</a></span>
<span class="normal"><a href="#__codelineno-0-442"> 442</a></span>
<span class="normal"><a href="#__codelineno-0-443"> 443</a></span>
<span class="normal"><a href="#__codelineno-0-444"> 444</a></span>
<span class="normal"><a href="#__codelineno-0-445"> 445</a></span>
<span class="normal"><a href="#__codelineno-0-446"> 446</a></span>
<span class="normal"><a href="#__codelineno-0-447"> 447</a></span>
<span class="normal"><a href="#__codelineno-0-448"> 448</a></span>
<span class="normal"><a href="#__codelineno-0-449"> 449</a></span>
<span class="normal"><a href="#__codelineno-0-450"> 450</a></span>
<span class="normal"><a href="#__codelineno-0-451"> 451</a></span>
<span class="normal"><a href="#__codelineno-0-452"> 452</a></span>
<span class="normal"><a href="#__codelineno-0-453"> 453</a></span>
<span class="normal"><a href="#__codelineno-0-454"> 454</a></span>
<span class="normal"><a href="#__codelineno-0-455"> 455</a></span>
<span class="normal"><a href="#__codelineno-0-456"> 456</a></span>
<span class="normal"><a href="#__codelineno-0-457"> 457</a></span>
<span class="normal"><a href="#__codelineno-0-458"> 458</a></span>
<span class="normal"><a href="#__codelineno-0-459"> 459</a></span>
<span class="normal"><a href="#__codelineno-0-460"> 460</a></span>
<span class="normal"><a href="#__codelineno-0-461"> 461</a></span>
<span class="normal"><a href="#__codelineno-0-462"> 462</a></span>
<span class="normal"><a href="#__codelineno-0-463"> 463</a></span>
<span class="normal"><a href="#__codelineno-0-464"> 464</a></span>
<span class="normal"><a href="#__codelineno-0-465"> 465</a></span>
<span class="normal"><a href="#__codelineno-0-466"> 466</a></span>
<span class="normal"><a href="#__codelineno-0-467"> 467</a></span>
<span class="normal"><a href="#__codelineno-0-468"> 468</a></span>
<span class="normal"><a href="#__codelineno-0-469"> 469</a></span>
<span class="normal"><a href="#__codelineno-0-470"> 470</a></span>
<span class="normal"><a href="#__codelineno-0-471"> 471</a></span>
<span class="normal"><a href="#__codelineno-0-472"> 472</a></span>
<span class="normal"><a href="#__codelineno-0-473"> 473</a></span>
<span class="normal"><a href="#__codelineno-0-474"> 474</a></span>
<span class="normal"><a href="#__codelineno-0-475"> 475</a></span>
<span class="normal"><a href="#__codelineno-0-476"> 476</a></span>
<span class="normal"><a href="#__codelineno-0-477"> 477</a></span>
<span class="normal"><a href="#__codelineno-0-478"> 478</a></span>
<span class="normal"><a href="#__codelineno-0-479"> 479</a></span>
<span class="normal"><a href="#__codelineno-0-480"> 480</a></span>
<span class="normal"><a href="#__codelineno-0-481"> 481</a></span>
<span class="normal"><a href="#__codelineno-0-482"> 482</a></span>
<span class="normal"><a href="#__codelineno-0-483"> 483</a></span>
<span class="normal"><a href="#__codelineno-0-484"> 484</a></span>
<span class="normal"><a href="#__codelineno-0-485"> 485</a></span>
<span class="normal"><a href="#__codelineno-0-486"> 486</a></span>
<span class="normal"><a href="#__codelineno-0-487"> 487</a></span>
<span class="normal"><a href="#__codelineno-0-488"> 488</a></span>
<span class="normal"><a href="#__codelineno-0-489"> 489</a></span>
<span class="normal"><a href="#__codelineno-0-490"> 490</a></span>
<span class="normal"><a href="#__codelineno-0-491"> 491</a></span>
<span class="normal"><a href="#__codelineno-0-492"> 492</a></span>
<span class="normal"><a href="#__codelineno-0-493"> 493</a></span>
<span class="normal"><a href="#__codelineno-0-494"> 494</a></span>
<span class="normal"><a href="#__codelineno-0-495"> 495</a></span>
<span class="normal"><a href="#__codelineno-0-496"> 496</a></span>
<span class="normal"><a href="#__codelineno-0-497"> 497</a></span>
<span class="normal"><a href="#__codelineno-0-498"> 498</a></span>
<span class="normal"><a href="#__codelineno-0-499"> 499</a></span>
<span class="normal"><a href="#__codelineno-0-500"> 500</a></span>
<span class="normal"><a href="#__codelineno-0-501"> 501</a></span>
<span class="normal"><a href="#__codelineno-0-502"> 502</a></span>
<span class="normal"><a href="#__codelineno-0-503"> 503</a></span>
<span class="normal"><a href="#__codelineno-0-504"> 504</a></span>
<span class="normal"><a href="#__codelineno-0-505"> 505</a></span>
<span class="normal"><a href="#__codelineno-0-506"> 506</a></span>
<span class="normal"><a href="#__codelineno-0-507"> 507</a></span>
<span class="normal"><a href="#__codelineno-0-508"> 508</a></span>
<span class="normal"><a href="#__codelineno-0-509"> 509</a></span>
<span class="normal"><a href="#__codelineno-0-510"> 510</a></span>
<span class="normal"><a href="#__codelineno-0-511"> 511</a></span>
<span class="normal"><a href="#__codelineno-0-512"> 512</a></span>
<span class="normal"><a href="#__codelineno-0-513"> 513</a></span>
<span class="normal"><a href="#__codelineno-0-514"> 514</a></span>
<span class="normal"><a href="#__codelineno-0-515"> 515</a></span>
<span class="normal"><a href="#__codelineno-0-516"> 516</a></span>
<span class="normal"><a href="#__codelineno-0-517"> 517</a></span>
<span class="normal"><a href="#__codelineno-0-518"> 518</a></span>
<span class="normal"><a href="#__codelineno-0-519"> 519</a></span>
<span class="normal"><a href="#__codelineno-0-520"> 520</a></span>
<span class="normal"><a href="#__codelineno-0-521"> 521</a></span>
<span class="normal"><a href="#__codelineno-0-522"> 522</a></span>
<span class="normal"><a href="#__codelineno-0-523"> 523</a></span>
<span class="normal"><a href="#__codelineno-0-524"> 524</a></span>
<span class="normal"><a href="#__codelineno-0-525"> 525</a></span>
<span class="normal"><a href="#__codelineno-0-526"> 526</a></span>
<span class="normal"><a href="#__codelineno-0-527"> 527</a></span>
<span class="normal"><a href="#__codelineno-0-528"> 528</a></span>
<span class="normal"><a href="#__codelineno-0-529"> 529</a></span>
<span class="normal"><a href="#__codelineno-0-530"> 530</a></span>
<span class="normal"><a href="#__codelineno-0-531"> 531</a></span>
<span class="normal"><a href="#__codelineno-0-532"> 532</a></span>
<span class="normal"><a href="#__codelineno-0-533"> 533</a></span>
<span class="normal"><a href="#__codelineno-0-534"> 534</a></span>
<span class="normal"><a href="#__codelineno-0-535"> 535</a></span>
<span class="normal"><a href="#__codelineno-0-536"> 536</a></span>
<span class="normal"><a href="#__codelineno-0-537"> 537</a></span>
<span class="normal"><a href="#__codelineno-0-538"> 538</a></span>
<span class="normal"><a href="#__codelineno-0-539"> 539</a></span>
<span class="normal"><a href="#__codelineno-0-540"> 540</a></span>
<span class="normal"><a href="#__codelineno-0-541"> 541</a></span>
<span class="normal"><a href="#__codelineno-0-542"> 542</a></span>
<span class="normal"><a href="#__codelineno-0-543"> 543</a></span>
<span class="normal"><a href="#__codelineno-0-544"> 544</a></span>
<span class="normal"><a href="#__codelineno-0-545"> 545</a></span>
<span class="normal"><a href="#__codelineno-0-546"> 546</a></span>
<span class="normal"><a href="#__codelineno-0-547"> 547</a></span>
<span class="normal"><a href="#__codelineno-0-548"> 548</a></span>
<span class="normal"><a href="#__codelineno-0-549"> 549</a></span>
<span class="normal"><a href="#__codelineno-0-550"> 550</a></span>
<span class="normal"><a href="#__codelineno-0-551"> 551</a></span>
<span class="normal"><a href="#__codelineno-0-552"> 552</a></span>
<span class="normal"><a href="#__codelineno-0-553"> 553</a></span>
<span class="normal"><a href="#__codelineno-0-554"> 554</a></span>
<span class="normal"><a href="#__codelineno-0-555"> 555</a></span>
<span class="normal"><a href="#__codelineno-0-556"> 556</a></span>
<span class="normal"><a href="#__codelineno-0-557"> 557</a></span>
<span class="normal"><a href="#__codelineno-0-558"> 558</a></span>
<span class="normal"><a href="#__codelineno-0-559"> 559</a></span>
<span class="normal"><a href="#__codelineno-0-560"> 560</a></span>
<span class="normal"><a href="#__codelineno-0-561"> 561</a></span>
<span class="normal"><a href="#__codelineno-0-562"> 562</a></span>
<span class="normal"><a href="#__codelineno-0-563"> 563</a></span>
<span class="normal"><a href="#__codelineno-0-564"> 564</a></span>
<span class="normal"><a href="#__codelineno-0-565"> 565</a></span>
<span class="normal"><a href="#__codelineno-0-566"> 566</a></span>
<span class="normal"><a href="#__codelineno-0-567"> 567</a></span>
<span class="normal"><a href="#__codelineno-0-568"> 568</a></span>
<span class="normal"><a href="#__codelineno-0-569"> 569</a></span>
<span class="normal"><a href="#__codelineno-0-570"> 570</a></span>
<span class="normal"><a href="#__codelineno-0-571"> 571</a></span>
<span class="normal"><a href="#__codelineno-0-572"> 572</a></span>
<span class="normal"><a href="#__codelineno-0-573"> 573</a></span>
<span class="normal"><a href="#__codelineno-0-574"> 574</a></span>
<span class="normal"><a href="#__codelineno-0-575"> 575</a></span>
<span class="normal"><a href="#__codelineno-0-576"> 576</a></span>
<span class="normal"><a href="#__codelineno-0-577"> 577</a></span>
<span class="normal"><a href="#__codelineno-0-578"> 578</a></span>
<span class="normal"><a href="#__codelineno-0-579"> 579</a></span>
<span class="normal"><a href="#__codelineno-0-580"> 580</a></span>
<span class="normal"><a href="#__codelineno-0-581"> 581</a></span>
<span class="normal"><a href="#__codelineno-0-582"> 582</a></span>
<span class="normal"><a href="#__codelineno-0-583"> 583</a></span>
<span class="normal"><a href="#__codelineno-0-584"> 584</a></span>
<span class="normal"><a href="#__codelineno-0-585"> 585</a></span>
<span class="normal"><a href="#__codelineno-0-586"> 586</a></span>
<span class="normal"><a href="#__codelineno-0-587"> 587</a></span>
<span class="normal"><a href="#__codelineno-0-588"> 588</a></span>
<span class="normal"><a href="#__codelineno-0-589"> 589</a></span>
<span class="normal"><a href="#__codelineno-0-590"> 590</a></span>
<span class="normal"><a href="#__codelineno-0-591"> 591</a></span>
<span class="normal"><a href="#__codelineno-0-592"> 592</a></span>
<span class="normal"><a href="#__codelineno-0-593"> 593</a></span>
<span class="normal"><a href="#__codelineno-0-594"> 594</a></span>
<span class="normal"><a href="#__codelineno-0-595"> 595</a></span>
<span class="normal"><a href="#__codelineno-0-596"> 596</a></span>
<span class="normal"><a href="#__codelineno-0-597"> 597</a></span>
<span class="normal"><a href="#__codelineno-0-598"> 598</a></span>
<span class="normal"><a href="#__codelineno-0-599"> 599</a></span>
<span class="normal"><a href="#__codelineno-0-600"> 600</a></span>
<span class="normal"><a href="#__codelineno-0-601"> 601</a></span>
<span class="normal"><a href="#__codelineno-0-602"> 602</a></span>
<span class="normal"><a href="#__codelineno-0-603"> 603</a></span>
<span class="normal"><a href="#__codelineno-0-604"> 604</a></span>
<span class="normal"><a href="#__codelineno-0-605"> 605</a></span>
<span class="normal"><a href="#__codelineno-0-606"> 606</a></span>
<span class="normal"><a href="#__codelineno-0-607"> 607</a></span>
<span class="normal"><a href="#__codelineno-0-608"> 608</a></span>
<span class="normal"><a href="#__codelineno-0-609"> 609</a></span>
<span class="normal"><a href="#__codelineno-0-610"> 610</a></span>
<span class="normal"><a href="#__codelineno-0-611"> 611</a></span>
<span class="normal"><a href="#__codelineno-0-612"> 612</a></span>
<span class="normal"><a href="#__codelineno-0-613"> 613</a></span>
<span class="normal"><a href="#__codelineno-0-614"> 614</a></span>
<span class="normal"><a href="#__codelineno-0-615"> 615</a></span>
<span class="normal"><a href="#__codelineno-0-616"> 616</a></span>
<span class="normal"><a href="#__codelineno-0-617"> 617</a></span>
<span class="normal"><a href="#__codelineno-0-618"> 618</a></span>
<span class="normal"><a href="#__codelineno-0-619"> 619</a></span>
<span class="normal"><a href="#__codelineno-0-620"> 620</a></span>
<span class="normal"><a href="#__codelineno-0-621"> 621</a></span>
<span class="normal"><a href="#__codelineno-0-622"> 622</a></span>
<span class="normal"><a href="#__codelineno-0-623"> 623</a></span>
<span class="normal"><a href="#__codelineno-0-624"> 624</a></span>
<span class="normal"><a href="#__codelineno-0-625"> 625</a></span>
<span class="normal"><a href="#__codelineno-0-626"> 626</a></span>
<span class="normal"><a href="#__codelineno-0-627"> 627</a></span>
<span class="normal"><a href="#__codelineno-0-628"> 628</a></span>
<span class="normal"><a href="#__codelineno-0-629"> 629</a></span>
<span class="normal"><a href="#__codelineno-0-630"> 630</a></span>
<span class="normal"><a href="#__codelineno-0-631"> 631</a></span>
<span class="normal"><a href="#__codelineno-0-632"> 632</a></span>
<span class="normal"><a href="#__codelineno-0-633"> 633</a></span>
<span class="normal"><a href="#__codelineno-0-634"> 634</a></span>
<span class="normal"><a href="#__codelineno-0-635"> 635</a></span>
<span class="normal"><a href="#__codelineno-0-636"> 636</a></span>
<span class="normal"><a href="#__codelineno-0-637"> 637</a></span>
<span class="normal"><a href="#__codelineno-0-638"> 638</a></span>
<span class="normal"><a href="#__codelineno-0-639"> 639</a></span>
<span class="normal"><a href="#__codelineno-0-640"> 640</a></span>
<span class="normal"><a href="#__codelineno-0-641"> 641</a></span>
<span class="normal"><a href="#__codelineno-0-642"> 642</a></span>
<span class="normal"><a href="#__codelineno-0-643"> 643</a></span>
<span class="normal"><a href="#__codelineno-0-644"> 644</a></span>
<span class="normal"><a href="#__codelineno-0-645"> 645</a></span>
<span class="normal"><a href="#__codelineno-0-646"> 646</a></span>
<span class="normal"><a href="#__codelineno-0-647"> 647</a></span>
<span class="normal"><a href="#__codelineno-0-648"> 648</a></span>
<span class="normal"><a href="#__codelineno-0-649"> 649</a></span>
<span class="normal"><a href="#__codelineno-0-650"> 650</a></span>
<span class="normal"><a href="#__codelineno-0-651"> 651</a></span>
<span class="normal"><a href="#__codelineno-0-652"> 652</a></span>
<span class="normal"><a href="#__codelineno-0-653"> 653</a></span>
<span class="normal"><a href="#__codelineno-0-654"> 654</a></span>
<span class="normal"><a href="#__codelineno-0-655"> 655</a></span>
<span class="normal"><a href="#__codelineno-0-656"> 656</a></span>
<span class="normal"><a href="#__codelineno-0-657"> 657</a></span>
<span class="normal"><a href="#__codelineno-0-658"> 658</a></span>
<span class="normal"><a href="#__codelineno-0-659"> 659</a></span>
<span class="normal"><a href="#__codelineno-0-660"> 660</a></span>
<span class="normal"><a href="#__codelineno-0-661"> 661</a></span>
<span class="normal"><a href="#__codelineno-0-662"> 662</a></span>
<span class="normal"><a href="#__codelineno-0-663"> 663</a></span>
<span class="normal"><a href="#__codelineno-0-664"> 664</a></span>
<span class="normal"><a href="#__codelineno-0-665"> 665</a></span>
<span class="normal"><a href="#__codelineno-0-666"> 666</a></span>
<span class="normal"><a href="#__codelineno-0-667"> 667</a></span>
<span class="normal"><a href="#__codelineno-0-668"> 668</a></span>
<span class="normal"><a href="#__codelineno-0-669"> 669</a></span>
<span class="normal"><a href="#__codelineno-0-670"> 670</a></span>
<span class="normal"><a href="#__codelineno-0-671"> 671</a></span>
<span class="normal"><a href="#__codelineno-0-672"> 672</a></span>
<span class="normal"><a href="#__codelineno-0-673"> 673</a></span>
<span class="normal"><a href="#__codelineno-0-674"> 674</a></span>
<span class="normal"><a href="#__codelineno-0-675"> 675</a></span>
<span class="normal"><a href="#__codelineno-0-676"> 676</a></span>
<span class="normal"><a href="#__codelineno-0-677"> 677</a></span>
<span class="normal"><a href="#__codelineno-0-678"> 678</a></span>
<span class="normal"><a href="#__codelineno-0-679"> 679</a></span>
<span class="normal"><a href="#__codelineno-0-680"> 680</a></span>
<span class="normal"><a href="#__codelineno-0-681"> 681</a></span>
<span class="normal"><a href="#__codelineno-0-682"> 682</a></span>
<span class="normal"><a href="#__codelineno-0-683"> 683</a></span>
<span class="normal"><a href="#__codelineno-0-684"> 684</a></span>
<span class="normal"><a href="#__codelineno-0-685"> 685</a></span>
<span class="normal"><a href="#__codelineno-0-686"> 686</a></span>
<span class="normal"><a href="#__codelineno-0-687"> 687</a></span>
<span class="normal"><a href="#__codelineno-0-688"> 688</a></span>
<span class="normal"><a href="#__codelineno-0-689"> 689</a></span>
<span class="normal"><a href="#__codelineno-0-690"> 690</a></span>
<span class="normal"><a href="#__codelineno-0-691"> 691</a></span>
<span class="normal"><a href="#__codelineno-0-692"> 692</a></span>
<span class="normal"><a href="#__codelineno-0-693"> 693</a></span>
<span class="normal"><a href="#__codelineno-0-694"> 694</a></span>
<span class="normal"><a href="#__codelineno-0-695"> 695</a></span>
<span class="normal"><a href="#__codelineno-0-696"> 696</a></span>
<span class="normal"><a href="#__codelineno-0-697"> 697</a></span>
<span class="normal"><a href="#__codelineno-0-698"> 698</a></span>
<span class="normal"><a href="#__codelineno-0-699"> 699</a></span>
<span class="normal"><a href="#__codelineno-0-700"> 700</a></span>
<span class="normal"><a href="#__codelineno-0-701"> 701</a></span>
<span class="normal"><a href="#__codelineno-0-702"> 702</a></span>
<span class="normal"><a href="#__codelineno-0-703"> 703</a></span>
<span class="normal"><a href="#__codelineno-0-704"> 704</a></span>
<span class="normal"><a href="#__codelineno-0-705"> 705</a></span>
<span class="normal"><a href="#__codelineno-0-706"> 706</a></span>
<span class="normal"><a href="#__codelineno-0-707"> 707</a></span>
<span class="normal"><a href="#__codelineno-0-708"> 708</a></span>
<span class="normal"><a href="#__codelineno-0-709"> 709</a></span>
<span class="normal"><a href="#__codelineno-0-710"> 710</a></span>
<span class="normal"><a href="#__codelineno-0-711"> 711</a></span>
<span class="normal"><a href="#__codelineno-0-712"> 712</a></span>
<span class="normal"><a href="#__codelineno-0-713"> 713</a></span>
<span class="normal"><a href="#__codelineno-0-714"> 714</a></span>
<span class="normal"><a href="#__codelineno-0-715"> 715</a></span>
<span class="normal"><a href="#__codelineno-0-716"> 716</a></span>
<span class="normal"><a href="#__codelineno-0-717"> 717</a></span>
<span class="normal"><a href="#__codelineno-0-718"> 718</a></span>
<span class="normal"><a href="#__codelineno-0-719"> 719</a></span>
<span class="normal"><a href="#__codelineno-0-720"> 720</a></span>
<span class="normal"><a href="#__codelineno-0-721"> 721</a></span>
<span class="normal"><a href="#__codelineno-0-722"> 722</a></span>
<span class="normal"><a href="#__codelineno-0-723"> 723</a></span>
<span class="normal"><a href="#__codelineno-0-724"> 724</a></span>
<span class="normal"><a href="#__codelineno-0-725"> 725</a></span>
<span class="normal"><a href="#__codelineno-0-726"> 726</a></span>
<span class="normal"><a href="#__codelineno-0-727"> 727</a></span>
<span class="normal"><a href="#__codelineno-0-728"> 728</a></span>
<span class="normal"><a href="#__codelineno-0-729"> 729</a></span>
<span class="normal"><a href="#__codelineno-0-730"> 730</a></span>
<span class="normal"><a href="#__codelineno-0-731"> 731</a></span>
<span class="normal"><a href="#__codelineno-0-732"> 732</a></span>
<span class="normal"><a href="#__codelineno-0-733"> 733</a></span>
<span class="normal"><a href="#__codelineno-0-734"> 734</a></span>
<span class="normal"><a href="#__codelineno-0-735"> 735</a></span>
<span class="normal"><a href="#__codelineno-0-736"> 736</a></span>
<span class="normal"><a href="#__codelineno-0-737"> 737</a></span>
<span class="normal"><a href="#__codelineno-0-738"> 738</a></span>
<span class="normal"><a href="#__codelineno-0-739"> 739</a></span>
<span class="normal"><a href="#__codelineno-0-740"> 740</a></span>
<span class="normal"><a href="#__codelineno-0-741"> 741</a></span>
<span class="normal"><a href="#__codelineno-0-742"> 742</a></span>
<span class="normal"><a href="#__codelineno-0-743"> 743</a></span>
<span class="normal"><a href="#__codelineno-0-744"> 744</a></span>
<span class="normal"><a href="#__codelineno-0-745"> 745</a></span>
<span class="normal"><a href="#__codelineno-0-746"> 746</a></span>
<span class="normal"><a href="#__codelineno-0-747"> 747</a></span>
<span class="normal"><a href="#__codelineno-0-748"> 748</a></span>
<span class="normal"><a href="#__codelineno-0-749"> 749</a></span>
<span class="normal"><a href="#__codelineno-0-750"> 750</a></span>
<span class="normal"><a href="#__codelineno-0-751"> 751</a></span>
<span class="normal"><a href="#__codelineno-0-752"> 752</a></span>
<span class="normal"><a href="#__codelineno-0-753"> 753</a></span>
<span class="normal"><a href="#__codelineno-0-754"> 754</a></span>
<span class="normal"><a href="#__codelineno-0-755"> 755</a></span>
<span class="normal"><a href="#__codelineno-0-756"> 756</a></span>
<span class="normal"><a href="#__codelineno-0-757"> 757</a></span>
<span class="normal"><a href="#__codelineno-0-758"> 758</a></span>
<span class="normal"><a href="#__codelineno-0-759"> 759</a></span>
<span class="normal"><a href="#__codelineno-0-760"> 760</a></span>
<span class="normal"><a href="#__codelineno-0-761"> 761</a></span>
<span class="normal"><a href="#__codelineno-0-762"> 762</a></span>
<span class="normal"><a href="#__codelineno-0-763"> 763</a></span>
<span class="normal"><a href="#__codelineno-0-764"> 764</a></span>
<span class="normal"><a href="#__codelineno-0-765"> 765</a></span>
<span class="normal"><a href="#__codelineno-0-766"> 766</a></span>
<span class="normal"><a href="#__codelineno-0-767"> 767</a></span>
<span class="normal"><a href="#__codelineno-0-768"> 768</a></span>
<span class="normal"><a href="#__codelineno-0-769"> 769</a></span>
<span class="normal"><a href="#__codelineno-0-770"> 770</a></span>
<span class="normal"><a href="#__codelineno-0-771"> 771</a></span>
<span class="normal"><a href="#__codelineno-0-772"> 772</a></span>
<span class="normal"><a href="#__codelineno-0-773"> 773</a></span>
<span class="normal"><a href="#__codelineno-0-774"> 774</a></span>
<span class="normal"><a href="#__codelineno-0-775"> 775</a></span>
<span class="normal"><a href="#__codelineno-0-776"> 776</a></span>
<span class="normal"><a href="#__codelineno-0-777"> 777</a></span>
<span class="normal"><a href="#__codelineno-0-778"> 778</a></span>
<span class="normal"><a href="#__codelineno-0-779"> 779</a></span>
<span class="normal"><a href="#__codelineno-0-780"> 780</a></span>
<span class="normal"><a href="#__codelineno-0-781"> 781</a></span>
<span class="normal"><a href="#__codelineno-0-782"> 782</a></span>
<span class="normal"><a href="#__codelineno-0-783"> 783</a></span>
<span class="normal"><a href="#__codelineno-0-784"> 784</a></span>
<span class="normal"><a href="#__codelineno-0-785"> 785</a></span>
<span class="normal"><a href="#__codelineno-0-786"> 786</a></span>
<span class="normal"><a href="#__codelineno-0-787"> 787</a></span>
<span class="normal"><a href="#__codelineno-0-788"> 788</a></span>
<span class="normal"><a href="#__codelineno-0-789"> 789</a></span>
<span class="normal"><a href="#__codelineno-0-790"> 790</a></span>
<span class="normal"><a href="#__codelineno-0-791"> 791</a></span>
<span class="normal"><a href="#__codelineno-0-792"> 792</a></span>
<span class="normal"><a href="#__codelineno-0-793"> 793</a></span>
<span class="normal"><a href="#__codelineno-0-794"> 794</a></span>
<span class="normal"><a href="#__codelineno-0-795"> 795</a></span>
<span class="normal"><a href="#__codelineno-0-796"> 796</a></span>
<span class="normal"><a href="#__codelineno-0-797"> 797</a></span>
<span class="normal"><a href="#__codelineno-0-798"> 798</a></span>
<span class="normal"><a href="#__codelineno-0-799"> 799</a></span>
<span class="normal"><a href="#__codelineno-0-800"> 800</a></span>
<span class="normal"><a href="#__codelineno-0-801"> 801</a></span>
<span class="normal"><a href="#__codelineno-0-802"> 802</a></span>
<span class="normal"><a href="#__codelineno-0-803"> 803</a></span>
<span class="normal"><a href="#__codelineno-0-804"> 804</a></span>
<span class="normal"><a href="#__codelineno-0-805"> 805</a></span>
<span class="normal"><a href="#__codelineno-0-806"> 806</a></span>
<span class="normal"><a href="#__codelineno-0-807"> 807</a></span>
<span class="normal"><a href="#__codelineno-0-808"> 808</a></span>
<span class="normal"><a href="#__codelineno-0-809"> 809</a></span>
<span class="normal"><a href="#__codelineno-0-810"> 810</a></span>
<span class="normal"><a href="#__codelineno-0-811"> 811</a></span>
<span class="normal"><a href="#__codelineno-0-812"> 812</a></span>
<span class="normal"><a href="#__codelineno-0-813"> 813</a></span>
<span class="normal"><a href="#__codelineno-0-814"> 814</a></span>
<span class="normal"><a href="#__codelineno-0-815"> 815</a></span>
<span class="normal"><a href="#__codelineno-0-816"> 816</a></span>
<span class="normal"><a href="#__codelineno-0-817"> 817</a></span>
<span class="normal"><a href="#__codelineno-0-818"> 818</a></span>
<span class="normal"><a href="#__codelineno-0-819"> 819</a></span>
<span class="normal"><a href="#__codelineno-0-820"> 820</a></span>
<span class="normal"><a href="#__codelineno-0-821"> 821</a></span>
<span class="normal"><a href="#__codelineno-0-822"> 822</a></span>
<span class="normal"><a href="#__codelineno-0-823"> 823</a></span>
<span class="normal"><a href="#__codelineno-0-824"> 824</a></span>
<span class="normal"><a href="#__codelineno-0-825"> 825</a></span>
<span class="normal"><a href="#__codelineno-0-826"> 826</a></span>
<span class="normal"><a href="#__codelineno-0-827"> 827</a></span>
<span class="normal"><a href="#__codelineno-0-828"> 828</a></span>
<span class="normal"><a href="#__codelineno-0-829"> 829</a></span>
<span class="normal"><a href="#__codelineno-0-830"> 830</a></span>
<span class="normal"><a href="#__codelineno-0-831"> 831</a></span>
<span class="normal"><a href="#__codelineno-0-832"> 832</a></span>
<span class="normal"><a href="#__codelineno-0-833"> 833</a></span>
<span class="normal"><a href="#__codelineno-0-834"> 834</a></span>
<span class="normal"><a href="#__codelineno-0-835"> 835</a></span>
<span class="normal"><a href="#__codelineno-0-836"> 836</a></span>
<span class="normal"><a href="#__codelineno-0-837"> 837</a></span>
<span class="normal"><a href="#__codelineno-0-838"> 838</a></span>
<span class="normal"><a href="#__codelineno-0-839"> 839</a></span>
<span class="normal"><a href="#__codelineno-0-840"> 840</a></span>
<span class="normal"><a href="#__codelineno-0-841"> 841</a></span>
<span class="normal"><a href="#__codelineno-0-842"> 842</a></span>
<span class="normal"><a href="#__codelineno-0-843"> 843</a></span>
<span class="normal"><a href="#__codelineno-0-844"> 844</a></span>
<span class="normal"><a href="#__codelineno-0-845"> 845</a></span>
<span class="normal"><a href="#__codelineno-0-846"> 846</a></span>
<span class="normal"><a href="#__codelineno-0-847"> 847</a></span>
<span class="normal"><a href="#__codelineno-0-848"> 848</a></span>
<span class="normal"><a href="#__codelineno-0-849"> 849</a></span>
<span class="normal"><a href="#__codelineno-0-850"> 850</a></span>
<span class="normal"><a href="#__codelineno-0-851"> 851</a></span>
<span class="normal"><a href="#__codelineno-0-852"> 852</a></span>
<span class="normal"><a href="#__codelineno-0-853"> 853</a></span>
<span class="normal"><a href="#__codelineno-0-854"> 854</a></span>
<span class="normal"><a href="#__codelineno-0-855"> 855</a></span>
<span class="normal"><a href="#__codelineno-0-856"> 856</a></span>
<span class="normal"><a href="#__codelineno-0-857"> 857</a></span>
<span class="normal"><a href="#__codelineno-0-858"> 858</a></span>
<span class="normal"><a href="#__codelineno-0-859"> 859</a></span>
<span class="normal"><a href="#__codelineno-0-860"> 860</a></span>
<span class="normal"><a href="#__codelineno-0-861"> 861</a></span>
<span class="normal"><a href="#__codelineno-0-862"> 862</a></span>
<span class="normal"><a href="#__codelineno-0-863"> 863</a></span>
<span class="normal"><a href="#__codelineno-0-864"> 864</a></span>
<span class="normal"><a href="#__codelineno-0-865"> 865</a></span>
<span class="normal"><a href="#__codelineno-0-866"> 866</a></span>
<span class="normal"><a href="#__codelineno-0-867"> 867</a></span>
<span class="normal"><a href="#__codelineno-0-868"> 868</a></span>
<span class="normal"><a href="#__codelineno-0-869"> 869</a></span>
<span class="normal"><a href="#__codelineno-0-870"> 870</a></span>
<span class="normal"><a href="#__codelineno-0-871"> 871</a></span>
<span class="normal"><a href="#__codelineno-0-872"> 872</a></span>
<span class="normal"><a href="#__codelineno-0-873"> 873</a></span>
<span class="normal"><a href="#__codelineno-0-874"> 874</a></span>
<span class="normal"><a href="#__codelineno-0-875"> 875</a></span>
<span class="normal"><a href="#__codelineno-0-876"> 876</a></span>
<span class="normal"><a href="#__codelineno-0-877"> 877</a></span>
<span class="normal"><a href="#__codelineno-0-878"> 878</a></span>
<span class="normal"><a href="#__codelineno-0-879"> 879</a></span>
<span class="normal"><a href="#__codelineno-0-880"> 880</a></span>
<span class="normal"><a href="#__codelineno-0-881"> 881</a></span>
<span class="normal"><a href="#__codelineno-0-882"> 882</a></span>
<span class="normal"><a href="#__codelineno-0-883"> 883</a></span>
<span class="normal"><a href="#__codelineno-0-884"> 884</a></span>
<span class="normal"><a href="#__codelineno-0-885"> 885</a></span>
<span class="normal"><a href="#__codelineno-0-886"> 886</a></span>
<span class="normal"><a href="#__codelineno-0-887"> 887</a></span>
<span class="normal"><a href="#__codelineno-0-888"> 888</a></span>
<span class="normal"><a href="#__codelineno-0-889"> 889</a></span>
<span class="normal"><a href="#__codelineno-0-890"> 890</a></span>
<span class="normal"><a href="#__codelineno-0-891"> 891</a></span>
<span class="normal"><a href="#__codelineno-0-892"> 892</a></span>
<span class="normal"><a href="#__codelineno-0-893"> 893</a></span>
<span class="normal"><a href="#__codelineno-0-894"> 894</a></span>
<span class="normal"><a href="#__codelineno-0-895"> 895</a></span>
<span class="normal"><a href="#__codelineno-0-896"> 896</a></span>
<span class="normal"><a href="#__codelineno-0-897"> 897</a></span>
<span class="normal"><a href="#__codelineno-0-898"> 898</a></span>
<span class="normal"><a href="#__codelineno-0-899"> 899</a></span>
<span class="normal"><a href="#__codelineno-0-900"> 900</a></span>
<span class="normal"><a href="#__codelineno-0-901"> 901</a></span>
<span class="normal"><a href="#__codelineno-0-902"> 902</a></span>
<span class="normal"><a href="#__codelineno-0-903"> 903</a></span>
<span class="normal"><a href="#__codelineno-0-904"> 904</a></span>
<span class="normal"><a href="#__codelineno-0-905"> 905</a></span>
<span class="normal"><a href="#__codelineno-0-906"> 906</a></span>
<span class="normal"><a href="#__codelineno-0-907"> 907</a></span>
<span class="normal"><a href="#__codelineno-0-908"> 908</a></span>
<span class="normal"><a href="#__codelineno-0-909"> 909</a></span>
<span class="normal"><a href="#__codelineno-0-910"> 910</a></span>
<span class="normal"><a href="#__codelineno-0-911"> 911</a></span>
<span class="normal"><a href="#__codelineno-0-912"> 912</a></span>
<span class="normal"><a href="#__codelineno-0-913"> 913</a></span>
<span class="normal"><a href="#__codelineno-0-914"> 914</a></span>
<span class="normal"><a href="#__codelineno-0-915"> 915</a></span>
<span class="normal"><a href="#__codelineno-0-916"> 916</a></span>
<span class="normal"><a href="#__codelineno-0-917"> 917</a></span>
<span class="normal"><a href="#__codelineno-0-918"> 918</a></span>
<span class="normal"><a href="#__codelineno-0-919"> 919</a></span>
<span class="normal"><a href="#__codelineno-0-920"> 920</a></span>
<span class="normal"><a href="#__codelineno-0-921"> 921</a></span>
<span class="normal"><a href="#__codelineno-0-922"> 922</a></span>
<span class="normal"><a href="#__codelineno-0-923"> 923</a></span>
<span class="normal"><a href="#__codelineno-0-924"> 924</a></span>
<span class="normal"><a href="#__codelineno-0-925"> 925</a></span>
<span class="normal"><a href="#__codelineno-0-926"> 926</a></span>
<span class="normal"><a href="#__codelineno-0-927"> 927</a></span>
<span class="normal"><a href="#__codelineno-0-928"> 928</a></span>
<span class="normal"><a href="#__codelineno-0-929"> 929</a></span>
<span class="normal"><a href="#__codelineno-0-930"> 930</a></span>
<span class="normal"><a href="#__codelineno-0-931"> 931</a></span>
<span class="normal"><a href="#__codelineno-0-932"> 932</a></span>
<span class="normal"><a href="#__codelineno-0-933"> 933</a></span>
<span class="normal"><a href="#__codelineno-0-934"> 934</a></span>
<span class="normal"><a href="#__codelineno-0-935"> 935</a></span>
<span class="normal"><a href="#__codelineno-0-936"> 936</a></span>
<span class="normal"><a href="#__codelineno-0-937"> 937</a></span>
<span class="normal"><a href="#__codelineno-0-938"> 938</a></span>
<span class="normal"><a href="#__codelineno-0-939"> 939</a></span>
<span class="normal"><a href="#__codelineno-0-940"> 940</a></span>
<span class="normal"><a href="#__codelineno-0-941"> 941</a></span>
<span class="normal"><a href="#__codelineno-0-942"> 942</a></span>
<span class="normal"><a href="#__codelineno-0-943"> 943</a></span>
<span class="normal"><a href="#__codelineno-0-944"> 944</a></span>
<span class="normal"><a href="#__codelineno-0-945"> 945</a></span>
<span class="normal"><a href="#__codelineno-0-946"> 946</a></span>
<span class="normal"><a href="#__codelineno-0-947"> 947</a></span>
<span class="normal"><a href="#__codelineno-0-948"> 948</a></span>
<span class="normal"><a href="#__codelineno-0-949"> 949</a></span>
<span class="normal"><a href="#__codelineno-0-950"> 950</a></span>
<span class="normal"><a href="#__codelineno-0-951"> 951</a></span>
<span class="normal"><a href="#__codelineno-0-952"> 952</a></span>
<span class="normal"><a href="#__codelineno-0-953"> 953</a></span>
<span class="normal"><a href="#__codelineno-0-954"> 954</a></span>
<span class="normal"><a href="#__codelineno-0-955"> 955</a></span>
<span class="normal"><a href="#__codelineno-0-956"> 956</a></span>
<span class="normal"><a href="#__codelineno-0-957"> 957</a></span>
<span class="normal"><a href="#__codelineno-0-958"> 958</a></span>
<span class="normal"><a href="#__codelineno-0-959"> 959</a></span>
<span class="normal"><a href="#__codelineno-0-960"> 960</a></span>
<span class="normal"><a href="#__codelineno-0-961"> 961</a></span>
<span class="normal"><a href="#__codelineno-0-962"> 962</a></span>
<span class="normal"><a href="#__codelineno-0-963"> 963</a></span>
<span class="normal"><a href="#__codelineno-0-964"> 964</a></span>
<span class="normal"><a href="#__codelineno-0-965"> 965</a></span>
<span class="normal"><a href="#__codelineno-0-966"> 966</a></span>
<span class="normal"><a href="#__codelineno-0-967"> 967</a></span>
<span class="normal"><a href="#__codelineno-0-968"> 968</a></span>
<span class="normal"><a href="#__codelineno-0-969"> 969</a></span>
<span class="normal"><a href="#__codelineno-0-970"> 970</a></span>
<span class="normal"><a href="#__codelineno-0-971"> 971</a></span>
<span class="normal"><a href="#__codelineno-0-972"> 972</a></span>
<span class="normal"><a href="#__codelineno-0-973"> 973</a></span>
<span class="normal"><a href="#__codelineno-0-974"> 974</a></span>
<span class="normal"><a href="#__codelineno-0-975"> 975</a></span>
<span class="normal"><a href="#__codelineno-0-976"> 976</a></span>
<span class="normal"><a href="#__codelineno-0-977"> 977</a></span>
<span class="normal"><a href="#__codelineno-0-978"> 978</a></span>
<span class="normal"><a href="#__codelineno-0-979"> 979</a></span>
<span class="normal"><a href="#__codelineno-0-980"> 980</a></span>
<span class="normal"><a href="#__codelineno-0-981"> 981</a></span>
<span class="normal"><a href="#__codelineno-0-982"> 982</a></span>
<span class="normal"><a href="#__codelineno-0-983"> 983</a></span>
<span class="normal"><a href="#__codelineno-0-984"> 984</a></span>
<span class="normal"><a href="#__codelineno-0-985"> 985</a></span>
<span class="normal"><a href="#__codelineno-0-986"> 986</a></span>
<span class="normal"><a href="#__codelineno-0-987"> 987</a></span>
<span class="normal"><a href="#__codelineno-0-988"> 988</a></span>
<span class="normal"><a href="#__codelineno-0-989"> 989</a></span>
<span class="normal"><a href="#__codelineno-0-990"> 990</a></span>
<span class="normal"><a href="#__codelineno-0-991"> 991</a></span>
<span class="normal"><a href="#__codelineno-0-992"> 992</a></span>
<span class="normal"><a href="#__codelineno-0-993"> 993</a></span>
<span class="normal"><a href="#__codelineno-0-994"> 994</a></span>
<span class="normal"><a href="#__codelineno-0-995"> 995</a></span>
<span class="normal"><a href="#__codelineno-0-996"> 996</a></span>
<span class="normal"><a href="#__codelineno-0-997"> 997</a></span>
<span class="normal"><a href="#__codelineno-0-998"> 998</a></span>
<span class="normal"><a href="#__codelineno-0-999"> 999</a></span>
<span class="normal"><a href="#__codelineno-0-1000">1000</a></span>
<span class="normal"><a href="#__codelineno-0-1001">1001</a></span>
<span class="normal"><a href="#__codelineno-0-1002">1002</a></span>
<span class="normal"><a href="#__codelineno-0-1003">1003</a></span>
<span class="normal"><a href="#__codelineno-0-1004">1004</a></span>
<span class="normal"><a href="#__codelineno-0-1005">1005</a></span>
<span class="normal"><a href="#__codelineno-0-1006">1006</a></span>
<span class="normal"><a href="#__codelineno-0-1007">1007</a></span>
<span class="normal"><a href="#__codelineno-0-1008">1008</a></span>
<span class="normal"><a href="#__codelineno-0-1009">1009</a></span>
<span class="normal"><a href="#__codelineno-0-1010">1010</a></span>
<span class="normal"><a href="#__codelineno-0-1011">1011</a></span>
<span class="normal"><a href="#__codelineno-0-1012">1012</a></span>
<span class="normal"><a href="#__codelineno-0-1013">1013</a></span>
<span class="normal"><a href="#__codelineno-0-1014">1014</a></span>
<span class="normal"><a href="#__codelineno-0-1015">1015</a></span>
<span class="normal"><a href="#__codelineno-0-1016">1016</a></span>
<span class="normal"><a href="#__codelineno-0-1017">1017</a></span>
<span class="normal"><a href="#__codelineno-0-1018">1018</a></span>
<span class="normal"><a href="#__codelineno-0-1019">1019</a></span>
<span class="normal"><a href="#__codelineno-0-1020">1020</a></span>
<span class="normal"><a href="#__codelineno-0-1021">1021</a></span>
<span class="normal"><a href="#__codelineno-0-1022">1022</a></span>
<span class="normal"><a href="#__codelineno-0-1023">1023</a></span>
<span class="normal"><a href="#__codelineno-0-1024">1024</a></span>
<span class="normal"><a href="#__codelineno-0-1025">1025</a></span>
<span class="normal"><a href="#__codelineno-0-1026">1026</a></span>
<span class="normal"><a href="#__codelineno-0-1027">1027</a></span>
<span class="normal"><a href="#__codelineno-0-1028">1028</a></span>
<span class="normal"><a href="#__codelineno-0-1029">1029</a></span>
<span class="normal"><a href="#__codelineno-0-1030">1030</a></span>
<span class="normal"><a href="#__codelineno-0-1031">1031</a></span>
<span class="normal"><a href="#__codelineno-0-1032">1032</a></span>
<span class="normal"><a href="#__codelineno-0-1033">1033</a></span>
<span class="normal"><a href="#__codelineno-0-1034">1034</a></span>
<span class="normal"><a href="#__codelineno-0-1035">1035</a></span>
<span class="normal"><a href="#__codelineno-0-1036">1036</a></span>
<span class="normal"><a href="#__codelineno-0-1037">1037</a></span>
<span class="normal"><a href="#__codelineno-0-1038">1038</a></span>
<span class="normal"><a href="#__codelineno-0-1039">1039</a></span>
<span class="normal"><a href="#__codelineno-0-1040">1040</a></span>
<span class="normal"><a href="#__codelineno-0-1041">1041</a></span>
<span class="normal"><a href="#__codelineno-0-1042">1042</a></span>
<span class="normal"><a href="#__codelineno-0-1043">1043</a></span>
<span class="normal"><a href="#__codelineno-0-1044">1044</a></span>
<span class="normal"><a href="#__codelineno-0-1045">1045</a></span>
<span class="normal"><a href="#__codelineno-0-1046">1046</a></span>
<span class="normal"><a href="#__codelineno-0-1047">1047</a></span>
<span class="normal"><a href="#__codelineno-0-1048">1048</a></span>
<span class="normal"><a href="#__codelineno-0-1049">1049</a></span>
<span class="normal"><a href="#__codelineno-0-1050">1050</a></span>
<span class="normal"><a href="#__codelineno-0-1051">1051</a></span>
<span class="normal"><a href="#__codelineno-0-1052">1052</a></span>
<span class="normal"><a href="#__codelineno-0-1053">1053</a></span>
<span class="normal"><a href="#__codelineno-0-1054">1054</a></span>
<span class="normal"><a href="#__codelineno-0-1055">1055</a></span>
<span class="normal"><a href="#__codelineno-0-1056">1056</a></span>
<span class="normal"><a href="#__codelineno-0-1057">1057</a></span>
<span class="normal"><a href="#__codelineno-0-1058">1058</a></span>
<span class="normal"><a href="#__codelineno-0-1059">1059</a></span>
<span class="normal"><a href="#__codelineno-0-1060">1060</a></span>
<span class="normal"><a href="#__codelineno-0-1061">1061</a></span>
<span class="normal"><a href="#__codelineno-0-1062">1062</a></span>
<span class="normal"><a href="#__codelineno-0-1063">1063</a></span>
<span class="normal"><a href="#__codelineno-0-1064">1064</a></span>
<span class="normal"><a href="#__codelineno-0-1065">1065</a></span>
<span class="normal"><a href="#__codelineno-0-1066">1066</a></span>
<span class="normal"><a href="#__codelineno-0-1067">1067</a></span>
<span class="normal"><a href="#__codelineno-0-1068">1068</a></span>
<span class="normal"><a href="#__codelineno-0-1069">1069</a></span>
<span class="normal"><a href="#__codelineno-0-1070">1070</a></span>
<span class="normal"><a href="#__codelineno-0-1071">1071</a></span>
<span class="normal"><a href="#__codelineno-0-1072">1072</a></span>
<span class="normal"><a href="#__codelineno-0-1073">1073</a></span>
<span class="normal"><a href="#__codelineno-0-1074">1074</a></span>
<span class="normal"><a href="#__codelineno-0-1075">1075</a></span>
<span class="normal"><a href="#__codelineno-0-1076">1076</a></span>
<span class="normal"><a href="#__codelineno-0-1077">1077</a></span>
<span class="normal"><a href="#__codelineno-0-1078">1078</a></span>
<span class="normal"><a href="#__codelineno-0-1079">1079</a></span>
<span class="normal"><a href="#__codelineno-0-1080">1080</a></span>
<span class="normal"><a href="#__codelineno-0-1081">1081</a></span>
<span class="normal"><a href="#__codelineno-0-1082">1082</a></span>
<span class="normal"><a href="#__codelineno-0-1083">1083</a></span>
<span class="normal"><a href="#__codelineno-0-1084">1084</a></span>
<span class="normal"><a href="#__codelineno-0-1085">1085</a></span>
<span class="normal"><a href="#__codelineno-0-1086">1086</a></span>
<span class="normal"><a href="#__codelineno-0-1087">1087</a></span>
<span class="normal"><a href="#__codelineno-0-1088">1088</a></span>
<span class="normal"><a href="#__codelineno-0-1089">1089</a></span>
<span class="normal"><a href="#__codelineno-0-1090">1090</a></span>
<span class="normal"><a href="#__codelineno-0-1091">1091</a></span>
<span class="normal"><a href="#__codelineno-0-1092">1092</a></span>
<span class="normal"><a href="#__codelineno-0-1093">1093</a></span>
<span class="normal"><a href="#__codelineno-0-1094">1094</a></span>
<span class="normal"><a href="#__codelineno-0-1095">1095</a></span>
<span class="normal"><a href="#__codelineno-0-1096">1096</a></span>
<span class="normal"><a href="#__codelineno-0-1097">1097</a></span>
<span class="normal"><a href="#__codelineno-0-1098">1098</a></span>
<span class="normal"><a href="#__codelineno-0-1099">1099</a></span></pre></div></td><td class="code"><div><pre><span></span><code><span id="__span-0-24"><a id="__codelineno-0-24" name="__codelineno-0-24"></a><span class="k">class</span><span class="w"> </span><span class="nc">GWADataLoader</span><span class="p">(</span><span class="nb">object</span><span class="p">):</span>
</span><span id="__span-0-25"><a id="__codelineno-0-25" name="__codelineno-0-25"></a><span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
</span><span id="__span-0-26"><a id="__codelineno-0-26" name="__codelineno-0-26"></a><span class="sd">    A class to load and manage multiple data sources for genetic association studies.</span>
</span><span id="__span-0-27"><a id="__codelineno-0-27" name="__codelineno-0-27"></a><span class="sd">    This class is designed to handle genotype matrices, summary statistics, LD matrices,</span>
</span><span id="__span-0-28"><a id="__codelineno-0-28" name="__codelineno-0-28"></a><span class="sd">    and annotation matrices. It also provides functionalities to filter samples and/or SNPs,</span>
</span><span id="__span-0-29"><a id="__codelineno-0-29" name="__codelineno-0-29"></a><span class="sd">    harmonize data sources, and compute LD matrices. This is all done in order to facilitate</span>
</span><span id="__span-0-30"><a id="__codelineno-0-30" name="__codelineno-0-30"></a><span class="sd">    downstream statistical genetics analyses that require multiple data sources to be aligned</span>
</span><span id="__span-0-31"><a id="__codelineno-0-31" name="__codelineno-0-31"></a><span class="sd">    and harmonized. The use cases include:</span>
</span><span id="__span-0-32"><a id="__codelineno-0-32" name="__codelineno-0-32"></a>
</span><span id="__span-0-33"><a id="__codelineno-0-33" name="__codelineno-0-33"></a><span class="sd">    * Summary statistics-based PRS computation</span>
</span><span id="__span-0-34"><a id="__codelineno-0-34" name="__codelineno-0-34"></a><span class="sd">    * Summary statistics-based heritability estimation.</span>
</span><span id="__span-0-35"><a id="__codelineno-0-35" name="__codelineno-0-35"></a><span class="sd">    * Complex trait simulation.</span>
</span><span id="__span-0-36"><a id="__codelineno-0-36" name="__codelineno-0-36"></a><span class="sd">    * Performing Genome-wide association tests.</span>
</span><span id="__span-0-37"><a id="__codelineno-0-37" name="__codelineno-0-37"></a>
</span><span id="__span-0-38"><a id="__codelineno-0-38" name="__codelineno-0-38"></a><span class="sd">    :ivar genotype: A dictionary of `GenotypeMatrix` objects, where the key is the chromosome number.</span>
</span><span id="__span-0-39"><a id="__codelineno-0-39" name="__codelineno-0-39"></a><span class="sd">    :ivar sample_table: A `SampleTable` object containing the sample information.</span>
</span><span id="__span-0-40"><a id="__codelineno-0-40" name="__codelineno-0-40"></a><span class="sd">    :ivar phenotype_likelihood: The likelihood of the phenotype (e.g. `gaussian`, `binomial`).</span>
</span><span id="__span-0-41"><a id="__codelineno-0-41" name="__codelineno-0-41"></a><span class="sd">    :ivar ld: A dictionary of `LDMatrix` objects, where the key is the chromosome number.</span>
</span><span id="__span-0-42"><a id="__codelineno-0-42" name="__codelineno-0-42"></a><span class="sd">    :ivar sumstats_table: A dictionary of `SumstatsTable` objects, where the key is the chromosome number.</span>
</span><span id="__span-0-43"><a id="__codelineno-0-43" name="__codelineno-0-43"></a><span class="sd">    :ivar annotation: A dictionary of `AnnotationMatrix` objects, where the key is the chromosome number.</span>
</span><span id="__span-0-44"><a id="__codelineno-0-44" name="__codelineno-0-44"></a><span class="sd">    :ivar backend: The backend software used for the computation. Currently, supports `xarray` and `plink`.</span>
</span><span id="__span-0-45"><a id="__codelineno-0-45" name="__codelineno-0-45"></a><span class="sd">    :ivar temp_dir: The temporary directory where we store intermediate files (if necessary).</span>
</span><span id="__span-0-46"><a id="__codelineno-0-46" name="__codelineno-0-46"></a><span class="sd">    :ivar output_dir: The output directory where we store the results of the computation.</span>
</span><span id="__span-0-47"><a id="__codelineno-0-47" name="__codelineno-0-47"></a><span class="sd">    &quot;&quot;&quot;</span>
</span><span id="__span-0-48"><a id="__codelineno-0-48" name="__codelineno-0-48"></a>
</span><span id="__span-0-49"><a id="__codelineno-0-49" name="__codelineno-0-49"></a>    <span class="k">def</span><span class="w"> </span><span class="fm">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span>
</span><span id="__span-0-50"><a id="__codelineno-0-50" name="__codelineno-0-50"></a>                 <span class="n">bed_files</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
</span><span id="__span-0-51"><a id="__codelineno-0-51" name="__codelineno-0-51"></a>                 <span class="n">phenotype_file</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
</span><span id="__span-0-52"><a id="__codelineno-0-52" name="__codelineno-0-52"></a>                 <span class="n">covariates_file</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
</span><span id="__span-0-53"><a id="__codelineno-0-53" name="__codelineno-0-53"></a>                 <span class="n">keep_samples</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
</span><span id="__span-0-54"><a id="__codelineno-0-54" name="__codelineno-0-54"></a>                 <span class="n">keep_file</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
</span><span id="__span-0-55"><a id="__codelineno-0-55" name="__codelineno-0-55"></a>                 <span class="n">extract_snps</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
</span><span id="__span-0-56"><a id="__codelineno-0-56" name="__codelineno-0-56"></a>                 <span class="n">extract_file</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
</span><span id="__span-0-57"><a id="__codelineno-0-57" name="__codelineno-0-57"></a>                 <span class="n">min_maf</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
</span><span id="__span-0-58"><a id="__codelineno-0-58" name="__codelineno-0-58"></a>                 <span class="n">min_mac</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
</span><span id="__span-0-59"><a id="__codelineno-0-59" name="__codelineno-0-59"></a>                 <span class="n">drop_duplicated</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
</span><span id="__span-0-60"><a id="__codelineno-0-60" name="__codelineno-0-60"></a>                 <span class="n">phenotype_likelihood</span><span class="o">=</span><span class="s1">&#39;gaussian&#39;</span><span class="p">,</span>
</span><span id="__span-0-61"><a id="__codelineno-0-61" name="__codelineno-0-61"></a>                 <span class="n">sumstats_files</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
</span><span id="__span-0-62"><a id="__codelineno-0-62" name="__codelineno-0-62"></a>                 <span class="n">sumstats_format</span><span class="o">=</span><span class="s1">&#39;magenpy&#39;</span><span class="p">,</span>
</span><span id="__span-0-63"><a id="__codelineno-0-63" name="__codelineno-0-63"></a>                 <span class="n">ld_store_files</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
</span><span id="__span-0-64"><a id="__codelineno-0-64" name="__codelineno-0-64"></a>                 <span class="n">annotation_files</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
</span><span id="__span-0-65"><a id="__codelineno-0-65" name="__codelineno-0-65"></a>                 <span class="n">annotation_format</span><span class="o">=</span><span class="s1">&#39;magenpy&#39;</span><span class="p">,</span>
</span><span id="__span-0-66"><a id="__codelineno-0-66" name="__codelineno-0-66"></a>                 <span class="n">backend</span><span class="o">=</span><span class="s1">&#39;xarray&#39;</span><span class="p">,</span>
</span><span id="__span-0-67"><a id="__codelineno-0-67" name="__codelineno-0-67"></a>                 <span class="n">temp_dir</span><span class="o">=</span><span class="s1">&#39;temp&#39;</span><span class="p">,</span>
</span><span id="__span-0-68"><a id="__codelineno-0-68" name="__codelineno-0-68"></a>                 <span class="n">output_dir</span><span class="o">=</span><span class="s1">&#39;output&#39;</span><span class="p">,</span>
</span><span id="__span-0-69"><a id="__codelineno-0-69" name="__codelineno-0-69"></a>                 <span class="n">threads</span><span class="o">=</span><span class="mi">1</span><span class="p">):</span>
</span><span id="__span-0-70"><a id="__codelineno-0-70" name="__codelineno-0-70"></a><span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
</span><span id="__span-0-71"><a id="__codelineno-0-71" name="__codelineno-0-71"></a><span class="sd">        Initialize the `GWADataLoader` object with the data sources required for</span>
</span><span id="__span-0-72"><a id="__codelineno-0-72" name="__codelineno-0-72"></a><span class="sd">        downstream statistical genetics analyses.</span>
</span><span id="__span-0-73"><a id="__codelineno-0-73" name="__codelineno-0-73"></a>
</span><span id="__span-0-74"><a id="__codelineno-0-74" name="__codelineno-0-74"></a><span class="sd">        :param bed_files: The path to the BED file(s). You may use a wildcard here to read files for multiple</span>
</span><span id="__span-0-75"><a id="__codelineno-0-75" name="__codelineno-0-75"></a><span class="sd">        chromosomes.</span>
</span><span id="__span-0-76"><a id="__codelineno-0-76" name="__codelineno-0-76"></a><span class="sd">        :param phenotype_file: The path to the phenotype file.</span>
</span><span id="__span-0-77"><a id="__codelineno-0-77" name="__codelineno-0-77"></a><span class="sd">        (Default: tab-separated file with `FID IID phenotype` columns).</span>
</span><span id="__span-0-78"><a id="__codelineno-0-78" name="__codelineno-0-78"></a><span class="sd">        :param covariates_file: The path to the covariates file.</span>
</span><span id="__span-0-79"><a id="__codelineno-0-79" name="__codelineno-0-79"></a><span class="sd">        (Default: tab-separated file starting with the `FID IID ...` columns and followed by the covariate columns).</span>
</span><span id="__span-0-80"><a id="__codelineno-0-80" name="__codelineno-0-80"></a><span class="sd">        :param keep_samples: A vector or list of sample IDs to keep when filtering the genotype matrix.</span>
</span><span id="__span-0-81"><a id="__codelineno-0-81" name="__codelineno-0-81"></a><span class="sd">        :param keep_file: A path to a plink-style keep file to select a subset of individuals.</span>
</span><span id="__span-0-82"><a id="__codelineno-0-82" name="__codelineno-0-82"></a><span class="sd">        :param extract_snps: A vector or list of SNP IDs to keep when filtering the genotype matrix.</span>
</span><span id="__span-0-83"><a id="__codelineno-0-83" name="__codelineno-0-83"></a><span class="sd">        :param extract_file: A path to a plink-style extract file to select a subset of SNPs.</span>
</span><span id="__span-0-84"><a id="__codelineno-0-84" name="__codelineno-0-84"></a><span class="sd">        :param min_maf: The minimum minor allele frequency cutoff.</span>
</span><span id="__span-0-85"><a id="__codelineno-0-85" name="__codelineno-0-85"></a><span class="sd">        :param min_mac: The minimum minor allele count cutoff.</span>
</span><span id="__span-0-86"><a id="__codelineno-0-86" name="__codelineno-0-86"></a><span class="sd">        :param drop_duplicated: If True, drop SNPs with duplicated rsID.</span>
</span><span id="__span-0-87"><a id="__codelineno-0-87" name="__codelineno-0-87"></a><span class="sd">        :param phenotype_likelihood: The likelihood of the phenotype (e.g. `gaussian`, `binomial`).</span>
</span><span id="__span-0-88"><a id="__codelineno-0-88" name="__codelineno-0-88"></a><span class="sd">        :param sumstats_files: The path to the summary statistics file(s). The path may be a wildcard.</span>
</span><span id="__span-0-89"><a id="__codelineno-0-89" name="__codelineno-0-89"></a><span class="sd">        :param sumstats_format: The format for the summary statistics. Currently, supports the following</span>
</span><span id="__span-0-90"><a id="__codelineno-0-90" name="__codelineno-0-90"></a><span class="sd">        formats: `plink1.9`, `plink2`, `magenpy`, `fastGWA`, `COJO`, `SAIGE`, or `GWASCatalog` for the standard</span>
</span><span id="__span-0-91"><a id="__codelineno-0-91" name="__codelineno-0-91"></a><span class="sd">        summary statistics format (also known as `ssf` or `gwas-ssf`).</span>
</span><span id="__span-0-92"><a id="__codelineno-0-92" name="__codelineno-0-92"></a><span class="sd">        :param ld_store_files: The path to the LD matrices. This may be a wildcard to accommodate reading data</span>
</span><span id="__span-0-93"><a id="__codelineno-0-93" name="__codelineno-0-93"></a><span class="sd">        for multiple chromosomes.</span>
</span><span id="__span-0-94"><a id="__codelineno-0-94" name="__codelineno-0-94"></a><span class="sd">        :param annotation_files: The path to the annotation file(s). The path may contain a wildcard.</span>
</span><span id="__span-0-95"><a id="__codelineno-0-95" name="__codelineno-0-95"></a><span class="sd">        :param annotation_format: The format for the summary statistics. Currently, supports the following</span>
</span><span id="__span-0-96"><a id="__codelineno-0-96" name="__codelineno-0-96"></a><span class="sd">        formats: `magenpy`, `ldsc`.</span>
</span><span id="__span-0-97"><a id="__codelineno-0-97" name="__codelineno-0-97"></a><span class="sd">        :param backend: The backend software used for computations with the genotype matrix. Currently, supports</span>
</span><span id="__span-0-98"><a id="__codelineno-0-98" name="__codelineno-0-98"></a><span class="sd">        `xarray` and `plink`.</span>
</span><span id="__span-0-99"><a id="__codelineno-0-99" name="__codelineno-0-99"></a><span class="sd">        :param temp_dir: The temporary directory where to store intermediate files.</span>
</span><span id="__span-0-100"><a id="__codelineno-0-100" name="__codelineno-0-100"></a><span class="sd">        :param output_dir: The output directory where to store the results of the computation.</span>
</span><span id="__span-0-101"><a id="__codelineno-0-101" name="__codelineno-0-101"></a><span class="sd">        :param threads: The number of threads to use for computations.</span>
</span><span id="__span-0-102"><a id="__codelineno-0-102" name="__codelineno-0-102"></a><span class="sd">        &quot;&quot;&quot;</span>
</span><span id="__span-0-103"><a id="__codelineno-0-103" name="__codelineno-0-103"></a>
</span><span id="__span-0-104"><a id="__codelineno-0-104" name="__codelineno-0-104"></a>        <span class="c1"># ------- Sanity checks -------</span>
</span><span id="__span-0-105"><a id="__codelineno-0-105" name="__codelineno-0-105"></a>
</span><span id="__span-0-106"><a id="__codelineno-0-106" name="__codelineno-0-106"></a>        <span class="k">assert</span> <span class="n">backend</span> <span class="ow">in</span> <span class="p">(</span><span class="s1">&#39;xarray&#39;</span><span class="p">,</span> <span class="s1">&#39;plink&#39;</span><span class="p">,</span> <span class="s1">&#39;bed-reader&#39;</span><span class="p">)</span>
</span><span id="__span-0-107"><a id="__codelineno-0-107" name="__codelineno-0-107"></a>        <span class="k">assert</span> <span class="n">phenotype_likelihood</span> <span class="ow">in</span> <span class="p">(</span><span class="s1">&#39;gaussian&#39;</span><span class="p">,</span> <span class="s1">&#39;binomial&#39;</span><span class="p">)</span>
</span><span id="__span-0-108"><a id="__codelineno-0-108" name="__codelineno-0-108"></a>
</span><span id="__span-0-109"><a id="__codelineno-0-109" name="__codelineno-0-109"></a>        <span class="c1"># ------- General options -------</span>
</span><span id="__span-0-110"><a id="__codelineno-0-110" name="__codelineno-0-110"></a>
</span><span id="__span-0-111"><a id="__codelineno-0-111" name="__codelineno-0-111"></a>        <span class="bp">self</span><span class="o">.</span><span class="n">backend</span> <span class="o">=</span> <span class="n">backend</span>
</span><span id="__span-0-112"><a id="__codelineno-0-112" name="__codelineno-0-112"></a>
</span><span id="__span-0-113"><a id="__codelineno-0-113" name="__codelineno-0-113"></a>        <span class="bp">self</span><span class="o">.</span><span class="n">temp_dir</span> <span class="o">=</span> <span class="n">temp_dir</span>
</span><span id="__span-0-114"><a id="__codelineno-0-114" name="__codelineno-0-114"></a>        <span class="bp">self</span><span class="o">.</span><span class="n">output_dir</span> <span class="o">=</span> <span class="n">output_dir</span>
</span><span id="__span-0-115"><a id="__codelineno-0-115" name="__codelineno-0-115"></a>        <span class="bp">self</span><span class="o">.</span><span class="n">cleanup_dir_list</span> <span class="o">=</span> <span class="p">[]</span>  <span class="c1"># Directories to clean up after execution.</span>
</span><span id="__span-0-116"><a id="__codelineno-0-116" name="__codelineno-0-116"></a>
</span><span id="__span-0-117"><a id="__codelineno-0-117" name="__codelineno-0-117"></a>        <span class="n">makedir</span><span class="p">([</span><span class="n">temp_dir</span><span class="p">,</span> <span class="n">output_dir</span><span class="p">])</span>
</span><span id="__span-0-118"><a id="__codelineno-0-118" name="__codelineno-0-118"></a>
</span><span id="__span-0-119"><a id="__codelineno-0-119" name="__codelineno-0-119"></a>        <span class="bp">self</span><span class="o">.</span><span class="n">threads</span> <span class="o">=</span> <span class="n">threads</span>
</span><span id="__span-0-120"><a id="__codelineno-0-120" name="__codelineno-0-120"></a>
</span><span id="__span-0-121"><a id="__codelineno-0-121" name="__codelineno-0-121"></a>        <span class="c1"># ------- General parameters -------</span>
</span><span id="__span-0-122"><a id="__codelineno-0-122" name="__codelineno-0-122"></a>
</span><span id="__span-0-123"><a id="__codelineno-0-123" name="__codelineno-0-123"></a>        <span class="bp">self</span><span class="o">.</span><span class="n">phenotype_likelihood</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="n">phenotype_likelihood</span>
</span><span id="__span-0-124"><a id="__codelineno-0-124" name="__codelineno-0-124"></a>
</span><span id="__span-0-125"><a id="__codelineno-0-125" name="__codelineno-0-125"></a>        <span class="bp">self</span><span class="o">.</span><span class="n">genotype</span><span class="p">:</span> <span class="n">Union</span><span class="p">[</span><span class="n">Dict</span><span class="p">[</span><span class="nb">int</span><span class="p">,</span> <span class="n">GenotypeMatrix</span><span class="p">],</span> <span class="kc">None</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span>
</span><span id="__span-0-126"><a id="__codelineno-0-126" name="__codelineno-0-126"></a>        <span class="bp">self</span><span class="o">.</span><span class="n">sample_table</span><span class="p">:</span> <span class="n">Union</span><span class="p">[</span><span class="n">SampleTable</span><span class="p">,</span> <span class="kc">None</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span>
</span><span id="__span-0-127"><a id="__codelineno-0-127" name="__codelineno-0-127"></a>        <span class="bp">self</span><span class="o">.</span><span class="n">ld</span><span class="p">:</span> <span class="n">Union</span><span class="p">[</span><span class="n">Dict</span><span class="p">[</span><span class="nb">int</span><span class="p">,</span> <span class="n">LDMatrix</span><span class="p">],</span> <span class="kc">None</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span>
</span><span id="__span-0-128"><a id="__codelineno-0-128" name="__codelineno-0-128"></a>        <span class="bp">self</span><span class="o">.</span><span class="n">sumstats_table</span><span class="p">:</span> <span class="n">Union</span><span class="p">[</span><span class="n">Dict</span><span class="p">[</span><span class="nb">int</span><span class="p">,</span> <span class="n">SumstatsTable</span><span class="p">],</span> <span class="kc">None</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span>
</span><span id="__span-0-129"><a id="__codelineno-0-129" name="__codelineno-0-129"></a>        <span class="bp">self</span><span class="o">.</span><span class="n">annotation</span><span class="p">:</span> <span class="n">Union</span><span class="p">[</span><span class="n">Dict</span><span class="p">[</span><span class="nb">int</span><span class="p">,</span> <span class="n">AnnotationMatrix</span><span class="p">],</span> <span class="kc">None</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span>
</span><span id="__span-0-130"><a id="__codelineno-0-130" name="__codelineno-0-130"></a>
</span><span id="__span-0-131"><a id="__codelineno-0-131" name="__codelineno-0-131"></a>        <span class="c1"># ------- Read data files -------</span>
</span><span id="__span-0-132"><a id="__codelineno-0-132" name="__codelineno-0-132"></a>
</span><span id="__span-0-133"><a id="__codelineno-0-133" name="__codelineno-0-133"></a>        <span class="bp">self</span><span class="o">.</span><span class="n">read_genotypes</span><span class="p">(</span><span class="n">bed_files</span><span class="p">,</span>
</span><span id="__span-0-134"><a id="__codelineno-0-134" name="__codelineno-0-134"></a>                            <span class="n">min_maf</span><span class="o">=</span><span class="n">min_maf</span><span class="p">,</span>
</span><span id="__span-0-135"><a id="__codelineno-0-135" name="__codelineno-0-135"></a>                            <span class="n">min_mac</span><span class="o">=</span><span class="n">min_mac</span><span class="p">,</span>
</span><span id="__span-0-136"><a id="__codelineno-0-136" name="__codelineno-0-136"></a>                            <span class="n">drop_duplicated</span><span class="o">=</span><span class="n">drop_duplicated</span><span class="p">)</span>
</span><span id="__span-0-137"><a id="__codelineno-0-137" name="__codelineno-0-137"></a>        <span class="bp">self</span><span class="o">.</span><span class="n">read_phenotype</span><span class="p">(</span><span class="n">phenotype_file</span><span class="p">)</span>
</span><span id="__span-0-138"><a id="__codelineno-0-138" name="__codelineno-0-138"></a>        <span class="bp">self</span><span class="o">.</span><span class="n">read_covariates</span><span class="p">(</span><span class="n">covariates_file</span><span class="p">)</span>
</span><span id="__span-0-139"><a id="__codelineno-0-139" name="__codelineno-0-139"></a>        <span class="bp">self</span><span class="o">.</span><span class="n">read_ld</span><span class="p">(</span><span class="n">ld_store_files</span><span class="p">)</span>
</span><span id="__span-0-140"><a id="__codelineno-0-140" name="__codelineno-0-140"></a>        <span class="bp">self</span><span class="o">.</span><span class="n">read_annotations</span><span class="p">(</span><span class="n">annotation_files</span><span class="p">,</span>
</span><span id="__span-0-141"><a id="__codelineno-0-141" name="__codelineno-0-141"></a>                              <span class="n">annot_format</span><span class="o">=</span><span class="n">annotation_format</span><span class="p">)</span>
</span><span id="__span-0-142"><a id="__codelineno-0-142" name="__codelineno-0-142"></a>        <span class="bp">self</span><span class="o">.</span><span class="n">read_summary_statistics</span><span class="p">(</span><span class="n">sumstats_files</span><span class="p">,</span>
</span><span id="__span-0-143"><a id="__codelineno-0-143" name="__codelineno-0-143"></a>                                     <span class="n">sumstats_format</span><span class="p">,</span>
</span><span id="__span-0-144"><a id="__codelineno-0-144" name="__codelineno-0-144"></a>                                     <span class="n">drop_duplicated</span><span class="o">=</span><span class="n">drop_duplicated</span><span class="p">)</span>
</span><span id="__span-0-145"><a id="__codelineno-0-145" name="__codelineno-0-145"></a>
</span><span id="__span-0-146"><a id="__codelineno-0-146" name="__codelineno-0-146"></a>        <span class="c1"># ------- Filter samples or SNPs -------</span>
</span><span id="__span-0-147"><a id="__codelineno-0-147" name="__codelineno-0-147"></a>
</span><span id="__span-0-148"><a id="__codelineno-0-148" name="__codelineno-0-148"></a>        <span class="k">if</span> <span class="n">extract_snps</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span> <span class="ow">or</span> <span class="n">extract_file</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
</span><span id="__span-0-149"><a id="__codelineno-0-149" name="__codelineno-0-149"></a>            <span class="bp">self</span><span class="o">.</span><span class="n">filter_snps</span><span class="p">(</span><span class="n">extract_snps</span><span class="o">=</span><span class="n">extract_snps</span><span class="p">,</span> <span class="n">extract_file</span><span class="o">=</span><span class="n">extract_file</span><span class="p">)</span>
</span><span id="__span-0-150"><a id="__codelineno-0-150" name="__codelineno-0-150"></a>
</span><span id="__span-0-151"><a id="__codelineno-0-151" name="__codelineno-0-151"></a>        <span class="k">if</span> <span class="n">keep_samples</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span> <span class="ow">or</span> <span class="n">keep_file</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
</span><span id="__span-0-152"><a id="__codelineno-0-152" name="__codelineno-0-152"></a>            <span class="bp">self</span><span class="o">.</span><span class="n">filter_samples</span><span class="p">(</span><span class="n">keep_samples</span><span class="o">=</span><span class="n">keep_samples</span><span class="p">,</span> <span class="n">keep_file</span><span class="o">=</span><span class="n">keep_file</span><span class="p">)</span>
</span><span id="__span-0-153"><a id="__codelineno-0-153" name="__codelineno-0-153"></a>
</span><span id="__span-0-154"><a id="__codelineno-0-154" name="__codelineno-0-154"></a>        <span class="c1"># ------- Harmonize data sources -------</span>
</span><span id="__span-0-155"><a id="__codelineno-0-155" name="__codelineno-0-155"></a>
</span><span id="__span-0-156"><a id="__codelineno-0-156" name="__codelineno-0-156"></a>        <span class="bp">self</span><span class="o">.</span><span class="n">harmonize_data</span><span class="p">()</span>
</span><span id="__span-0-157"><a id="__codelineno-0-157" name="__codelineno-0-157"></a>
</span><span id="__span-0-158"><a id="__codelineno-0-158" name="__codelineno-0-158"></a>    <span class="nd">@property</span>
</span><span id="__span-0-159"><a id="__codelineno-0-159" name="__codelineno-0-159"></a>    <span class="k">def</span><span class="w"> </span><span class="nf">samples</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
</span><span id="__span-0-160"><a id="__codelineno-0-160" name="__codelineno-0-160"></a><span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
</span><span id="__span-0-161"><a id="__codelineno-0-161" name="__codelineno-0-161"></a><span class="sd">        :return: The list of samples retained in the sample table.</span>
</span><span id="__span-0-162"><a id="__codelineno-0-162" name="__codelineno-0-162"></a><span class="sd">        &quot;&quot;&quot;</span>
</span><span id="__span-0-163"><a id="__codelineno-0-163" name="__codelineno-0-163"></a>        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">sample_table</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
</span><span id="__span-0-164"><a id="__codelineno-0-164" name="__codelineno-0-164"></a>            <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">sample_table</span><span class="o">.</span><span class="n">iid</span>
</span><span id="__span-0-165"><a id="__codelineno-0-165" name="__codelineno-0-165"></a>
</span><span id="__span-0-166"><a id="__codelineno-0-166" name="__codelineno-0-166"></a>    <span class="nd">@property</span>
</span><span id="__span-0-167"><a id="__codelineno-0-167" name="__codelineno-0-167"></a>    <span class="k">def</span><span class="w"> </span><span class="nf">sample_size</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
</span><span id="__span-0-168"><a id="__codelineno-0-168" name="__codelineno-0-168"></a><span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
</span><span id="__span-0-169"><a id="__codelineno-0-169" name="__codelineno-0-169"></a>
</span><span id="__span-0-170"><a id="__codelineno-0-170" name="__codelineno-0-170"></a><span class="sd">        !!! seealso &quot;See Also&quot;</span>
</span><span id="__span-0-171"><a id="__codelineno-0-171" name="__codelineno-0-171"></a><span class="sd">            * [n][magenpy.GWADataLoader.GWADataLoader.n]</span>
</span><span id="__span-0-172"><a id="__codelineno-0-172" name="__codelineno-0-172"></a>
</span><span id="__span-0-173"><a id="__codelineno-0-173" name="__codelineno-0-173"></a><span class="sd">        :return: The number of samples in the genotype matrix.</span>
</span><span id="__span-0-174"><a id="__codelineno-0-174" name="__codelineno-0-174"></a>
</span><span id="__span-0-175"><a id="__codelineno-0-175" name="__codelineno-0-175"></a><span class="sd">        &quot;&quot;&quot;</span>
</span><span id="__span-0-176"><a id="__codelineno-0-176" name="__codelineno-0-176"></a>        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">sample_table</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
</span><span id="__span-0-177"><a id="__codelineno-0-177" name="__codelineno-0-177"></a>            <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">sample_table</span><span class="o">.</span><span class="n">n</span>
</span><span id="__span-0-178"><a id="__codelineno-0-178" name="__codelineno-0-178"></a>        <span class="k">elif</span> <span class="bp">self</span><span class="o">.</span><span class="n">sumstats_table</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
</span><span id="__span-0-179"><a id="__codelineno-0-179" name="__codelineno-0-179"></a>            <span class="k">return</span> <span class="n">np</span><span class="o">.</span><span class="n">max</span><span class="p">([</span><span class="n">np</span><span class="o">.</span><span class="n">max</span><span class="p">(</span><span class="n">ss</span><span class="o">.</span><span class="n">n_per_snp</span><span class="p">)</span> <span class="k">for</span> <span class="n">ss</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">sumstats_table</span><span class="o">.</span><span class="n">values</span><span class="p">()])</span>
</span><span id="__span-0-180"><a id="__codelineno-0-180" name="__codelineno-0-180"></a>        <span class="k">else</span><span class="p">:</span>
</span><span id="__span-0-181"><a id="__codelineno-0-181" name="__codelineno-0-181"></a>            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;Information about the sample size is not available!&quot;</span><span class="p">)</span>
</span><span id="__span-0-182"><a id="__codelineno-0-182" name="__codelineno-0-182"></a>
</span><span id="__span-0-183"><a id="__codelineno-0-183" name="__codelineno-0-183"></a>    <span class="nd">@property</span>
</span><span id="__span-0-184"><a id="__codelineno-0-184" name="__codelineno-0-184"></a>    <span class="k">def</span><span class="w"> </span><span class="nf">n</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
</span><span id="__span-0-185"><a id="__codelineno-0-185" name="__codelineno-0-185"></a><span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
</span><span id="__span-0-186"><a id="__codelineno-0-186" name="__codelineno-0-186"></a><span class="sd">        !!! seealso &quot;See Also&quot;</span>
</span><span id="__span-0-187"><a id="__codelineno-0-187" name="__codelineno-0-187"></a><span class="sd">            * [sample_size][magenpy.GWADataLoader.GWADataLoader.sample_size]</span>
</span><span id="__span-0-188"><a id="__codelineno-0-188" name="__codelineno-0-188"></a>
</span><span id="__span-0-189"><a id="__codelineno-0-189" name="__codelineno-0-189"></a><span class="sd">        :return: The number of samples in the genotype matrix.</span>
</span><span id="__span-0-190"><a id="__codelineno-0-190" name="__codelineno-0-190"></a><span class="sd">        &quot;&quot;&quot;</span>
</span><span id="__span-0-191"><a id="__codelineno-0-191" name="__codelineno-0-191"></a>
</span><span id="__span-0-192"><a id="__codelineno-0-192" name="__codelineno-0-192"></a>        <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">sample_size</span>
</span><span id="__span-0-193"><a id="__codelineno-0-193" name="__codelineno-0-193"></a>
</span><span id="__span-0-194"><a id="__codelineno-0-194" name="__codelineno-0-194"></a>    <span class="nd">@property</span>
</span><span id="__span-0-195"><a id="__codelineno-0-195" name="__codelineno-0-195"></a>    <span class="k">def</span><span class="w"> </span><span class="nf">snps</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
</span><span id="__span-0-196"><a id="__codelineno-0-196" name="__codelineno-0-196"></a><span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
</span><span id="__span-0-197"><a id="__codelineno-0-197" name="__codelineno-0-197"></a><span class="sd">        :return: The list of SNP rsIDs retained in each chromosome.</span>
</span><span id="__span-0-198"><a id="__codelineno-0-198" name="__codelineno-0-198"></a><span class="sd">        :rtype: dict</span>
</span><span id="__span-0-199"><a id="__codelineno-0-199" name="__codelineno-0-199"></a><span class="sd">        &quot;&quot;&quot;</span>
</span><span id="__span-0-200"><a id="__codelineno-0-200" name="__codelineno-0-200"></a>        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">genotype</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
</span><span id="__span-0-201"><a id="__codelineno-0-201" name="__codelineno-0-201"></a>            <span class="k">return</span> <span class="p">{</span><span class="n">c</span><span class="p">:</span> <span class="n">g</span><span class="o">.</span><span class="n">snps</span> <span class="k">for</span> <span class="n">c</span><span class="p">,</span> <span class="n">g</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">genotype</span><span class="o">.</span><span class="n">items</span><span class="p">()}</span>
</span><span id="__span-0-202"><a id="__codelineno-0-202" name="__codelineno-0-202"></a>        <span class="k">elif</span> <span class="bp">self</span><span class="o">.</span><span class="n">sumstats_table</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
</span><span id="__span-0-203"><a id="__codelineno-0-203" name="__codelineno-0-203"></a>            <span class="k">return</span> <span class="p">{</span><span class="n">c</span><span class="p">:</span> <span class="n">s</span><span class="o">.</span><span class="n">snps</span> <span class="k">for</span> <span class="n">c</span><span class="p">,</span> <span class="n">s</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">sumstats_table</span><span class="o">.</span><span class="n">items</span><span class="p">()}</span>
</span><span id="__span-0-204"><a id="__codelineno-0-204" name="__codelineno-0-204"></a>        <span class="k">elif</span> <span class="bp">self</span><span class="o">.</span><span class="n">ld</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
</span><span id="__span-0-205"><a id="__codelineno-0-205" name="__codelineno-0-205"></a>            <span class="k">return</span> <span class="p">{</span><span class="n">c</span><span class="p">:</span> <span class="n">l</span><span class="o">.</span><span class="n">snps</span> <span class="k">for</span> <span class="n">c</span><span class="p">,</span> <span class="n">l</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">ld</span><span class="o">.</span><span class="n">items</span><span class="p">()}</span>
</span><span id="__span-0-206"><a id="__codelineno-0-206" name="__codelineno-0-206"></a>        <span class="k">elif</span> <span class="bp">self</span><span class="o">.</span><span class="n">annotation</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
</span><span id="__span-0-207"><a id="__codelineno-0-207" name="__codelineno-0-207"></a>            <span class="k">return</span> <span class="p">{</span><span class="n">c</span><span class="p">:</span> <span class="n">a</span><span class="o">.</span><span class="n">snps</span> <span class="k">for</span> <span class="n">c</span><span class="p">,</span> <span class="n">a</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">annotation</span><span class="o">.</span><span class="n">items</span><span class="p">()}</span>
</span><span id="__span-0-208"><a id="__codelineno-0-208" name="__codelineno-0-208"></a>        <span class="k">else</span><span class="p">:</span>
</span><span id="__span-0-209"><a id="__codelineno-0-209" name="__codelineno-0-209"></a>            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;GWADataLoader instance is not properly initialized!&quot;</span><span class="p">)</span>
</span><span id="__span-0-210"><a id="__codelineno-0-210" name="__codelineno-0-210"></a>
</span><span id="__span-0-211"><a id="__codelineno-0-211" name="__codelineno-0-211"></a>    <span class="nd">@property</span>
</span><span id="__span-0-212"><a id="__codelineno-0-212" name="__codelineno-0-212"></a>    <span class="k">def</span><span class="w"> </span><span class="nf">m</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
</span><span id="__span-0-213"><a id="__codelineno-0-213" name="__codelineno-0-213"></a><span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
</span><span id="__span-0-214"><a id="__codelineno-0-214" name="__codelineno-0-214"></a><span class="sd">        !!! seealso &quot;See Also&quot;</span>
</span><span id="__span-0-215"><a id="__codelineno-0-215" name="__codelineno-0-215"></a><span class="sd">            * [n_snps][magenpy.GWADataLoader.GWADataLoader.n_snps]</span>
</span><span id="__span-0-216"><a id="__codelineno-0-216" name="__codelineno-0-216"></a>
</span><span id="__span-0-217"><a id="__codelineno-0-217" name="__codelineno-0-217"></a><span class="sd">        :return: The number of variants in the harmonized data sources.</span>
</span><span id="__span-0-218"><a id="__codelineno-0-218" name="__codelineno-0-218"></a><span class="sd">        &quot;&quot;&quot;</span>
</span><span id="__span-0-219"><a id="__codelineno-0-219" name="__codelineno-0-219"></a>        <span class="k">return</span> <span class="nb">sum</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">shapes</span><span class="o">.</span><span class="n">values</span><span class="p">())</span>
</span><span id="__span-0-220"><a id="__codelineno-0-220" name="__codelineno-0-220"></a>
</span><span id="__span-0-221"><a id="__codelineno-0-221" name="__codelineno-0-221"></a>    <span class="nd">@property</span>
</span><span id="__span-0-222"><a id="__codelineno-0-222" name="__codelineno-0-222"></a>    <span class="k">def</span><span class="w"> </span><span class="nf">n_snps</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
</span><span id="__span-0-223"><a id="__codelineno-0-223" name="__codelineno-0-223"></a><span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
</span><span id="__span-0-224"><a id="__codelineno-0-224" name="__codelineno-0-224"></a><span class="sd">        !!! seealso &quot;See Also&quot;</span>
</span><span id="__span-0-225"><a id="__codelineno-0-225" name="__codelineno-0-225"></a><span class="sd">            * [m][magenpy.GWADataLoader.GWADataLoader.m]</span>
</span><span id="__span-0-226"><a id="__codelineno-0-226" name="__codelineno-0-226"></a>
</span><span id="__span-0-227"><a id="__codelineno-0-227" name="__codelineno-0-227"></a><span class="sd">        :return: The number of variants in the harmonized data sources.</span>
</span><span id="__span-0-228"><a id="__codelineno-0-228" name="__codelineno-0-228"></a><span class="sd">        &quot;&quot;&quot;</span>
</span><span id="__span-0-229"><a id="__codelineno-0-229" name="__codelineno-0-229"></a>        <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">m</span>
</span><span id="__span-0-230"><a id="__codelineno-0-230" name="__codelineno-0-230"></a>
</span><span id="__span-0-231"><a id="__codelineno-0-231" name="__codelineno-0-231"></a>    <span class="nd">@property</span>
</span><span id="__span-0-232"><a id="__codelineno-0-232" name="__codelineno-0-232"></a>    <span class="k">def</span><span class="w"> </span><span class="nf">shapes</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
</span><span id="__span-0-233"><a id="__codelineno-0-233" name="__codelineno-0-233"></a><span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
</span><span id="__span-0-234"><a id="__codelineno-0-234" name="__codelineno-0-234"></a><span class="sd">        :return: A dictionary where the key is the chromosome number and the value is</span>
</span><span id="__span-0-235"><a id="__codelineno-0-235" name="__codelineno-0-235"></a><span class="sd">        the number of variants on that chromosome.</span>
</span><span id="__span-0-236"><a id="__codelineno-0-236" name="__codelineno-0-236"></a><span class="sd">        &quot;&quot;&quot;</span>
</span><span id="__span-0-237"><a id="__codelineno-0-237" name="__codelineno-0-237"></a>        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">genotype</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
</span><span id="__span-0-238"><a id="__codelineno-0-238" name="__codelineno-0-238"></a>            <span class="k">return</span> <span class="p">{</span><span class="n">c</span><span class="p">:</span> <span class="n">g</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span> <span class="k">for</span> <span class="n">c</span><span class="p">,</span> <span class="n">g</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">genotype</span><span class="o">.</span><span class="n">items</span><span class="p">()}</span>
</span><span id="__span-0-239"><a id="__codelineno-0-239" name="__codelineno-0-239"></a>        <span class="k">elif</span> <span class="bp">self</span><span class="o">.</span><span class="n">sumstats_table</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
</span><span id="__span-0-240"><a id="__codelineno-0-240" name="__codelineno-0-240"></a>            <span class="k">return</span> <span class="p">{</span><span class="n">c</span><span class="p">:</span> <span class="n">s</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="k">for</span> <span class="n">c</span><span class="p">,</span> <span class="n">s</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">sumstats_table</span><span class="o">.</span><span class="n">items</span><span class="p">()}</span>
</span><span id="__span-0-241"><a id="__codelineno-0-241" name="__codelineno-0-241"></a>        <span class="k">elif</span> <span class="bp">self</span><span class="o">.</span><span class="n">ld</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
</span><span id="__span-0-242"><a id="__codelineno-0-242" name="__codelineno-0-242"></a>            <span class="k">return</span> <span class="p">{</span><span class="n">c</span><span class="p">:</span> <span class="n">l</span><span class="o">.</span><span class="n">n_snps</span> <span class="k">for</span> <span class="n">c</span><span class="p">,</span> <span class="n">l</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">ld</span><span class="o">.</span><span class="n">items</span><span class="p">()}</span>
</span><span id="__span-0-243"><a id="__codelineno-0-243" name="__codelineno-0-243"></a>        <span class="k">elif</span> <span class="bp">self</span><span class="o">.</span><span class="n">annotation</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
</span><span id="__span-0-244"><a id="__codelineno-0-244" name="__codelineno-0-244"></a>            <span class="k">return</span> <span class="p">{</span><span class="n">c</span><span class="p">:</span> <span class="n">a</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="k">for</span> <span class="n">c</span><span class="p">,</span> <span class="n">a</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">annotation</span><span class="o">.</span><span class="n">items</span><span class="p">()}</span>
</span><span id="__span-0-245"><a id="__codelineno-0-245" name="__codelineno-0-245"></a>        <span class="k">else</span><span class="p">:</span>
</span><span id="__span-0-246"><a id="__codelineno-0-246" name="__codelineno-0-246"></a>            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;GWADataLoader instance is not properly initialized!&quot;</span><span class="p">)</span>
</span><span id="__span-0-247"><a id="__codelineno-0-247" name="__codelineno-0-247"></a>
</span><span id="__span-0-248"><a id="__codelineno-0-248" name="__codelineno-0-248"></a>    <span class="nd">@property</span>
</span><span id="__span-0-249"><a id="__codelineno-0-249" name="__codelineno-0-249"></a>    <span class="k">def</span><span class="w"> </span><span class="nf">chromosomes</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
</span><span id="__span-0-250"><a id="__codelineno-0-250" name="__codelineno-0-250"></a><span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
</span><span id="__span-0-251"><a id="__codelineno-0-251" name="__codelineno-0-251"></a><span class="sd">        :return: The list of chromosomes that were loaded to `GWADataLoader`.</span>
</span><span id="__span-0-252"><a id="__codelineno-0-252" name="__codelineno-0-252"></a><span class="sd">        &quot;&quot;&quot;</span>
</span><span id="__span-0-253"><a id="__codelineno-0-253" name="__codelineno-0-253"></a>        <span class="k">return</span> <span class="nb">sorted</span><span class="p">(</span><span class="nb">list</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">shapes</span><span class="o">.</span><span class="n">keys</span><span class="p">()))</span>
</span><span id="__span-0-254"><a id="__codelineno-0-254" name="__codelineno-0-254"></a>
</span><span id="__span-0-255"><a id="__codelineno-0-255" name="__codelineno-0-255"></a>    <span class="nd">@property</span>
</span><span id="__span-0-256"><a id="__codelineno-0-256" name="__codelineno-0-256"></a>    <span class="k">def</span><span class="w"> </span><span class="nf">n_annotations</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
</span><span id="__span-0-257"><a id="__codelineno-0-257" name="__codelineno-0-257"></a><span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
</span><span id="__span-0-258"><a id="__codelineno-0-258" name="__codelineno-0-258"></a><span class="sd">        :return: The number of annotations included in the annotation matrices.</span>
</span><span id="__span-0-259"><a id="__codelineno-0-259" name="__codelineno-0-259"></a><span class="sd">        &quot;&quot;&quot;</span>
</span><span id="__span-0-260"><a id="__codelineno-0-260" name="__codelineno-0-260"></a>        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">annotation</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
</span><span id="__span-0-261"><a id="__codelineno-0-261" name="__codelineno-0-261"></a>            <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">annotation</span><span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">chromosomes</span><span class="p">[</span><span class="mi">0</span><span class="p">]]</span><span class="o">.</span><span class="n">n_annotations</span>
</span><span id="__span-0-262"><a id="__codelineno-0-262" name="__codelineno-0-262"></a>
</span><span id="__span-0-263"><a id="__codelineno-0-263" name="__codelineno-0-263"></a>    <span class="k">def</span><span class="w"> </span><span class="nf">filter_snps</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">extract_snps</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">extract_file</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">chromosome</span><span class="o">=</span><span class="kc">None</span><span class="p">):</span>
</span><span id="__span-0-264"><a id="__codelineno-0-264" name="__codelineno-0-264"></a><span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
</span><span id="__span-0-265"><a id="__codelineno-0-265" name="__codelineno-0-265"></a><span class="sd">        Filter the SNP set from all the GWADataLoader objects.</span>
</span><span id="__span-0-266"><a id="__codelineno-0-266" name="__codelineno-0-266"></a><span class="sd">        :param extract_snps: A list or array of SNP rsIDs to keep.</span>
</span><span id="__span-0-267"><a id="__codelineno-0-267" name="__codelineno-0-267"></a><span class="sd">        :param extract_file: A path to a plink-style file with SNP rsIDs to keep.</span>
</span><span id="__span-0-268"><a id="__codelineno-0-268" name="__codelineno-0-268"></a><span class="sd">        :param chromosome: Chromosome number. If specified, applies the filter to that chromosome only.</span>
</span><span id="__span-0-269"><a id="__codelineno-0-269" name="__codelineno-0-269"></a><span class="sd">        &quot;&quot;&quot;</span>
</span><span id="__span-0-270"><a id="__codelineno-0-270" name="__codelineno-0-270"></a>
</span><span id="__span-0-271"><a id="__codelineno-0-271" name="__codelineno-0-271"></a>        <span class="k">if</span> <span class="n">extract_snps</span> <span class="ow">is</span> <span class="kc">None</span> <span class="ow">and</span> <span class="n">extract_file</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
</span><span id="__span-0-272"><a id="__codelineno-0-272" name="__codelineno-0-272"></a>            <span class="k">return</span>
</span><span id="__span-0-273"><a id="__codelineno-0-273" name="__codelineno-0-273"></a>
</span><span id="__span-0-274"><a id="__codelineno-0-274" name="__codelineno-0-274"></a>        <span class="k">if</span> <span class="n">chromosome</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
</span><span id="__span-0-275"><a id="__codelineno-0-275" name="__codelineno-0-275"></a>            <span class="n">chroms</span> <span class="o">=</span> <span class="p">[</span><span class="n">chromosome</span><span class="p">]</span>
</span><span id="__span-0-276"><a id="__codelineno-0-276" name="__codelineno-0-276"></a>        <span class="k">else</span><span class="p">:</span>
</span><span id="__span-0-277"><a id="__codelineno-0-277" name="__codelineno-0-277"></a>            <span class="n">chroms</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">chromosomes</span>
</span><span id="__span-0-278"><a id="__codelineno-0-278" name="__codelineno-0-278"></a>
</span><span id="__span-0-279"><a id="__codelineno-0-279" name="__codelineno-0-279"></a>        <span class="k">if</span> <span class="n">extract_snps</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
</span><span id="__span-0-280"><a id="__codelineno-0-280" name="__codelineno-0-280"></a>            <span class="kn">from</span><span class="w"> </span><span class="nn">.parsers.misc_parsers</span><span class="w"> </span><span class="kn">import</span> <span class="n">read_snp_filter_file</span>
</span><span id="__span-0-281"><a id="__codelineno-0-281" name="__codelineno-0-281"></a>            <span class="n">extract_snps</span> <span class="o">=</span> <span class="n">read_snp_filter_file</span><span class="p">(</span><span class="n">extract_file</span><span class="p">)</span>
</span><span id="__span-0-282"><a id="__codelineno-0-282" name="__codelineno-0-282"></a>
</span><span id="__span-0-283"><a id="__codelineno-0-283" name="__codelineno-0-283"></a>        <span class="k">for</span> <span class="n">c</span> <span class="ow">in</span> <span class="n">chroms</span><span class="p">:</span>
</span><span id="__span-0-284"><a id="__codelineno-0-284" name="__codelineno-0-284"></a>
</span><span id="__span-0-285"><a id="__codelineno-0-285" name="__codelineno-0-285"></a>            <span class="c1"># Filter the genotype matrix:</span>
</span><span id="__span-0-286"><a id="__codelineno-0-286" name="__codelineno-0-286"></a>            <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">genotype</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span> <span class="ow">and</span> <span class="n">c</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">genotype</span><span class="p">:</span>
</span><span id="__span-0-287"><a id="__codelineno-0-287" name="__codelineno-0-287"></a>                <span class="bp">self</span><span class="o">.</span><span class="n">genotype</span><span class="p">[</span><span class="n">c</span><span class="p">]</span><span class="o">.</span><span class="n">filter_snps</span><span class="p">(</span><span class="n">extract_snps</span><span class="o">=</span><span class="n">extract_snps</span><span class="p">)</span>
</span><span id="__span-0-288"><a id="__codelineno-0-288" name="__codelineno-0-288"></a>
</span><span id="__span-0-289"><a id="__codelineno-0-289" name="__codelineno-0-289"></a>                <span class="c1"># If no SNPs remain in the genotype matrix for that chromosome, then remove it:</span>
</span><span id="__span-0-290"><a id="__codelineno-0-290" name="__codelineno-0-290"></a>                <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">genotype</span><span class="p">[</span><span class="n">c</span><span class="p">]</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span> <span class="o">&lt;</span> <span class="mi">1</span><span class="p">:</span>
</span><span id="__span-0-291"><a id="__codelineno-0-291" name="__codelineno-0-291"></a>                    <span class="k">del</span> <span class="bp">self</span><span class="o">.</span><span class="n">genotype</span><span class="p">[</span><span class="n">c</span><span class="p">]</span>
</span><span id="__span-0-292"><a id="__codelineno-0-292" name="__codelineno-0-292"></a>
</span><span id="__span-0-293"><a id="__codelineno-0-293" name="__codelineno-0-293"></a>            <span class="c1"># Filter the summary statistics table:</span>
</span><span id="__span-0-294"><a id="__codelineno-0-294" name="__codelineno-0-294"></a>            <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">sumstats_table</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span> <span class="ow">and</span> <span class="n">c</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">sumstats_table</span><span class="p">:</span>
</span><span id="__span-0-295"><a id="__codelineno-0-295" name="__codelineno-0-295"></a>                <span class="bp">self</span><span class="o">.</span><span class="n">sumstats_table</span><span class="p">[</span><span class="n">c</span><span class="p">]</span><span class="o">.</span><span class="n">filter_snps</span><span class="p">(</span><span class="n">extract_snps</span><span class="o">=</span><span class="n">extract_snps</span><span class="p">)</span>
</span><span id="__span-0-296"><a id="__codelineno-0-296" name="__codelineno-0-296"></a>
</span><span id="__span-0-297"><a id="__codelineno-0-297" name="__codelineno-0-297"></a>                <span class="c1"># If no SNPs remain in the summary statistics table for that chromosome, then remove it:</span>
</span><span id="__span-0-298"><a id="__codelineno-0-298" name="__codelineno-0-298"></a>                <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">sumstats_table</span><span class="p">[</span><span class="n">c</span><span class="p">]</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="o">&lt;</span> <span class="mi">1</span><span class="p">:</span>
</span><span id="__span-0-299"><a id="__codelineno-0-299" name="__codelineno-0-299"></a>                    <span class="k">del</span> <span class="bp">self</span><span class="o">.</span><span class="n">sumstats_table</span><span class="p">[</span><span class="n">c</span><span class="p">]</span>
</span><span id="__span-0-300"><a id="__codelineno-0-300" name="__codelineno-0-300"></a>
</span><span id="__span-0-301"><a id="__codelineno-0-301" name="__codelineno-0-301"></a>            <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">ld</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span> <span class="ow">and</span> <span class="n">c</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">ld</span><span class="p">:</span>
</span><span id="__span-0-302"><a id="__codelineno-0-302" name="__codelineno-0-302"></a>                <span class="bp">self</span><span class="o">.</span><span class="n">ld</span><span class="p">[</span><span class="n">c</span><span class="p">]</span><span class="o">.</span><span class="n">filter_snps</span><span class="p">(</span><span class="n">extract_snps</span><span class="o">=</span><span class="n">extract_snps</span><span class="p">)</span>
</span><span id="__span-0-303"><a id="__codelineno-0-303" name="__codelineno-0-303"></a>
</span><span id="__span-0-304"><a id="__codelineno-0-304" name="__codelineno-0-304"></a>                <span class="c1"># If no SNPs remain in the summary statistics table for that chromosome, then remove it:</span>
</span><span id="__span-0-305"><a id="__codelineno-0-305" name="__codelineno-0-305"></a>                <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">ld</span><span class="p">[</span><span class="n">c</span><span class="p">]</span><span class="o">.</span><span class="n">n_snps</span> <span class="o">&lt;</span> <span class="mi">1</span><span class="p">:</span>
</span><span id="__span-0-306"><a id="__codelineno-0-306" name="__codelineno-0-306"></a>                    <span class="k">del</span> <span class="bp">self</span><span class="o">.</span><span class="n">ld</span><span class="p">[</span><span class="n">c</span><span class="p">]</span>
</span><span id="__span-0-307"><a id="__codelineno-0-307" name="__codelineno-0-307"></a>
</span><span id="__span-0-308"><a id="__codelineno-0-308" name="__codelineno-0-308"></a>            <span class="c1"># Filter the annotation matrix:</span>
</span><span id="__span-0-309"><a id="__codelineno-0-309" name="__codelineno-0-309"></a>            <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">annotation</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span> <span class="ow">and</span> <span class="n">c</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">annotation</span><span class="p">:</span>
</span><span id="__span-0-310"><a id="__codelineno-0-310" name="__codelineno-0-310"></a>                <span class="bp">self</span><span class="o">.</span><span class="n">annotation</span><span class="p">[</span><span class="n">c</span><span class="p">]</span><span class="o">.</span><span class="n">filter_snps</span><span class="p">(</span><span class="n">extract_snps</span><span class="o">=</span><span class="n">extract_snps</span><span class="p">)</span>
</span><span id="__span-0-311"><a id="__codelineno-0-311" name="__codelineno-0-311"></a>
</span><span id="__span-0-312"><a id="__codelineno-0-312" name="__codelineno-0-312"></a>                <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">annotation</span><span class="p">[</span><span class="n">c</span><span class="p">]</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="o">&lt;</span> <span class="mi">1</span><span class="p">:</span>
</span><span id="__span-0-313"><a id="__codelineno-0-313" name="__codelineno-0-313"></a>                    <span class="k">del</span> <span class="bp">self</span><span class="o">.</span><span class="n">annotation</span><span class="p">[</span><span class="n">c</span><span class="p">]</span>
</span><span id="__span-0-314"><a id="__codelineno-0-314" name="__codelineno-0-314"></a>
</span><span id="__span-0-315"><a id="__codelineno-0-315" name="__codelineno-0-315"></a>    <span class="k">def</span><span class="w"> </span><span class="nf">filter_samples</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">keep_samples</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">keep_file</span><span class="o">=</span><span class="kc">None</span><span class="p">):</span>
</span><span id="__span-0-316"><a id="__codelineno-0-316" name="__codelineno-0-316"></a><span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
</span><span id="__span-0-317"><a id="__codelineno-0-317" name="__codelineno-0-317"></a><span class="sd">        Filter samples from the samples table. User must specify</span>
</span><span id="__span-0-318"><a id="__codelineno-0-318" name="__codelineno-0-318"></a><span class="sd">        either a list of samples to keep or the path to a file</span>
</span><span id="__span-0-319"><a id="__codelineno-0-319" name="__codelineno-0-319"></a><span class="sd">        with the list of samples to keep.</span>
</span><span id="__span-0-320"><a id="__codelineno-0-320" name="__codelineno-0-320"></a>
</span><span id="__span-0-321"><a id="__codelineno-0-321" name="__codelineno-0-321"></a><span class="sd">        :param keep_samples: A list or array of sample IDs to keep.</span>
</span><span id="__span-0-322"><a id="__codelineno-0-322" name="__codelineno-0-322"></a><span class="sd">        :param keep_file: The path to a file with the list of samples to keep.</span>
</span><span id="__span-0-323"><a id="__codelineno-0-323" name="__codelineno-0-323"></a><span class="sd">        &quot;&quot;&quot;</span>
</span><span id="__span-0-324"><a id="__codelineno-0-324" name="__codelineno-0-324"></a>
</span><span id="__span-0-325"><a id="__codelineno-0-325" name="__codelineno-0-325"></a>        <span class="bp">self</span><span class="o">.</span><span class="n">sample_table</span><span class="o">.</span><span class="n">filter_samples</span><span class="p">(</span><span class="n">keep_samples</span><span class="o">=</span><span class="n">keep_samples</span><span class="p">,</span> <span class="n">keep_file</span><span class="o">=</span><span class="n">keep_file</span><span class="p">)</span>
</span><span id="__span-0-326"><a id="__codelineno-0-326" name="__codelineno-0-326"></a>        <span class="bp">self</span><span class="o">.</span><span class="n">sync_sample_tables</span><span class="p">()</span>
</span><span id="__span-0-327"><a id="__codelineno-0-327" name="__codelineno-0-327"></a>
</span><span id="__span-0-328"><a id="__codelineno-0-328" name="__codelineno-0-328"></a>    <span class="k">def</span><span class="w"> </span><span class="nf">read_annotations</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">annot_path</span><span class="p">,</span>
</span><span id="__span-0-329"><a id="__codelineno-0-329" name="__codelineno-0-329"></a>                         <span class="n">annot_format</span><span class="o">=</span><span class="s1">&#39;magenpy&#39;</span><span class="p">,</span>
</span><span id="__span-0-330"><a id="__codelineno-0-330" name="__codelineno-0-330"></a>                         <span class="n">parser</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
</span><span id="__span-0-331"><a id="__codelineno-0-331" name="__codelineno-0-331"></a>                         <span class="o">**</span><span class="n">parse_kwargs</span><span class="p">):</span>
</span><span id="__span-0-332"><a id="__codelineno-0-332" name="__codelineno-0-332"></a><span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
</span><span id="__span-0-333"><a id="__codelineno-0-333" name="__codelineno-0-333"></a><span class="sd">        Read the annotation matrix from file. Annotations are a set of features associated</span>
</span><span id="__span-0-334"><a id="__codelineno-0-334" name="__codelineno-0-334"></a><span class="sd">        with each SNP and are generally represented in table format.</span>
</span><span id="__span-0-335"><a id="__codelineno-0-335" name="__codelineno-0-335"></a><span class="sd">        Consult the documentation for `AnnotationMatrix` for more details.</span>
</span><span id="__span-0-336"><a id="__codelineno-0-336" name="__codelineno-0-336"></a>
</span><span id="__span-0-337"><a id="__codelineno-0-337" name="__codelineno-0-337"></a><span class="sd">        :param annot_path: The path to the annotation file(s). The path may contain a wildcard.</span>
</span><span id="__span-0-338"><a id="__codelineno-0-338" name="__codelineno-0-338"></a><span class="sd">        :param annot_format: The format for the summary statistics. Currently, supports the following</span>
</span><span id="__span-0-339"><a id="__codelineno-0-339" name="__codelineno-0-339"></a><span class="sd">         formats: `magenpy`, `ldsc`.</span>
</span><span id="__span-0-340"><a id="__codelineno-0-340" name="__codelineno-0-340"></a><span class="sd">        :param parser: If the annotation file does not follow any of the formats above, you can create</span>
</span><span id="__span-0-341"><a id="__codelineno-0-341" name="__codelineno-0-341"></a><span class="sd">        your own parser by inheriting from the base `AnnotationMatrixParser` class and passing it here as an argument.</span>
</span><span id="__span-0-342"><a id="__codelineno-0-342" name="__codelineno-0-342"></a><span class="sd">        :param parse_kwargs: keyword arguments for the parser. These are mainly parameters that will be passed to</span>
</span><span id="__span-0-343"><a id="__codelineno-0-343" name="__codelineno-0-343"></a><span class="sd">        `pandas.read_csv` function, such as the delimiter, header information, etc.</span>
</span><span id="__span-0-344"><a id="__codelineno-0-344" name="__codelineno-0-344"></a><span class="sd">        &quot;&quot;&quot;</span>
</span><span id="__span-0-345"><a id="__codelineno-0-345" name="__codelineno-0-345"></a>
</span><span id="__span-0-346"><a id="__codelineno-0-346" name="__codelineno-0-346"></a>        <span class="k">if</span> <span class="n">annot_path</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
</span><span id="__span-0-347"><a id="__codelineno-0-347" name="__codelineno-0-347"></a>            <span class="k">return</span>
</span><span id="__span-0-348"><a id="__codelineno-0-348" name="__codelineno-0-348"></a>
</span><span id="__span-0-349"><a id="__codelineno-0-349" name="__codelineno-0-349"></a>        <span class="c1"># Find all the relevant files in the path passed by the user:</span>
</span><span id="__span-0-350"><a id="__codelineno-0-350" name="__codelineno-0-350"></a>        <span class="k">if</span> <span class="ow">not</span> <span class="n">iterable</span><span class="p">(</span><span class="n">annot_path</span><span class="p">):</span>
</span><span id="__span-0-351"><a id="__codelineno-0-351" name="__codelineno-0-351"></a>            <span class="n">annot_files</span> <span class="o">=</span> <span class="n">get_filenames</span><span class="p">(</span><span class="n">annot_path</span><span class="p">,</span> <span class="n">extension</span><span class="o">=</span><span class="s1">&#39;.annot&#39;</span><span class="p">)</span>
</span><span id="__span-0-352"><a id="__codelineno-0-352" name="__codelineno-0-352"></a>        <span class="k">else</span><span class="p">:</span>
</span><span id="__span-0-353"><a id="__codelineno-0-353" name="__codelineno-0-353"></a>            <span class="n">annot_files</span> <span class="o">=</span> <span class="n">annot_path</span>
</span><span id="__span-0-354"><a id="__codelineno-0-354" name="__codelineno-0-354"></a>
</span><span id="__span-0-355"><a id="__codelineno-0-355" name="__codelineno-0-355"></a>        <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">annot_files</span><span class="p">)</span> <span class="o">&lt;</span> <span class="mi">1</span><span class="p">:</span>
</span><span id="__span-0-356"><a id="__codelineno-0-356" name="__codelineno-0-356"></a>            <span class="n">logger</span><span class="o">.</span><span class="n">warning</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;No annotation files were found at: </span><span class="si">{</span><span class="n">annot_path</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
</span><span id="__span-0-357"><a id="__codelineno-0-357" name="__codelineno-0-357"></a>            <span class="k">return</span>
</span><span id="__span-0-358"><a id="__codelineno-0-358" name="__codelineno-0-358"></a>
</span><span id="__span-0-359"><a id="__codelineno-0-359" name="__codelineno-0-359"></a>        <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;&gt; Reading annotation file...&quot;</span><span class="p">)</span>
</span><span id="__span-0-360"><a id="__codelineno-0-360" name="__codelineno-0-360"></a>
</span><span id="__span-0-361"><a id="__codelineno-0-361" name="__codelineno-0-361"></a>        <span class="bp">self</span><span class="o">.</span><span class="n">annotation</span> <span class="o">=</span> <span class="p">{}</span>
</span><span id="__span-0-362"><a id="__codelineno-0-362" name="__codelineno-0-362"></a>
</span><span id="__span-0-363"><a id="__codelineno-0-363" name="__codelineno-0-363"></a>        <span class="k">for</span> <span class="n">annot_file</span> <span class="ow">in</span> <span class="n">tqdm</span><span class="p">(</span><span class="n">annot_files</span><span class="p">,</span>
</span><span id="__span-0-364"><a id="__codelineno-0-364" name="__codelineno-0-364"></a>                               <span class="n">total</span><span class="o">=</span><span class="nb">len</span><span class="p">(</span><span class="n">annot_files</span><span class="p">),</span>
</span><span id="__span-0-365"><a id="__codelineno-0-365" name="__codelineno-0-365"></a>                               <span class="n">desc</span><span class="o">=</span><span class="s2">&quot;Reading annotation files&quot;</span><span class="p">):</span>
</span><span id="__span-0-366"><a id="__codelineno-0-366" name="__codelineno-0-366"></a>            <span class="n">annot_mat</span> <span class="o">=</span> <span class="n">AnnotationMatrix</span><span class="o">.</span><span class="n">from_file</span><span class="p">(</span><span class="n">annot_file</span><span class="p">,</span>
</span><span id="__span-0-367"><a id="__codelineno-0-367" name="__codelineno-0-367"></a>                                                   <span class="n">annot_format</span><span class="o">=</span><span class="n">annot_format</span><span class="p">,</span>
</span><span id="__span-0-368"><a id="__codelineno-0-368" name="__codelineno-0-368"></a>                                                   <span class="n">annot_parser</span><span class="o">=</span><span class="n">parser</span><span class="p">,</span>
</span><span id="__span-0-369"><a id="__codelineno-0-369" name="__codelineno-0-369"></a>                                                   <span class="o">**</span><span class="n">parse_kwargs</span><span class="p">)</span>
</span><span id="__span-0-370"><a id="__codelineno-0-370" name="__codelineno-0-370"></a>            <span class="bp">self</span><span class="o">.</span><span class="n">annotation</span><span class="p">[</span><span class="n">annot_mat</span><span class="o">.</span><span class="n">chromosome</span><span class="p">]</span> <span class="o">=</span> <span class="n">annot_mat</span>
</span><span id="__span-0-371"><a id="__codelineno-0-371" name="__codelineno-0-371"></a>
</span><span id="__span-0-372"><a id="__codelineno-0-372" name="__codelineno-0-372"></a>    <span class="k">def</span><span class="w"> </span><span class="nf">read_genotypes</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span>
</span><span id="__span-0-373"><a id="__codelineno-0-373" name="__codelineno-0-373"></a>                       <span class="n">bed_paths</span><span class="p">,</span>
</span><span id="__span-0-374"><a id="__codelineno-0-374" name="__codelineno-0-374"></a>                       <span class="n">keep_samples</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
</span><span id="__span-0-375"><a id="__codelineno-0-375" name="__codelineno-0-375"></a>                       <span class="n">keep_file</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
</span><span id="__span-0-376"><a id="__codelineno-0-376" name="__codelineno-0-376"></a>                       <span class="n">extract_snps</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
</span><span id="__span-0-377"><a id="__codelineno-0-377" name="__codelineno-0-377"></a>                       <span class="n">extract_file</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
</span><span id="__span-0-378"><a id="__codelineno-0-378" name="__codelineno-0-378"></a>                       <span class="n">min_maf</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
</span><span id="__span-0-379"><a id="__codelineno-0-379" name="__codelineno-0-379"></a>                       <span class="n">min_mac</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span>
</span><span id="__span-0-380"><a id="__codelineno-0-380" name="__codelineno-0-380"></a>                       <span class="n">drop_duplicated</span><span class="o">=</span><span class="kc">True</span><span class="p">):</span>
</span><span id="__span-0-381"><a id="__codelineno-0-381" name="__codelineno-0-381"></a><span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
</span><span id="__span-0-382"><a id="__codelineno-0-382" name="__codelineno-0-382"></a><span class="sd">        Read the genotype matrix and/or associated metadata from plink&#39;s BED file format.</span>
</span><span id="__span-0-383"><a id="__codelineno-0-383" name="__codelineno-0-383"></a><span class="sd">        Consult the documentation for `GenotypeMatrix` for more details.</span>
</span><span id="__span-0-384"><a id="__codelineno-0-384" name="__codelineno-0-384"></a>
</span><span id="__span-0-385"><a id="__codelineno-0-385" name="__codelineno-0-385"></a><span class="sd">        :param bed_paths: The path to the BED file(s). You may use a wildcard here to read files for multiple</span>
</span><span id="__span-0-386"><a id="__codelineno-0-386" name="__codelineno-0-386"></a><span class="sd">        chromosomes.</span>
</span><span id="__span-0-387"><a id="__codelineno-0-387" name="__codelineno-0-387"></a><span class="sd">        :param keep_samples: A vector or list of sample IDs to keep when filtering the genotype matrix.</span>
</span><span id="__span-0-388"><a id="__codelineno-0-388" name="__codelineno-0-388"></a><span class="sd">        :param keep_file: A path to a plink-style file containing sample IDs to keep.</span>
</span><span id="__span-0-389"><a id="__codelineno-0-389" name="__codelineno-0-389"></a><span class="sd">        :param extract_snps: A vector or list of SNP IDs to keep when filtering the genotype matrix.</span>
</span><span id="__span-0-390"><a id="__codelineno-0-390" name="__codelineno-0-390"></a><span class="sd">        :param extract_file: A path to a plink-style file containing SNP IDs to keep.</span>
</span><span id="__span-0-391"><a id="__codelineno-0-391" name="__codelineno-0-391"></a><span class="sd">        :param min_maf: The minimum minor allele frequency cutoff.</span>
</span><span id="__span-0-392"><a id="__codelineno-0-392" name="__codelineno-0-392"></a><span class="sd">        :param min_mac: The minimum minor allele count cutoff.</span>
</span><span id="__span-0-393"><a id="__codelineno-0-393" name="__codelineno-0-393"></a><span class="sd">        :param drop_duplicated: If True, drop SNPs with duplicated rsID.</span>
</span><span id="__span-0-394"><a id="__codelineno-0-394" name="__codelineno-0-394"></a><span class="sd">        &quot;&quot;&quot;</span>
</span><span id="__span-0-395"><a id="__codelineno-0-395" name="__codelineno-0-395"></a>
</span><span id="__span-0-396"><a id="__codelineno-0-396" name="__codelineno-0-396"></a>        <span class="k">if</span> <span class="n">bed_paths</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
</span><span id="__span-0-397"><a id="__codelineno-0-397" name="__codelineno-0-397"></a>            <span class="k">return</span>
</span><span id="__span-0-398"><a id="__codelineno-0-398" name="__codelineno-0-398"></a>
</span><span id="__span-0-399"><a id="__codelineno-0-399" name="__codelineno-0-399"></a>        <span class="c1"># Find all the relevant files in the path passed by the user:</span>
</span><span id="__span-0-400"><a id="__codelineno-0-400" name="__codelineno-0-400"></a>        <span class="k">if</span> <span class="ow">not</span> <span class="n">iterable</span><span class="p">(</span><span class="n">bed_paths</span><span class="p">):</span>
</span><span id="__span-0-401"><a id="__codelineno-0-401" name="__codelineno-0-401"></a>            <span class="n">bed_files</span> <span class="o">=</span> <span class="n">get_filenames</span><span class="p">(</span><span class="n">bed_paths</span><span class="p">,</span> <span class="n">extension</span><span class="o">=</span><span class="s1">&#39;.bed&#39;</span><span class="p">)</span>
</span><span id="__span-0-402"><a id="__codelineno-0-402" name="__codelineno-0-402"></a>        <span class="k">else</span><span class="p">:</span>
</span><span id="__span-0-403"><a id="__codelineno-0-403" name="__codelineno-0-403"></a>            <span class="n">bed_files</span> <span class="o">=</span> <span class="n">bed_paths</span>
</span><span id="__span-0-404"><a id="__codelineno-0-404" name="__codelineno-0-404"></a>
</span><span id="__span-0-405"><a id="__codelineno-0-405" name="__codelineno-0-405"></a>        <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">bed_files</span><span class="p">)</span> <span class="o">&lt;</span> <span class="mi">1</span><span class="p">:</span>
</span><span id="__span-0-406"><a id="__codelineno-0-406" name="__codelineno-0-406"></a>            <span class="n">logger</span><span class="o">.</span><span class="n">warning</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;No BED files were found at: </span><span class="si">{</span><span class="n">bed_paths</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
</span><span id="__span-0-407"><a id="__codelineno-0-407" name="__codelineno-0-407"></a>            <span class="k">return</span>
</span><span id="__span-0-408"><a id="__codelineno-0-408" name="__codelineno-0-408"></a>
</span><span id="__span-0-409"><a id="__codelineno-0-409" name="__codelineno-0-409"></a>        <span class="c1"># Depending on the backend, select the `GenotypeMatrix` class:</span>
</span><span id="__span-0-410"><a id="__codelineno-0-410" name="__codelineno-0-410"></a>        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">backend</span> <span class="o">==</span> <span class="s1">&#39;xarray&#39;</span><span class="p">:</span>
</span><span id="__span-0-411"><a id="__codelineno-0-411" name="__codelineno-0-411"></a>            <span class="n">gmat_class</span> <span class="o">=</span> <span class="n">xarrayGenotypeMatrix</span>
</span><span id="__span-0-412"><a id="__codelineno-0-412" name="__codelineno-0-412"></a>        <span class="k">elif</span> <span class="bp">self</span><span class="o">.</span><span class="n">backend</span> <span class="o">==</span> <span class="s1">&#39;bed-reader&#39;</span><span class="p">:</span>
</span><span id="__span-0-413"><a id="__codelineno-0-413" name="__codelineno-0-413"></a>            <span class="n">gmat_class</span> <span class="o">=</span> <span class="n">bedReaderGenotypeMatrix</span>
</span><span id="__span-0-414"><a id="__codelineno-0-414" name="__codelineno-0-414"></a>        <span class="k">else</span><span class="p">:</span>
</span><span id="__span-0-415"><a id="__codelineno-0-415" name="__codelineno-0-415"></a>            <span class="n">gmat_class</span> <span class="o">=</span> <span class="n">plinkBEDGenotypeMatrix</span>
</span><span id="__span-0-416"><a id="__codelineno-0-416" name="__codelineno-0-416"></a>
</span><span id="__span-0-417"><a id="__codelineno-0-417" name="__codelineno-0-417"></a>        <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;&gt; Reading genotype metadata...&quot;</span><span class="p">)</span>
</span><span id="__span-0-418"><a id="__codelineno-0-418" name="__codelineno-0-418"></a>
</span><span id="__span-0-419"><a id="__codelineno-0-419" name="__codelineno-0-419"></a>        <span class="bp">self</span><span class="o">.</span><span class="n">genotype</span> <span class="o">=</span> <span class="p">{}</span>
</span><span id="__span-0-420"><a id="__codelineno-0-420" name="__codelineno-0-420"></a>
</span><span id="__span-0-421"><a id="__codelineno-0-421" name="__codelineno-0-421"></a>        <span class="k">for</span> <span class="n">bfile</span> <span class="ow">in</span> <span class="n">tqdm</span><span class="p">(</span><span class="n">bed_files</span><span class="p">,</span>
</span><span id="__span-0-422"><a id="__codelineno-0-422" name="__codelineno-0-422"></a>                          <span class="n">total</span><span class="o">=</span><span class="nb">len</span><span class="p">(</span><span class="n">bed_files</span><span class="p">),</span>
</span><span id="__span-0-423"><a id="__codelineno-0-423" name="__codelineno-0-423"></a>                          <span class="n">desc</span><span class="o">=</span><span class="s2">&quot;Reading genotype metadata&quot;</span><span class="p">):</span>
</span><span id="__span-0-424"><a id="__codelineno-0-424" name="__codelineno-0-424"></a>            <span class="c1"># Read BED file and update the genotypes dictionary:</span>
</span><span id="__span-0-425"><a id="__codelineno-0-425" name="__codelineno-0-425"></a>            <span class="bp">self</span><span class="o">.</span><span class="n">genotype</span><span class="o">.</span><span class="n">update</span><span class="p">(</span><span class="n">gmat_class</span><span class="o">.</span><span class="n">from_file</span><span class="p">(</span><span class="n">bfile</span><span class="p">,</span>
</span><span id="__span-0-426"><a id="__codelineno-0-426" name="__codelineno-0-426"></a>                                                      <span class="n">temp_dir</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">temp_dir</span><span class="p">,</span>
</span><span id="__span-0-427"><a id="__codelineno-0-427" name="__codelineno-0-427"></a>                                                      <span class="n">threads</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">threads</span><span class="p">)</span><span class="o">.</span><span class="n">split_by_chromosome</span><span class="p">())</span>
</span><span id="__span-0-428"><a id="__codelineno-0-428" name="__codelineno-0-428"></a>
</span><span id="__span-0-429"><a id="__codelineno-0-429" name="__codelineno-0-429"></a>        <span class="c1"># After reading the genotype matrices, apply some standard filters:</span>
</span><span id="__span-0-430"><a id="__codelineno-0-430" name="__codelineno-0-430"></a>        <span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="p">(</span><span class="n">c</span><span class="p">,</span> <span class="n">g</span><span class="p">)</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">genotype</span><span class="o">.</span><span class="n">items</span><span class="p">()):</span>
</span><span id="__span-0-431"><a id="__codelineno-0-431" name="__codelineno-0-431"></a>
</span><span id="__span-0-432"><a id="__codelineno-0-432" name="__codelineno-0-432"></a>            <span class="c1"># Filter the genotype matrix to keep certain subsample:</span>
</span><span id="__span-0-433"><a id="__codelineno-0-433" name="__codelineno-0-433"></a>            <span class="k">if</span> <span class="n">keep_samples</span> <span class="ow">or</span> <span class="n">keep_file</span><span class="p">:</span>
</span><span id="__span-0-434"><a id="__codelineno-0-434" name="__codelineno-0-434"></a>                <span class="n">g</span><span class="o">.</span><span class="n">filter_samples</span><span class="p">(</span><span class="n">keep_samples</span><span class="o">=</span><span class="n">keep_samples</span><span class="p">,</span> <span class="n">keep_file</span><span class="o">=</span><span class="n">keep_file</span><span class="p">)</span>
</span><span id="__span-0-435"><a id="__codelineno-0-435" name="__codelineno-0-435"></a>
</span><span id="__span-0-436"><a id="__codelineno-0-436" name="__codelineno-0-436"></a>            <span class="c1"># Filter the genotype matrix to keep certain SNPs</span>
</span><span id="__span-0-437"><a id="__codelineno-0-437" name="__codelineno-0-437"></a>            <span class="k">if</span> <span class="n">extract_snps</span> <span class="ow">or</span> <span class="n">extract_file</span><span class="p">:</span>
</span><span id="__span-0-438"><a id="__codelineno-0-438" name="__codelineno-0-438"></a>                <span class="n">g</span><span class="o">.</span><span class="n">filter_snps</span><span class="p">(</span><span class="n">extract_snps</span><span class="o">=</span><span class="n">extract_snps</span><span class="p">,</span> <span class="n">extract_file</span><span class="o">=</span><span class="n">extract_file</span><span class="p">)</span>
</span><span id="__span-0-439"><a id="__codelineno-0-439" name="__codelineno-0-439"></a>
</span><span id="__span-0-440"><a id="__codelineno-0-440" name="__codelineno-0-440"></a>            <span class="c1"># Drop duplicated SNP IDs</span>
</span><span id="__span-0-441"><a id="__codelineno-0-441" name="__codelineno-0-441"></a>            <span class="k">if</span> <span class="n">drop_duplicated</span><span class="p">:</span>
</span><span id="__span-0-442"><a id="__codelineno-0-442" name="__codelineno-0-442"></a>                <span class="n">g</span><span class="o">.</span><span class="n">drop_duplicated_snps</span><span class="p">()</span>
</span><span id="__span-0-443"><a id="__codelineno-0-443" name="__codelineno-0-443"></a>
</span><span id="__span-0-444"><a id="__codelineno-0-444" name="__codelineno-0-444"></a>            <span class="c1"># Filter SNPs by minor allele frequency and/or count:</span>
</span><span id="__span-0-445"><a id="__codelineno-0-445" name="__codelineno-0-445"></a>            <span class="n">g</span><span class="o">.</span><span class="n">filter_by_allele_frequency</span><span class="p">(</span><span class="n">min_maf</span><span class="o">=</span><span class="n">min_maf</span><span class="p">,</span> <span class="n">min_mac</span><span class="o">=</span><span class="n">min_mac</span><span class="p">)</span>
</span><span id="__span-0-446"><a id="__codelineno-0-446" name="__codelineno-0-446"></a>
</span><span id="__span-0-447"><a id="__codelineno-0-447" name="__codelineno-0-447"></a>            <span class="k">if</span> <span class="n">i</span> <span class="o">==</span> <span class="mi">0</span><span class="p">:</span>
</span><span id="__span-0-448"><a id="__codelineno-0-448" name="__codelineno-0-448"></a>                <span class="bp">self</span><span class="o">.</span><span class="n">sample_table</span> <span class="o">=</span> <span class="n">g</span><span class="o">.</span><span class="n">sample_table</span>
</span><span id="__span-0-449"><a id="__codelineno-0-449" name="__codelineno-0-449"></a>
</span><span id="__span-0-450"><a id="__codelineno-0-450" name="__codelineno-0-450"></a>    <span class="k">def</span><span class="w"> </span><span class="nf">read_phenotype</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">phenotype_file</span><span class="p">,</span> <span class="n">drop_na</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="o">**</span><span class="n">read_csv_kwargs</span><span class="p">):</span>
</span><span id="__span-0-451"><a id="__codelineno-0-451" name="__codelineno-0-451"></a><span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
</span><span id="__span-0-452"><a id="__codelineno-0-452" name="__codelineno-0-452"></a><span class="sd">        Read the phenotype file and integrate it with the sample tables and genotype matrices.</span>
</span><span id="__span-0-453"><a id="__codelineno-0-453" name="__codelineno-0-453"></a>
</span><span id="__span-0-454"><a id="__codelineno-0-454" name="__codelineno-0-454"></a><span class="sd">        :param phenotype_file: The path to the phenotype file</span>
</span><span id="__span-0-455"><a id="__codelineno-0-455" name="__codelineno-0-455"></a><span class="sd">        (Default: tab-separated file with `FID IID phenotype` columns). If different, supply</span>
</span><span id="__span-0-456"><a id="__codelineno-0-456" name="__codelineno-0-456"></a><span class="sd">        details as additional arguments to this function.</span>
</span><span id="__span-0-457"><a id="__codelineno-0-457" name="__codelineno-0-457"></a><span class="sd">        :param drop_na: Drop samples with missing phenotype information.</span>
</span><span id="__span-0-458"><a id="__codelineno-0-458" name="__codelineno-0-458"></a><span class="sd">        :param read_csv_kwargs: keyword arguments for the `read_csv` function of `pandas`.</span>
</span><span id="__span-0-459"><a id="__codelineno-0-459" name="__codelineno-0-459"></a><span class="sd">        &quot;&quot;&quot;</span>
</span><span id="__span-0-460"><a id="__codelineno-0-460" name="__codelineno-0-460"></a>
</span><span id="__span-0-461"><a id="__codelineno-0-461" name="__codelineno-0-461"></a>        <span class="k">if</span> <span class="n">phenotype_file</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
</span><span id="__span-0-462"><a id="__codelineno-0-462" name="__codelineno-0-462"></a>            <span class="k">return</span>
</span><span id="__span-0-463"><a id="__codelineno-0-463" name="__codelineno-0-463"></a>
</span><span id="__span-0-464"><a id="__codelineno-0-464" name="__codelineno-0-464"></a>        <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;&gt; Reading phenotype file...&quot;</span><span class="p">)</span>
</span><span id="__span-0-465"><a id="__codelineno-0-465" name="__codelineno-0-465"></a>
</span><span id="__span-0-466"><a id="__codelineno-0-466" name="__codelineno-0-466"></a>        <span class="k">assert</span> <span class="bp">self</span><span class="o">.</span><span class="n">sample_table</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span>
</span><span id="__span-0-467"><a id="__codelineno-0-467" name="__codelineno-0-467"></a>
</span><span id="__span-0-468"><a id="__codelineno-0-468" name="__codelineno-0-468"></a>        <span class="bp">self</span><span class="o">.</span><span class="n">sample_table</span><span class="o">.</span><span class="n">read_phenotype_file</span><span class="p">(</span><span class="n">phenotype_file</span><span class="p">,</span> <span class="n">drop_na</span><span class="o">=</span><span class="n">drop_na</span><span class="p">,</span> <span class="o">**</span><span class="n">read_csv_kwargs</span><span class="p">)</span>
</span><span id="__span-0-469"><a id="__codelineno-0-469" name="__codelineno-0-469"></a>        <span class="bp">self</span><span class="o">.</span><span class="n">sync_sample_tables</span><span class="p">()</span>
</span><span id="__span-0-470"><a id="__codelineno-0-470" name="__codelineno-0-470"></a>
</span><span id="__span-0-471"><a id="__codelineno-0-471" name="__codelineno-0-471"></a>    <span class="k">def</span><span class="w"> </span><span class="nf">set_phenotype</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">new_phenotype</span><span class="p">,</span> <span class="n">phenotype_likelihood</span><span class="o">=</span><span class="kc">None</span><span class="p">):</span>
</span><span id="__span-0-472"><a id="__codelineno-0-472" name="__codelineno-0-472"></a><span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
</span><span id="__span-0-473"><a id="__codelineno-0-473" name="__codelineno-0-473"></a><span class="sd">        A convenience method to update the phenotype column for the samples.</span>
</span><span id="__span-0-474"><a id="__codelineno-0-474" name="__codelineno-0-474"></a><span class="sd">        :param new_phenotype: A vector or list of phenotype values.</span>
</span><span id="__span-0-475"><a id="__codelineno-0-475" name="__codelineno-0-475"></a><span class="sd">        :param phenotype_likelihood: The phenotype likelihood (e.g. `binomial`, `gaussian`). Optional.</span>
</span><span id="__span-0-476"><a id="__codelineno-0-476" name="__codelineno-0-476"></a><span class="sd">        &quot;&quot;&quot;</span>
</span><span id="__span-0-477"><a id="__codelineno-0-477" name="__codelineno-0-477"></a>
</span><span id="__span-0-478"><a id="__codelineno-0-478" name="__codelineno-0-478"></a>        <span class="bp">self</span><span class="o">.</span><span class="n">sample_table</span><span class="o">.</span><span class="n">set_phenotype</span><span class="p">(</span><span class="n">new_phenotype</span><span class="p">,</span>
</span><span id="__span-0-479"><a id="__codelineno-0-479" name="__codelineno-0-479"></a>                                        <span class="n">phenotype_likelihood</span><span class="o">=</span><span class="n">phenotype_likelihood</span> <span class="ow">or</span> <span class="bp">self</span><span class="o">.</span><span class="n">phenotype_likelihood</span><span class="p">)</span>
</span><span id="__span-0-480"><a id="__codelineno-0-480" name="__codelineno-0-480"></a>        <span class="bp">self</span><span class="o">.</span><span class="n">sync_sample_tables</span><span class="p">()</span>
</span><span id="__span-0-481"><a id="__codelineno-0-481" name="__codelineno-0-481"></a>
</span><span id="__span-0-482"><a id="__codelineno-0-482" name="__codelineno-0-482"></a>    <span class="k">def</span><span class="w"> </span><span class="nf">read_covariates</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">covariates_file</span><span class="p">,</span> <span class="o">**</span><span class="n">read_csv_kwargs</span><span class="p">):</span>
</span><span id="__span-0-483"><a id="__codelineno-0-483" name="__codelineno-0-483"></a><span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
</span><span id="__span-0-484"><a id="__codelineno-0-484" name="__codelineno-0-484"></a><span class="sd">        Read the covariates file and integrate it with the sample tables and genotype matrices.</span>
</span><span id="__span-0-485"><a id="__codelineno-0-485" name="__codelineno-0-485"></a>
</span><span id="__span-0-486"><a id="__codelineno-0-486" name="__codelineno-0-486"></a><span class="sd">        :param covariates_file: The path to the covariates file</span>
</span><span id="__span-0-487"><a id="__codelineno-0-487" name="__codelineno-0-487"></a><span class="sd">        (Default: tab-separated file starting with the `FID IID ...` columns and followed by the covariate columns).</span>
</span><span id="__span-0-488"><a id="__codelineno-0-488" name="__codelineno-0-488"></a><span class="sd">        :param read_csv_kwargs: keyword arguments for the `read_csv` function of `pandas`.</span>
</span><span id="__span-0-489"><a id="__codelineno-0-489" name="__codelineno-0-489"></a><span class="sd">        &quot;&quot;&quot;</span>
</span><span id="__span-0-490"><a id="__codelineno-0-490" name="__codelineno-0-490"></a>
</span><span id="__span-0-491"><a id="__codelineno-0-491" name="__codelineno-0-491"></a>        <span class="k">if</span> <span class="n">covariates_file</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
</span><span id="__span-0-492"><a id="__codelineno-0-492" name="__codelineno-0-492"></a>            <span class="k">return</span>
</span><span id="__span-0-493"><a id="__codelineno-0-493" name="__codelineno-0-493"></a>
</span><span id="__span-0-494"><a id="__codelineno-0-494" name="__codelineno-0-494"></a>        <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;&gt; Reading covariates file...&quot;</span><span class="p">)</span>
</span><span id="__span-0-495"><a id="__codelineno-0-495" name="__codelineno-0-495"></a>
</span><span id="__span-0-496"><a id="__codelineno-0-496" name="__codelineno-0-496"></a>        <span class="k">assert</span> <span class="bp">self</span><span class="o">.</span><span class="n">sample_table</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span>
</span><span id="__span-0-497"><a id="__codelineno-0-497" name="__codelineno-0-497"></a>
</span><span id="__span-0-498"><a id="__codelineno-0-498" name="__codelineno-0-498"></a>        <span class="bp">self</span><span class="o">.</span><span class="n">sample_table</span><span class="o">.</span><span class="n">read_covariates_file</span><span class="p">(</span><span class="n">covariates_file</span><span class="p">,</span> <span class="o">**</span><span class="n">read_csv_kwargs</span><span class="p">)</span>
</span><span id="__span-0-499"><a id="__codelineno-0-499" name="__codelineno-0-499"></a>        <span class="bp">self</span><span class="o">.</span><span class="n">sync_sample_tables</span><span class="p">()</span>
</span><span id="__span-0-500"><a id="__codelineno-0-500" name="__codelineno-0-500"></a>
</span><span id="__span-0-501"><a id="__codelineno-0-501" name="__codelineno-0-501"></a>    <span class="k">def</span><span class="w"> </span><span class="nf">read_summary_statistics</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span>
</span><span id="__span-0-502"><a id="__codelineno-0-502" name="__codelineno-0-502"></a>                                <span class="n">sumstats_path</span><span class="p">,</span>
</span><span id="__span-0-503"><a id="__codelineno-0-503" name="__codelineno-0-503"></a>                                <span class="n">sumstats_format</span><span class="o">=</span><span class="s1">&#39;magenpy&#39;</span><span class="p">,</span>
</span><span id="__span-0-504"><a id="__codelineno-0-504" name="__codelineno-0-504"></a>                                <span class="n">parser</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
</span><span id="__span-0-505"><a id="__codelineno-0-505" name="__codelineno-0-505"></a>                                <span class="n">drop_duplicated</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
</span><span id="__span-0-506"><a id="__codelineno-0-506" name="__codelineno-0-506"></a>                                <span class="o">**</span><span class="n">parse_kwargs</span><span class="p">):</span>
</span><span id="__span-0-507"><a id="__codelineno-0-507" name="__codelineno-0-507"></a><span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
</span><span id="__span-0-508"><a id="__codelineno-0-508" name="__codelineno-0-508"></a><span class="sd">        Read GWAS summary statistics file(s) and parse them to `SumstatsTable` objects.</span>
</span><span id="__span-0-509"><a id="__codelineno-0-509" name="__codelineno-0-509"></a>
</span><span id="__span-0-510"><a id="__codelineno-0-510" name="__codelineno-0-510"></a><span class="sd">        :param sumstats_path: The path to the summary statistics file(s). The path may be a wildcard.</span>
</span><span id="__span-0-511"><a id="__codelineno-0-511" name="__codelineno-0-511"></a><span class="sd">        :param sumstats_format: The format for the summary statistics. Currently supports the following</span>
</span><span id="__span-0-512"><a id="__codelineno-0-512" name="__codelineno-0-512"></a><span class="sd">         formats: `plink1.9`, `plink2`, `magenpy`, `fastGWA`, `COJO`, `SAIGE`, or `GWASCatalog` for the standard</span>
</span><span id="__span-0-513"><a id="__codelineno-0-513" name="__codelineno-0-513"></a><span class="sd">         summary statistics format (also known as `ssf` or `gwas-ssf`).</span>
</span><span id="__span-0-514"><a id="__codelineno-0-514" name="__codelineno-0-514"></a><span class="sd">        :param parser: If the summary statistics file does not follow any of the formats above, you can create</span>
</span><span id="__span-0-515"><a id="__codelineno-0-515" name="__codelineno-0-515"></a><span class="sd">        your own parser by inheriting from the base `SumstatsParser` class and passing it here as an argument.</span>
</span><span id="__span-0-516"><a id="__codelineno-0-516" name="__codelineno-0-516"></a><span class="sd">        :param drop_duplicated: Drop SNPs with duplicated rsIDs.</span>
</span><span id="__span-0-517"><a id="__codelineno-0-517" name="__codelineno-0-517"></a><span class="sd">        :param parse_kwargs: keyword arguments for the parser. These are mainly parameters that will be passed to</span>
</span><span id="__span-0-518"><a id="__codelineno-0-518" name="__codelineno-0-518"></a><span class="sd">        `pandas.read_csv` function, such as the delimiter, header information, etc.</span>
</span><span id="__span-0-519"><a id="__codelineno-0-519" name="__codelineno-0-519"></a><span class="sd">        &quot;&quot;&quot;</span>
</span><span id="__span-0-520"><a id="__codelineno-0-520" name="__codelineno-0-520"></a>
</span><span id="__span-0-521"><a id="__codelineno-0-521" name="__codelineno-0-521"></a>        <span class="k">if</span> <span class="n">sumstats_path</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
</span><span id="__span-0-522"><a id="__codelineno-0-522" name="__codelineno-0-522"></a>            <span class="k">return</span>
</span><span id="__span-0-523"><a id="__codelineno-0-523" name="__codelineno-0-523"></a>
</span><span id="__span-0-524"><a id="__codelineno-0-524" name="__codelineno-0-524"></a>        <span class="k">if</span> <span class="ow">not</span> <span class="n">iterable</span><span class="p">(</span><span class="n">sumstats_path</span><span class="p">):</span>
</span><span id="__span-0-525"><a id="__codelineno-0-525" name="__codelineno-0-525"></a>            <span class="n">sumstats_files</span> <span class="o">=</span> <span class="n">get_filenames</span><span class="p">(</span><span class="n">sumstats_path</span><span class="p">)</span>
</span><span id="__span-0-526"><a id="__codelineno-0-526" name="__codelineno-0-526"></a>
</span><span id="__span-0-527"><a id="__codelineno-0-527" name="__codelineno-0-527"></a>            <span class="kn">from</span><span class="w"> </span><span class="nn">.utils.system_utils</span><span class="w"> </span><span class="kn">import</span> <span class="n">valid_url</span>
</span><span id="__span-0-528"><a id="__codelineno-0-528" name="__codelineno-0-528"></a>            <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">sumstats_files</span><span class="p">)</span> <span class="o">&lt;</span> <span class="mi">1</span> <span class="ow">and</span> <span class="n">valid_url</span><span class="p">(</span><span class="n">sumstats_path</span><span class="p">):</span>
</span><span id="__span-0-529"><a id="__codelineno-0-529" name="__codelineno-0-529"></a>                <span class="n">sumstats_files</span> <span class="o">=</span> <span class="p">[</span><span class="n">sumstats_path</span><span class="p">]</span>
</span><span id="__span-0-530"><a id="__codelineno-0-530" name="__codelineno-0-530"></a>        <span class="k">else</span><span class="p">:</span>
</span><span id="__span-0-531"><a id="__codelineno-0-531" name="__codelineno-0-531"></a>            <span class="n">sumstats_files</span> <span class="o">=</span> <span class="n">sumstats_path</span>
</span><span id="__span-0-532"><a id="__codelineno-0-532" name="__codelineno-0-532"></a>
</span><span id="__span-0-533"><a id="__codelineno-0-533" name="__codelineno-0-533"></a>        <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">sumstats_files</span><span class="p">)</span> <span class="o">&lt;</span> <span class="mi">1</span><span class="p">:</span>
</span><span id="__span-0-534"><a id="__codelineno-0-534" name="__codelineno-0-534"></a>            <span class="n">logger</span><span class="o">.</span><span class="n">warning</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;No summary statistics files were found at: </span><span class="si">{</span><span class="n">sumstats_path</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
</span><span id="__span-0-535"><a id="__codelineno-0-535" name="__codelineno-0-535"></a>            <span class="k">return</span>
</span><span id="__span-0-536"><a id="__codelineno-0-536" name="__codelineno-0-536"></a>
</span><span id="__span-0-537"><a id="__codelineno-0-537" name="__codelineno-0-537"></a>        <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;&gt; Reading summary statistics...&quot;</span><span class="p">)</span>
</span><span id="__span-0-538"><a id="__codelineno-0-538" name="__codelineno-0-538"></a>
</span><span id="__span-0-539"><a id="__codelineno-0-539" name="__codelineno-0-539"></a>        <span class="bp">self</span><span class="o">.</span><span class="n">sumstats_table</span> <span class="o">=</span> <span class="p">{}</span>
</span><span id="__span-0-540"><a id="__codelineno-0-540" name="__codelineno-0-540"></a>
</span><span id="__span-0-541"><a id="__codelineno-0-541" name="__codelineno-0-541"></a>        <span class="k">for</span> <span class="n">f</span> <span class="ow">in</span> <span class="n">tqdm</span><span class="p">(</span><span class="n">sumstats_files</span><span class="p">,</span>
</span><span id="__span-0-542"><a id="__codelineno-0-542" name="__codelineno-0-542"></a>                      <span class="n">total</span><span class="o">=</span><span class="nb">len</span><span class="p">(</span><span class="n">sumstats_files</span><span class="p">),</span>
</span><span id="__span-0-543"><a id="__codelineno-0-543" name="__codelineno-0-543"></a>                      <span class="n">desc</span><span class="o">=</span><span class="s2">&quot;Reading summary statistics&quot;</span><span class="p">):</span>
</span><span id="__span-0-544"><a id="__codelineno-0-544" name="__codelineno-0-544"></a>
</span><span id="__span-0-545"><a id="__codelineno-0-545" name="__codelineno-0-545"></a>            <span class="n">ss_tab</span> <span class="o">=</span> <span class="n">SumstatsTable</span><span class="o">.</span><span class="n">from_file</span><span class="p">(</span><span class="n">f</span><span class="p">,</span>
</span><span id="__span-0-546"><a id="__codelineno-0-546" name="__codelineno-0-546"></a>                                             <span class="n">sumstats_format</span><span class="o">=</span><span class="n">sumstats_format</span><span class="p">,</span>
</span><span id="__span-0-547"><a id="__codelineno-0-547" name="__codelineno-0-547"></a>                                             <span class="n">parser</span><span class="o">=</span><span class="n">parser</span><span class="p">,</span>
</span><span id="__span-0-548"><a id="__codelineno-0-548" name="__codelineno-0-548"></a>                                             <span class="o">**</span><span class="n">parse_kwargs</span><span class="p">)</span>
</span><span id="__span-0-549"><a id="__codelineno-0-549" name="__codelineno-0-549"></a>
</span><span id="__span-0-550"><a id="__codelineno-0-550" name="__codelineno-0-550"></a>            <span class="k">if</span> <span class="n">drop_duplicated</span><span class="p">:</span>
</span><span id="__span-0-551"><a id="__codelineno-0-551" name="__codelineno-0-551"></a>                <span class="n">ss_tab</span><span class="o">.</span><span class="n">drop_duplicates</span><span class="p">()</span>
</span><span id="__span-0-552"><a id="__codelineno-0-552" name="__codelineno-0-552"></a>
</span><span id="__span-0-553"><a id="__codelineno-0-553" name="__codelineno-0-553"></a>            <span class="k">if</span> <span class="s1">&#39;CHR&#39;</span> <span class="ow">in</span> <span class="n">ss_tab</span><span class="o">.</span><span class="n">table</span><span class="o">.</span><span class="n">columns</span><span class="p">:</span>
</span><span id="__span-0-554"><a id="__codelineno-0-554" name="__codelineno-0-554"></a>                <span class="bp">self</span><span class="o">.</span><span class="n">sumstats_table</span><span class="o">.</span><span class="n">update</span><span class="p">(</span><span class="n">ss_tab</span><span class="o">.</span><span class="n">split_by_chromosome</span><span class="p">())</span>
</span><span id="__span-0-555"><a id="__codelineno-0-555" name="__codelineno-0-555"></a>            <span class="k">else</span><span class="p">:</span>
</span><span id="__span-0-556"><a id="__codelineno-0-556" name="__codelineno-0-556"></a>                <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">genotype</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
</span><span id="__span-0-557"><a id="__codelineno-0-557" name="__codelineno-0-557"></a>                    <span class="n">ref_table</span> <span class="o">=</span> <span class="p">{</span><span class="n">c</span><span class="p">:</span> <span class="n">g</span><span class="o">.</span><span class="n">snps</span> <span class="k">for</span> <span class="n">c</span><span class="p">,</span> <span class="n">g</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">genotype</span><span class="o">.</span><span class="n">items</span><span class="p">()}</span>
</span><span id="__span-0-558"><a id="__codelineno-0-558" name="__codelineno-0-558"></a>                <span class="k">elif</span> <span class="bp">self</span><span class="o">.</span><span class="n">ld</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
</span><span id="__span-0-559"><a id="__codelineno-0-559" name="__codelineno-0-559"></a>                    <span class="n">ref_table</span> <span class="o">=</span> <span class="p">{</span><span class="n">c</span><span class="p">:</span> <span class="n">ld</span><span class="o">.</span><span class="n">snps</span> <span class="k">for</span> <span class="n">c</span><span class="p">,</span> <span class="n">ld</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">ld</span><span class="o">.</span><span class="n">items</span><span class="p">()}</span>
</span><span id="__span-0-560"><a id="__codelineno-0-560" name="__codelineno-0-560"></a>                <span class="k">else</span><span class="p">:</span>
</span><span id="__span-0-561"><a id="__codelineno-0-561" name="__codelineno-0-561"></a>                    <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;Cannot index summary statistics tables without chromosome information!&quot;</span><span class="p">)</span>
</span><span id="__span-0-562"><a id="__codelineno-0-562" name="__codelineno-0-562"></a>
</span><span id="__span-0-563"><a id="__codelineno-0-563" name="__codelineno-0-563"></a>                <span class="bp">self</span><span class="o">.</span><span class="n">sumstats_table</span><span class="o">.</span><span class="n">update</span><span class="p">(</span><span class="n">ss_tab</span><span class="o">.</span><span class="n">split_by_chromosome</span><span class="p">(</span><span class="n">snps_per_chrom</span><span class="o">=</span><span class="n">ref_table</span><span class="p">))</span>
</span><span id="__span-0-564"><a id="__codelineno-0-564" name="__codelineno-0-564"></a>
</span><span id="__span-0-565"><a id="__codelineno-0-565" name="__codelineno-0-565"></a>        <span class="c1"># If SNP information is not present in the sumstats tables, try to impute it</span>
</span><span id="__span-0-566"><a id="__codelineno-0-566" name="__codelineno-0-566"></a>        <span class="c1"># using other reference tables:</span>
</span><span id="__span-0-567"><a id="__codelineno-0-567" name="__codelineno-0-567"></a>
</span><span id="__span-0-568"><a id="__codelineno-0-568" name="__codelineno-0-568"></a>        <span class="n">missing_snp</span> <span class="o">=</span> <span class="nb">any</span><span class="p">(</span><span class="s1">&#39;SNP&#39;</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">ss</span><span class="o">.</span><span class="n">table</span><span class="o">.</span><span class="n">columns</span> <span class="k">for</span> <span class="n">ss</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">sumstats_table</span><span class="o">.</span><span class="n">values</span><span class="p">())</span>
</span><span id="__span-0-569"><a id="__codelineno-0-569" name="__codelineno-0-569"></a>
</span><span id="__span-0-570"><a id="__codelineno-0-570" name="__codelineno-0-570"></a>        <span class="k">if</span> <span class="n">missing_snp</span> <span class="ow">and</span> <span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">genotype</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span> <span class="ow">or</span> <span class="bp">self</span><span class="o">.</span><span class="n">ld</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">):</span>
</span><span id="__span-0-571"><a id="__codelineno-0-571" name="__codelineno-0-571"></a>
</span><span id="__span-0-572"><a id="__codelineno-0-572" name="__codelineno-0-572"></a>            <span class="n">ref_table</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">to_snp_table</span><span class="p">(</span><span class="n">col_subset</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;CHR&#39;</span><span class="p">,</span> <span class="s1">&#39;POS&#39;</span><span class="p">,</span> <span class="s1">&#39;SNP&#39;</span><span class="p">],</span> <span class="n">per_chromosome</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
</span><span id="__span-0-573"><a id="__codelineno-0-573" name="__codelineno-0-573"></a>
</span><span id="__span-0-574"><a id="__codelineno-0-574" name="__codelineno-0-574"></a>            <span class="k">for</span> <span class="n">c</span><span class="p">,</span> <span class="n">ss</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">sumstats_table</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
</span><span id="__span-0-575"><a id="__codelineno-0-575" name="__codelineno-0-575"></a>                <span class="k">if</span> <span class="s1">&#39;SNP&#39;</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">ss</span><span class="o">.</span><span class="n">table</span><span class="o">.</span><span class="n">columns</span> <span class="ow">and</span> <span class="n">c</span> <span class="ow">in</span> <span class="n">ref_table</span><span class="p">:</span>
</span><span id="__span-0-576"><a id="__codelineno-0-576" name="__codelineno-0-576"></a>                    <span class="n">ss</span><span class="o">.</span><span class="n">infer_snp_id</span><span class="p">(</span><span class="n">ref_table</span><span class="p">[</span><span class="n">c</span><span class="p">],</span> <span class="n">allow_na</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
</span><span id="__span-0-577"><a id="__codelineno-0-577" name="__codelineno-0-577"></a>
</span><span id="__span-0-578"><a id="__codelineno-0-578" name="__codelineno-0-578"></a>    <span class="k">def</span><span class="w"> </span><span class="nf">read_ld</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">ld_store_paths</span><span class="p">):</span>
</span><span id="__span-0-579"><a id="__codelineno-0-579" name="__codelineno-0-579"></a><span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
</span><span id="__span-0-580"><a id="__codelineno-0-580" name="__codelineno-0-580"></a><span class="sd">        Read the LD matrix files stored on-disk in Zarr array format.</span>
</span><span id="__span-0-581"><a id="__codelineno-0-581" name="__codelineno-0-581"></a><span class="sd">        :param ld_store_paths: The path to the LD matrices. This may be a wildcard to accommodate reading data</span>
</span><span id="__span-0-582"><a id="__codelineno-0-582" name="__codelineno-0-582"></a><span class="sd">        for multiple chromosomes.</span>
</span><span id="__span-0-583"><a id="__codelineno-0-583" name="__codelineno-0-583"></a><span class="sd">        &quot;&quot;&quot;</span>
</span><span id="__span-0-584"><a id="__codelineno-0-584" name="__codelineno-0-584"></a>
</span><span id="__span-0-585"><a id="__codelineno-0-585" name="__codelineno-0-585"></a>        <span class="k">if</span> <span class="n">ld_store_paths</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
</span><span id="__span-0-586"><a id="__codelineno-0-586" name="__codelineno-0-586"></a>            <span class="k">return</span>
</span><span id="__span-0-587"><a id="__codelineno-0-587" name="__codelineno-0-587"></a>
</span><span id="__span-0-588"><a id="__codelineno-0-588" name="__codelineno-0-588"></a>        <span class="k">if</span> <span class="ow">not</span> <span class="n">iterable</span><span class="p">(</span><span class="n">ld_store_paths</span><span class="p">):</span>
</span><span id="__span-0-589"><a id="__codelineno-0-589" name="__codelineno-0-589"></a>            <span class="k">if</span> <span class="s1">&#39;s3://&#39;</span> <span class="ow">in</span> <span class="n">ld_store_paths</span><span class="p">:</span>
</span><span id="__span-0-590"><a id="__codelineno-0-590" name="__codelineno-0-590"></a>                <span class="kn">from</span><span class="w"> </span><span class="nn">.utils.system_utils</span><span class="w"> </span><span class="kn">import</span> <span class="n">glob_s3_path</span>
</span><span id="__span-0-591"><a id="__codelineno-0-591" name="__codelineno-0-591"></a>                <span class="n">ld_store_files</span> <span class="o">=</span> <span class="n">glob_s3_path</span><span class="p">(</span><span class="n">ld_store_paths</span><span class="p">)</span>
</span><span id="__span-0-592"><a id="__codelineno-0-592" name="__codelineno-0-592"></a>            <span class="k">else</span><span class="p">:</span>
</span><span id="__span-0-593"><a id="__codelineno-0-593" name="__codelineno-0-593"></a>                <span class="n">ld_store_files</span> <span class="o">=</span> <span class="n">get_filenames</span><span class="p">(</span><span class="n">ld_store_paths</span><span class="p">,</span> <span class="n">extension</span><span class="o">=</span><span class="s1">&#39;.zgroup&#39;</span><span class="p">)</span>
</span><span id="__span-0-594"><a id="__codelineno-0-594" name="__codelineno-0-594"></a>        <span class="k">else</span><span class="p">:</span>
</span><span id="__span-0-595"><a id="__codelineno-0-595" name="__codelineno-0-595"></a>            <span class="n">ld_store_files</span> <span class="o">=</span> <span class="n">ld_store_paths</span>
</span><span id="__span-0-596"><a id="__codelineno-0-596" name="__codelineno-0-596"></a>
</span><span id="__span-0-597"><a id="__codelineno-0-597" name="__codelineno-0-597"></a>        <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">ld_store_files</span><span class="p">)</span> <span class="o">&lt;</span> <span class="mi">1</span><span class="p">:</span>
</span><span id="__span-0-598"><a id="__codelineno-0-598" name="__codelineno-0-598"></a>            <span class="n">logger</span><span class="o">.</span><span class="n">warning</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;No LD matrix files were found at: </span><span class="si">{</span><span class="n">ld_store_paths</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
</span><span id="__span-0-599"><a id="__codelineno-0-599" name="__codelineno-0-599"></a>            <span class="k">return</span>
</span><span id="__span-0-600"><a id="__codelineno-0-600" name="__codelineno-0-600"></a>
</span><span id="__span-0-601"><a id="__codelineno-0-601" name="__codelineno-0-601"></a>        <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;&gt; Reading LD metadata...&quot;</span><span class="p">)</span>
</span><span id="__span-0-602"><a id="__codelineno-0-602" name="__codelineno-0-602"></a>
</span><span id="__span-0-603"><a id="__codelineno-0-603" name="__codelineno-0-603"></a>        <span class="bp">self</span><span class="o">.</span><span class="n">ld</span> <span class="o">=</span> <span class="p">{}</span>
</span><span id="__span-0-604"><a id="__codelineno-0-604" name="__codelineno-0-604"></a>
</span><span id="__span-0-605"><a id="__codelineno-0-605" name="__codelineno-0-605"></a>        <span class="k">for</span> <span class="n">f</span> <span class="ow">in</span> <span class="n">tqdm</span><span class="p">(</span><span class="n">ld_store_files</span><span class="p">,</span>
</span><span id="__span-0-606"><a id="__codelineno-0-606" name="__codelineno-0-606"></a>                      <span class="n">total</span><span class="o">=</span><span class="nb">len</span><span class="p">(</span><span class="n">ld_store_files</span><span class="p">),</span>
</span><span id="__span-0-607"><a id="__codelineno-0-607" name="__codelineno-0-607"></a>                      <span class="n">desc</span><span class="o">=</span><span class="s2">&quot;Reading LD metadata&quot;</span><span class="p">):</span>
</span><span id="__span-0-608"><a id="__codelineno-0-608" name="__codelineno-0-608"></a>            <span class="n">z</span> <span class="o">=</span> <span class="n">LDMatrix</span><span class="o">.</span><span class="n">from_path</span><span class="p">(</span><span class="n">f</span><span class="p">)</span>
</span><span id="__span-0-609"><a id="__codelineno-0-609" name="__codelineno-0-609"></a>            <span class="bp">self</span><span class="o">.</span><span class="n">ld</span><span class="p">[</span><span class="n">z</span><span class="o">.</span><span class="n">chromosome</span><span class="p">]</span> <span class="o">=</span> <span class="n">z</span>
</span><span id="__span-0-610"><a id="__codelineno-0-610" name="__codelineno-0-610"></a>
</span><span id="__span-0-611"><a id="__codelineno-0-611" name="__codelineno-0-611"></a>    <span class="k">def</span><span class="w"> </span><span class="nf">load_ld</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
</span><span id="__span-0-612"><a id="__codelineno-0-612" name="__codelineno-0-612"></a><span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
</span><span id="__span-0-613"><a id="__codelineno-0-613" name="__codelineno-0-613"></a><span class="sd">        A utility method to load the LD matrices to memory from on-disk storage.</span>
</span><span id="__span-0-614"><a id="__codelineno-0-614" name="__codelineno-0-614"></a><span class="sd">        &quot;&quot;&quot;</span>
</span><span id="__span-0-615"><a id="__codelineno-0-615" name="__codelineno-0-615"></a>        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">ld</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
</span><span id="__span-0-616"><a id="__codelineno-0-616" name="__codelineno-0-616"></a>            <span class="k">for</span> <span class="n">ld</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">ld</span><span class="o">.</span><span class="n">values</span><span class="p">():</span>
</span><span id="__span-0-617"><a id="__codelineno-0-617" name="__codelineno-0-617"></a>                <span class="n">ld</span><span class="o">.</span><span class="n">load</span><span class="p">()</span>
</span><span id="__span-0-618"><a id="__codelineno-0-618" name="__codelineno-0-618"></a>
</span><span id="__span-0-619"><a id="__codelineno-0-619" name="__codelineno-0-619"></a>    <span class="k">def</span><span class="w"> </span><span class="nf">release_ld</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
</span><span id="__span-0-620"><a id="__codelineno-0-620" name="__codelineno-0-620"></a><span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
</span><span id="__span-0-621"><a id="__codelineno-0-621" name="__codelineno-0-621"></a><span class="sd">        A utility function to release the LD matrices from memory.</span>
</span><span id="__span-0-622"><a id="__codelineno-0-622" name="__codelineno-0-622"></a><span class="sd">        &quot;&quot;&quot;</span>
</span><span id="__span-0-623"><a id="__codelineno-0-623" name="__codelineno-0-623"></a>        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">ld</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
</span><span id="__span-0-624"><a id="__codelineno-0-624" name="__codelineno-0-624"></a>            <span class="k">for</span> <span class="n">ld</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">ld</span><span class="o">.</span><span class="n">values</span><span class="p">():</span>
</span><span id="__span-0-625"><a id="__codelineno-0-625" name="__codelineno-0-625"></a>                <span class="n">ld</span><span class="o">.</span><span class="n">release</span><span class="p">()</span>
</span><span id="__span-0-626"><a id="__codelineno-0-626" name="__codelineno-0-626"></a>
</span><span id="__span-0-627"><a id="__codelineno-0-627" name="__codelineno-0-627"></a>    <span class="k">def</span><span class="w"> </span><span class="nf">compute_ld</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span>
</span><span id="__span-0-628"><a id="__codelineno-0-628" name="__codelineno-0-628"></a>                   <span class="n">estimator</span><span class="p">,</span>
</span><span id="__span-0-629"><a id="__codelineno-0-629" name="__codelineno-0-629"></a>                   <span class="n">output_dir</span><span class="p">,</span>
</span><span id="__span-0-630"><a id="__codelineno-0-630" name="__codelineno-0-630"></a>                   <span class="n">dtype</span><span class="o">=</span><span class="s1">&#39;int16&#39;</span><span class="p">,</span>
</span><span id="__span-0-631"><a id="__codelineno-0-631" name="__codelineno-0-631"></a>                   <span class="n">compressor_name</span><span class="o">=</span><span class="s1">&#39;zstd&#39;</span><span class="p">,</span>
</span><span id="__span-0-632"><a id="__codelineno-0-632" name="__codelineno-0-632"></a>                   <span class="n">compression_level</span><span class="o">=</span><span class="mi">7</span><span class="p">,</span>
</span><span id="__span-0-633"><a id="__codelineno-0-633" name="__codelineno-0-633"></a>                   <span class="n">compute_spectral_properties</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span>
</span><span id="__span-0-634"><a id="__codelineno-0-634" name="__codelineno-0-634"></a>                   <span class="o">**</span><span class="n">ld_kwargs</span><span class="p">):</span>
</span><span id="__span-0-635"><a id="__codelineno-0-635" name="__codelineno-0-635"></a><span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
</span><span id="__span-0-636"><a id="__codelineno-0-636" name="__codelineno-0-636"></a><span class="sd">        Compute the Linkage-Disequilibrium (LD) matrix or SNP-by-SNP Pearson</span>
</span><span id="__span-0-637"><a id="__codelineno-0-637" name="__codelineno-0-637"></a><span class="sd">        correlation matrix between genetic variants. This function only considers correlations</span>
</span><span id="__span-0-638"><a id="__codelineno-0-638" name="__codelineno-0-638"></a><span class="sd">        between SNPs on the same chromosome. This is a utility function that calls the</span>
</span><span id="__span-0-639"><a id="__codelineno-0-639" name="__codelineno-0-639"></a><span class="sd">        `.compute_ld()` method of the `GenotypeMatrix` objects associated with</span>
</span><span id="__span-0-640"><a id="__codelineno-0-640" name="__codelineno-0-640"></a><span class="sd">        GWADataLoader.</span>
</span><span id="__span-0-641"><a id="__codelineno-0-641" name="__codelineno-0-641"></a>
</span><span id="__span-0-642"><a id="__codelineno-0-642" name="__codelineno-0-642"></a><span class="sd">        :param estimator: The estimator for the LD matrix. We currently support</span>
</span><span id="__span-0-643"><a id="__codelineno-0-643" name="__codelineno-0-643"></a><span class="sd">        4 different estimators: `sample`, `windowed`, `shrinkage`, and `block`.</span>
</span><span id="__span-0-644"><a id="__codelineno-0-644" name="__codelineno-0-644"></a><span class="sd">        :param output_dir: The output directory where the Zarr array containing the</span>
</span><span id="__span-0-645"><a id="__codelineno-0-645" name="__codelineno-0-645"></a><span class="sd">        entries of the LD matrix will be stored.</span>
</span><span id="__span-0-646"><a id="__codelineno-0-646" name="__codelineno-0-646"></a><span class="sd">        :param dtype: The data type for the entries of the LD matrix (supported data types are float32, float64</span>
</span><span id="__span-0-647"><a id="__codelineno-0-647" name="__codelineno-0-647"></a><span class="sd">        and integer quantized data types int8 and int16).</span>
</span><span id="__span-0-648"><a id="__codelineno-0-648" name="__codelineno-0-648"></a><span class="sd">        :param compressor_name: The name of the compression algorithm to use for the LD matrix.</span>
</span><span id="__span-0-649"><a id="__codelineno-0-649" name="__codelineno-0-649"></a><span class="sd">        :param compression_level: The compression level to use for the entries of the LD matrix (1-9).</span>
</span><span id="__span-0-650"><a id="__codelineno-0-650" name="__codelineno-0-650"></a><span class="sd">        :param compute_spectral_properties: If True, compute the spectral properties of the LD matrix.</span>
</span><span id="__span-0-651"><a id="__codelineno-0-651" name="__codelineno-0-651"></a><span class="sd">        :param ld_kwargs: keyword arguments for the various LD estimators. Consult</span>
</span><span id="__span-0-652"><a id="__codelineno-0-652" name="__codelineno-0-652"></a><span class="sd">        the implementations of `WindowedLD`, `ShrinkageLD`, and `BlockLD` for details.</span>
</span><span id="__span-0-653"><a id="__codelineno-0-653" name="__codelineno-0-653"></a><span class="sd">        &quot;&quot;&quot;</span>
</span><span id="__span-0-654"><a id="__codelineno-0-654" name="__codelineno-0-654"></a>
</span><span id="__span-0-655"><a id="__codelineno-0-655" name="__codelineno-0-655"></a>        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">genotype</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
</span><span id="__span-0-656"><a id="__codelineno-0-656" name="__codelineno-0-656"></a>            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;Cannot compute LD without genotype data.&quot;</span><span class="p">)</span>
</span><span id="__span-0-657"><a id="__codelineno-0-657" name="__codelineno-0-657"></a>
</span><span id="__span-0-658"><a id="__codelineno-0-658" name="__codelineno-0-658"></a>        <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;&gt; Computing LD matrix...&quot;</span><span class="p">)</span>
</span><span id="__span-0-659"><a id="__codelineno-0-659" name="__codelineno-0-659"></a>
</span><span id="__span-0-660"><a id="__codelineno-0-660" name="__codelineno-0-660"></a>        <span class="bp">self</span><span class="o">.</span><span class="n">ld</span> <span class="o">=</span> <span class="p">{</span>
</span><span id="__span-0-661"><a id="__codelineno-0-661" name="__codelineno-0-661"></a>            <span class="n">c</span><span class="p">:</span> <span class="n">g</span><span class="o">.</span><span class="n">compute_ld</span><span class="p">(</span><span class="n">estimator</span><span class="p">,</span>
</span><span id="__span-0-662"><a id="__codelineno-0-662" name="__codelineno-0-662"></a>                            <span class="n">output_dir</span><span class="p">,</span>
</span><span id="__span-0-663"><a id="__codelineno-0-663" name="__codelineno-0-663"></a>                            <span class="n">dtype</span><span class="o">=</span><span class="n">dtype</span><span class="p">,</span>
</span><span id="__span-0-664"><a id="__codelineno-0-664" name="__codelineno-0-664"></a>                            <span class="n">compressor_name</span><span class="o">=</span><span class="n">compressor_name</span><span class="p">,</span>
</span><span id="__span-0-665"><a id="__codelineno-0-665" name="__codelineno-0-665"></a>                            <span class="n">compression_level</span><span class="o">=</span><span class="n">compression_level</span><span class="p">,</span>
</span><span id="__span-0-666"><a id="__codelineno-0-666" name="__codelineno-0-666"></a>                            <span class="n">compute_spectral_properties</span><span class="o">=</span><span class="n">compute_spectral_properties</span><span class="p">,</span>
</span><span id="__span-0-667"><a id="__codelineno-0-667" name="__codelineno-0-667"></a>                            <span class="o">**</span><span class="n">ld_kwargs</span><span class="p">)</span>
</span><span id="__span-0-668"><a id="__codelineno-0-668" name="__codelineno-0-668"></a>            <span class="k">for</span> <span class="n">c</span><span class="p">,</span> <span class="n">g</span> <span class="ow">in</span> <span class="n">tqdm</span><span class="p">(</span><span class="nb">sorted</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">genotype</span><span class="o">.</span><span class="n">items</span><span class="p">(),</span> <span class="n">key</span><span class="o">=</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">x</span><span class="p">[</span><span class="mi">0</span><span class="p">]),</span>
</span><span id="__span-0-669"><a id="__codelineno-0-669" name="__codelineno-0-669"></a>                             <span class="n">total</span><span class="o">=</span><span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">genotype</span><span class="p">),</span>
</span><span id="__span-0-670"><a id="__codelineno-0-670" name="__codelineno-0-670"></a>                             <span class="n">desc</span><span class="o">=</span><span class="s1">&#39;Computing LD matrices&#39;</span><span class="p">)</span>
</span><span id="__span-0-671"><a id="__codelineno-0-671" name="__codelineno-0-671"></a>        <span class="p">}</span>
</span><span id="__span-0-672"><a id="__codelineno-0-672" name="__codelineno-0-672"></a>
</span><span id="__span-0-673"><a id="__codelineno-0-673" name="__codelineno-0-673"></a>    <span class="k">def</span><span class="w"> </span><span class="nf">get_ld_matrices</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
</span><span id="__span-0-674"><a id="__codelineno-0-674" name="__codelineno-0-674"></a><span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
</span><span id="__span-0-675"><a id="__codelineno-0-675" name="__codelineno-0-675"></a><span class="sd">        :return: A dictionary containing the chromosome ID as key and corresponding LD matrices</span>
</span><span id="__span-0-676"><a id="__codelineno-0-676" name="__codelineno-0-676"></a><span class="sd">        as value.</span>
</span><span id="__span-0-677"><a id="__codelineno-0-677" name="__codelineno-0-677"></a><span class="sd">        &quot;&quot;&quot;</span>
</span><span id="__span-0-678"><a id="__codelineno-0-678" name="__codelineno-0-678"></a>        <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">ld</span>
</span><span id="__span-0-679"><a id="__codelineno-0-679" name="__codelineno-0-679"></a>
</span><span id="__span-0-680"><a id="__codelineno-0-680" name="__codelineno-0-680"></a>    <span class="k">def</span><span class="w"> </span><span class="nf">harmonize_data</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
</span><span id="__span-0-681"><a id="__codelineno-0-681" name="__codelineno-0-681"></a><span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
</span><span id="__span-0-682"><a id="__codelineno-0-682" name="__codelineno-0-682"></a><span class="sd">        This method ensures that the data sources (reference genotype,</span>
</span><span id="__span-0-683"><a id="__codelineno-0-683" name="__codelineno-0-683"></a><span class="sd">        LD matrices, summary statistics, annotations) are all aligned in terms of the</span>
</span><span id="__span-0-684"><a id="__codelineno-0-684" name="__codelineno-0-684"></a><span class="sd">        set of variants that they operate on as well as the designation of the effect allele for</span>
</span><span id="__span-0-685"><a id="__codelineno-0-685" name="__codelineno-0-685"></a><span class="sd">        each variant.</span>
</span><span id="__span-0-686"><a id="__codelineno-0-686" name="__codelineno-0-686"></a>
</span><span id="__span-0-687"><a id="__codelineno-0-687" name="__codelineno-0-687"></a><span class="sd">        !!! note</span>
</span><span id="__span-0-688"><a id="__codelineno-0-688" name="__codelineno-0-688"></a><span class="sd">            This method is called automatically during the initialization of the `GWADataLoader` object.</span>
</span><span id="__span-0-689"><a id="__codelineno-0-689" name="__codelineno-0-689"></a><span class="sd">            However, if you read or manipulate the data sources after initialization,</span>
</span><span id="__span-0-690"><a id="__codelineno-0-690" name="__codelineno-0-690"></a><span class="sd">            you may need to call this method again to ensure that the data sources remain aligned.</span>
</span><span id="__span-0-691"><a id="__codelineno-0-691" name="__codelineno-0-691"></a>
</span><span id="__span-0-692"><a id="__codelineno-0-692" name="__codelineno-0-692"></a><span class="sd">        !!! warning</span>
</span><span id="__span-0-693"><a id="__codelineno-0-693" name="__codelineno-0-693"></a><span class="sd">            Harmonization for now depends on having SNP rsID be present in all the resources. Hopefully</span>
</span><span id="__span-0-694"><a id="__codelineno-0-694" name="__codelineno-0-694"></a><span class="sd">            this requirement will be relaxed in the future.</span>
</span><span id="__span-0-695"><a id="__codelineno-0-695" name="__codelineno-0-695"></a>
</span><span id="__span-0-696"><a id="__codelineno-0-696" name="__codelineno-0-696"></a><span class="sd">        &quot;&quot;&quot;</span>
</span><span id="__span-0-697"><a id="__codelineno-0-697" name="__codelineno-0-697"></a>
</span><span id="__span-0-698"><a id="__codelineno-0-698" name="__codelineno-0-698"></a>        <span class="n">data_sources</span> <span class="o">=</span> <span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">genotype</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">sumstats_table</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">ld</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">annotation</span><span class="p">)</span>
</span><span id="__span-0-699"><a id="__codelineno-0-699" name="__codelineno-0-699"></a>        <span class="n">initialized_data_sources</span> <span class="o">=</span> <span class="p">[</span><span class="n">ds</span> <span class="k">for</span> <span class="n">ds</span> <span class="ow">in</span> <span class="n">data_sources</span> <span class="k">if</span> <span class="n">ds</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">]</span>
</span><span id="__span-0-700"><a id="__codelineno-0-700" name="__codelineno-0-700"></a>
</span><span id="__span-0-701"><a id="__codelineno-0-701" name="__codelineno-0-701"></a>        <span class="c1"># If less than two data sources are present, skip harmonization...</span>
</span><span id="__span-0-702"><a id="__codelineno-0-702" name="__codelineno-0-702"></a>        <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">initialized_data_sources</span><span class="p">)</span> <span class="o">&lt;</span> <span class="mi">2</span><span class="p">:</span>
</span><span id="__span-0-703"><a id="__codelineno-0-703" name="__codelineno-0-703"></a>            <span class="k">return</span>
</span><span id="__span-0-704"><a id="__codelineno-0-704" name="__codelineno-0-704"></a>
</span><span id="__span-0-705"><a id="__codelineno-0-705" name="__codelineno-0-705"></a>        <span class="c1"># Get the chromosomes information from all the data sources:</span>
</span><span id="__span-0-706"><a id="__codelineno-0-706" name="__codelineno-0-706"></a>        <span class="n">chromosomes</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="nb">set</span><span class="o">.</span><span class="n">union</span><span class="p">(</span><span class="o">*</span><span class="p">[</span><span class="nb">set</span><span class="p">(</span><span class="n">ds</span><span class="o">.</span><span class="n">keys</span><span class="p">())</span> <span class="k">for</span> <span class="n">ds</span> <span class="ow">in</span> <span class="n">initialized_data_sources</span><span class="p">]))</span>
</span><span id="__span-0-707"><a id="__codelineno-0-707" name="__codelineno-0-707"></a>
</span><span id="__span-0-708"><a id="__codelineno-0-708" name="__codelineno-0-708"></a>        <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;&gt; Harmonizing data...&quot;</span><span class="p">)</span>
</span><span id="__span-0-709"><a id="__codelineno-0-709" name="__codelineno-0-709"></a>
</span><span id="__span-0-710"><a id="__codelineno-0-710" name="__codelineno-0-710"></a>        <span class="k">for</span> <span class="n">c</span> <span class="ow">in</span> <span class="n">tqdm</span><span class="p">(</span><span class="n">chromosomes</span><span class="p">,</span>
</span><span id="__span-0-711"><a id="__codelineno-0-711" name="__codelineno-0-711"></a>                      <span class="n">total</span><span class="o">=</span><span class="nb">len</span><span class="p">(</span><span class="n">chromosomes</span><span class="p">),</span>
</span><span id="__span-0-712"><a id="__codelineno-0-712" name="__codelineno-0-712"></a>                      <span class="n">desc</span><span class="o">=</span><span class="s1">&#39;Harmonizing data&#39;</span><span class="p">):</span>
</span><span id="__span-0-713"><a id="__codelineno-0-713" name="__codelineno-0-713"></a>
</span><span id="__span-0-714"><a id="__codelineno-0-714" name="__codelineno-0-714"></a>            <span class="c1"># Which initialized data sources have information for chromosome `c`</span>
</span><span id="__span-0-715"><a id="__codelineno-0-715" name="__codelineno-0-715"></a>            <span class="n">miss_chroms</span> <span class="o">=</span> <span class="p">[</span><span class="n">c</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">ds</span> <span class="k">for</span> <span class="n">ds</span> <span class="ow">in</span> <span class="n">initialized_data_sources</span><span class="p">]</span>
</span><span id="__span-0-716"><a id="__codelineno-0-716" name="__codelineno-0-716"></a>
</span><span id="__span-0-717"><a id="__codelineno-0-717" name="__codelineno-0-717"></a>            <span class="k">if</span> <span class="nb">sum</span><span class="p">(</span><span class="n">miss_chroms</span><span class="p">)</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">:</span>
</span><span id="__span-0-718"><a id="__codelineno-0-718" name="__codelineno-0-718"></a>                <span class="c1"># If the chromosome data only exists for some data sources but not others, remove the chromosome</span>
</span><span id="__span-0-719"><a id="__codelineno-0-719" name="__codelineno-0-719"></a>                <span class="c1"># from all data source.</span>
</span><span id="__span-0-720"><a id="__codelineno-0-720" name="__codelineno-0-720"></a>                <span class="c1"># Is this the best way to handle the missingness? Should we just post a warning?</span>
</span><span id="__span-0-721"><a id="__codelineno-0-721" name="__codelineno-0-721"></a>                <span class="n">logger</span><span class="o">.</span><span class="n">debug</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Chromosome </span><span class="si">{</span><span class="n">c</span><span class="si">}</span><span class="s2"> is missing in some data sources. &quot;</span>
</span><span id="__span-0-722"><a id="__codelineno-0-722" name="__codelineno-0-722"></a>                             <span class="sa">f</span><span class="s2">&quot;Removing it from all data sources.&quot;</span><span class="p">)</span>
</span><span id="__span-0-723"><a id="__codelineno-0-723" name="__codelineno-0-723"></a>                <span class="k">for</span> <span class="n">ds</span> <span class="ow">in</span> <span class="n">initialized_data_sources</span><span class="p">:</span>
</span><span id="__span-0-724"><a id="__codelineno-0-724" name="__codelineno-0-724"></a>                    <span class="k">if</span> <span class="n">c</span> <span class="ow">in</span> <span class="n">ds</span><span class="p">:</span>
</span><span id="__span-0-725"><a id="__codelineno-0-725" name="__codelineno-0-725"></a>                        <span class="k">del</span> <span class="n">ds</span><span class="p">[</span><span class="n">c</span><span class="p">]</span>
</span><span id="__span-0-726"><a id="__codelineno-0-726" name="__codelineno-0-726"></a>
</span><span id="__span-0-727"><a id="__codelineno-0-727" name="__codelineno-0-727"></a>            <span class="k">else</span><span class="p">:</span>
</span><span id="__span-0-728"><a id="__codelineno-0-728" name="__codelineno-0-728"></a>
</span><span id="__span-0-729"><a id="__codelineno-0-729" name="__codelineno-0-729"></a>                <span class="c1"># Find the set of SNPs that are shared across all data sources (exclude missing values):</span>
</span><span id="__span-0-730"><a id="__codelineno-0-730" name="__codelineno-0-730"></a>                <span class="n">common_snps</span> <span class="o">=</span> <span class="n">intersect_multiple_arrays</span><span class="p">([</span><span class="n">ds</span><span class="p">[</span><span class="n">c</span><span class="p">]</span><span class="o">.</span><span class="n">snps</span> <span class="k">for</span> <span class="n">ds</span> <span class="ow">in</span> <span class="n">initialized_data_sources</span><span class="p">])</span>
</span><span id="__span-0-731"><a id="__codelineno-0-731" name="__codelineno-0-731"></a>
</span><span id="__span-0-732"><a id="__codelineno-0-732" name="__codelineno-0-732"></a>                <span class="c1"># If necessary, filter the data sources to only have the common SNPs:</span>
</span><span id="__span-0-733"><a id="__codelineno-0-733" name="__codelineno-0-733"></a>                <span class="k">for</span> <span class="n">ds</span> <span class="ow">in</span> <span class="n">initialized_data_sources</span><span class="p">:</span>
</span><span id="__span-0-734"><a id="__codelineno-0-734" name="__codelineno-0-734"></a>                    <span class="k">if</span> <span class="n">ds</span><span class="p">[</span><span class="n">c</span><span class="p">]</span><span class="o">.</span><span class="n">n_snps</span> <span class="o">!=</span> <span class="nb">len</span><span class="p">(</span><span class="n">common_snps</span><span class="p">):</span>
</span><span id="__span-0-735"><a id="__codelineno-0-735" name="__codelineno-0-735"></a>                        <span class="n">ds</span><span class="p">[</span><span class="n">c</span><span class="p">]</span><span class="o">.</span><span class="n">filter_snps</span><span class="p">(</span><span class="n">extract_snps</span><span class="o">=</span><span class="n">common_snps</span><span class="p">)</span>
</span><span id="__span-0-736"><a id="__codelineno-0-736" name="__codelineno-0-736"></a>
</span><span id="__span-0-737"><a id="__codelineno-0-737" name="__codelineno-0-737"></a>                <span class="c1"># Harmonize the summary statistics data with either genotype or LD reference.</span>
</span><span id="__span-0-738"><a id="__codelineno-0-738" name="__codelineno-0-738"></a>                <span class="c1"># This procedure checks for flips in the effect allele between data sources.</span>
</span><span id="__span-0-739"><a id="__codelineno-0-739" name="__codelineno-0-739"></a>                <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">sumstats_table</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
</span><span id="__span-0-740"><a id="__codelineno-0-740" name="__codelineno-0-740"></a>
</span><span id="__span-0-741"><a id="__codelineno-0-741" name="__codelineno-0-741"></a>                    <span class="n">id_cols</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">sumstats_table</span><span class="p">[</span><span class="n">c</span><span class="p">]</span><span class="o">.</span><span class="n">identifier_cols</span>
</span><span id="__span-0-742"><a id="__codelineno-0-742" name="__codelineno-0-742"></a>
</span><span id="__span-0-743"><a id="__codelineno-0-743" name="__codelineno-0-743"></a>                    <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">genotype</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
</span><span id="__span-0-744"><a id="__codelineno-0-744" name="__codelineno-0-744"></a>                        <span class="bp">self</span><span class="o">.</span><span class="n">sumstats_table</span><span class="p">[</span><span class="n">c</span><span class="p">]</span><span class="o">.</span><span class="n">match</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">genotype</span><span class="p">[</span><span class="n">c</span><span class="p">]</span><span class="o">.</span><span class="n">get_snp_table</span><span class="p">(</span>
</span><span id="__span-0-745"><a id="__codelineno-0-745" name="__codelineno-0-745"></a>                            <span class="n">col_subset</span><span class="o">=</span><span class="n">id_cols</span> <span class="o">+</span> <span class="p">[</span><span class="s1">&#39;A1&#39;</span><span class="p">,</span> <span class="s1">&#39;A2&#39;</span><span class="p">]</span>
</span><span id="__span-0-746"><a id="__codelineno-0-746" name="__codelineno-0-746"></a>                        <span class="p">))</span>
</span><span id="__span-0-747"><a id="__codelineno-0-747" name="__codelineno-0-747"></a>                    <span class="k">elif</span> <span class="bp">self</span><span class="o">.</span><span class="n">ld</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
</span><span id="__span-0-748"><a id="__codelineno-0-748" name="__codelineno-0-748"></a>                        <span class="bp">self</span><span class="o">.</span><span class="n">sumstats_table</span><span class="p">[</span><span class="n">c</span><span class="p">]</span><span class="o">.</span><span class="n">match</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">ld</span><span class="p">[</span><span class="n">c</span><span class="p">]</span><span class="o">.</span><span class="n">to_snp_table</span><span class="p">(</span>
</span><span id="__span-0-749"><a id="__codelineno-0-749" name="__codelineno-0-749"></a>                            <span class="n">col_subset</span><span class="o">=</span><span class="n">id_cols</span> <span class="o">+</span> <span class="p">[</span><span class="s1">&#39;A1&#39;</span><span class="p">,</span> <span class="s1">&#39;A2&#39;</span><span class="p">]</span>
</span><span id="__span-0-750"><a id="__codelineno-0-750" name="__codelineno-0-750"></a>                        <span class="p">))</span>
</span><span id="__span-0-751"><a id="__codelineno-0-751" name="__codelineno-0-751"></a>
</span><span id="__span-0-752"><a id="__codelineno-0-752" name="__codelineno-0-752"></a>                    <span class="c1"># If during the allele matching process we discover incompatibilities,</span>
</span><span id="__span-0-753"><a id="__codelineno-0-753" name="__codelineno-0-753"></a>                    <span class="c1"># we filter those SNPs:</span>
</span><span id="__span-0-754"><a id="__codelineno-0-754" name="__codelineno-0-754"></a>                    <span class="k">for</span> <span class="n">ds</span> <span class="ow">in</span> <span class="n">initialized_data_sources</span><span class="p">:</span>
</span><span id="__span-0-755"><a id="__codelineno-0-755" name="__codelineno-0-755"></a>                        <span class="k">if</span> <span class="n">ds</span><span class="p">[</span><span class="n">c</span><span class="p">]</span><span class="o">.</span><span class="n">n_snps</span> <span class="o">!=</span> <span class="bp">self</span><span class="o">.</span><span class="n">sumstats_table</span><span class="p">[</span><span class="n">c</span><span class="p">]</span><span class="o">.</span><span class="n">n_snps</span><span class="p">:</span>
</span><span id="__span-0-756"><a id="__codelineno-0-756" name="__codelineno-0-756"></a>                            <span class="n">ds</span><span class="p">[</span><span class="n">c</span><span class="p">]</span><span class="o">.</span><span class="n">filter_snps</span><span class="p">(</span><span class="n">extract_snps</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">sumstats_table</span><span class="p">[</span><span class="n">c</span><span class="p">]</span><span class="o">.</span><span class="n">snps</span><span class="p">)</span>
</span><span id="__span-0-757"><a id="__codelineno-0-757" name="__codelineno-0-757"></a>
</span><span id="__span-0-758"><a id="__codelineno-0-758" name="__codelineno-0-758"></a>    <span class="k">def</span><span class="w"> </span><span class="nf">perform_gwas</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="o">**</span><span class="n">gwa_kwargs</span><span class="p">):</span>
</span><span id="__span-0-759"><a id="__codelineno-0-759" name="__codelineno-0-759"></a><span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
</span><span id="__span-0-760"><a id="__codelineno-0-760" name="__codelineno-0-760"></a><span class="sd">        Perform genome-wide association testing of all variants against the phenotype.</span>
</span><span id="__span-0-761"><a id="__codelineno-0-761" name="__codelineno-0-761"></a><span class="sd">        This is a utility function that calls the `.perform_gwas()` method of the</span>
</span><span id="__span-0-762"><a id="__codelineno-0-762" name="__codelineno-0-762"></a><span class="sd">        `GenotypeMatrix` objects associated with GWADataLoader.</span>
</span><span id="__span-0-763"><a id="__codelineno-0-763" name="__codelineno-0-763"></a>
</span><span id="__span-0-764"><a id="__codelineno-0-764" name="__codelineno-0-764"></a><span class="sd">        :param gwa_kwargs: Keyword arguments to pass to the GWA functions. Consult stats.gwa.utils</span>
</span><span id="__span-0-765"><a id="__codelineno-0-765" name="__codelineno-0-765"></a><span class="sd">        for relevant keyword arguments for each backend.</span>
</span><span id="__span-0-766"><a id="__codelineno-0-766" name="__codelineno-0-766"></a><span class="sd">        &quot;&quot;&quot;</span>
</span><span id="__span-0-767"><a id="__codelineno-0-767" name="__codelineno-0-767"></a>
</span><span id="__span-0-768"><a id="__codelineno-0-768" name="__codelineno-0-768"></a>        <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;&gt; Performing GWAS...&quot;</span><span class="p">)</span>
</span><span id="__span-0-769"><a id="__codelineno-0-769" name="__codelineno-0-769"></a>
</span><span id="__span-0-770"><a id="__codelineno-0-770" name="__codelineno-0-770"></a>        <span class="bp">self</span><span class="o">.</span><span class="n">sumstats_table</span> <span class="o">=</span> <span class="p">{</span>
</span><span id="__span-0-771"><a id="__codelineno-0-771" name="__codelineno-0-771"></a>            <span class="n">c</span><span class="p">:</span> <span class="n">g</span><span class="o">.</span><span class="n">perform_gwas</span><span class="p">(</span><span class="o">**</span><span class="n">gwa_kwargs</span><span class="p">)</span>
</span><span id="__span-0-772"><a id="__codelineno-0-772" name="__codelineno-0-772"></a>            <span class="k">for</span> <span class="n">c</span><span class="p">,</span> <span class="n">g</span> <span class="ow">in</span> <span class="n">tqdm</span><span class="p">(</span><span class="nb">sorted</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">genotype</span><span class="o">.</span><span class="n">items</span><span class="p">(),</span> <span class="n">key</span><span class="o">=</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">x</span><span class="p">[</span><span class="mi">0</span><span class="p">]),</span>
</span><span id="__span-0-773"><a id="__codelineno-0-773" name="__codelineno-0-773"></a>                             <span class="n">total</span><span class="o">=</span><span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">genotype</span><span class="p">),</span>
</span><span id="__span-0-774"><a id="__codelineno-0-774" name="__codelineno-0-774"></a>                             <span class="n">desc</span><span class="o">=</span><span class="s1">&#39;Performing GWAS&#39;</span><span class="p">)</span>
</span><span id="__span-0-775"><a id="__codelineno-0-775" name="__codelineno-0-775"></a>        <span class="p">}</span>
</span><span id="__span-0-776"><a id="__codelineno-0-776" name="__codelineno-0-776"></a>
</span><span id="__span-0-777"><a id="__codelineno-0-777" name="__codelineno-0-777"></a>    <span class="k">def</span><span class="w"> </span><span class="nf">score</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">beta</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">standardize_genotype</span><span class="o">=</span><span class="kc">False</span><span class="p">):</span>
</span><span id="__span-0-778"><a id="__codelineno-0-778" name="__codelineno-0-778"></a><span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
</span><span id="__span-0-779"><a id="__codelineno-0-779" name="__codelineno-0-779"></a><span class="sd">        Perform linear scoring, i.e. multiply the genotype matrix by the vector of effect sizes, `beta`.</span>
</span><span id="__span-0-780"><a id="__codelineno-0-780" name="__codelineno-0-780"></a>
</span><span id="__span-0-781"><a id="__codelineno-0-781" name="__codelineno-0-781"></a><span class="sd">        :param beta: A dictionary where the keys are the chromosome numbers and the</span>
</span><span id="__span-0-782"><a id="__codelineno-0-782" name="__codelineno-0-782"></a><span class="sd">        values are a vector of effect sizes for each variant on that chromosome. If the</span>
</span><span id="__span-0-783"><a id="__codelineno-0-783" name="__codelineno-0-783"></a><span class="sd">        betas are not provided, we use the marginal betas by default (if those are available).</span>
</span><span id="__span-0-784"><a id="__codelineno-0-784" name="__codelineno-0-784"></a><span class="sd">        :param standardize_genotype: If True, standardize the genotype matrix before scoring.</span>
</span><span id="__span-0-785"><a id="__codelineno-0-785" name="__codelineno-0-785"></a><span class="sd">        &quot;&quot;&quot;</span>
</span><span id="__span-0-786"><a id="__codelineno-0-786" name="__codelineno-0-786"></a>
</span><span id="__span-0-787"><a id="__codelineno-0-787" name="__codelineno-0-787"></a>        <span class="k">if</span> <span class="n">beta</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
</span><span id="__span-0-788"><a id="__codelineno-0-788" name="__codelineno-0-788"></a>            <span class="k">try</span><span class="p">:</span>
</span><span id="__span-0-789"><a id="__codelineno-0-789" name="__codelineno-0-789"></a>                <span class="n">beta</span> <span class="o">=</span> <span class="p">{</span><span class="n">c</span><span class="p">:</span> <span class="n">s</span><span class="o">.</span><span class="n">marginal_beta</span> <span class="ow">or</span> <span class="n">s</span><span class="o">.</span><span class="n">get_snp_pseudo_corr</span><span class="p">()</span> <span class="k">for</span> <span class="n">c</span><span class="p">,</span> <span class="n">s</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">sumstats_table</span><span class="o">.</span><span class="n">items</span><span class="p">()}</span>
</span><span id="__span-0-790"><a id="__codelineno-0-790" name="__codelineno-0-790"></a>            <span class="k">except</span> <span class="ne">Exception</span><span class="p">:</span>
</span><span id="__span-0-791"><a id="__codelineno-0-791" name="__codelineno-0-791"></a>                <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;To perform linear scoring, you must &quot;</span>
</span><span id="__span-0-792"><a id="__codelineno-0-792" name="__codelineno-0-792"></a>                                 <span class="s2">&quot;provide effect size estimates (BETA)!&quot;</span><span class="p">)</span>
</span><span id="__span-0-793"><a id="__codelineno-0-793" name="__codelineno-0-793"></a>
</span><span id="__span-0-794"><a id="__codelineno-0-794" name="__codelineno-0-794"></a>        <span class="c1"># Here, we have a very ugly way of accounting for</span>
</span><span id="__span-0-795"><a id="__codelineno-0-795" name="__codelineno-0-795"></a>        <span class="c1"># the fact that the chromosomes may be coded differently between the genotype</span>
</span><span id="__span-0-796"><a id="__codelineno-0-796" name="__codelineno-0-796"></a>        <span class="c1"># and the beta dictionary. Maybe we can find a better solution in the future.</span>
</span><span id="__span-0-797"><a id="__codelineno-0-797" name="__codelineno-0-797"></a>        <span class="n">common_chr_g</span><span class="p">,</span> <span class="n">common_chr_b</span> <span class="o">=</span> <span class="n">match_chromosomes</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">genotype</span><span class="o">.</span><span class="n">keys</span><span class="p">(),</span> <span class="n">beta</span><span class="o">.</span><span class="n">keys</span><span class="p">(),</span> <span class="n">return_both</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
</span><span id="__span-0-798"><a id="__codelineno-0-798" name="__codelineno-0-798"></a>
</span><span id="__span-0-799"><a id="__codelineno-0-799" name="__codelineno-0-799"></a>        <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">common_chr_g</span><span class="p">)</span> <span class="o">&lt;</span> <span class="mi">1</span><span class="p">:</span>
</span><span id="__span-0-800"><a id="__codelineno-0-800" name="__codelineno-0-800"></a>            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;No common chromosomes found between &quot;</span>
</span><span id="__span-0-801"><a id="__codelineno-0-801" name="__codelineno-0-801"></a>                             <span class="s2">&quot;the genotype and the effect size estimates!&quot;</span><span class="p">)</span>
</span><span id="__span-0-802"><a id="__codelineno-0-802" name="__codelineno-0-802"></a>
</span><span id="__span-0-803"><a id="__codelineno-0-803" name="__codelineno-0-803"></a>        <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;&gt; Generating polygenic scores...&quot;</span><span class="p">)</span>
</span><span id="__span-0-804"><a id="__codelineno-0-804" name="__codelineno-0-804"></a>
</span><span id="__span-0-805"><a id="__codelineno-0-805" name="__codelineno-0-805"></a>        <span class="n">pgs</span> <span class="o">=</span> <span class="kc">None</span>
</span><span id="__span-0-806"><a id="__codelineno-0-806" name="__codelineno-0-806"></a>
</span><span id="__span-0-807"><a id="__codelineno-0-807" name="__codelineno-0-807"></a>        <span class="k">for</span> <span class="n">c_g</span><span class="p">,</span> <span class="n">c_b</span> <span class="ow">in</span> <span class="n">tqdm</span><span class="p">(</span><span class="nb">zip</span><span class="p">(</span><span class="n">common_chr_g</span><span class="p">,</span> <span class="n">common_chr_b</span><span class="p">),</span>
</span><span id="__span-0-808"><a id="__codelineno-0-808" name="__codelineno-0-808"></a>                             <span class="n">total</span><span class="o">=</span><span class="nb">len</span><span class="p">(</span><span class="n">common_chr_g</span><span class="p">),</span>
</span><span id="__span-0-809"><a id="__codelineno-0-809" name="__codelineno-0-809"></a>                             <span class="n">desc</span><span class="o">=</span><span class="s1">&#39;Generating polygenic scores&#39;</span><span class="p">):</span>
</span><span id="__span-0-810"><a id="__codelineno-0-810" name="__codelineno-0-810"></a>
</span><span id="__span-0-811"><a id="__codelineno-0-811" name="__codelineno-0-811"></a>            <span class="k">if</span> <span class="n">pgs</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
</span><span id="__span-0-812"><a id="__codelineno-0-812" name="__codelineno-0-812"></a>                <span class="n">pgs</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">genotype</span><span class="p">[</span><span class="n">c_g</span><span class="p">]</span><span class="o">.</span><span class="n">score</span><span class="p">(</span><span class="n">beta</span><span class="p">[</span><span class="n">c_b</span><span class="p">],</span> <span class="n">standardize_genotype</span><span class="o">=</span><span class="n">standardize_genotype</span><span class="p">)</span>
</span><span id="__span-0-813"><a id="__codelineno-0-813" name="__codelineno-0-813"></a>            <span class="k">else</span><span class="p">:</span>
</span><span id="__span-0-814"><a id="__codelineno-0-814" name="__codelineno-0-814"></a>                <span class="n">pgs</span> <span class="o">+=</span> <span class="bp">self</span><span class="o">.</span><span class="n">genotype</span><span class="p">[</span><span class="n">c_g</span><span class="p">]</span><span class="o">.</span><span class="n">score</span><span class="p">(</span><span class="n">beta</span><span class="p">[</span><span class="n">c_b</span><span class="p">],</span> <span class="n">standardize_genotype</span><span class="o">=</span><span class="n">standardize_genotype</span><span class="p">)</span>
</span><span id="__span-0-815"><a id="__codelineno-0-815" name="__codelineno-0-815"></a>
</span><span id="__span-0-816"><a id="__codelineno-0-816" name="__codelineno-0-816"></a>        <span class="c1"># If we only have a single set of betas, flatten the PGS vector:</span>
</span><span id="__span-0-817"><a id="__codelineno-0-817" name="__codelineno-0-817"></a>        <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">pgs</span><span class="o">.</span><span class="n">shape</span><span class="p">)</span> <span class="o">&gt;</span> <span class="mi">1</span> <span class="ow">and</span> <span class="n">pgs</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span> <span class="o">==</span> <span class="mi">1</span><span class="p">:</span>
</span><span id="__span-0-818"><a id="__codelineno-0-818" name="__codelineno-0-818"></a>            <span class="n">pgs</span> <span class="o">=</span> <span class="n">pgs</span><span class="o">.</span><span class="n">flatten</span><span class="p">()</span>
</span><span id="__span-0-819"><a id="__codelineno-0-819" name="__codelineno-0-819"></a>
</span><span id="__span-0-820"><a id="__codelineno-0-820" name="__codelineno-0-820"></a>        <span class="k">return</span> <span class="n">pgs</span>
</span><span id="__span-0-821"><a id="__codelineno-0-821" name="__codelineno-0-821"></a>
</span><span id="__span-0-822"><a id="__codelineno-0-822" name="__codelineno-0-822"></a>    <span class="k">def</span><span class="w"> </span><span class="nf">predict</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">beta</span><span class="o">=</span><span class="kc">None</span><span class="p">):</span>
</span><span id="__span-0-823"><a id="__codelineno-0-823" name="__codelineno-0-823"></a><span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
</span><span id="__span-0-824"><a id="__codelineno-0-824" name="__codelineno-0-824"></a><span class="sd">        Predict the phenotype for the genotyped samples using the provided effect size</span>
</span><span id="__span-0-825"><a id="__codelineno-0-825" name="__codelineno-0-825"></a><span class="sd">        estimates `beta`. For quantitative traits, this is equivalent to performing</span>
</span><span id="__span-0-826"><a id="__codelineno-0-826" name="__codelineno-0-826"></a><span class="sd">        linear scoring. For binary phenotypes, we transform the output using probit link function.</span>
</span><span id="__span-0-827"><a id="__codelineno-0-827" name="__codelineno-0-827"></a>
</span><span id="__span-0-828"><a id="__codelineno-0-828" name="__codelineno-0-828"></a><span class="sd">        :param beta: A dictionary where the keys are the chromosome numbers and the</span>
</span><span id="__span-0-829"><a id="__codelineno-0-829" name="__codelineno-0-829"></a><span class="sd">        values are a vector of effect sizes for each variant on that chromosome. If the</span>
</span><span id="__span-0-830"><a id="__codelineno-0-830" name="__codelineno-0-830"></a><span class="sd">        betas are not provided, we use the marginal betas by default (if those are available).</span>
</span><span id="__span-0-831"><a id="__codelineno-0-831" name="__codelineno-0-831"></a><span class="sd">        &quot;&quot;&quot;</span>
</span><span id="__span-0-832"><a id="__codelineno-0-832" name="__codelineno-0-832"></a>
</span><span id="__span-0-833"><a id="__codelineno-0-833" name="__codelineno-0-833"></a>        <span class="c1"># Perform linear scoring:</span>
</span><span id="__span-0-834"><a id="__codelineno-0-834" name="__codelineno-0-834"></a>        <span class="n">pgs</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">score</span><span class="p">(</span><span class="n">beta</span><span class="p">)</span>
</span><span id="__span-0-835"><a id="__codelineno-0-835" name="__codelineno-0-835"></a>
</span><span id="__span-0-836"><a id="__codelineno-0-836" name="__codelineno-0-836"></a>        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">phenotype_likelihood</span> <span class="o">==</span> <span class="s1">&#39;binomial&#39;</span><span class="p">:</span>
</span><span id="__span-0-837"><a id="__codelineno-0-837" name="__codelineno-0-837"></a>            <span class="c1"># Apply probit link function:</span>
</span><span id="__span-0-838"><a id="__codelineno-0-838" name="__codelineno-0-838"></a>            <span class="kn">from</span><span class="w"> </span><span class="nn">scipy.stats</span><span class="w"> </span><span class="kn">import</span> <span class="n">norm</span>
</span><span id="__span-0-839"><a id="__codelineno-0-839" name="__codelineno-0-839"></a>            <span class="n">pgs</span> <span class="o">=</span> <span class="n">norm</span><span class="o">.</span><span class="n">cdf</span><span class="p">(</span><span class="n">pgs</span><span class="p">)</span>
</span><span id="__span-0-840"><a id="__codelineno-0-840" name="__codelineno-0-840"></a>
</span><span id="__span-0-841"><a id="__codelineno-0-841" name="__codelineno-0-841"></a>        <span class="k">return</span> <span class="n">pgs</span>
</span><span id="__span-0-842"><a id="__codelineno-0-842" name="__codelineno-0-842"></a>
</span><span id="__span-0-843"><a id="__codelineno-0-843" name="__codelineno-0-843"></a>    <span class="k">def</span><span class="w"> </span><span class="nf">to_individual_table</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
</span><span id="__span-0-844"><a id="__codelineno-0-844" name="__codelineno-0-844"></a><span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
</span><span id="__span-0-845"><a id="__codelineno-0-845" name="__codelineno-0-845"></a><span class="sd">        :return: A plink-style dataframe of individual IDs, in the form of</span>
</span><span id="__span-0-846"><a id="__codelineno-0-846" name="__codelineno-0-846"></a><span class="sd">        Family ID (FID) and Individual ID (IID).</span>
</span><span id="__span-0-847"><a id="__codelineno-0-847" name="__codelineno-0-847"></a><span class="sd">        &quot;&quot;&quot;</span>
</span><span id="__span-0-848"><a id="__codelineno-0-848" name="__codelineno-0-848"></a>        <span class="k">assert</span> <span class="bp">self</span><span class="o">.</span><span class="n">sample_table</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span>
</span><span id="__span-0-849"><a id="__codelineno-0-849" name="__codelineno-0-849"></a>
</span><span id="__span-0-850"><a id="__codelineno-0-850" name="__codelineno-0-850"></a>        <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">sample_table</span><span class="o">.</span><span class="n">get_individual_table</span><span class="p">()</span>
</span><span id="__span-0-851"><a id="__codelineno-0-851" name="__codelineno-0-851"></a>
</span><span id="__span-0-852"><a id="__codelineno-0-852" name="__codelineno-0-852"></a>    <span class="k">def</span><span class="w"> </span><span class="nf">to_phenotype_table</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
</span><span id="__span-0-853"><a id="__codelineno-0-853" name="__codelineno-0-853"></a><span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
</span><span id="__span-0-854"><a id="__codelineno-0-854" name="__codelineno-0-854"></a><span class="sd">        :return: A plink-style dataframe with each individual&#39;s Family ID (FID),</span>
</span><span id="__span-0-855"><a id="__codelineno-0-855" name="__codelineno-0-855"></a><span class="sd">        Individual ID (IID), and phenotype value.</span>
</span><span id="__span-0-856"><a id="__codelineno-0-856" name="__codelineno-0-856"></a><span class="sd">        &quot;&quot;&quot;</span>
</span><span id="__span-0-857"><a id="__codelineno-0-857" name="__codelineno-0-857"></a>
</span><span id="__span-0-858"><a id="__codelineno-0-858" name="__codelineno-0-858"></a>        <span class="k">assert</span> <span class="bp">self</span><span class="o">.</span><span class="n">sample_table</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span>
</span><span id="__span-0-859"><a id="__codelineno-0-859" name="__codelineno-0-859"></a>
</span><span id="__span-0-860"><a id="__codelineno-0-860" name="__codelineno-0-860"></a>        <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">sample_table</span><span class="o">.</span><span class="n">get_phenotype_table</span><span class="p">()</span>
</span><span id="__span-0-861"><a id="__codelineno-0-861" name="__codelineno-0-861"></a>
</span><span id="__span-0-862"><a id="__codelineno-0-862" name="__codelineno-0-862"></a>    <span class="k">def</span><span class="w"> </span><span class="nf">to_snp_table</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">col_subset</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">per_chromosome</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">resource</span><span class="o">=</span><span class="s1">&#39;auto&#39;</span><span class="p">):</span>
</span><span id="__span-0-863"><a id="__codelineno-0-863" name="__codelineno-0-863"></a><span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
</span><span id="__span-0-864"><a id="__codelineno-0-864" name="__codelineno-0-864"></a><span class="sd">        Get a dataframe of SNP data for all variants</span>
</span><span id="__span-0-865"><a id="__codelineno-0-865" name="__codelineno-0-865"></a><span class="sd">        across different chromosomes.</span>
</span><span id="__span-0-866"><a id="__codelineno-0-866" name="__codelineno-0-866"></a>
</span><span id="__span-0-867"><a id="__codelineno-0-867" name="__codelineno-0-867"></a><span class="sd">        :param col_subset: The subset of columns to obtain.</span>
</span><span id="__span-0-868"><a id="__codelineno-0-868" name="__codelineno-0-868"></a><span class="sd">        :param per_chromosome: If True, returns a dictionary where the key</span>
</span><span id="__span-0-869"><a id="__codelineno-0-869" name="__codelineno-0-869"></a><span class="sd">        is the chromosome number and the value is the SNP table per</span>
</span><span id="__span-0-870"><a id="__codelineno-0-870" name="__codelineno-0-870"></a><span class="sd">        chromosome.</span>
</span><span id="__span-0-871"><a id="__codelineno-0-871" name="__codelineno-0-871"></a><span class="sd">        :param resource: The data source to extract the SNP table from. By default, the method</span>
</span><span id="__span-0-872"><a id="__codelineno-0-872" name="__codelineno-0-872"></a><span class="sd">        will try to extract the SNP table from the genotype matrix. If the genotype matrix is not</span>
</span><span id="__span-0-873"><a id="__codelineno-0-873" name="__codelineno-0-873"></a><span class="sd">        available, then it will try to extract the SNP information from the LD matrix or the summary</span>
</span><span id="__span-0-874"><a id="__codelineno-0-874" name="__codelineno-0-874"></a><span class="sd">        statistics table. Possible values: `auto`, `genotype`, `ld`, `sumstats`.</span>
</span><span id="__span-0-875"><a id="__codelineno-0-875" name="__codelineno-0-875"></a>
</span><span id="__span-0-876"><a id="__codelineno-0-876" name="__codelineno-0-876"></a><span class="sd">        :return: A dataframe (or dictionary of dataframes) of SNP data.</span>
</span><span id="__span-0-877"><a id="__codelineno-0-877" name="__codelineno-0-877"></a><span class="sd">        &quot;&quot;&quot;</span>
</span><span id="__span-0-878"><a id="__codelineno-0-878" name="__codelineno-0-878"></a>
</span><span id="__span-0-879"><a id="__codelineno-0-879" name="__codelineno-0-879"></a>        <span class="c1"># Sanity checks:</span>
</span><span id="__span-0-880"><a id="__codelineno-0-880" name="__codelineno-0-880"></a>        <span class="k">assert</span> <span class="n">resource</span> <span class="ow">in</span> <span class="p">(</span><span class="s1">&#39;auto&#39;</span><span class="p">,</span> <span class="s1">&#39;genotype&#39;</span><span class="p">,</span> <span class="s1">&#39;ld&#39;</span><span class="p">,</span> <span class="s1">&#39;sumstats&#39;</span><span class="p">)</span>
</span><span id="__span-0-881"><a id="__codelineno-0-881" name="__codelineno-0-881"></a>
</span><span id="__span-0-882"><a id="__codelineno-0-882" name="__codelineno-0-882"></a>        <span class="k">if</span> <span class="n">resource</span> <span class="o">!=</span> <span class="s1">&#39;auto&#39;</span><span class="p">:</span>
</span><span id="__span-0-883"><a id="__codelineno-0-883" name="__codelineno-0-883"></a>            <span class="k">if</span> <span class="n">resource</span> <span class="o">==</span> <span class="s1">&#39;genotype&#39;</span> <span class="ow">and</span> <span class="bp">self</span><span class="o">.</span><span class="n">genotype</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
</span><span id="__span-0-884"><a id="__codelineno-0-884" name="__codelineno-0-884"></a>                <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;Genotype matrix is not available!&quot;</span><span class="p">)</span>
</span><span id="__span-0-885"><a id="__codelineno-0-885" name="__codelineno-0-885"></a>            <span class="k">if</span> <span class="n">resource</span> <span class="o">==</span> <span class="s1">&#39;ld&#39;</span> <span class="ow">and</span> <span class="bp">self</span><span class="o">.</span><span class="n">ld</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
</span><span id="__span-0-886"><a id="__codelineno-0-886" name="__codelineno-0-886"></a>                <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;LD matrix is not available!&quot;</span><span class="p">)</span>
</span><span id="__span-0-887"><a id="__codelineno-0-887" name="__codelineno-0-887"></a>            <span class="k">if</span> <span class="n">resource</span> <span class="o">==</span> <span class="s1">&#39;sumstats&#39;</span> <span class="ow">and</span> <span class="bp">self</span><span class="o">.</span><span class="n">sumstats_table</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
</span><span id="__span-0-888"><a id="__codelineno-0-888" name="__codelineno-0-888"></a>                <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;Summary statistics table is not available!&quot;</span><span class="p">)</span>
</span><span id="__span-0-889"><a id="__codelineno-0-889" name="__codelineno-0-889"></a>        <span class="k">else</span><span class="p">:</span>
</span><span id="__span-0-890"><a id="__codelineno-0-890" name="__codelineno-0-890"></a>            <span class="k">if</span> <span class="nb">all</span><span class="p">(</span><span class="n">ds</span> <span class="ow">is</span> <span class="kc">None</span> <span class="k">for</span> <span class="n">ds</span> <span class="ow">in</span> <span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">genotype</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">ld</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">sumstats_table</span><span class="p">)):</span>
</span><span id="__span-0-891"><a id="__codelineno-0-891" name="__codelineno-0-891"></a>                <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;No data sources available to extract SNP data from!&quot;</span><span class="p">)</span>
</span><span id="__span-0-892"><a id="__codelineno-0-892" name="__codelineno-0-892"></a>
</span><span id="__span-0-893"><a id="__codelineno-0-893" name="__codelineno-0-893"></a>        <span class="c1"># Extract the SNP data:</span>
</span><span id="__span-0-894"><a id="__codelineno-0-894" name="__codelineno-0-894"></a>
</span><span id="__span-0-895"><a id="__codelineno-0-895" name="__codelineno-0-895"></a>        <span class="n">snp_tables</span> <span class="o">=</span> <span class="p">{}</span>
</span><span id="__span-0-896"><a id="__codelineno-0-896" name="__codelineno-0-896"></a>
</span><span id="__span-0-897"><a id="__codelineno-0-897" name="__codelineno-0-897"></a>        <span class="k">if</span> <span class="n">resource</span> <span class="ow">in</span> <span class="p">(</span><span class="s1">&#39;auto&#39;</span><span class="p">,</span> <span class="s1">&#39;genotype&#39;</span><span class="p">)</span> <span class="ow">and</span> <span class="bp">self</span><span class="o">.</span><span class="n">genotype</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
</span><span id="__span-0-898"><a id="__codelineno-0-898" name="__codelineno-0-898"></a>            <span class="k">for</span> <span class="n">c</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">chromosomes</span><span class="p">:</span>
</span><span id="__span-0-899"><a id="__codelineno-0-899" name="__codelineno-0-899"></a>                <span class="n">snp_tables</span><span class="p">[</span><span class="n">c</span><span class="p">]</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">genotype</span><span class="p">[</span><span class="n">c</span><span class="p">]</span><span class="o">.</span><span class="n">get_snp_table</span><span class="p">(</span><span class="n">col_subset</span><span class="o">=</span><span class="n">col_subset</span><span class="p">)</span>
</span><span id="__span-0-900"><a id="__codelineno-0-900" name="__codelineno-0-900"></a>        <span class="k">elif</span> <span class="n">resource</span> <span class="ow">in</span> <span class="p">(</span><span class="s1">&#39;auto&#39;</span><span class="p">,</span> <span class="s1">&#39;ld&#39;</span><span class="p">)</span> <span class="ow">and</span> <span class="bp">self</span><span class="o">.</span><span class="n">ld</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
</span><span id="__span-0-901"><a id="__codelineno-0-901" name="__codelineno-0-901"></a>            <span class="k">for</span> <span class="n">c</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">chromosomes</span><span class="p">:</span>
</span><span id="__span-0-902"><a id="__codelineno-0-902" name="__codelineno-0-902"></a>                <span class="n">snp_tables</span><span class="p">[</span><span class="n">c</span><span class="p">]</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">ld</span><span class="p">[</span><span class="n">c</span><span class="p">]</span><span class="o">.</span><span class="n">to_snp_table</span><span class="p">(</span><span class="n">col_subset</span><span class="o">=</span><span class="n">col_subset</span><span class="p">)</span>
</span><span id="__span-0-903"><a id="__codelineno-0-903" name="__codelineno-0-903"></a>        <span class="k">else</span><span class="p">:</span>
</span><span id="__span-0-904"><a id="__codelineno-0-904" name="__codelineno-0-904"></a>            <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">to_summary_statistics_table</span><span class="p">(</span><span class="n">col_subset</span><span class="o">=</span><span class="n">col_subset</span><span class="p">,</span>
</span><span id="__span-0-905"><a id="__codelineno-0-905" name="__codelineno-0-905"></a>                                                    <span class="n">per_chromosome</span><span class="o">=</span><span class="n">per_chromosome</span><span class="p">)</span>
</span><span id="__span-0-906"><a id="__codelineno-0-906" name="__codelineno-0-906"></a>
</span><span id="__span-0-907"><a id="__codelineno-0-907" name="__codelineno-0-907"></a>        <span class="k">if</span> <span class="n">per_chromosome</span><span class="p">:</span>
</span><span id="__span-0-908"><a id="__codelineno-0-908" name="__codelineno-0-908"></a>            <span class="k">return</span> <span class="n">snp_tables</span>
</span><span id="__span-0-909"><a id="__codelineno-0-909" name="__codelineno-0-909"></a>        <span class="k">else</span><span class="p">:</span>
</span><span id="__span-0-910"><a id="__codelineno-0-910" name="__codelineno-0-910"></a>            <span class="k">return</span> <span class="n">pd</span><span class="o">.</span><span class="n">concat</span><span class="p">(</span><span class="nb">list</span><span class="p">(</span><span class="n">snp_tables</span><span class="o">.</span><span class="n">values</span><span class="p">()))</span>
</span><span id="__span-0-911"><a id="__codelineno-0-911" name="__codelineno-0-911"></a>
</span><span id="__span-0-912"><a id="__codelineno-0-912" name="__codelineno-0-912"></a>    <span class="k">def</span><span class="w"> </span><span class="nf">to_summary_statistics_table</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">col_subset</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">per_chromosome</span><span class="o">=</span><span class="kc">False</span><span class="p">):</span>
</span><span id="__span-0-913"><a id="__codelineno-0-913" name="__codelineno-0-913"></a><span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
</span><span id="__span-0-914"><a id="__codelineno-0-914" name="__codelineno-0-914"></a><span class="sd">        Get a dataframe of the GWAS summary statistics for all variants</span>
</span><span id="__span-0-915"><a id="__codelineno-0-915" name="__codelineno-0-915"></a><span class="sd">        across different chromosomes.</span>
</span><span id="__span-0-916"><a id="__codelineno-0-916" name="__codelineno-0-916"></a>
</span><span id="__span-0-917"><a id="__codelineno-0-917" name="__codelineno-0-917"></a><span class="sd">        :param col_subset: The subset of columns (or summary statistics) to obtain.</span>
</span><span id="__span-0-918"><a id="__codelineno-0-918" name="__codelineno-0-918"></a><span class="sd">        :param per_chromosome: If True, returns a dictionary where the key</span>
</span><span id="__span-0-919"><a id="__codelineno-0-919" name="__codelineno-0-919"></a><span class="sd">        is the chromosome number and the value is the summary statistics table per</span>
</span><span id="__span-0-920"><a id="__codelineno-0-920" name="__codelineno-0-920"></a><span class="sd">        chromosome.</span>
</span><span id="__span-0-921"><a id="__codelineno-0-921" name="__codelineno-0-921"></a>
</span><span id="__span-0-922"><a id="__codelineno-0-922" name="__codelineno-0-922"></a><span class="sd">        :return: A dataframe (or dictionary of dataframes) of summary statistics.</span>
</span><span id="__span-0-923"><a id="__codelineno-0-923" name="__codelineno-0-923"></a><span class="sd">        &quot;&quot;&quot;</span>
</span><span id="__span-0-924"><a id="__codelineno-0-924" name="__codelineno-0-924"></a>
</span><span id="__span-0-925"><a id="__codelineno-0-925" name="__codelineno-0-925"></a>        <span class="k">assert</span> <span class="bp">self</span><span class="o">.</span><span class="n">sumstats_table</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span>
</span><span id="__span-0-926"><a id="__codelineno-0-926" name="__codelineno-0-926"></a>
</span><span id="__span-0-927"><a id="__codelineno-0-927" name="__codelineno-0-927"></a>        <span class="n">snp_tables</span> <span class="o">=</span> <span class="p">{}</span>
</span><span id="__span-0-928"><a id="__codelineno-0-928" name="__codelineno-0-928"></a>
</span><span id="__span-0-929"><a id="__codelineno-0-929" name="__codelineno-0-929"></a>        <span class="k">for</span> <span class="n">c</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">chromosomes</span><span class="p">:</span>
</span><span id="__span-0-930"><a id="__codelineno-0-930" name="__codelineno-0-930"></a>            <span class="n">snp_tables</span><span class="p">[</span><span class="n">c</span><span class="p">]</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">sumstats_table</span><span class="p">[</span><span class="n">c</span><span class="p">]</span><span class="o">.</span><span class="n">to_table</span><span class="p">(</span><span class="n">col_subset</span><span class="o">=</span><span class="n">col_subset</span><span class="p">)</span>
</span><span id="__span-0-931"><a id="__codelineno-0-931" name="__codelineno-0-931"></a>
</span><span id="__span-0-932"><a id="__codelineno-0-932" name="__codelineno-0-932"></a>        <span class="k">if</span> <span class="n">per_chromosome</span><span class="p">:</span>
</span><span id="__span-0-933"><a id="__codelineno-0-933" name="__codelineno-0-933"></a>            <span class="k">return</span> <span class="n">snp_tables</span>
</span><span id="__span-0-934"><a id="__codelineno-0-934" name="__codelineno-0-934"></a>        <span class="k">else</span><span class="p">:</span>
</span><span id="__span-0-935"><a id="__codelineno-0-935" name="__codelineno-0-935"></a>            <span class="k">return</span> <span class="n">pd</span><span class="o">.</span><span class="n">concat</span><span class="p">(</span><span class="nb">list</span><span class="p">(</span><span class="n">snp_tables</span><span class="o">.</span><span class="n">values</span><span class="p">()))</span>
</span><span id="__span-0-936"><a id="__codelineno-0-936" name="__codelineno-0-936"></a>
</span><span id="__span-0-937"><a id="__codelineno-0-937" name="__codelineno-0-937"></a>    <span class="k">def</span><span class="w"> </span><span class="nf">sync_sample_tables</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
</span><span id="__span-0-938"><a id="__codelineno-0-938" name="__codelineno-0-938"></a><span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
</span><span id="__span-0-939"><a id="__codelineno-0-939" name="__codelineno-0-939"></a><span class="sd">        A utility method to sync the sample tables of the</span>
</span><span id="__span-0-940"><a id="__codelineno-0-940" name="__codelineno-0-940"></a><span class="sd">        `GenotypeMatrix` objects with the sample table under</span>
</span><span id="__span-0-941"><a id="__codelineno-0-941" name="__codelineno-0-941"></a><span class="sd">        the `GWADataLoader` object. This is especially important</span>
</span><span id="__span-0-942"><a id="__codelineno-0-942" name="__codelineno-0-942"></a><span class="sd">        when setting new phenotypes (from the simulators) or reading</span>
</span><span id="__span-0-943"><a id="__codelineno-0-943" name="__codelineno-0-943"></a><span class="sd">        covariates files, etc.</span>
</span><span id="__span-0-944"><a id="__codelineno-0-944" name="__codelineno-0-944"></a><span class="sd">        &quot;&quot;&quot;</span>
</span><span id="__span-0-945"><a id="__codelineno-0-945" name="__codelineno-0-945"></a>
</span><span id="__span-0-946"><a id="__codelineno-0-946" name="__codelineno-0-946"></a>        <span class="k">for</span> <span class="n">c</span><span class="p">,</span> <span class="n">g</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">genotype</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
</span><span id="__span-0-947"><a id="__codelineno-0-947" name="__codelineno-0-947"></a>            <span class="n">g</span><span class="o">.</span><span class="n">set_sample_table</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">sample_table</span><span class="p">)</span>
</span><span id="__span-0-948"><a id="__codelineno-0-948" name="__codelineno-0-948"></a>
</span><span id="__span-0-949"><a id="__codelineno-0-949" name="__codelineno-0-949"></a>    <span class="k">def</span><span class="w"> </span><span class="nf">split_by_chromosome</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
</span><span id="__span-0-950"><a id="__codelineno-0-950" name="__codelineno-0-950"></a><span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
</span><span id="__span-0-951"><a id="__codelineno-0-951" name="__codelineno-0-951"></a><span class="sd">        A utility method to split a GWADataLoader object by chromosome ID, such that</span>
</span><span id="__span-0-952"><a id="__codelineno-0-952" name="__codelineno-0-952"></a><span class="sd">        we would have one `GWADataLoader` object per chromosome. The method returns a dictionary</span>
</span><span id="__span-0-953"><a id="__codelineno-0-953" name="__codelineno-0-953"></a><span class="sd">        where the key is the chromosome number and the value is the `GWADataLoader` object corresponding</span>
</span><span id="__span-0-954"><a id="__codelineno-0-954" name="__codelineno-0-954"></a><span class="sd">        to that chromosome only.</span>
</span><span id="__span-0-955"><a id="__codelineno-0-955" name="__codelineno-0-955"></a><span class="sd">        &quot;&quot;&quot;</span>
</span><span id="__span-0-956"><a id="__codelineno-0-956" name="__codelineno-0-956"></a>
</span><span id="__span-0-957"><a id="__codelineno-0-957" name="__codelineno-0-957"></a>        <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">chromosomes</span><span class="p">)</span> <span class="o">==</span> <span class="mi">1</span><span class="p">:</span>
</span><span id="__span-0-958"><a id="__codelineno-0-958" name="__codelineno-0-958"></a>            <span class="k">return</span> <span class="p">{</span><span class="bp">self</span><span class="o">.</span><span class="n">chromosomes</span><span class="p">[</span><span class="mi">0</span><span class="p">]:</span> <span class="bp">self</span><span class="p">}</span>
</span><span id="__span-0-959"><a id="__codelineno-0-959" name="__codelineno-0-959"></a>
</span><span id="__span-0-960"><a id="__codelineno-0-960" name="__codelineno-0-960"></a>        <span class="k">else</span><span class="p">:</span>
</span><span id="__span-0-961"><a id="__codelineno-0-961" name="__codelineno-0-961"></a>            <span class="n">split_dict</span> <span class="o">=</span> <span class="p">{}</span>
</span><span id="__span-0-962"><a id="__codelineno-0-962" name="__codelineno-0-962"></a>
</span><span id="__span-0-963"><a id="__codelineno-0-963" name="__codelineno-0-963"></a>            <span class="k">for</span> <span class="n">c</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">chromosomes</span><span class="p">:</span>
</span><span id="__span-0-964"><a id="__codelineno-0-964" name="__codelineno-0-964"></a>                <span class="n">split_dict</span><span class="p">[</span><span class="n">c</span><span class="p">]</span> <span class="o">=</span> <span class="n">copy</span><span class="o">.</span><span class="n">copy</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span>
</span><span id="__span-0-965"><a id="__codelineno-0-965" name="__codelineno-0-965"></a>
</span><span id="__span-0-966"><a id="__codelineno-0-966" name="__codelineno-0-966"></a>                <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">genotype</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span> <span class="ow">and</span> <span class="n">c</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">genotype</span><span class="p">:</span>
</span><span id="__span-0-967"><a id="__codelineno-0-967" name="__codelineno-0-967"></a>                    <span class="n">split_dict</span><span class="p">[</span><span class="n">c</span><span class="p">]</span><span class="o">.</span><span class="n">genotype</span> <span class="o">=</span> <span class="p">{</span><span class="n">c</span><span class="p">:</span> <span class="bp">self</span><span class="o">.</span><span class="n">genotype</span><span class="p">[</span><span class="n">c</span><span class="p">]}</span>
</span><span id="__span-0-968"><a id="__codelineno-0-968" name="__codelineno-0-968"></a>                <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">sumstats_table</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span> <span class="ow">and</span> <span class="n">c</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">sumstats_table</span><span class="p">:</span>
</span><span id="__span-0-969"><a id="__codelineno-0-969" name="__codelineno-0-969"></a>                    <span class="n">split_dict</span><span class="p">[</span><span class="n">c</span><span class="p">]</span><span class="o">.</span><span class="n">sumstats_table</span> <span class="o">=</span> <span class="p">{</span><span class="n">c</span><span class="p">:</span> <span class="bp">self</span><span class="o">.</span><span class="n">sumstats_table</span><span class="p">[</span><span class="n">c</span><span class="p">]}</span>
</span><span id="__span-0-970"><a id="__codelineno-0-970" name="__codelineno-0-970"></a>                <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">ld</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span> <span class="ow">and</span> <span class="n">c</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">ld</span><span class="p">:</span>
</span><span id="__span-0-971"><a id="__codelineno-0-971" name="__codelineno-0-971"></a>                    <span class="n">split_dict</span><span class="p">[</span><span class="n">c</span><span class="p">]</span><span class="o">.</span><span class="n">ld</span> <span class="o">=</span> <span class="p">{</span><span class="n">c</span><span class="p">:</span> <span class="bp">self</span><span class="o">.</span><span class="n">ld</span><span class="p">[</span><span class="n">c</span><span class="p">]}</span>
</span><span id="__span-0-972"><a id="__codelineno-0-972" name="__codelineno-0-972"></a>                <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">annotation</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span> <span class="ow">and</span> <span class="n">c</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">annotation</span><span class="p">:</span>
</span><span id="__span-0-973"><a id="__codelineno-0-973" name="__codelineno-0-973"></a>                    <span class="n">split_dict</span><span class="p">[</span><span class="n">c</span><span class="p">]</span><span class="o">.</span><span class="n">annotation</span> <span class="o">=</span> <span class="p">{</span><span class="n">c</span><span class="p">:</span> <span class="bp">self</span><span class="o">.</span><span class="n">annotation</span><span class="p">[</span><span class="n">c</span><span class="p">]}</span>
</span><span id="__span-0-974"><a id="__codelineno-0-974" name="__codelineno-0-974"></a>
</span><span id="__span-0-975"><a id="__codelineno-0-975" name="__codelineno-0-975"></a>            <span class="k">return</span> <span class="n">split_dict</span>
</span><span id="__span-0-976"><a id="__codelineno-0-976" name="__codelineno-0-976"></a>
</span><span id="__span-0-977"><a id="__codelineno-0-977" name="__codelineno-0-977"></a>    <span class="k">def</span><span class="w"> </span><span class="nf">split_by_samples</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">proportions</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">groups</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">keep_original</span><span class="o">=</span><span class="kc">True</span><span class="p">):</span>
</span><span id="__span-0-978"><a id="__codelineno-0-978" name="__codelineno-0-978"></a><span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
</span><span id="__span-0-979"><a id="__codelineno-0-979" name="__codelineno-0-979"></a><span class="sd">        Split the `GWADataLoader` object by samples, if genotype or sample data</span>
</span><span id="__span-0-980"><a id="__codelineno-0-980" name="__codelineno-0-980"></a><span class="sd">        is available. The user must provide a list or proportion of samples in each split,</span>
</span><span id="__span-0-981"><a id="__codelineno-0-981" name="__codelineno-0-981"></a><span class="sd">        and the method will return a list of `GWADataLoader` objects with only the samples</span>
</span><span id="__span-0-982"><a id="__codelineno-0-982" name="__codelineno-0-982"></a><span class="sd">        designated for each split. This may be a useful utility for training/testing split or some</span>
</span><span id="__span-0-983"><a id="__codelineno-0-983" name="__codelineno-0-983"></a><span class="sd">        other downstream tasks.</span>
</span><span id="__span-0-984"><a id="__codelineno-0-984" name="__codelineno-0-984"></a>
</span><span id="__span-0-985"><a id="__codelineno-0-985" name="__codelineno-0-985"></a><span class="sd">        :param proportions: A list with the proportion of samples in each split. Must add to 1.</span>
</span><span id="__span-0-986"><a id="__codelineno-0-986" name="__codelineno-0-986"></a><span class="sd">        :param groups: A list of lists containing the sample IDs in each split.</span>
</span><span id="__span-0-987"><a id="__codelineno-0-987" name="__codelineno-0-987"></a><span class="sd">        :param keep_original: If True, keep the original `GWADataLoader` object and do not</span>
</span><span id="__span-0-988"><a id="__codelineno-0-988" name="__codelineno-0-988"></a><span class="sd">        transform it in the splitting process.</span>
</span><span id="__span-0-989"><a id="__codelineno-0-989" name="__codelineno-0-989"></a><span class="sd">        &quot;&quot;&quot;</span>
</span><span id="__span-0-990"><a id="__codelineno-0-990" name="__codelineno-0-990"></a>
</span><span id="__span-0-991"><a id="__codelineno-0-991" name="__codelineno-0-991"></a>        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">sample_table</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
</span><span id="__span-0-992"><a id="__codelineno-0-992" name="__codelineno-0-992"></a>            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;The sample table is not set!&quot;</span><span class="p">)</span>
</span><span id="__span-0-993"><a id="__codelineno-0-993" name="__codelineno-0-993"></a>
</span><span id="__span-0-994"><a id="__codelineno-0-994" name="__codelineno-0-994"></a>        <span class="k">if</span> <span class="n">groups</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
</span><span id="__span-0-995"><a id="__codelineno-0-995" name="__codelineno-0-995"></a>            <span class="k">if</span> <span class="n">proportions</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
</span><span id="__span-0-996"><a id="__codelineno-0-996" name="__codelineno-0-996"></a>                <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;To split a `GWADataloader` object by samples, the user must provide either the list &quot;</span>
</span><span id="__span-0-997"><a id="__codelineno-0-997" name="__codelineno-0-997"></a>                                 <span class="s2">&quot;or proportion of individuals in each split.&quot;</span><span class="p">)</span>
</span><span id="__span-0-998"><a id="__codelineno-0-998" name="__codelineno-0-998"></a>            <span class="k">else</span><span class="p">:</span>
</span><span id="__span-0-999"><a id="__codelineno-0-999" name="__codelineno-0-999"></a>
</span><span id="__span-0-1000"><a id="__codelineno-0-1000" name="__codelineno-0-1000"></a>                <span class="c1"># Assign each sample to a different split randomly by drawing from a multinomial:</span>
</span><span id="__span-0-1001"><a id="__codelineno-0-1001" name="__codelineno-0-1001"></a>                <span class="n">random_split</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">random</span><span class="o">.</span><span class="n">multinomial</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="n">proportions</span><span class="p">,</span> <span class="n">size</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">sample_size</span><span class="p">)</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="nb">bool</span><span class="p">)</span>
</span><span id="__span-0-1002"><a id="__codelineno-0-1002" name="__codelineno-0-1002"></a>                <span class="c1"># Extract the individuals in each group from the multinomial sample:</span>
</span><span id="__span-0-1003"><a id="__codelineno-0-1003" name="__codelineno-0-1003"></a>                <span class="n">groups</span> <span class="o">=</span> <span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">samples</span><span class="p">[</span><span class="n">random_split</span><span class="p">[:,</span> <span class="n">i</span><span class="p">]]</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">random_split</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">1</span><span class="p">])]</span>
</span><span id="__span-0-1004"><a id="__codelineno-0-1004" name="__codelineno-0-1004"></a>
</span><span id="__span-0-1005"><a id="__codelineno-0-1005" name="__codelineno-0-1005"></a>        <span class="n">gdls</span> <span class="o">=</span> <span class="p">[]</span>
</span><span id="__span-0-1006"><a id="__codelineno-0-1006" name="__codelineno-0-1006"></a>        <span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="n">g</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">groups</span><span class="p">):</span>
</span><span id="__span-0-1007"><a id="__codelineno-0-1007" name="__codelineno-0-1007"></a>
</span><span id="__span-0-1008"><a id="__codelineno-0-1008" name="__codelineno-0-1008"></a>            <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">g</span><span class="p">)</span> <span class="o">&lt;</span> <span class="mi">1</span><span class="p">:</span>
</span><span id="__span-0-1009"><a id="__codelineno-0-1009" name="__codelineno-0-1009"></a>                <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Group </span><span class="si">{</span><span class="n">i</span><span class="si">}</span><span class="s2"> is empty! Please ensure that all splits have at least one sample.&quot;</span><span class="p">)</span>
</span><span id="__span-0-1010"><a id="__codelineno-0-1010" name="__codelineno-0-1010"></a>
</span><span id="__span-0-1011"><a id="__codelineno-0-1011" name="__codelineno-0-1011"></a>            <span class="k">if</span> <span class="p">(</span><span class="n">i</span> <span class="o">+</span> <span class="mi">1</span><span class="p">)</span> <span class="o">==</span> <span class="nb">len</span><span class="p">(</span><span class="n">groups</span><span class="p">)</span> <span class="ow">and</span> <span class="ow">not</span> <span class="n">keep_original</span><span class="p">:</span>
</span><span id="__span-0-1012"><a id="__codelineno-0-1012" name="__codelineno-0-1012"></a>                <span class="n">new_gdl</span> <span class="o">=</span> <span class="bp">self</span>
</span><span id="__span-0-1013"><a id="__codelineno-0-1013" name="__codelineno-0-1013"></a>            <span class="k">else</span><span class="p">:</span>
</span><span id="__span-0-1014"><a id="__codelineno-0-1014" name="__codelineno-0-1014"></a>                <span class="n">new_gdl</span> <span class="o">=</span> <span class="n">copy</span><span class="o">.</span><span class="n">deepcopy</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span>
</span><span id="__span-0-1015"><a id="__codelineno-0-1015" name="__codelineno-0-1015"></a>
</span><span id="__span-0-1016"><a id="__codelineno-0-1016" name="__codelineno-0-1016"></a>            <span class="n">new_gdl</span><span class="o">.</span><span class="n">filter_samples</span><span class="p">(</span><span class="n">keep_samples</span><span class="o">=</span><span class="n">g</span><span class="p">)</span>
</span><span id="__span-0-1017"><a id="__codelineno-0-1017" name="__codelineno-0-1017"></a>
</span><span id="__span-0-1018"><a id="__codelineno-0-1018" name="__codelineno-0-1018"></a>            <span class="n">gdls</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">new_gdl</span><span class="p">)</span>
</span><span id="__span-0-1019"><a id="__codelineno-0-1019" name="__codelineno-0-1019"></a>
</span><span id="__span-0-1020"><a id="__codelineno-0-1020" name="__codelineno-0-1020"></a>        <span class="k">return</span> <span class="n">gdls</span>
</span><span id="__span-0-1021"><a id="__codelineno-0-1021" name="__codelineno-0-1021"></a>
</span><span id="__span-0-1022"><a id="__codelineno-0-1022" name="__codelineno-0-1022"></a>    <span class="k">def</span><span class="w"> </span><span class="nf">align_with</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">other_gdls</span><span class="p">,</span> <span class="n">axis</span><span class="o">=</span><span class="s1">&#39;SNP&#39;</span><span class="p">,</span> <span class="n">how</span><span class="o">=</span><span class="s1">&#39;inner&#39;</span><span class="p">):</span>
</span><span id="__span-0-1023"><a id="__codelineno-0-1023" name="__codelineno-0-1023"></a><span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
</span><span id="__span-0-1024"><a id="__codelineno-0-1024" name="__codelineno-0-1024"></a><span class="sd">        Align the `GWADataLoader` object with other GDL objects to have the same</span>
</span><span id="__span-0-1025"><a id="__codelineno-0-1025" name="__codelineno-0-1025"></a><span class="sd">        set of SNPs or samples. This utility method is meant to enable the user to</span>
</span><span id="__span-0-1026"><a id="__codelineno-0-1026" name="__codelineno-0-1026"></a><span class="sd">        align multiple data sources for downstream analyses.</span>
</span><span id="__span-0-1027"><a id="__codelineno-0-1027" name="__codelineno-0-1027"></a>
</span><span id="__span-0-1028"><a id="__codelineno-0-1028" name="__codelineno-0-1028"></a><span class="sd">        :param other_gdls: A `GWADataLoader` or list of `GWADataLoader` objects.</span>
</span><span id="__span-0-1029"><a id="__codelineno-0-1029" name="__codelineno-0-1029"></a><span class="sd">        :param axis: The axis on which to perform the alignment (can be `sample` for aligning individuals or</span>
</span><span id="__span-0-1030"><a id="__codelineno-0-1030" name="__codelineno-0-1030"></a><span class="sd">        `SNP` for aligning variants across the datasets).</span>
</span><span id="__span-0-1031"><a id="__codelineno-0-1031" name="__codelineno-0-1031"></a><span class="sd">        :param how: The type of join to perform across the datasets. For now, we support an inner join sort</span>
</span><span id="__span-0-1032"><a id="__codelineno-0-1032" name="__codelineno-0-1032"></a><span class="sd">        of operation.</span>
</span><span id="__span-0-1033"><a id="__codelineno-0-1033" name="__codelineno-0-1033"></a>
</span><span id="__span-0-1034"><a id="__codelineno-0-1034" name="__codelineno-0-1034"></a><span class="sd">        !!! warning</span>
</span><span id="__span-0-1035"><a id="__codelineno-0-1035" name="__codelineno-0-1035"></a><span class="sd">            Experimental for now, would like to add more features here in the near future.</span>
</span><span id="__span-0-1036"><a id="__codelineno-0-1036" name="__codelineno-0-1036"></a>
</span><span id="__span-0-1037"><a id="__codelineno-0-1037" name="__codelineno-0-1037"></a><span class="sd">        &quot;&quot;&quot;</span>
</span><span id="__span-0-1038"><a id="__codelineno-0-1038" name="__codelineno-0-1038"></a>
</span><span id="__span-0-1039"><a id="__codelineno-0-1039" name="__codelineno-0-1039"></a>        <span class="k">if</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">other_gdls</span><span class="p">,</span> <span class="n">GWADataLoader</span><span class="p">):</span>
</span><span id="__span-0-1040"><a id="__codelineno-0-1040" name="__codelineno-0-1040"></a>            <span class="n">other_gdls</span> <span class="o">=</span> <span class="p">[</span><span class="n">other_gdls</span><span class="p">]</span>
</span><span id="__span-0-1041"><a id="__codelineno-0-1041" name="__codelineno-0-1041"></a>
</span><span id="__span-0-1042"><a id="__codelineno-0-1042" name="__codelineno-0-1042"></a>        <span class="k">assert</span> <span class="nb">all</span><span class="p">([</span><span class="nb">isinstance</span><span class="p">(</span><span class="n">gdl</span><span class="p">,</span> <span class="n">GWADataLoader</span><span class="p">)</span> <span class="k">for</span> <span class="n">gdl</span> <span class="ow">in</span> <span class="n">other_gdls</span><span class="p">])</span>
</span><span id="__span-0-1043"><a id="__codelineno-0-1043" name="__codelineno-0-1043"></a>
</span><span id="__span-0-1044"><a id="__codelineno-0-1044" name="__codelineno-0-1044"></a>        <span class="k">if</span> <span class="n">axis</span> <span class="o">==</span> <span class="s1">&#39;SNP&#39;</span><span class="p">:</span>
</span><span id="__span-0-1045"><a id="__codelineno-0-1045" name="__codelineno-0-1045"></a>            <span class="c1"># Ensure that all the GDLs have the same set of SNPs.</span>
</span><span id="__span-0-1046"><a id="__codelineno-0-1046" name="__codelineno-0-1046"></a>            <span class="c1"># This may be useful if the goal is to select a common set of variants</span>
</span><span id="__span-0-1047"><a id="__codelineno-0-1047" name="__codelineno-0-1047"></a>            <span class="c1"># that are shared across different datasets.</span>
</span><span id="__span-0-1048"><a id="__codelineno-0-1048" name="__codelineno-0-1048"></a>            <span class="k">for</span> <span class="n">c</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">chromosomes</span><span class="p">:</span>
</span><span id="__span-0-1049"><a id="__codelineno-0-1049" name="__codelineno-0-1049"></a>                <span class="n">common_snps</span> <span class="o">=</span> <span class="nb">set</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">snps</span><span class="p">[</span><span class="n">c</span><span class="p">])</span>
</span><span id="__span-0-1050"><a id="__codelineno-0-1050" name="__codelineno-0-1050"></a>                <span class="k">for</span> <span class="n">gdl</span> <span class="ow">in</span> <span class="n">other_gdls</span><span class="p">:</span>
</span><span id="__span-0-1051"><a id="__codelineno-0-1051" name="__codelineno-0-1051"></a>                    <span class="n">common_snps</span> <span class="o">=</span> <span class="n">common_snps</span><span class="o">.</span><span class="n">intersection</span><span class="p">(</span><span class="nb">set</span><span class="p">(</span><span class="n">gdl</span><span class="o">.</span><span class="n">snps</span><span class="p">[</span><span class="n">c</span><span class="p">]))</span>
</span><span id="__span-0-1052"><a id="__codelineno-0-1052" name="__codelineno-0-1052"></a>
</span><span id="__span-0-1053"><a id="__codelineno-0-1053" name="__codelineno-0-1053"></a>                <span class="n">common_snps</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="nb">list</span><span class="p">(</span><span class="n">common_snps</span><span class="p">))</span>
</span><span id="__span-0-1054"><a id="__codelineno-0-1054" name="__codelineno-0-1054"></a>
</span><span id="__span-0-1055"><a id="__codelineno-0-1055" name="__codelineno-0-1055"></a>                <span class="k">for</span> <span class="n">gdl</span> <span class="ow">in</span> <span class="n">other_gdls</span><span class="p">:</span>
</span><span id="__span-0-1056"><a id="__codelineno-0-1056" name="__codelineno-0-1056"></a>                    <span class="n">gdl</span><span class="o">.</span><span class="n">filter_snps</span><span class="p">(</span><span class="n">extract_snps</span><span class="o">=</span><span class="n">common_snps</span><span class="p">,</span> <span class="n">chromosome</span><span class="o">=</span><span class="n">c</span><span class="p">)</span>
</span><span id="__span-0-1057"><a id="__codelineno-0-1057" name="__codelineno-0-1057"></a>
</span><span id="__span-0-1058"><a id="__codelineno-0-1058" name="__codelineno-0-1058"></a>                <span class="bp">self</span><span class="o">.</span><span class="n">filter_snps</span><span class="p">(</span><span class="n">extract_snps</span><span class="o">=</span><span class="n">common_snps</span><span class="p">,</span> <span class="n">chromosome</span><span class="o">=</span><span class="n">c</span><span class="p">)</span>
</span><span id="__span-0-1059"><a id="__codelineno-0-1059" name="__codelineno-0-1059"></a>
</span><span id="__span-0-1060"><a id="__codelineno-0-1060" name="__codelineno-0-1060"></a>        <span class="k">elif</span> <span class="n">axis</span> <span class="o">==</span> <span class="s1">&#39;sample&#39;</span><span class="p">:</span>
</span><span id="__span-0-1061"><a id="__codelineno-0-1061" name="__codelineno-0-1061"></a>            <span class="c1"># Ensure that all the GDLs have the same set of samples.</span>
</span><span id="__span-0-1062"><a id="__codelineno-0-1062" name="__codelineno-0-1062"></a>            <span class="c1"># This may be useful when different GDLs have different covariates, phenotypes,</span>
</span><span id="__span-0-1063"><a id="__codelineno-0-1063" name="__codelineno-0-1063"></a>            <span class="c1"># or other information pertaining to the individuals.</span>
</span><span id="__span-0-1064"><a id="__codelineno-0-1064" name="__codelineno-0-1064"></a>
</span><span id="__span-0-1065"><a id="__codelineno-0-1065" name="__codelineno-0-1065"></a>            <span class="n">common_samples</span> <span class="o">=</span> <span class="nb">set</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">samples</span><span class="p">)</span>
</span><span id="__span-0-1066"><a id="__codelineno-0-1066" name="__codelineno-0-1066"></a>
</span><span id="__span-0-1067"><a id="__codelineno-0-1067" name="__codelineno-0-1067"></a>            <span class="k">for</span> <span class="n">gdl</span> <span class="ow">in</span> <span class="n">other_gdls</span><span class="p">:</span>
</span><span id="__span-0-1068"><a id="__codelineno-0-1068" name="__codelineno-0-1068"></a>                <span class="n">common_samples</span> <span class="o">=</span> <span class="n">common_samples</span><span class="o">.</span><span class="n">intersection</span><span class="p">(</span><span class="nb">set</span><span class="p">(</span><span class="n">gdl</span><span class="o">.</span><span class="n">samples</span><span class="p">))</span>
</span><span id="__span-0-1069"><a id="__codelineno-0-1069" name="__codelineno-0-1069"></a>
</span><span id="__span-0-1070"><a id="__codelineno-0-1070" name="__codelineno-0-1070"></a>            <span class="n">common_samples</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="nb">list</span><span class="p">(</span><span class="n">common_samples</span><span class="p">))</span>
</span><span id="__span-0-1071"><a id="__codelineno-0-1071" name="__codelineno-0-1071"></a>
</span><span id="__span-0-1072"><a id="__codelineno-0-1072" name="__codelineno-0-1072"></a>            <span class="k">for</span> <span class="n">gdl</span> <span class="ow">in</span> <span class="n">other_gdls</span><span class="p">:</span>
</span><span id="__span-0-1073"><a id="__codelineno-0-1073" name="__codelineno-0-1073"></a>                <span class="n">gdl</span><span class="o">.</span><span class="n">filter_samples</span><span class="p">(</span><span class="n">keep_samples</span><span class="o">=</span><span class="n">common_samples</span><span class="p">)</span>
</span><span id="__span-0-1074"><a id="__codelineno-0-1074" name="__codelineno-0-1074"></a>
</span><span id="__span-0-1075"><a id="__codelineno-0-1075" name="__codelineno-0-1075"></a>            <span class="bp">self</span><span class="o">.</span><span class="n">filter_samples</span><span class="p">(</span><span class="n">keep_samples</span><span class="o">=</span><span class="n">common_samples</span><span class="p">)</span>
</span><span id="__span-0-1076"><a id="__codelineno-0-1076" name="__codelineno-0-1076"></a>
</span><span id="__span-0-1077"><a id="__codelineno-0-1077" name="__codelineno-0-1077"></a>        <span class="k">else</span><span class="p">:</span>
</span><span id="__span-0-1078"><a id="__codelineno-0-1078" name="__codelineno-0-1078"></a>            <span class="k">raise</span> <span class="ne">KeyError</span><span class="p">(</span><span class="s2">&quot;Alignment axis can only be either &#39;SNP&#39; or &#39;sample&#39;!&quot;</span><span class="p">)</span>
</span><span id="__span-0-1079"><a id="__codelineno-0-1079" name="__codelineno-0-1079"></a>
</span><span id="__span-0-1080"><a id="__codelineno-0-1080" name="__codelineno-0-1080"></a>    <span class="k">def</span><span class="w"> </span><span class="nf">cleanup</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
</span><span id="__span-0-1081"><a id="__codelineno-0-1081" name="__codelineno-0-1081"></a><span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
</span><span id="__span-0-1082"><a id="__codelineno-0-1082" name="__codelineno-0-1082"></a><span class="sd">        Clean up all temporary files and directories</span>
</span><span id="__span-0-1083"><a id="__codelineno-0-1083" name="__codelineno-0-1083"></a><span class="sd">        &quot;&quot;&quot;</span>
</span><span id="__span-0-1084"><a id="__codelineno-0-1084" name="__codelineno-0-1084"></a>
</span><span id="__span-0-1085"><a id="__codelineno-0-1085" name="__codelineno-0-1085"></a>        <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;&gt; Cleaning up workspace.&quot;</span><span class="p">)</span>
</span><span id="__span-0-1086"><a id="__codelineno-0-1086" name="__codelineno-0-1086"></a>
</span><span id="__span-0-1087"><a id="__codelineno-0-1087" name="__codelineno-0-1087"></a>        <span class="k">for</span> <span class="n">tmpdir</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">cleanup_dir_list</span><span class="p">:</span>
</span><span id="__span-0-1088"><a id="__codelineno-0-1088" name="__codelineno-0-1088"></a>            <span class="k">try</span><span class="p">:</span>
</span><span id="__span-0-1089"><a id="__codelineno-0-1089" name="__codelineno-0-1089"></a>                <span class="n">tmpdir</span><span class="o">.</span><span class="n">cleanup</span><span class="p">()</span>
</span><span id="__span-0-1090"><a id="__codelineno-0-1090" name="__codelineno-0-1090"></a>            <span class="k">except</span> <span class="ne">FileNotFoundError</span><span class="p">:</span>
</span><span id="__span-0-1091"><a id="__codelineno-0-1091" name="__codelineno-0-1091"></a>                <span class="k">continue</span>
</span><span id="__span-0-1092"><a id="__codelineno-0-1092" name="__codelineno-0-1092"></a>
</span><span id="__span-0-1093"><a id="__codelineno-0-1093" name="__codelineno-0-1093"></a>        <span class="c1"># Clean up the temporary files associated with the genotype matrices:</span>
</span><span id="__span-0-1094"><a id="__codelineno-0-1094" name="__codelineno-0-1094"></a>        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">genotype</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
</span><span id="__span-0-1095"><a id="__codelineno-0-1095" name="__codelineno-0-1095"></a>            <span class="k">for</span> <span class="n">g</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">genotype</span><span class="o">.</span><span class="n">values</span><span class="p">():</span>
</span><span id="__span-0-1096"><a id="__codelineno-0-1096" name="__codelineno-0-1096"></a>                <span class="n">g</span><span class="o">.</span><span class="n">cleanup</span><span class="p">()</span>
</span><span id="__span-0-1097"><a id="__codelineno-0-1097" name="__codelineno-0-1097"></a>
</span><span id="__span-0-1098"><a id="__codelineno-0-1098" name="__codelineno-0-1098"></a>        <span class="c1"># Release the LD data from memory:</span>
</span><span id="__span-0-1099"><a id="__codelineno-0-1099" name="__codelineno-0-1099"></a>        <span class="bp">self</span><span class="o">.</span><span class="n">release_ld</span><span class="p">()</span>
</span></code></pre></div></td></tr></table></div>
              </details>



  <div class="doc doc-children">







<div class="doc doc-object doc-attribute">



<h2 id="magenpy.GWADataLoader.GWADataLoader.chromosomes" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">chromosomes</span></code>

  <span class="doc doc-labels">
      <small class="doc doc-label doc-label-property"><code>property</code></small>
  </span>

<a href="#magenpy.GWADataLoader.GWADataLoader.chromosomes" class="headerlink" title="Permanent link">&para;</a></h2>


    <div class="doc doc-contents ">

        


    <p><span class="doc-section-title">Returns:</span></p>
    <table>
      <thead>
        <tr>
          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
            </td>
            <td>
              <div class="doc-md-description">
                <p>The list of chromosomes that were loaded to <code>GWADataLoader</code>.</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>

    </div>

</div>

<div class="doc doc-object doc-attribute">



<h2 id="magenpy.GWADataLoader.GWADataLoader.m" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">m</span></code>

  <span class="doc doc-labels">
      <small class="doc doc-label doc-label-property"><code>property</code></small>
  </span>

<a href="#magenpy.GWADataLoader.GWADataLoader.m" class="headerlink" title="Permanent link">&para;</a></h2>


    <div class="doc doc-contents ">

        <div class="admonition seealso">
<p class="admonition-title">See Also</p>
<ul>
<li><a class="autorefs autorefs-internal" title="&lt;code class=&quot;highlight language-python&quot;&gt;&lt;span class=&quot;n&quot;&gt;n_snps&lt;/span&gt;&lt;/code&gt;

  &lt;span class=&quot;doc doc-labels&quot;&gt;
      &lt;small class=&quot;doc doc-label doc-label-property&quot;&gt;&lt;code&gt;property&lt;/code&gt;&lt;/small&gt;
  &lt;/span&gt;" href="#magenpy.GWADataLoader.GWADataLoader.n_snps">n_snps</a></li>
</ul>
</div>


    <p><span class="doc-section-title">Returns:</span></p>
    <table>
      <thead>
        <tr>
          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
            </td>
            <td>
              <div class="doc-md-description">
                <p>The number of variants in the harmonized data sources.</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>

    </div>

</div>

<div class="doc doc-object doc-attribute">



<h2 id="magenpy.GWADataLoader.GWADataLoader.n" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">n</span></code>

  <span class="doc doc-labels">
      <small class="doc doc-label doc-label-property"><code>property</code></small>
  </span>

<a href="#magenpy.GWADataLoader.GWADataLoader.n" class="headerlink" title="Permanent link">&para;</a></h2>


    <div class="doc doc-contents ">

        <div class="admonition seealso">
<p class="admonition-title">See Also</p>
<ul>
<li><a class="autorefs autorefs-internal" title="&lt;code class=&quot;highlight language-python&quot;&gt;&lt;span class=&quot;n&quot;&gt;sample_size&lt;/span&gt;&lt;/code&gt;

  &lt;span class=&quot;doc doc-labels&quot;&gt;
      &lt;small class=&quot;doc doc-label doc-label-property&quot;&gt;&lt;code&gt;property&lt;/code&gt;&lt;/small&gt;
  &lt;/span&gt;" href="#magenpy.GWADataLoader.GWADataLoader.sample_size">sample_size</a></li>
</ul>
</div>


    <p><span class="doc-section-title">Returns:</span></p>
    <table>
      <thead>
        <tr>
          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
            </td>
            <td>
              <div class="doc-md-description">
                <p>The number of samples in the genotype matrix.</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>

    </div>

</div>

<div class="doc doc-object doc-attribute">



<h2 id="magenpy.GWADataLoader.GWADataLoader.n_annotations" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">n_annotations</span></code>

  <span class="doc doc-labels">
      <small class="doc doc-label doc-label-property"><code>property</code></small>
  </span>

<a href="#magenpy.GWADataLoader.GWADataLoader.n_annotations" class="headerlink" title="Permanent link">&para;</a></h2>


    <div class="doc doc-contents ">

        


    <p><span class="doc-section-title">Returns:</span></p>
    <table>
      <thead>
        <tr>
          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
            </td>
            <td>
              <div class="doc-md-description">
                <p>The number of annotations included in the annotation matrices.</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>

    </div>

</div>

<div class="doc doc-object doc-attribute">



<h2 id="magenpy.GWADataLoader.GWADataLoader.n_snps" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">n_snps</span></code>

  <span class="doc doc-labels">
      <small class="doc doc-label doc-label-property"><code>property</code></small>
  </span>

<a href="#magenpy.GWADataLoader.GWADataLoader.n_snps" class="headerlink" title="Permanent link">&para;</a></h2>


    <div class="doc doc-contents ">

        <div class="admonition seealso">
<p class="admonition-title">See Also</p>
<ul>
<li><a class="autorefs autorefs-internal" title="&lt;code class=&quot;highlight language-python&quot;&gt;&lt;span class=&quot;n&quot;&gt;m&lt;/span&gt;&lt;/code&gt;

  &lt;span class=&quot;doc doc-labels&quot;&gt;
      &lt;small class=&quot;doc doc-label doc-label-property&quot;&gt;&lt;code&gt;property&lt;/code&gt;&lt;/small&gt;
  &lt;/span&gt;" href="#magenpy.GWADataLoader.GWADataLoader.m">m</a></li>
</ul>
</div>


    <p><span class="doc-section-title">Returns:</span></p>
    <table>
      <thead>
        <tr>
          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
            </td>
            <td>
              <div class="doc-md-description">
                <p>The number of variants in the harmonized data sources.</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>

    </div>

</div>

<div class="doc doc-object doc-attribute">



<h2 id="magenpy.GWADataLoader.GWADataLoader.sample_size" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">sample_size</span></code>

  <span class="doc doc-labels">
      <small class="doc doc-label doc-label-property"><code>property</code></small>
  </span>

<a href="#magenpy.GWADataLoader.GWADataLoader.sample_size" class="headerlink" title="Permanent link">&para;</a></h2>


    <div class="doc doc-contents ">

        <div class="admonition seealso">
<p class="admonition-title">See Also</p>
<ul>
<li><a class="autorefs autorefs-internal" title="&lt;code class=&quot;highlight language-python&quot;&gt;&lt;span class=&quot;n&quot;&gt;n&lt;/span&gt;&lt;/code&gt;

  &lt;span class=&quot;doc doc-labels&quot;&gt;
      &lt;small class=&quot;doc doc-label doc-label-property&quot;&gt;&lt;code&gt;property&lt;/code&gt;&lt;/small&gt;
  &lt;/span&gt;" href="#magenpy.GWADataLoader.GWADataLoader.n">n</a></li>
</ul>
</div>


    <p><span class="doc-section-title">Returns:</span></p>
    <table>
      <thead>
        <tr>
          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
            </td>
            <td>
              <div class="doc-md-description">
                <p>The number of samples in the genotype matrix.</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>

    </div>

</div>

<div class="doc doc-object doc-attribute">



<h2 id="magenpy.GWADataLoader.GWADataLoader.samples" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">samples</span></code>

  <span class="doc doc-labels">
      <small class="doc doc-label doc-label-property"><code>property</code></small>
  </span>

<a href="#magenpy.GWADataLoader.GWADataLoader.samples" class="headerlink" title="Permanent link">&para;</a></h2>


    <div class="doc doc-contents ">

        


    <p><span class="doc-section-title">Returns:</span></p>
    <table>
      <thead>
        <tr>
          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
            </td>
            <td>
              <div class="doc-md-description">
                <p>The list of samples retained in the sample table.</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>

    </div>

</div>

<div class="doc doc-object doc-attribute">



<h2 id="magenpy.GWADataLoader.GWADataLoader.shapes" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">shapes</span></code>

  <span class="doc doc-labels">
      <small class="doc doc-label doc-label-property"><code>property</code></small>
  </span>

<a href="#magenpy.GWADataLoader.GWADataLoader.shapes" class="headerlink" title="Permanent link">&para;</a></h2>


    <div class="doc doc-contents ">

        


    <p><span class="doc-section-title">Returns:</span></p>
    <table>
      <thead>
        <tr>
          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
            </td>
            <td>
              <div class="doc-md-description">
                <p>A dictionary where the key is the chromosome number and the value is the number of variants on that chromosome.</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>

    </div>

</div>

<div class="doc doc-object doc-attribute">



<h2 id="magenpy.GWADataLoader.GWADataLoader.snps" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">snps</span></code>

  <span class="doc doc-labels">
      <small class="doc doc-label doc-label-property"><code>property</code></small>
  </span>

<a href="#magenpy.GWADataLoader.GWADataLoader.snps" class="headerlink" title="Permanent link">&para;</a></h2>


    <div class="doc doc-contents ">

        


    <p><span class="doc-section-title">Returns:</span></p>
    <table>
      <thead>
        <tr>
          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                  <code>dict</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>The list of SNP rsIDs retained in each chromosome.</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>

    </div>

</div>



<div class="doc doc-object doc-function">


<h2 id="magenpy.GWADataLoader.GWADataLoader.__init__" class="doc doc-heading">
            <code class="highlight language-python"><span class="fm">__init__</span><span class="p">(</span><span class="n">bed_files</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">phenotype_file</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">covariates_file</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">keep_samples</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">keep_file</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">extract_snps</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">extract_file</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">min_maf</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">min_mac</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">drop_duplicated</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">phenotype_likelihood</span><span class="o">=</span><span class="s1">&#39;gaussian&#39;</span><span class="p">,</span> <span class="n">sumstats_files</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">sumstats_format</span><span class="o">=</span><span class="s1">&#39;magenpy&#39;</span><span class="p">,</span> <span class="n">ld_store_files</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">annotation_files</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">annotation_format</span><span class="o">=</span><span class="s1">&#39;magenpy&#39;</span><span class="p">,</span> <span class="n">backend</span><span class="o">=</span><span class="s1">&#39;xarray&#39;</span><span class="p">,</span> <span class="n">temp_dir</span><span class="o">=</span><span class="s1">&#39;temp&#39;</span><span class="p">,</span> <span class="n">output_dir</span><span class="o">=</span><span class="s1">&#39;output&#39;</span><span class="p">,</span> <span class="n">threads</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span></code>

<a href="#magenpy.GWADataLoader.GWADataLoader.__init__" class="headerlink" title="Permanent link">&para;</a></h2>


    <div class="doc doc-contents ">

        <p>Initialize the <code>GWADataLoader</code> object with the data sources required for
downstream statistical genetics analyses.</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                <code>bed_files</code>
            </td>
            <td>
            </td>
            <td>
              <div class="doc-md-description">
                <p>The path to the BED file(s). You may use a wildcard here to read files for multiple chromosomes.</p>
              </div>
            </td>
            <td>
                  <code>None</code>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>phenotype_file</code>
            </td>
            <td>
            </td>
            <td>
              <div class="doc-md-description">
                <p>The path to the phenotype file. (Default: tab-separated file with <code>FID IID phenotype</code> columns).</p>
              </div>
            </td>
            <td>
                  <code>None</code>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>covariates_file</code>
            </td>
            <td>
            </td>
            <td>
              <div class="doc-md-description">
                <p>The path to the covariates file. (Default: tab-separated file starting with the <code>FID IID ...</code> columns and followed by the covariate columns).</p>
              </div>
            </td>
            <td>
                  <code>None</code>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>keep_samples</code>
            </td>
            <td>
            </td>
            <td>
              <div class="doc-md-description">
                <p>A vector or list of sample IDs to keep when filtering the genotype matrix.</p>
              </div>
            </td>
            <td>
                  <code>None</code>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>keep_file</code>
            </td>
            <td>
            </td>
            <td>
              <div class="doc-md-description">
                <p>A path to a plink-style keep file to select a subset of individuals.</p>
              </div>
            </td>
            <td>
                  <code>None</code>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>extract_snps</code>
            </td>
            <td>
            </td>
            <td>
              <div class="doc-md-description">
                <p>A vector or list of SNP IDs to keep when filtering the genotype matrix.</p>
              </div>
            </td>
            <td>
                  <code>None</code>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>extract_file</code>
            </td>
            <td>
            </td>
            <td>
              <div class="doc-md-description">
                <p>A path to a plink-style extract file to select a subset of SNPs.</p>
              </div>
            </td>
            <td>
                  <code>None</code>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>min_maf</code>
            </td>
            <td>
            </td>
            <td>
              <div class="doc-md-description">
                <p>The minimum minor allele frequency cutoff.</p>
              </div>
            </td>
            <td>
                  <code>None</code>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>min_mac</code>
            </td>
            <td>
            </td>
            <td>
              <div class="doc-md-description">
                <p>The minimum minor allele count cutoff.</p>
              </div>
            </td>
            <td>
                  <code>None</code>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>drop_duplicated</code>
            </td>
            <td>
            </td>
            <td>
              <div class="doc-md-description">
                <p>If True, drop SNPs with duplicated rsID.</p>
              </div>
            </td>
            <td>
                  <code>True</code>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>phenotype_likelihood</code>
            </td>
            <td>
            </td>
            <td>
              <div class="doc-md-description">
                <p>The likelihood of the phenotype (e.g. <code>gaussian</code>, <code>binomial</code>).</p>
              </div>
            </td>
            <td>
                  <code>&#39;gaussian&#39;</code>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>sumstats_files</code>
            </td>
            <td>
            </td>
            <td>
              <div class="doc-md-description">
                <p>The path to the summary statistics file(s). The path may be a wildcard.</p>
              </div>
            </td>
            <td>
                  <code>None</code>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>sumstats_format</code>
            </td>
            <td>
            </td>
            <td>
              <div class="doc-md-description">
                <p>The format for the summary statistics. Currently, supports the following formats: <code>plink1.9</code>, <code>plink2</code>, <code>magenpy</code>, <code>fastGWA</code>, <code>COJO</code>, <code>SAIGE</code>, or <code>GWASCatalog</code> for the standard summary statistics format (also known as <code>ssf</code> or <code>gwas-ssf</code>).</p>
              </div>
            </td>
            <td>
                  <code>&#39;magenpy&#39;</code>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>ld_store_files</code>
            </td>
            <td>
            </td>
            <td>
              <div class="doc-md-description">
                <p>The path to the LD matrices. This may be a wildcard to accommodate reading data for multiple chromosomes.</p>
              </div>
            </td>
            <td>
                  <code>None</code>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>annotation_files</code>
            </td>
            <td>
            </td>
            <td>
              <div class="doc-md-description">
                <p>The path to the annotation file(s). The path may contain a wildcard.</p>
              </div>
            </td>
            <td>
                  <code>None</code>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>annotation_format</code>
            </td>
            <td>
            </td>
            <td>
              <div class="doc-md-description">
                <p>The format for the summary statistics. Currently, supports the following formats: <code>magenpy</code>, <code>ldsc</code>.</p>
              </div>
            </td>
            <td>
                  <code>&#39;magenpy&#39;</code>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>backend</code>
            </td>
            <td>
            </td>
            <td>
              <div class="doc-md-description">
                <p>The backend software used for computations with the genotype matrix. Currently, supports <code>xarray</code> and <code>plink</code>.</p>
              </div>
            </td>
            <td>
                  <code>&#39;xarray&#39;</code>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>temp_dir</code>
            </td>
            <td>
            </td>
            <td>
              <div class="doc-md-description">
                <p>The temporary directory where to store intermediate files.</p>
              </div>
            </td>
            <td>
                  <code>&#39;temp&#39;</code>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>output_dir</code>
            </td>
            <td>
            </td>
            <td>
              <div class="doc-md-description">
                <p>The output directory where to store the results of the computation.</p>
              </div>
            </td>
            <td>
                  <code>&#39;output&#39;</code>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>threads</code>
            </td>
            <td>
            </td>
            <td>
              <div class="doc-md-description">
                <p>The number of threads to use for computations.</p>
              </div>
            </td>
            <td>
                  <code>1</code>
            </td>
          </tr>
      </tbody>
    </table>


            <details class="quote">
              <summary>Source code in <code>magenpy/GWADataLoader.py</code></summary>
              <div class="language-python highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal"><a href="#__codelineno-0-49"> 49</a></span>
<span class="normal"><a href="#__codelineno-0-50"> 50</a></span>
<span class="normal"><a href="#__codelineno-0-51"> 51</a></span>
<span class="normal"><a href="#__codelineno-0-52"> 52</a></span>
<span class="normal"><a href="#__codelineno-0-53"> 53</a></span>
<span class="normal"><a href="#__codelineno-0-54"> 54</a></span>
<span class="normal"><a href="#__codelineno-0-55"> 55</a></span>
<span class="normal"><a href="#__codelineno-0-56"> 56</a></span>
<span class="normal"><a href="#__codelineno-0-57"> 57</a></span>
<span class="normal"><a href="#__codelineno-0-58"> 58</a></span>
<span class="normal"><a href="#__codelineno-0-59"> 59</a></span>
<span class="normal"><a href="#__codelineno-0-60"> 60</a></span>
<span class="normal"><a href="#__codelineno-0-61"> 61</a></span>
<span class="normal"><a href="#__codelineno-0-62"> 62</a></span>
<span class="normal"><a href="#__codelineno-0-63"> 63</a></span>
<span class="normal"><a href="#__codelineno-0-64"> 64</a></span>
<span class="normal"><a href="#__codelineno-0-65"> 65</a></span>
<span class="normal"><a href="#__codelineno-0-66"> 66</a></span>
<span class="normal"><a href="#__codelineno-0-67"> 67</a></span>
<span class="normal"><a href="#__codelineno-0-68"> 68</a></span>
<span class="normal"><a href="#__codelineno-0-69"> 69</a></span>
<span class="normal"><a href="#__codelineno-0-70"> 70</a></span>
<span class="normal"><a href="#__codelineno-0-71"> 71</a></span>
<span class="normal"><a href="#__codelineno-0-72"> 72</a></span>
<span class="normal"><a href="#__codelineno-0-73"> 73</a></span>
<span class="normal"><a href="#__codelineno-0-74"> 74</a></span>
<span class="normal"><a href="#__codelineno-0-75"> 75</a></span>
<span class="normal"><a href="#__codelineno-0-76"> 76</a></span>
<span class="normal"><a href="#__codelineno-0-77"> 77</a></span>
<span class="normal"><a href="#__codelineno-0-78"> 78</a></span>
<span class="normal"><a href="#__codelineno-0-79"> 79</a></span>
<span class="normal"><a href="#__codelineno-0-80"> 80</a></span>
<span class="normal"><a href="#__codelineno-0-81"> 81</a></span>
<span class="normal"><a href="#__codelineno-0-82"> 82</a></span>
<span class="normal"><a href="#__codelineno-0-83"> 83</a></span>
<span class="normal"><a href="#__codelineno-0-84"> 84</a></span>
<span class="normal"><a href="#__codelineno-0-85"> 85</a></span>
<span class="normal"><a href="#__codelineno-0-86"> 86</a></span>
<span class="normal"><a href="#__codelineno-0-87"> 87</a></span>
<span class="normal"><a href="#__codelineno-0-88"> 88</a></span>
<span class="normal"><a href="#__codelineno-0-89"> 89</a></span>
<span class="normal"><a href="#__codelineno-0-90"> 90</a></span>
<span class="normal"><a href="#__codelineno-0-91"> 91</a></span>
<span class="normal"><a href="#__codelineno-0-92"> 92</a></span>
<span class="normal"><a href="#__codelineno-0-93"> 93</a></span>
<span class="normal"><a href="#__codelineno-0-94"> 94</a></span>
<span class="normal"><a href="#__codelineno-0-95"> 95</a></span>
<span class="normal"><a href="#__codelineno-0-96"> 96</a></span>
<span class="normal"><a href="#__codelineno-0-97"> 97</a></span>
<span class="normal"><a href="#__codelineno-0-98"> 98</a></span>
<span class="normal"><a href="#__codelineno-0-99"> 99</a></span>
<span class="normal"><a href="#__codelineno-0-100">100</a></span>
<span class="normal"><a href="#__codelineno-0-101">101</a></span>
<span class="normal"><a href="#__codelineno-0-102">102</a></span>
<span class="normal"><a href="#__codelineno-0-103">103</a></span>
<span class="normal"><a href="#__codelineno-0-104">104</a></span>
<span class="normal"><a href="#__codelineno-0-105">105</a></span>
<span class="normal"><a href="#__codelineno-0-106">106</a></span>
<span class="normal"><a href="#__codelineno-0-107">107</a></span>
<span class="normal"><a href="#__codelineno-0-108">108</a></span>
<span class="normal"><a href="#__codelineno-0-109">109</a></span>
<span class="normal"><a href="#__codelineno-0-110">110</a></span>
<span class="normal"><a href="#__codelineno-0-111">111</a></span>
<span class="normal"><a href="#__codelineno-0-112">112</a></span>
<span class="normal"><a href="#__codelineno-0-113">113</a></span>
<span class="normal"><a href="#__codelineno-0-114">114</a></span>
<span class="normal"><a href="#__codelineno-0-115">115</a></span>
<span class="normal"><a href="#__codelineno-0-116">116</a></span>
<span class="normal"><a href="#__codelineno-0-117">117</a></span>
<span class="normal"><a href="#__codelineno-0-118">118</a></span>
<span class="normal"><a href="#__codelineno-0-119">119</a></span>
<span class="normal"><a href="#__codelineno-0-120">120</a></span>
<span class="normal"><a href="#__codelineno-0-121">121</a></span>
<span class="normal"><a href="#__codelineno-0-122">122</a></span>
<span class="normal"><a href="#__codelineno-0-123">123</a></span>
<span class="normal"><a href="#__codelineno-0-124">124</a></span>
<span class="normal"><a href="#__codelineno-0-125">125</a></span>
<span class="normal"><a href="#__codelineno-0-126">126</a></span>
<span class="normal"><a href="#__codelineno-0-127">127</a></span>
<span class="normal"><a href="#__codelineno-0-128">128</a></span>
<span class="normal"><a href="#__codelineno-0-129">129</a></span>
<span class="normal"><a href="#__codelineno-0-130">130</a></span>
<span class="normal"><a href="#__codelineno-0-131">131</a></span>
<span class="normal"><a href="#__codelineno-0-132">132</a></span>
<span class="normal"><a href="#__codelineno-0-133">133</a></span>
<span class="normal"><a href="#__codelineno-0-134">134</a></span>
<span class="normal"><a href="#__codelineno-0-135">135</a></span>
<span class="normal"><a href="#__codelineno-0-136">136</a></span>
<span class="normal"><a href="#__codelineno-0-137">137</a></span>
<span class="normal"><a href="#__codelineno-0-138">138</a></span>
<span class="normal"><a href="#__codelineno-0-139">139</a></span>
<span class="normal"><a href="#__codelineno-0-140">140</a></span>
<span class="normal"><a href="#__codelineno-0-141">141</a></span>
<span class="normal"><a href="#__codelineno-0-142">142</a></span>
<span class="normal"><a href="#__codelineno-0-143">143</a></span>
<span class="normal"><a href="#__codelineno-0-144">144</a></span>
<span class="normal"><a href="#__codelineno-0-145">145</a></span>
<span class="normal"><a href="#__codelineno-0-146">146</a></span>
<span class="normal"><a href="#__codelineno-0-147">147</a></span>
<span class="normal"><a href="#__codelineno-0-148">148</a></span>
<span class="normal"><a href="#__codelineno-0-149">149</a></span>
<span class="normal"><a href="#__codelineno-0-150">150</a></span>
<span class="normal"><a href="#__codelineno-0-151">151</a></span>
<span class="normal"><a href="#__codelineno-0-152">152</a></span>
<span class="normal"><a href="#__codelineno-0-153">153</a></span>
<span class="normal"><a href="#__codelineno-0-154">154</a></span>
<span class="normal"><a href="#__codelineno-0-155">155</a></span>
<span class="normal"><a href="#__codelineno-0-156">156</a></span></pre></div></td><td class="code"><div><pre><span></span><code><span id="__span-0-49"><a id="__codelineno-0-49" name="__codelineno-0-49"></a><span class="k">def</span><span class="w"> </span><span class="fm">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span>
</span><span id="__span-0-50"><a id="__codelineno-0-50" name="__codelineno-0-50"></a>             <span class="n">bed_files</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
</span><span id="__span-0-51"><a id="__codelineno-0-51" name="__codelineno-0-51"></a>             <span class="n">phenotype_file</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
</span><span id="__span-0-52"><a id="__codelineno-0-52" name="__codelineno-0-52"></a>             <span class="n">covariates_file</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
</span><span id="__span-0-53"><a id="__codelineno-0-53" name="__codelineno-0-53"></a>             <span class="n">keep_samples</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
</span><span id="__span-0-54"><a id="__codelineno-0-54" name="__codelineno-0-54"></a>             <span class="n">keep_file</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
</span><span id="__span-0-55"><a id="__codelineno-0-55" name="__codelineno-0-55"></a>             <span class="n">extract_snps</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
</span><span id="__span-0-56"><a id="__codelineno-0-56" name="__codelineno-0-56"></a>             <span class="n">extract_file</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
</span><span id="__span-0-57"><a id="__codelineno-0-57" name="__codelineno-0-57"></a>             <span class="n">min_maf</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
</span><span id="__span-0-58"><a id="__codelineno-0-58" name="__codelineno-0-58"></a>             <span class="n">min_mac</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
</span><span id="__span-0-59"><a id="__codelineno-0-59" name="__codelineno-0-59"></a>             <span class="n">drop_duplicated</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
</span><span id="__span-0-60"><a id="__codelineno-0-60" name="__codelineno-0-60"></a>             <span class="n">phenotype_likelihood</span><span class="o">=</span><span class="s1">&#39;gaussian&#39;</span><span class="p">,</span>
</span><span id="__span-0-61"><a id="__codelineno-0-61" name="__codelineno-0-61"></a>             <span class="n">sumstats_files</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
</span><span id="__span-0-62"><a id="__codelineno-0-62" name="__codelineno-0-62"></a>             <span class="n">sumstats_format</span><span class="o">=</span><span class="s1">&#39;magenpy&#39;</span><span class="p">,</span>
</span><span id="__span-0-63"><a id="__codelineno-0-63" name="__codelineno-0-63"></a>             <span class="n">ld_store_files</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
</span><span id="__span-0-64"><a id="__codelineno-0-64" name="__codelineno-0-64"></a>             <span class="n">annotation_files</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
</span><span id="__span-0-65"><a id="__codelineno-0-65" name="__codelineno-0-65"></a>             <span class="n">annotation_format</span><span class="o">=</span><span class="s1">&#39;magenpy&#39;</span><span class="p">,</span>
</span><span id="__span-0-66"><a id="__codelineno-0-66" name="__codelineno-0-66"></a>             <span class="n">backend</span><span class="o">=</span><span class="s1">&#39;xarray&#39;</span><span class="p">,</span>
</span><span id="__span-0-67"><a id="__codelineno-0-67" name="__codelineno-0-67"></a>             <span class="n">temp_dir</span><span class="o">=</span><span class="s1">&#39;temp&#39;</span><span class="p">,</span>
</span><span id="__span-0-68"><a id="__codelineno-0-68" name="__codelineno-0-68"></a>             <span class="n">output_dir</span><span class="o">=</span><span class="s1">&#39;output&#39;</span><span class="p">,</span>
</span><span id="__span-0-69"><a id="__codelineno-0-69" name="__codelineno-0-69"></a>             <span class="n">threads</span><span class="o">=</span><span class="mi">1</span><span class="p">):</span>
</span><span id="__span-0-70"><a id="__codelineno-0-70" name="__codelineno-0-70"></a><span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
</span><span id="__span-0-71"><a id="__codelineno-0-71" name="__codelineno-0-71"></a><span class="sd">    Initialize the `GWADataLoader` object with the data sources required for</span>
</span><span id="__span-0-72"><a id="__codelineno-0-72" name="__codelineno-0-72"></a><span class="sd">    downstream statistical genetics analyses.</span>
</span><span id="__span-0-73"><a id="__codelineno-0-73" name="__codelineno-0-73"></a>
</span><span id="__span-0-74"><a id="__codelineno-0-74" name="__codelineno-0-74"></a><span class="sd">    :param bed_files: The path to the BED file(s). You may use a wildcard here to read files for multiple</span>
</span><span id="__span-0-75"><a id="__codelineno-0-75" name="__codelineno-0-75"></a><span class="sd">    chromosomes.</span>
</span><span id="__span-0-76"><a id="__codelineno-0-76" name="__codelineno-0-76"></a><span class="sd">    :param phenotype_file: The path to the phenotype file.</span>
</span><span id="__span-0-77"><a id="__codelineno-0-77" name="__codelineno-0-77"></a><span class="sd">    (Default: tab-separated file with `FID IID phenotype` columns).</span>
</span><span id="__span-0-78"><a id="__codelineno-0-78" name="__codelineno-0-78"></a><span class="sd">    :param covariates_file: The path to the covariates file.</span>
</span><span id="__span-0-79"><a id="__codelineno-0-79" name="__codelineno-0-79"></a><span class="sd">    (Default: tab-separated file starting with the `FID IID ...` columns and followed by the covariate columns).</span>
</span><span id="__span-0-80"><a id="__codelineno-0-80" name="__codelineno-0-80"></a><span class="sd">    :param keep_samples: A vector or list of sample IDs to keep when filtering the genotype matrix.</span>
</span><span id="__span-0-81"><a id="__codelineno-0-81" name="__codelineno-0-81"></a><span class="sd">    :param keep_file: A path to a plink-style keep file to select a subset of individuals.</span>
</span><span id="__span-0-82"><a id="__codelineno-0-82" name="__codelineno-0-82"></a><span class="sd">    :param extract_snps: A vector or list of SNP IDs to keep when filtering the genotype matrix.</span>
</span><span id="__span-0-83"><a id="__codelineno-0-83" name="__codelineno-0-83"></a><span class="sd">    :param extract_file: A path to a plink-style extract file to select a subset of SNPs.</span>
</span><span id="__span-0-84"><a id="__codelineno-0-84" name="__codelineno-0-84"></a><span class="sd">    :param min_maf: The minimum minor allele frequency cutoff.</span>
</span><span id="__span-0-85"><a id="__codelineno-0-85" name="__codelineno-0-85"></a><span class="sd">    :param min_mac: The minimum minor allele count cutoff.</span>
</span><span id="__span-0-86"><a id="__codelineno-0-86" name="__codelineno-0-86"></a><span class="sd">    :param drop_duplicated: If True, drop SNPs with duplicated rsID.</span>
</span><span id="__span-0-87"><a id="__codelineno-0-87" name="__codelineno-0-87"></a><span class="sd">    :param phenotype_likelihood: The likelihood of the phenotype (e.g. `gaussian`, `binomial`).</span>
</span><span id="__span-0-88"><a id="__codelineno-0-88" name="__codelineno-0-88"></a><span class="sd">    :param sumstats_files: The path to the summary statistics file(s). The path may be a wildcard.</span>
</span><span id="__span-0-89"><a id="__codelineno-0-89" name="__codelineno-0-89"></a><span class="sd">    :param sumstats_format: The format for the summary statistics. Currently, supports the following</span>
</span><span id="__span-0-90"><a id="__codelineno-0-90" name="__codelineno-0-90"></a><span class="sd">    formats: `plink1.9`, `plink2`, `magenpy`, `fastGWA`, `COJO`, `SAIGE`, or `GWASCatalog` for the standard</span>
</span><span id="__span-0-91"><a id="__codelineno-0-91" name="__codelineno-0-91"></a><span class="sd">    summary statistics format (also known as `ssf` or `gwas-ssf`).</span>
</span><span id="__span-0-92"><a id="__codelineno-0-92" name="__codelineno-0-92"></a><span class="sd">    :param ld_store_files: The path to the LD matrices. This may be a wildcard to accommodate reading data</span>
</span><span id="__span-0-93"><a id="__codelineno-0-93" name="__codelineno-0-93"></a><span class="sd">    for multiple chromosomes.</span>
</span><span id="__span-0-94"><a id="__codelineno-0-94" name="__codelineno-0-94"></a><span class="sd">    :param annotation_files: The path to the annotation file(s). The path may contain a wildcard.</span>
</span><span id="__span-0-95"><a id="__codelineno-0-95" name="__codelineno-0-95"></a><span class="sd">    :param annotation_format: The format for the summary statistics. Currently, supports the following</span>
</span><span id="__span-0-96"><a id="__codelineno-0-96" name="__codelineno-0-96"></a><span class="sd">    formats: `magenpy`, `ldsc`.</span>
</span><span id="__span-0-97"><a id="__codelineno-0-97" name="__codelineno-0-97"></a><span class="sd">    :param backend: The backend software used for computations with the genotype matrix. Currently, supports</span>
</span><span id="__span-0-98"><a id="__codelineno-0-98" name="__codelineno-0-98"></a><span class="sd">    `xarray` and `plink`.</span>
</span><span id="__span-0-99"><a id="__codelineno-0-99" name="__codelineno-0-99"></a><span class="sd">    :param temp_dir: The temporary directory where to store intermediate files.</span>
</span><span id="__span-0-100"><a id="__codelineno-0-100" name="__codelineno-0-100"></a><span class="sd">    :param output_dir: The output directory where to store the results of the computation.</span>
</span><span id="__span-0-101"><a id="__codelineno-0-101" name="__codelineno-0-101"></a><span class="sd">    :param threads: The number of threads to use for computations.</span>
</span><span id="__span-0-102"><a id="__codelineno-0-102" name="__codelineno-0-102"></a><span class="sd">    &quot;&quot;&quot;</span>
</span><span id="__span-0-103"><a id="__codelineno-0-103" name="__codelineno-0-103"></a>
</span><span id="__span-0-104"><a id="__codelineno-0-104" name="__codelineno-0-104"></a>    <span class="c1"># ------- Sanity checks -------</span>
</span><span id="__span-0-105"><a id="__codelineno-0-105" name="__codelineno-0-105"></a>
</span><span id="__span-0-106"><a id="__codelineno-0-106" name="__codelineno-0-106"></a>    <span class="k">assert</span> <span class="n">backend</span> <span class="ow">in</span> <span class="p">(</span><span class="s1">&#39;xarray&#39;</span><span class="p">,</span> <span class="s1">&#39;plink&#39;</span><span class="p">,</span> <span class="s1">&#39;bed-reader&#39;</span><span class="p">)</span>
</span><span id="__span-0-107"><a id="__codelineno-0-107" name="__codelineno-0-107"></a>    <span class="k">assert</span> <span class="n">phenotype_likelihood</span> <span class="ow">in</span> <span class="p">(</span><span class="s1">&#39;gaussian&#39;</span><span class="p">,</span> <span class="s1">&#39;binomial&#39;</span><span class="p">)</span>
</span><span id="__span-0-108"><a id="__codelineno-0-108" name="__codelineno-0-108"></a>
</span><span id="__span-0-109"><a id="__codelineno-0-109" name="__codelineno-0-109"></a>    <span class="c1"># ------- General options -------</span>
</span><span id="__span-0-110"><a id="__codelineno-0-110" name="__codelineno-0-110"></a>
</span><span id="__span-0-111"><a id="__codelineno-0-111" name="__codelineno-0-111"></a>    <span class="bp">self</span><span class="o">.</span><span class="n">backend</span> <span class="o">=</span> <span class="n">backend</span>
</span><span id="__span-0-112"><a id="__codelineno-0-112" name="__codelineno-0-112"></a>
</span><span id="__span-0-113"><a id="__codelineno-0-113" name="__codelineno-0-113"></a>    <span class="bp">self</span><span class="o">.</span><span class="n">temp_dir</span> <span class="o">=</span> <span class="n">temp_dir</span>
</span><span id="__span-0-114"><a id="__codelineno-0-114" name="__codelineno-0-114"></a>    <span class="bp">self</span><span class="o">.</span><span class="n">output_dir</span> <span class="o">=</span> <span class="n">output_dir</span>
</span><span id="__span-0-115"><a id="__codelineno-0-115" name="__codelineno-0-115"></a>    <span class="bp">self</span><span class="o">.</span><span class="n">cleanup_dir_list</span> <span class="o">=</span> <span class="p">[]</span>  <span class="c1"># Directories to clean up after execution.</span>
</span><span id="__span-0-116"><a id="__codelineno-0-116" name="__codelineno-0-116"></a>
</span><span id="__span-0-117"><a id="__codelineno-0-117" name="__codelineno-0-117"></a>    <span class="n">makedir</span><span class="p">([</span><span class="n">temp_dir</span><span class="p">,</span> <span class="n">output_dir</span><span class="p">])</span>
</span><span id="__span-0-118"><a id="__codelineno-0-118" name="__codelineno-0-118"></a>
</span><span id="__span-0-119"><a id="__codelineno-0-119" name="__codelineno-0-119"></a>    <span class="bp">self</span><span class="o">.</span><span class="n">threads</span> <span class="o">=</span> <span class="n">threads</span>
</span><span id="__span-0-120"><a id="__codelineno-0-120" name="__codelineno-0-120"></a>
</span><span id="__span-0-121"><a id="__codelineno-0-121" name="__codelineno-0-121"></a>    <span class="c1"># ------- General parameters -------</span>
</span><span id="__span-0-122"><a id="__codelineno-0-122" name="__codelineno-0-122"></a>
</span><span id="__span-0-123"><a id="__codelineno-0-123" name="__codelineno-0-123"></a>    <span class="bp">self</span><span class="o">.</span><span class="n">phenotype_likelihood</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="n">phenotype_likelihood</span>
</span><span id="__span-0-124"><a id="__codelineno-0-124" name="__codelineno-0-124"></a>
</span><span id="__span-0-125"><a id="__codelineno-0-125" name="__codelineno-0-125"></a>    <span class="bp">self</span><span class="o">.</span><span class="n">genotype</span><span class="p">:</span> <span class="n">Union</span><span class="p">[</span><span class="n">Dict</span><span class="p">[</span><span class="nb">int</span><span class="p">,</span> <span class="n">GenotypeMatrix</span><span class="p">],</span> <span class="kc">None</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span>
</span><span id="__span-0-126"><a id="__codelineno-0-126" name="__codelineno-0-126"></a>    <span class="bp">self</span><span class="o">.</span><span class="n">sample_table</span><span class="p">:</span> <span class="n">Union</span><span class="p">[</span><span class="n">SampleTable</span><span class="p">,</span> <span class="kc">None</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span>
</span><span id="__span-0-127"><a id="__codelineno-0-127" name="__codelineno-0-127"></a>    <span class="bp">self</span><span class="o">.</span><span class="n">ld</span><span class="p">:</span> <span class="n">Union</span><span class="p">[</span><span class="n">Dict</span><span class="p">[</span><span class="nb">int</span><span class="p">,</span> <span class="n">LDMatrix</span><span class="p">],</span> <span class="kc">None</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span>
</span><span id="__span-0-128"><a id="__codelineno-0-128" name="__codelineno-0-128"></a>    <span class="bp">self</span><span class="o">.</span><span class="n">sumstats_table</span><span class="p">:</span> <span class="n">Union</span><span class="p">[</span><span class="n">Dict</span><span class="p">[</span><span class="nb">int</span><span class="p">,</span> <span class="n">SumstatsTable</span><span class="p">],</span> <span class="kc">None</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span>
</span><span id="__span-0-129"><a id="__codelineno-0-129" name="__codelineno-0-129"></a>    <span class="bp">self</span><span class="o">.</span><span class="n">annotation</span><span class="p">:</span> <span class="n">Union</span><span class="p">[</span><span class="n">Dict</span><span class="p">[</span><span class="nb">int</span><span class="p">,</span> <span class="n">AnnotationMatrix</span><span class="p">],</span> <span class="kc">None</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span>
</span><span id="__span-0-130"><a id="__codelineno-0-130" name="__codelineno-0-130"></a>
</span><span id="__span-0-131"><a id="__codelineno-0-131" name="__codelineno-0-131"></a>    <span class="c1"># ------- Read data files -------</span>
</span><span id="__span-0-132"><a id="__codelineno-0-132" name="__codelineno-0-132"></a>
</span><span id="__span-0-133"><a id="__codelineno-0-133" name="__codelineno-0-133"></a>    <span class="bp">self</span><span class="o">.</span><span class="n">read_genotypes</span><span class="p">(</span><span class="n">bed_files</span><span class="p">,</span>
</span><span id="__span-0-134"><a id="__codelineno-0-134" name="__codelineno-0-134"></a>                        <span class="n">min_maf</span><span class="o">=</span><span class="n">min_maf</span><span class="p">,</span>
</span><span id="__span-0-135"><a id="__codelineno-0-135" name="__codelineno-0-135"></a>                        <span class="n">min_mac</span><span class="o">=</span><span class="n">min_mac</span><span class="p">,</span>
</span><span id="__span-0-136"><a id="__codelineno-0-136" name="__codelineno-0-136"></a>                        <span class="n">drop_duplicated</span><span class="o">=</span><span class="n">drop_duplicated</span><span class="p">)</span>
</span><span id="__span-0-137"><a id="__codelineno-0-137" name="__codelineno-0-137"></a>    <span class="bp">self</span><span class="o">.</span><span class="n">read_phenotype</span><span class="p">(</span><span class="n">phenotype_file</span><span class="p">)</span>
</span><span id="__span-0-138"><a id="__codelineno-0-138" name="__codelineno-0-138"></a>    <span class="bp">self</span><span class="o">.</span><span class="n">read_covariates</span><span class="p">(</span><span class="n">covariates_file</span><span class="p">)</span>
</span><span id="__span-0-139"><a id="__codelineno-0-139" name="__codelineno-0-139"></a>    <span class="bp">self</span><span class="o">.</span><span class="n">read_ld</span><span class="p">(</span><span class="n">ld_store_files</span><span class="p">)</span>
</span><span id="__span-0-140"><a id="__codelineno-0-140" name="__codelineno-0-140"></a>    <span class="bp">self</span><span class="o">.</span><span class="n">read_annotations</span><span class="p">(</span><span class="n">annotation_files</span><span class="p">,</span>
</span><span id="__span-0-141"><a id="__codelineno-0-141" name="__codelineno-0-141"></a>                          <span class="n">annot_format</span><span class="o">=</span><span class="n">annotation_format</span><span class="p">)</span>
</span><span id="__span-0-142"><a id="__codelineno-0-142" name="__codelineno-0-142"></a>    <span class="bp">self</span><span class="o">.</span><span class="n">read_summary_statistics</span><span class="p">(</span><span class="n">sumstats_files</span><span class="p">,</span>
</span><span id="__span-0-143"><a id="__codelineno-0-143" name="__codelineno-0-143"></a>                                 <span class="n">sumstats_format</span><span class="p">,</span>
</span><span id="__span-0-144"><a id="__codelineno-0-144" name="__codelineno-0-144"></a>                                 <span class="n">drop_duplicated</span><span class="o">=</span><span class="n">drop_duplicated</span><span class="p">)</span>
</span><span id="__span-0-145"><a id="__codelineno-0-145" name="__codelineno-0-145"></a>
</span><span id="__span-0-146"><a id="__codelineno-0-146" name="__codelineno-0-146"></a>    <span class="c1"># ------- Filter samples or SNPs -------</span>
</span><span id="__span-0-147"><a id="__codelineno-0-147" name="__codelineno-0-147"></a>
</span><span id="__span-0-148"><a id="__codelineno-0-148" name="__codelineno-0-148"></a>    <span class="k">if</span> <span class="n">extract_snps</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span> <span class="ow">or</span> <span class="n">extract_file</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
</span><span id="__span-0-149"><a id="__codelineno-0-149" name="__codelineno-0-149"></a>        <span class="bp">self</span><span class="o">.</span><span class="n">filter_snps</span><span class="p">(</span><span class="n">extract_snps</span><span class="o">=</span><span class="n">extract_snps</span><span class="p">,</span> <span class="n">extract_file</span><span class="o">=</span><span class="n">extract_file</span><span class="p">)</span>
</span><span id="__span-0-150"><a id="__codelineno-0-150" name="__codelineno-0-150"></a>
</span><span id="__span-0-151"><a id="__codelineno-0-151" name="__codelineno-0-151"></a>    <span class="k">if</span> <span class="n">keep_samples</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span> <span class="ow">or</span> <span class="n">keep_file</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
</span><span id="__span-0-152"><a id="__codelineno-0-152" name="__codelineno-0-152"></a>        <span class="bp">self</span><span class="o">.</span><span class="n">filter_samples</span><span class="p">(</span><span class="n">keep_samples</span><span class="o">=</span><span class="n">keep_samples</span><span class="p">,</span> <span class="n">keep_file</span><span class="o">=</span><span class="n">keep_file</span><span class="p">)</span>
</span><span id="__span-0-153"><a id="__codelineno-0-153" name="__codelineno-0-153"></a>
</span><span id="__span-0-154"><a id="__codelineno-0-154" name="__codelineno-0-154"></a>    <span class="c1"># ------- Harmonize data sources -------</span>
</span><span id="__span-0-155"><a id="__codelineno-0-155" name="__codelineno-0-155"></a>
</span><span id="__span-0-156"><a id="__codelineno-0-156" name="__codelineno-0-156"></a>    <span class="bp">self</span><span class="o">.</span><span class="n">harmonize_data</span><span class="p">()</span>
</span></code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h2 id="magenpy.GWADataLoader.GWADataLoader.align_with" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">align_with</span><span class="p">(</span><span class="n">other_gdls</span><span class="p">,</span> <span class="n">axis</span><span class="o">=</span><span class="s1">&#39;SNP&#39;</span><span class="p">,</span> <span class="n">how</span><span class="o">=</span><span class="s1">&#39;inner&#39;</span><span class="p">)</span></code>

<a href="#magenpy.GWADataLoader.GWADataLoader.align_with" class="headerlink" title="Permanent link">&para;</a></h2>


    <div class="doc doc-contents ">

        <p>Align the <code>GWADataLoader</code> object with other GDL objects to have the same
set of SNPs or samples. This utility method is meant to enable the user to
align multiple data sources for downstream analyses.</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                <code>other_gdls</code>
            </td>
            <td>
            </td>
            <td>
              <div class="doc-md-description">
                <p>A <code>GWADataLoader</code> or list of <code>GWADataLoader</code> objects.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>axis</code>
            </td>
            <td>
            </td>
            <td>
              <div class="doc-md-description">
                <p>The axis on which to perform the alignment (can be <code>sample</code> for aligning individuals or <code>SNP</code> for aligning variants across the datasets).</p>
              </div>
            </td>
            <td>
                  <code>&#39;SNP&#39;</code>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>how</code>
            </td>
            <td>
            </td>
            <td>
              <div class="doc-md-description">
                <p>The type of join to perform across the datasets. For now, we support an inner join sort of operation.  !!! warning Experimental for now, would like to add more features here in the near future.</p>
              </div>
            </td>
            <td>
                  <code>&#39;inner&#39;</code>
            </td>
          </tr>
      </tbody>
    </table>


            <details class="quote">
              <summary>Source code in <code>magenpy/GWADataLoader.py</code></summary>
              <div class="language-python highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal"><a href="#__codelineno-0-1022">1022</a></span>
<span class="normal"><a href="#__codelineno-0-1023">1023</a></span>
<span class="normal"><a href="#__codelineno-0-1024">1024</a></span>
<span class="normal"><a href="#__codelineno-0-1025">1025</a></span>
<span class="normal"><a href="#__codelineno-0-1026">1026</a></span>
<span class="normal"><a href="#__codelineno-0-1027">1027</a></span>
<span class="normal"><a href="#__codelineno-0-1028">1028</a></span>
<span class="normal"><a href="#__codelineno-0-1029">1029</a></span>
<span class="normal"><a href="#__codelineno-0-1030">1030</a></span>
<span class="normal"><a href="#__codelineno-0-1031">1031</a></span>
<span class="normal"><a href="#__codelineno-0-1032">1032</a></span>
<span class="normal"><a href="#__codelineno-0-1033">1033</a></span>
<span class="normal"><a href="#__codelineno-0-1034">1034</a></span>
<span class="normal"><a href="#__codelineno-0-1035">1035</a></span>
<span class="normal"><a href="#__codelineno-0-1036">1036</a></span>
<span class="normal"><a href="#__codelineno-0-1037">1037</a></span>
<span class="normal"><a href="#__codelineno-0-1038">1038</a></span>
<span class="normal"><a href="#__codelineno-0-1039">1039</a></span>
<span class="normal"><a href="#__codelineno-0-1040">1040</a></span>
<span class="normal"><a href="#__codelineno-0-1041">1041</a></span>
<span class="normal"><a href="#__codelineno-0-1042">1042</a></span>
<span class="normal"><a href="#__codelineno-0-1043">1043</a></span>
<span class="normal"><a href="#__codelineno-0-1044">1044</a></span>
<span class="normal"><a href="#__codelineno-0-1045">1045</a></span>
<span class="normal"><a href="#__codelineno-0-1046">1046</a></span>
<span class="normal"><a href="#__codelineno-0-1047">1047</a></span>
<span class="normal"><a href="#__codelineno-0-1048">1048</a></span>
<span class="normal"><a href="#__codelineno-0-1049">1049</a></span>
<span class="normal"><a href="#__codelineno-0-1050">1050</a></span>
<span class="normal"><a href="#__codelineno-0-1051">1051</a></span>
<span class="normal"><a href="#__codelineno-0-1052">1052</a></span>
<span class="normal"><a href="#__codelineno-0-1053">1053</a></span>
<span class="normal"><a href="#__codelineno-0-1054">1054</a></span>
<span class="normal"><a href="#__codelineno-0-1055">1055</a></span>
<span class="normal"><a href="#__codelineno-0-1056">1056</a></span>
<span class="normal"><a href="#__codelineno-0-1057">1057</a></span>
<span class="normal"><a href="#__codelineno-0-1058">1058</a></span>
<span class="normal"><a href="#__codelineno-0-1059">1059</a></span>
<span class="normal"><a href="#__codelineno-0-1060">1060</a></span>
<span class="normal"><a href="#__codelineno-0-1061">1061</a></span>
<span class="normal"><a href="#__codelineno-0-1062">1062</a></span>
<span class="normal"><a href="#__codelineno-0-1063">1063</a></span>
<span class="normal"><a href="#__codelineno-0-1064">1064</a></span>
<span class="normal"><a href="#__codelineno-0-1065">1065</a></span>
<span class="normal"><a href="#__codelineno-0-1066">1066</a></span>
<span class="normal"><a href="#__codelineno-0-1067">1067</a></span>
<span class="normal"><a href="#__codelineno-0-1068">1068</a></span>
<span class="normal"><a href="#__codelineno-0-1069">1069</a></span>
<span class="normal"><a href="#__codelineno-0-1070">1070</a></span>
<span class="normal"><a href="#__codelineno-0-1071">1071</a></span>
<span class="normal"><a href="#__codelineno-0-1072">1072</a></span>
<span class="normal"><a href="#__codelineno-0-1073">1073</a></span>
<span class="normal"><a href="#__codelineno-0-1074">1074</a></span>
<span class="normal"><a href="#__codelineno-0-1075">1075</a></span>
<span class="normal"><a href="#__codelineno-0-1076">1076</a></span>
<span class="normal"><a href="#__codelineno-0-1077">1077</a></span>
<span class="normal"><a href="#__codelineno-0-1078">1078</a></span></pre></div></td><td class="code"><div><pre><span></span><code><span id="__span-0-1022"><a id="__codelineno-0-1022" name="__codelineno-0-1022"></a><span class="k">def</span><span class="w"> </span><span class="nf">align_with</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">other_gdls</span><span class="p">,</span> <span class="n">axis</span><span class="o">=</span><span class="s1">&#39;SNP&#39;</span><span class="p">,</span> <span class="n">how</span><span class="o">=</span><span class="s1">&#39;inner&#39;</span><span class="p">):</span>
</span><span id="__span-0-1023"><a id="__codelineno-0-1023" name="__codelineno-0-1023"></a><span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
</span><span id="__span-0-1024"><a id="__codelineno-0-1024" name="__codelineno-0-1024"></a><span class="sd">    Align the `GWADataLoader` object with other GDL objects to have the same</span>
</span><span id="__span-0-1025"><a id="__codelineno-0-1025" name="__codelineno-0-1025"></a><span class="sd">    set of SNPs or samples. This utility method is meant to enable the user to</span>
</span><span id="__span-0-1026"><a id="__codelineno-0-1026" name="__codelineno-0-1026"></a><span class="sd">    align multiple data sources for downstream analyses.</span>
</span><span id="__span-0-1027"><a id="__codelineno-0-1027" name="__codelineno-0-1027"></a>
</span><span id="__span-0-1028"><a id="__codelineno-0-1028" name="__codelineno-0-1028"></a><span class="sd">    :param other_gdls: A `GWADataLoader` or list of `GWADataLoader` objects.</span>
</span><span id="__span-0-1029"><a id="__codelineno-0-1029" name="__codelineno-0-1029"></a><span class="sd">    :param axis: The axis on which to perform the alignment (can be `sample` for aligning individuals or</span>
</span><span id="__span-0-1030"><a id="__codelineno-0-1030" name="__codelineno-0-1030"></a><span class="sd">    `SNP` for aligning variants across the datasets).</span>
</span><span id="__span-0-1031"><a id="__codelineno-0-1031" name="__codelineno-0-1031"></a><span class="sd">    :param how: The type of join to perform across the datasets. For now, we support an inner join sort</span>
</span><span id="__span-0-1032"><a id="__codelineno-0-1032" name="__codelineno-0-1032"></a><span class="sd">    of operation.</span>
</span><span id="__span-0-1033"><a id="__codelineno-0-1033" name="__codelineno-0-1033"></a>
</span><span id="__span-0-1034"><a id="__codelineno-0-1034" name="__codelineno-0-1034"></a><span class="sd">    !!! warning</span>
</span><span id="__span-0-1035"><a id="__codelineno-0-1035" name="__codelineno-0-1035"></a><span class="sd">        Experimental for now, would like to add more features here in the near future.</span>
</span><span id="__span-0-1036"><a id="__codelineno-0-1036" name="__codelineno-0-1036"></a>
</span><span id="__span-0-1037"><a id="__codelineno-0-1037" name="__codelineno-0-1037"></a><span class="sd">    &quot;&quot;&quot;</span>
</span><span id="__span-0-1038"><a id="__codelineno-0-1038" name="__codelineno-0-1038"></a>
</span><span id="__span-0-1039"><a id="__codelineno-0-1039" name="__codelineno-0-1039"></a>    <span class="k">if</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">other_gdls</span><span class="p">,</span> <span class="n">GWADataLoader</span><span class="p">):</span>
</span><span id="__span-0-1040"><a id="__codelineno-0-1040" name="__codelineno-0-1040"></a>        <span class="n">other_gdls</span> <span class="o">=</span> <span class="p">[</span><span class="n">other_gdls</span><span class="p">]</span>
</span><span id="__span-0-1041"><a id="__codelineno-0-1041" name="__codelineno-0-1041"></a>
</span><span id="__span-0-1042"><a id="__codelineno-0-1042" name="__codelineno-0-1042"></a>    <span class="k">assert</span> <span class="nb">all</span><span class="p">([</span><span class="nb">isinstance</span><span class="p">(</span><span class="n">gdl</span><span class="p">,</span> <span class="n">GWADataLoader</span><span class="p">)</span> <span class="k">for</span> <span class="n">gdl</span> <span class="ow">in</span> <span class="n">other_gdls</span><span class="p">])</span>
</span><span id="__span-0-1043"><a id="__codelineno-0-1043" name="__codelineno-0-1043"></a>
</span><span id="__span-0-1044"><a id="__codelineno-0-1044" name="__codelineno-0-1044"></a>    <span class="k">if</span> <span class="n">axis</span> <span class="o">==</span> <span class="s1">&#39;SNP&#39;</span><span class="p">:</span>
</span><span id="__span-0-1045"><a id="__codelineno-0-1045" name="__codelineno-0-1045"></a>        <span class="c1"># Ensure that all the GDLs have the same set of SNPs.</span>
</span><span id="__span-0-1046"><a id="__codelineno-0-1046" name="__codelineno-0-1046"></a>        <span class="c1"># This may be useful if the goal is to select a common set of variants</span>
</span><span id="__span-0-1047"><a id="__codelineno-0-1047" name="__codelineno-0-1047"></a>        <span class="c1"># that are shared across different datasets.</span>
</span><span id="__span-0-1048"><a id="__codelineno-0-1048" name="__codelineno-0-1048"></a>        <span class="k">for</span> <span class="n">c</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">chromosomes</span><span class="p">:</span>
</span><span id="__span-0-1049"><a id="__codelineno-0-1049" name="__codelineno-0-1049"></a>            <span class="n">common_snps</span> <span class="o">=</span> <span class="nb">set</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">snps</span><span class="p">[</span><span class="n">c</span><span class="p">])</span>
</span><span id="__span-0-1050"><a id="__codelineno-0-1050" name="__codelineno-0-1050"></a>            <span class="k">for</span> <span class="n">gdl</span> <span class="ow">in</span> <span class="n">other_gdls</span><span class="p">:</span>
</span><span id="__span-0-1051"><a id="__codelineno-0-1051" name="__codelineno-0-1051"></a>                <span class="n">common_snps</span> <span class="o">=</span> <span class="n">common_snps</span><span class="o">.</span><span class="n">intersection</span><span class="p">(</span><span class="nb">set</span><span class="p">(</span><span class="n">gdl</span><span class="o">.</span><span class="n">snps</span><span class="p">[</span><span class="n">c</span><span class="p">]))</span>
</span><span id="__span-0-1052"><a id="__codelineno-0-1052" name="__codelineno-0-1052"></a>
</span><span id="__span-0-1053"><a id="__codelineno-0-1053" name="__codelineno-0-1053"></a>            <span class="n">common_snps</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="nb">list</span><span class="p">(</span><span class="n">common_snps</span><span class="p">))</span>
</span><span id="__span-0-1054"><a id="__codelineno-0-1054" name="__codelineno-0-1054"></a>
</span><span id="__span-0-1055"><a id="__codelineno-0-1055" name="__codelineno-0-1055"></a>            <span class="k">for</span> <span class="n">gdl</span> <span class="ow">in</span> <span class="n">other_gdls</span><span class="p">:</span>
</span><span id="__span-0-1056"><a id="__codelineno-0-1056" name="__codelineno-0-1056"></a>                <span class="n">gdl</span><span class="o">.</span><span class="n">filter_snps</span><span class="p">(</span><span class="n">extract_snps</span><span class="o">=</span><span class="n">common_snps</span><span class="p">,</span> <span class="n">chromosome</span><span class="o">=</span><span class="n">c</span><span class="p">)</span>
</span><span id="__span-0-1057"><a id="__codelineno-0-1057" name="__codelineno-0-1057"></a>
</span><span id="__span-0-1058"><a id="__codelineno-0-1058" name="__codelineno-0-1058"></a>            <span class="bp">self</span><span class="o">.</span><span class="n">filter_snps</span><span class="p">(</span><span class="n">extract_snps</span><span class="o">=</span><span class="n">common_snps</span><span class="p">,</span> <span class="n">chromosome</span><span class="o">=</span><span class="n">c</span><span class="p">)</span>
</span><span id="__span-0-1059"><a id="__codelineno-0-1059" name="__codelineno-0-1059"></a>
</span><span id="__span-0-1060"><a id="__codelineno-0-1060" name="__codelineno-0-1060"></a>    <span class="k">elif</span> <span class="n">axis</span> <span class="o">==</span> <span class="s1">&#39;sample&#39;</span><span class="p">:</span>
</span><span id="__span-0-1061"><a id="__codelineno-0-1061" name="__codelineno-0-1061"></a>        <span class="c1"># Ensure that all the GDLs have the same set of samples.</span>
</span><span id="__span-0-1062"><a id="__codelineno-0-1062" name="__codelineno-0-1062"></a>        <span class="c1"># This may be useful when different GDLs have different covariates, phenotypes,</span>
</span><span id="__span-0-1063"><a id="__codelineno-0-1063" name="__codelineno-0-1063"></a>        <span class="c1"># or other information pertaining to the individuals.</span>
</span><span id="__span-0-1064"><a id="__codelineno-0-1064" name="__codelineno-0-1064"></a>
</span><span id="__span-0-1065"><a id="__codelineno-0-1065" name="__codelineno-0-1065"></a>        <span class="n">common_samples</span> <span class="o">=</span> <span class="nb">set</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">samples</span><span class="p">)</span>
</span><span id="__span-0-1066"><a id="__codelineno-0-1066" name="__codelineno-0-1066"></a>
</span><span id="__span-0-1067"><a id="__codelineno-0-1067" name="__codelineno-0-1067"></a>        <span class="k">for</span> <span class="n">gdl</span> <span class="ow">in</span> <span class="n">other_gdls</span><span class="p">:</span>
</span><span id="__span-0-1068"><a id="__codelineno-0-1068" name="__codelineno-0-1068"></a>            <span class="n">common_samples</span> <span class="o">=</span> <span class="n">common_samples</span><span class="o">.</span><span class="n">intersection</span><span class="p">(</span><span class="nb">set</span><span class="p">(</span><span class="n">gdl</span><span class="o">.</span><span class="n">samples</span><span class="p">))</span>
</span><span id="__span-0-1069"><a id="__codelineno-0-1069" name="__codelineno-0-1069"></a>
</span><span id="__span-0-1070"><a id="__codelineno-0-1070" name="__codelineno-0-1070"></a>        <span class="n">common_samples</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="nb">list</span><span class="p">(</span><span class="n">common_samples</span><span class="p">))</span>
</span><span id="__span-0-1071"><a id="__codelineno-0-1071" name="__codelineno-0-1071"></a>
</span><span id="__span-0-1072"><a id="__codelineno-0-1072" name="__codelineno-0-1072"></a>        <span class="k">for</span> <span class="n">gdl</span> <span class="ow">in</span> <span class="n">other_gdls</span><span class="p">:</span>
</span><span id="__span-0-1073"><a id="__codelineno-0-1073" name="__codelineno-0-1073"></a>            <span class="n">gdl</span><span class="o">.</span><span class="n">filter_samples</span><span class="p">(</span><span class="n">keep_samples</span><span class="o">=</span><span class="n">common_samples</span><span class="p">)</span>
</span><span id="__span-0-1074"><a id="__codelineno-0-1074" name="__codelineno-0-1074"></a>
</span><span id="__span-0-1075"><a id="__codelineno-0-1075" name="__codelineno-0-1075"></a>        <span class="bp">self</span><span class="o">.</span><span class="n">filter_samples</span><span class="p">(</span><span class="n">keep_samples</span><span class="o">=</span><span class="n">common_samples</span><span class="p">)</span>
</span><span id="__span-0-1076"><a id="__codelineno-0-1076" name="__codelineno-0-1076"></a>
</span><span id="__span-0-1077"><a id="__codelineno-0-1077" name="__codelineno-0-1077"></a>    <span class="k">else</span><span class="p">:</span>
</span><span id="__span-0-1078"><a id="__codelineno-0-1078" name="__codelineno-0-1078"></a>        <span class="k">raise</span> <span class="ne">KeyError</span><span class="p">(</span><span class="s2">&quot;Alignment axis can only be either &#39;SNP&#39; or &#39;sample&#39;!&quot;</span><span class="p">)</span>
</span></code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h2 id="magenpy.GWADataLoader.GWADataLoader.cleanup" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">cleanup</span><span class="p">()</span></code>

<a href="#magenpy.GWADataLoader.GWADataLoader.cleanup" class="headerlink" title="Permanent link">&para;</a></h2>


    <div class="doc doc-contents ">

        <p>Clean up all temporary files and directories</p>


            <details class="quote">
              <summary>Source code in <code>magenpy/GWADataLoader.py</code></summary>
              <div class="language-python highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal"><a href="#__codelineno-0-1080">1080</a></span>
<span class="normal"><a href="#__codelineno-0-1081">1081</a></span>
<span class="normal"><a href="#__codelineno-0-1082">1082</a></span>
<span class="normal"><a href="#__codelineno-0-1083">1083</a></span>
<span class="normal"><a href="#__codelineno-0-1084">1084</a></span>
<span class="normal"><a href="#__codelineno-0-1085">1085</a></span>
<span class="normal"><a href="#__codelineno-0-1086">1086</a></span>
<span class="normal"><a href="#__codelineno-0-1087">1087</a></span>
<span class="normal"><a href="#__codelineno-0-1088">1088</a></span>
<span class="normal"><a href="#__codelineno-0-1089">1089</a></span>
<span class="normal"><a href="#__codelineno-0-1090">1090</a></span>
<span class="normal"><a href="#__codelineno-0-1091">1091</a></span>
<span class="normal"><a href="#__codelineno-0-1092">1092</a></span>
<span class="normal"><a href="#__codelineno-0-1093">1093</a></span>
<span class="normal"><a href="#__codelineno-0-1094">1094</a></span>
<span class="normal"><a href="#__codelineno-0-1095">1095</a></span>
<span class="normal"><a href="#__codelineno-0-1096">1096</a></span>
<span class="normal"><a href="#__codelineno-0-1097">1097</a></span>
<span class="normal"><a href="#__codelineno-0-1098">1098</a></span>
<span class="normal"><a href="#__codelineno-0-1099">1099</a></span></pre></div></td><td class="code"><div><pre><span></span><code><span id="__span-0-1080"><a id="__codelineno-0-1080" name="__codelineno-0-1080"></a><span class="k">def</span><span class="w"> </span><span class="nf">cleanup</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
</span><span id="__span-0-1081"><a id="__codelineno-0-1081" name="__codelineno-0-1081"></a><span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
</span><span id="__span-0-1082"><a id="__codelineno-0-1082" name="__codelineno-0-1082"></a><span class="sd">    Clean up all temporary files and directories</span>
</span><span id="__span-0-1083"><a id="__codelineno-0-1083" name="__codelineno-0-1083"></a><span class="sd">    &quot;&quot;&quot;</span>
</span><span id="__span-0-1084"><a id="__codelineno-0-1084" name="__codelineno-0-1084"></a>
</span><span id="__span-0-1085"><a id="__codelineno-0-1085" name="__codelineno-0-1085"></a>    <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;&gt; Cleaning up workspace.&quot;</span><span class="p">)</span>
</span><span id="__span-0-1086"><a id="__codelineno-0-1086" name="__codelineno-0-1086"></a>
</span><span id="__span-0-1087"><a id="__codelineno-0-1087" name="__codelineno-0-1087"></a>    <span class="k">for</span> <span class="n">tmpdir</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">cleanup_dir_list</span><span class="p">:</span>
</span><span id="__span-0-1088"><a id="__codelineno-0-1088" name="__codelineno-0-1088"></a>        <span class="k">try</span><span class="p">:</span>
</span><span id="__span-0-1089"><a id="__codelineno-0-1089" name="__codelineno-0-1089"></a>            <span class="n">tmpdir</span><span class="o">.</span><span class="n">cleanup</span><span class="p">()</span>
</span><span id="__span-0-1090"><a id="__codelineno-0-1090" name="__codelineno-0-1090"></a>        <span class="k">except</span> <span class="ne">FileNotFoundError</span><span class="p">:</span>
</span><span id="__span-0-1091"><a id="__codelineno-0-1091" name="__codelineno-0-1091"></a>            <span class="k">continue</span>
</span><span id="__span-0-1092"><a id="__codelineno-0-1092" name="__codelineno-0-1092"></a>
</span><span id="__span-0-1093"><a id="__codelineno-0-1093" name="__codelineno-0-1093"></a>    <span class="c1"># Clean up the temporary files associated with the genotype matrices:</span>
</span><span id="__span-0-1094"><a id="__codelineno-0-1094" name="__codelineno-0-1094"></a>    <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">genotype</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
</span><span id="__span-0-1095"><a id="__codelineno-0-1095" name="__codelineno-0-1095"></a>        <span class="k">for</span> <span class="n">g</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">genotype</span><span class="o">.</span><span class="n">values</span><span class="p">():</span>
</span><span id="__span-0-1096"><a id="__codelineno-0-1096" name="__codelineno-0-1096"></a>            <span class="n">g</span><span class="o">.</span><span class="n">cleanup</span><span class="p">()</span>
</span><span id="__span-0-1097"><a id="__codelineno-0-1097" name="__codelineno-0-1097"></a>
</span><span id="__span-0-1098"><a id="__codelineno-0-1098" name="__codelineno-0-1098"></a>    <span class="c1"># Release the LD data from memory:</span>
</span><span id="__span-0-1099"><a id="__codelineno-0-1099" name="__codelineno-0-1099"></a>    <span class="bp">self</span><span class="o">.</span><span class="n">release_ld</span><span class="p">()</span>
</span></code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h2 id="magenpy.GWADataLoader.GWADataLoader.compute_ld" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">compute_ld</span><span class="p">(</span><span class="n">estimator</span><span class="p">,</span> <span class="n">output_dir</span><span class="p">,</span> <span class="n">dtype</span><span class="o">=</span><span class="s1">&#39;int16&#39;</span><span class="p">,</span> <span class="n">compressor_name</span><span class="o">=</span><span class="s1">&#39;zstd&#39;</span><span class="p">,</span> <span class="n">compression_level</span><span class="o">=</span><span class="mi">7</span><span class="p">,</span> <span class="n">compute_spectral_properties</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="o">**</span><span class="n">ld_kwargs</span><span class="p">)</span></code>

<a href="#magenpy.GWADataLoader.GWADataLoader.compute_ld" class="headerlink" title="Permanent link">&para;</a></h2>


    <div class="doc doc-contents ">

        <p>Compute the Linkage-Disequilibrium (LD) matrix or SNP-by-SNP Pearson
correlation matrix between genetic variants. This function only considers correlations
between SNPs on the same chromosome. This is a utility function that calls the
<code>.compute_ld()</code> method of the <code>GenotypeMatrix</code> objects associated with
GWADataLoader.</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                <code>estimator</code>
            </td>
            <td>
            </td>
            <td>
              <div class="doc-md-description">
                <p>The estimator for the LD matrix. We currently support 4 different estimators: <code>sample</code>, <code>windowed</code>, <code>shrinkage</code>, and <code>block</code>.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>output_dir</code>
            </td>
            <td>
            </td>
            <td>
              <div class="doc-md-description">
                <p>The output directory where the Zarr array containing the entries of the LD matrix will be stored.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>dtype</code>
            </td>
            <td>
            </td>
            <td>
              <div class="doc-md-description">
                <p>The data type for the entries of the LD matrix (supported data types are float32, float64 and integer quantized data types int8 and int16).</p>
              </div>
            </td>
            <td>
                  <code>&#39;int16&#39;</code>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>compressor_name</code>
            </td>
            <td>
            </td>
            <td>
              <div class="doc-md-description">
                <p>The name of the compression algorithm to use for the LD matrix.</p>
              </div>
            </td>
            <td>
                  <code>&#39;zstd&#39;</code>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>compression_level</code>
            </td>
            <td>
            </td>
            <td>
              <div class="doc-md-description">
                <p>The compression level to use for the entries of the LD matrix (1-9).</p>
              </div>
            </td>
            <td>
                  <code>7</code>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>compute_spectral_properties</code>
            </td>
            <td>
            </td>
            <td>
              <div class="doc-md-description">
                <p>If True, compute the spectral properties of the LD matrix.</p>
              </div>
            </td>
            <td>
                  <code>False</code>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>ld_kwargs</code>
            </td>
            <td>
            </td>
            <td>
              <div class="doc-md-description">
                <p>keyword arguments for the various LD estimators. Consult the implementations of <code>WindowedLD</code>, <code>ShrinkageLD</code>, and <code>BlockLD</code> for details.</p>
              </div>
            </td>
            <td>
                  <code>{}</code>
            </td>
          </tr>
      </tbody>
    </table>


            <details class="quote">
              <summary>Source code in <code>magenpy/GWADataLoader.py</code></summary>
              <div class="language-python highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal"><a href="#__codelineno-0-627">627</a></span>
<span class="normal"><a href="#__codelineno-0-628">628</a></span>
<span class="normal"><a href="#__codelineno-0-629">629</a></span>
<span class="normal"><a href="#__codelineno-0-630">630</a></span>
<span class="normal"><a href="#__codelineno-0-631">631</a></span>
<span class="normal"><a href="#__codelineno-0-632">632</a></span>
<span class="normal"><a href="#__codelineno-0-633">633</a></span>
<span class="normal"><a href="#__codelineno-0-634">634</a></span>
<span class="normal"><a href="#__codelineno-0-635">635</a></span>
<span class="normal"><a href="#__codelineno-0-636">636</a></span>
<span class="normal"><a href="#__codelineno-0-637">637</a></span>
<span class="normal"><a href="#__codelineno-0-638">638</a></span>
<span class="normal"><a href="#__codelineno-0-639">639</a></span>
<span class="normal"><a href="#__codelineno-0-640">640</a></span>
<span class="normal"><a href="#__codelineno-0-641">641</a></span>
<span class="normal"><a href="#__codelineno-0-642">642</a></span>
<span class="normal"><a href="#__codelineno-0-643">643</a></span>
<span class="normal"><a href="#__codelineno-0-644">644</a></span>
<span class="normal"><a href="#__codelineno-0-645">645</a></span>
<span class="normal"><a href="#__codelineno-0-646">646</a></span>
<span class="normal"><a href="#__codelineno-0-647">647</a></span>
<span class="normal"><a href="#__codelineno-0-648">648</a></span>
<span class="normal"><a href="#__codelineno-0-649">649</a></span>
<span class="normal"><a href="#__codelineno-0-650">650</a></span>
<span class="normal"><a href="#__codelineno-0-651">651</a></span>
<span class="normal"><a href="#__codelineno-0-652">652</a></span>
<span class="normal"><a href="#__codelineno-0-653">653</a></span>
<span class="normal"><a href="#__codelineno-0-654">654</a></span>
<span class="normal"><a href="#__codelineno-0-655">655</a></span>
<span class="normal"><a href="#__codelineno-0-656">656</a></span>
<span class="normal"><a href="#__codelineno-0-657">657</a></span>
<span class="normal"><a href="#__codelineno-0-658">658</a></span>
<span class="normal"><a href="#__codelineno-0-659">659</a></span>
<span class="normal"><a href="#__codelineno-0-660">660</a></span>
<span class="normal"><a href="#__codelineno-0-661">661</a></span>
<span class="normal"><a href="#__codelineno-0-662">662</a></span>
<span class="normal"><a href="#__codelineno-0-663">663</a></span>
<span class="normal"><a href="#__codelineno-0-664">664</a></span>
<span class="normal"><a href="#__codelineno-0-665">665</a></span>
<span class="normal"><a href="#__codelineno-0-666">666</a></span>
<span class="normal"><a href="#__codelineno-0-667">667</a></span>
<span class="normal"><a href="#__codelineno-0-668">668</a></span>
<span class="normal"><a href="#__codelineno-0-669">669</a></span>
<span class="normal"><a href="#__codelineno-0-670">670</a></span>
<span class="normal"><a href="#__codelineno-0-671">671</a></span></pre></div></td><td class="code"><div><pre><span></span><code><span id="__span-0-627"><a id="__codelineno-0-627" name="__codelineno-0-627"></a><span class="k">def</span><span class="w"> </span><span class="nf">compute_ld</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span>
</span><span id="__span-0-628"><a id="__codelineno-0-628" name="__codelineno-0-628"></a>               <span class="n">estimator</span><span class="p">,</span>
</span><span id="__span-0-629"><a id="__codelineno-0-629" name="__codelineno-0-629"></a>               <span class="n">output_dir</span><span class="p">,</span>
</span><span id="__span-0-630"><a id="__codelineno-0-630" name="__codelineno-0-630"></a>               <span class="n">dtype</span><span class="o">=</span><span class="s1">&#39;int16&#39;</span><span class="p">,</span>
</span><span id="__span-0-631"><a id="__codelineno-0-631" name="__codelineno-0-631"></a>               <span class="n">compressor_name</span><span class="o">=</span><span class="s1">&#39;zstd&#39;</span><span class="p">,</span>
</span><span id="__span-0-632"><a id="__codelineno-0-632" name="__codelineno-0-632"></a>               <span class="n">compression_level</span><span class="o">=</span><span class="mi">7</span><span class="p">,</span>
</span><span id="__span-0-633"><a id="__codelineno-0-633" name="__codelineno-0-633"></a>               <span class="n">compute_spectral_properties</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span>
</span><span id="__span-0-634"><a id="__codelineno-0-634" name="__codelineno-0-634"></a>               <span class="o">**</span><span class="n">ld_kwargs</span><span class="p">):</span>
</span><span id="__span-0-635"><a id="__codelineno-0-635" name="__codelineno-0-635"></a><span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
</span><span id="__span-0-636"><a id="__codelineno-0-636" name="__codelineno-0-636"></a><span class="sd">    Compute the Linkage-Disequilibrium (LD) matrix or SNP-by-SNP Pearson</span>
</span><span id="__span-0-637"><a id="__codelineno-0-637" name="__codelineno-0-637"></a><span class="sd">    correlation matrix between genetic variants. This function only considers correlations</span>
</span><span id="__span-0-638"><a id="__codelineno-0-638" name="__codelineno-0-638"></a><span class="sd">    between SNPs on the same chromosome. This is a utility function that calls the</span>
</span><span id="__span-0-639"><a id="__codelineno-0-639" name="__codelineno-0-639"></a><span class="sd">    `.compute_ld()` method of the `GenotypeMatrix` objects associated with</span>
</span><span id="__span-0-640"><a id="__codelineno-0-640" name="__codelineno-0-640"></a><span class="sd">    GWADataLoader.</span>
</span><span id="__span-0-641"><a id="__codelineno-0-641" name="__codelineno-0-641"></a>
</span><span id="__span-0-642"><a id="__codelineno-0-642" name="__codelineno-0-642"></a><span class="sd">    :param estimator: The estimator for the LD matrix. We currently support</span>
</span><span id="__span-0-643"><a id="__codelineno-0-643" name="__codelineno-0-643"></a><span class="sd">    4 different estimators: `sample`, `windowed`, `shrinkage`, and `block`.</span>
</span><span id="__span-0-644"><a id="__codelineno-0-644" name="__codelineno-0-644"></a><span class="sd">    :param output_dir: The output directory where the Zarr array containing the</span>
</span><span id="__span-0-645"><a id="__codelineno-0-645" name="__codelineno-0-645"></a><span class="sd">    entries of the LD matrix will be stored.</span>
</span><span id="__span-0-646"><a id="__codelineno-0-646" name="__codelineno-0-646"></a><span class="sd">    :param dtype: The data type for the entries of the LD matrix (supported data types are float32, float64</span>
</span><span id="__span-0-647"><a id="__codelineno-0-647" name="__codelineno-0-647"></a><span class="sd">    and integer quantized data types int8 and int16).</span>
</span><span id="__span-0-648"><a id="__codelineno-0-648" name="__codelineno-0-648"></a><span class="sd">    :param compressor_name: The name of the compression algorithm to use for the LD matrix.</span>
</span><span id="__span-0-649"><a id="__codelineno-0-649" name="__codelineno-0-649"></a><span class="sd">    :param compression_level: The compression level to use for the entries of the LD matrix (1-9).</span>
</span><span id="__span-0-650"><a id="__codelineno-0-650" name="__codelineno-0-650"></a><span class="sd">    :param compute_spectral_properties: If True, compute the spectral properties of the LD matrix.</span>
</span><span id="__span-0-651"><a id="__codelineno-0-651" name="__codelineno-0-651"></a><span class="sd">    :param ld_kwargs: keyword arguments for the various LD estimators. Consult</span>
</span><span id="__span-0-652"><a id="__codelineno-0-652" name="__codelineno-0-652"></a><span class="sd">    the implementations of `WindowedLD`, `ShrinkageLD`, and `BlockLD` for details.</span>
</span><span id="__span-0-653"><a id="__codelineno-0-653" name="__codelineno-0-653"></a><span class="sd">    &quot;&quot;&quot;</span>
</span><span id="__span-0-654"><a id="__codelineno-0-654" name="__codelineno-0-654"></a>
</span><span id="__span-0-655"><a id="__codelineno-0-655" name="__codelineno-0-655"></a>    <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">genotype</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
</span><span id="__span-0-656"><a id="__codelineno-0-656" name="__codelineno-0-656"></a>        <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;Cannot compute LD without genotype data.&quot;</span><span class="p">)</span>
</span><span id="__span-0-657"><a id="__codelineno-0-657" name="__codelineno-0-657"></a>
</span><span id="__span-0-658"><a id="__codelineno-0-658" name="__codelineno-0-658"></a>    <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;&gt; Computing LD matrix...&quot;</span><span class="p">)</span>
</span><span id="__span-0-659"><a id="__codelineno-0-659" name="__codelineno-0-659"></a>
</span><span id="__span-0-660"><a id="__codelineno-0-660" name="__codelineno-0-660"></a>    <span class="bp">self</span><span class="o">.</span><span class="n">ld</span> <span class="o">=</span> <span class="p">{</span>
</span><span id="__span-0-661"><a id="__codelineno-0-661" name="__codelineno-0-661"></a>        <span class="n">c</span><span class="p">:</span> <span class="n">g</span><span class="o">.</span><span class="n">compute_ld</span><span class="p">(</span><span class="n">estimator</span><span class="p">,</span>
</span><span id="__span-0-662"><a id="__codelineno-0-662" name="__codelineno-0-662"></a>                        <span class="n">output_dir</span><span class="p">,</span>
</span><span id="__span-0-663"><a id="__codelineno-0-663" name="__codelineno-0-663"></a>                        <span class="n">dtype</span><span class="o">=</span><span class="n">dtype</span><span class="p">,</span>
</span><span id="__span-0-664"><a id="__codelineno-0-664" name="__codelineno-0-664"></a>                        <span class="n">compressor_name</span><span class="o">=</span><span class="n">compressor_name</span><span class="p">,</span>
</span><span id="__span-0-665"><a id="__codelineno-0-665" name="__codelineno-0-665"></a>                        <span class="n">compression_level</span><span class="o">=</span><span class="n">compression_level</span><span class="p">,</span>
</span><span id="__span-0-666"><a id="__codelineno-0-666" name="__codelineno-0-666"></a>                        <span class="n">compute_spectral_properties</span><span class="o">=</span><span class="n">compute_spectral_properties</span><span class="p">,</span>
</span><span id="__span-0-667"><a id="__codelineno-0-667" name="__codelineno-0-667"></a>                        <span class="o">**</span><span class="n">ld_kwargs</span><span class="p">)</span>
</span><span id="__span-0-668"><a id="__codelineno-0-668" name="__codelineno-0-668"></a>        <span class="k">for</span> <span class="n">c</span><span class="p">,</span> <span class="n">g</span> <span class="ow">in</span> <span class="n">tqdm</span><span class="p">(</span><span class="nb">sorted</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">genotype</span><span class="o">.</span><span class="n">items</span><span class="p">(),</span> <span class="n">key</span><span class="o">=</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">x</span><span class="p">[</span><span class="mi">0</span><span class="p">]),</span>
</span><span id="__span-0-669"><a id="__codelineno-0-669" name="__codelineno-0-669"></a>                         <span class="n">total</span><span class="o">=</span><span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">genotype</span><span class="p">),</span>
</span><span id="__span-0-670"><a id="__codelineno-0-670" name="__codelineno-0-670"></a>                         <span class="n">desc</span><span class="o">=</span><span class="s1">&#39;Computing LD matrices&#39;</span><span class="p">)</span>
</span><span id="__span-0-671"><a id="__codelineno-0-671" name="__codelineno-0-671"></a>    <span class="p">}</span>
</span></code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h2 id="magenpy.GWADataLoader.GWADataLoader.filter_samples" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">filter_samples</span><span class="p">(</span><span class="n">keep_samples</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">keep_file</span><span class="o">=</span><span class="kc">None</span><span class="p">)</span></code>

<a href="#magenpy.GWADataLoader.GWADataLoader.filter_samples" class="headerlink" title="Permanent link">&para;</a></h2>


    <div class="doc doc-contents ">

        <p>Filter samples from the samples table. User must specify
either a list of samples to keep or the path to a file
with the list of samples to keep.</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                <code>keep_samples</code>
            </td>
            <td>
            </td>
            <td>
              <div class="doc-md-description">
                <p>A list or array of sample IDs to keep.</p>
              </div>
            </td>
            <td>
                  <code>None</code>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>keep_file</code>
            </td>
            <td>
            </td>
            <td>
              <div class="doc-md-description">
                <p>The path to a file with the list of samples to keep.</p>
              </div>
            </td>
            <td>
                  <code>None</code>
            </td>
          </tr>
      </tbody>
    </table>


            <details class="quote">
              <summary>Source code in <code>magenpy/GWADataLoader.py</code></summary>
              <div class="language-python highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal"><a href="#__codelineno-0-315">315</a></span>
<span class="normal"><a href="#__codelineno-0-316">316</a></span>
<span class="normal"><a href="#__codelineno-0-317">317</a></span>
<span class="normal"><a href="#__codelineno-0-318">318</a></span>
<span class="normal"><a href="#__codelineno-0-319">319</a></span>
<span class="normal"><a href="#__codelineno-0-320">320</a></span>
<span class="normal"><a href="#__codelineno-0-321">321</a></span>
<span class="normal"><a href="#__codelineno-0-322">322</a></span>
<span class="normal"><a href="#__codelineno-0-323">323</a></span>
<span class="normal"><a href="#__codelineno-0-324">324</a></span>
<span class="normal"><a href="#__codelineno-0-325">325</a></span>
<span class="normal"><a href="#__codelineno-0-326">326</a></span></pre></div></td><td class="code"><div><pre><span></span><code><span id="__span-0-315"><a id="__codelineno-0-315" name="__codelineno-0-315"></a><span class="k">def</span><span class="w"> </span><span class="nf">filter_samples</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">keep_samples</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">keep_file</span><span class="o">=</span><span class="kc">None</span><span class="p">):</span>
</span><span id="__span-0-316"><a id="__codelineno-0-316" name="__codelineno-0-316"></a><span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
</span><span id="__span-0-317"><a id="__codelineno-0-317" name="__codelineno-0-317"></a><span class="sd">    Filter samples from the samples table. User must specify</span>
</span><span id="__span-0-318"><a id="__codelineno-0-318" name="__codelineno-0-318"></a><span class="sd">    either a list of samples to keep or the path to a file</span>
</span><span id="__span-0-319"><a id="__codelineno-0-319" name="__codelineno-0-319"></a><span class="sd">    with the list of samples to keep.</span>
</span><span id="__span-0-320"><a id="__codelineno-0-320" name="__codelineno-0-320"></a>
</span><span id="__span-0-321"><a id="__codelineno-0-321" name="__codelineno-0-321"></a><span class="sd">    :param keep_samples: A list or array of sample IDs to keep.</span>
</span><span id="__span-0-322"><a id="__codelineno-0-322" name="__codelineno-0-322"></a><span class="sd">    :param keep_file: The path to a file with the list of samples to keep.</span>
</span><span id="__span-0-323"><a id="__codelineno-0-323" name="__codelineno-0-323"></a><span class="sd">    &quot;&quot;&quot;</span>
</span><span id="__span-0-324"><a id="__codelineno-0-324" name="__codelineno-0-324"></a>
</span><span id="__span-0-325"><a id="__codelineno-0-325" name="__codelineno-0-325"></a>    <span class="bp">self</span><span class="o">.</span><span class="n">sample_table</span><span class="o">.</span><span class="n">filter_samples</span><span class="p">(</span><span class="n">keep_samples</span><span class="o">=</span><span class="n">keep_samples</span><span class="p">,</span> <span class="n">keep_file</span><span class="o">=</span><span class="n">keep_file</span><span class="p">)</span>
</span><span id="__span-0-326"><a id="__codelineno-0-326" name="__codelineno-0-326"></a>    <span class="bp">self</span><span class="o">.</span><span class="n">sync_sample_tables</span><span class="p">()</span>
</span></code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h2 id="magenpy.GWADataLoader.GWADataLoader.filter_snps" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">filter_snps</span><span class="p">(</span><span class="n">extract_snps</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">extract_file</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">chromosome</span><span class="o">=</span><span class="kc">None</span><span class="p">)</span></code>

<a href="#magenpy.GWADataLoader.GWADataLoader.filter_snps" class="headerlink" title="Permanent link">&para;</a></h2>


    <div class="doc doc-contents ">

        <p>Filter the SNP set from all the GWADataLoader objects.</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                <code>extract_snps</code>
            </td>
            <td>
            </td>
            <td>
              <div class="doc-md-description">
                <p>A list or array of SNP rsIDs to keep.</p>
              </div>
            </td>
            <td>
                  <code>None</code>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>extract_file</code>
            </td>
            <td>
            </td>
            <td>
              <div class="doc-md-description">
                <p>A path to a plink-style file with SNP rsIDs to keep.</p>
              </div>
            </td>
            <td>
                  <code>None</code>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>chromosome</code>
            </td>
            <td>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Chromosome number. If specified, applies the filter to that chromosome only.</p>
              </div>
            </td>
            <td>
                  <code>None</code>
            </td>
          </tr>
      </tbody>
    </table>


            <details class="quote">
              <summary>Source code in <code>magenpy/GWADataLoader.py</code></summary>
              <div class="language-python highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal"><a href="#__codelineno-0-263">263</a></span>
<span class="normal"><a href="#__codelineno-0-264">264</a></span>
<span class="normal"><a href="#__codelineno-0-265">265</a></span>
<span class="normal"><a href="#__codelineno-0-266">266</a></span>
<span class="normal"><a href="#__codelineno-0-267">267</a></span>
<span class="normal"><a href="#__codelineno-0-268">268</a></span>
<span class="normal"><a href="#__codelineno-0-269">269</a></span>
<span class="normal"><a href="#__codelineno-0-270">270</a></span>
<span class="normal"><a href="#__codelineno-0-271">271</a></span>
<span class="normal"><a href="#__codelineno-0-272">272</a></span>
<span class="normal"><a href="#__codelineno-0-273">273</a></span>
<span class="normal"><a href="#__codelineno-0-274">274</a></span>
<span class="normal"><a href="#__codelineno-0-275">275</a></span>
<span class="normal"><a href="#__codelineno-0-276">276</a></span>
<span class="normal"><a href="#__codelineno-0-277">277</a></span>
<span class="normal"><a href="#__codelineno-0-278">278</a></span>
<span class="normal"><a href="#__codelineno-0-279">279</a></span>
<span class="normal"><a href="#__codelineno-0-280">280</a></span>
<span class="normal"><a href="#__codelineno-0-281">281</a></span>
<span class="normal"><a href="#__codelineno-0-282">282</a></span>
<span class="normal"><a href="#__codelineno-0-283">283</a></span>
<span class="normal"><a href="#__codelineno-0-284">284</a></span>
<span class="normal"><a href="#__codelineno-0-285">285</a></span>
<span class="normal"><a href="#__codelineno-0-286">286</a></span>
<span class="normal"><a href="#__codelineno-0-287">287</a></span>
<span class="normal"><a href="#__codelineno-0-288">288</a></span>
<span class="normal"><a href="#__codelineno-0-289">289</a></span>
<span class="normal"><a href="#__codelineno-0-290">290</a></span>
<span class="normal"><a href="#__codelineno-0-291">291</a></span>
<span class="normal"><a href="#__codelineno-0-292">292</a></span>
<span class="normal"><a href="#__codelineno-0-293">293</a></span>
<span class="normal"><a href="#__codelineno-0-294">294</a></span>
<span class="normal"><a href="#__codelineno-0-295">295</a></span>
<span class="normal"><a href="#__codelineno-0-296">296</a></span>
<span class="normal"><a href="#__codelineno-0-297">297</a></span>
<span class="normal"><a href="#__codelineno-0-298">298</a></span>
<span class="normal"><a href="#__codelineno-0-299">299</a></span>
<span class="normal"><a href="#__codelineno-0-300">300</a></span>
<span class="normal"><a href="#__codelineno-0-301">301</a></span>
<span class="normal"><a href="#__codelineno-0-302">302</a></span>
<span class="normal"><a href="#__codelineno-0-303">303</a></span>
<span class="normal"><a href="#__codelineno-0-304">304</a></span>
<span class="normal"><a href="#__codelineno-0-305">305</a></span>
<span class="normal"><a href="#__codelineno-0-306">306</a></span>
<span class="normal"><a href="#__codelineno-0-307">307</a></span>
<span class="normal"><a href="#__codelineno-0-308">308</a></span>
<span class="normal"><a href="#__codelineno-0-309">309</a></span>
<span class="normal"><a href="#__codelineno-0-310">310</a></span>
<span class="normal"><a href="#__codelineno-0-311">311</a></span>
<span class="normal"><a href="#__codelineno-0-312">312</a></span>
<span class="normal"><a href="#__codelineno-0-313">313</a></span></pre></div></td><td class="code"><div><pre><span></span><code><span id="__span-0-263"><a id="__codelineno-0-263" name="__codelineno-0-263"></a><span class="k">def</span><span class="w"> </span><span class="nf">filter_snps</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">extract_snps</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">extract_file</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">chromosome</span><span class="o">=</span><span class="kc">None</span><span class="p">):</span>
</span><span id="__span-0-264"><a id="__codelineno-0-264" name="__codelineno-0-264"></a><span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
</span><span id="__span-0-265"><a id="__codelineno-0-265" name="__codelineno-0-265"></a><span class="sd">    Filter the SNP set from all the GWADataLoader objects.</span>
</span><span id="__span-0-266"><a id="__codelineno-0-266" name="__codelineno-0-266"></a><span class="sd">    :param extract_snps: A list or array of SNP rsIDs to keep.</span>
</span><span id="__span-0-267"><a id="__codelineno-0-267" name="__codelineno-0-267"></a><span class="sd">    :param extract_file: A path to a plink-style file with SNP rsIDs to keep.</span>
</span><span id="__span-0-268"><a id="__codelineno-0-268" name="__codelineno-0-268"></a><span class="sd">    :param chromosome: Chromosome number. If specified, applies the filter to that chromosome only.</span>
</span><span id="__span-0-269"><a id="__codelineno-0-269" name="__codelineno-0-269"></a><span class="sd">    &quot;&quot;&quot;</span>
</span><span id="__span-0-270"><a id="__codelineno-0-270" name="__codelineno-0-270"></a>
</span><span id="__span-0-271"><a id="__codelineno-0-271" name="__codelineno-0-271"></a>    <span class="k">if</span> <span class="n">extract_snps</span> <span class="ow">is</span> <span class="kc">None</span> <span class="ow">and</span> <span class="n">extract_file</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
</span><span id="__span-0-272"><a id="__codelineno-0-272" name="__codelineno-0-272"></a>        <span class="k">return</span>
</span><span id="__span-0-273"><a id="__codelineno-0-273" name="__codelineno-0-273"></a>
</span><span id="__span-0-274"><a id="__codelineno-0-274" name="__codelineno-0-274"></a>    <span class="k">if</span> <span class="n">chromosome</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
</span><span id="__span-0-275"><a id="__codelineno-0-275" name="__codelineno-0-275"></a>        <span class="n">chroms</span> <span class="o">=</span> <span class="p">[</span><span class="n">chromosome</span><span class="p">]</span>
</span><span id="__span-0-276"><a id="__codelineno-0-276" name="__codelineno-0-276"></a>    <span class="k">else</span><span class="p">:</span>
</span><span id="__span-0-277"><a id="__codelineno-0-277" name="__codelineno-0-277"></a>        <span class="n">chroms</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">chromosomes</span>
</span><span id="__span-0-278"><a id="__codelineno-0-278" name="__codelineno-0-278"></a>
</span><span id="__span-0-279"><a id="__codelineno-0-279" name="__codelineno-0-279"></a>    <span class="k">if</span> <span class="n">extract_snps</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
</span><span id="__span-0-280"><a id="__codelineno-0-280" name="__codelineno-0-280"></a>        <span class="kn">from</span><span class="w"> </span><span class="nn">.parsers.misc_parsers</span><span class="w"> </span><span class="kn">import</span> <span class="n">read_snp_filter_file</span>
</span><span id="__span-0-281"><a id="__codelineno-0-281" name="__codelineno-0-281"></a>        <span class="n">extract_snps</span> <span class="o">=</span> <span class="n">read_snp_filter_file</span><span class="p">(</span><span class="n">extract_file</span><span class="p">)</span>
</span><span id="__span-0-282"><a id="__codelineno-0-282" name="__codelineno-0-282"></a>
</span><span id="__span-0-283"><a id="__codelineno-0-283" name="__codelineno-0-283"></a>    <span class="k">for</span> <span class="n">c</span> <span class="ow">in</span> <span class="n">chroms</span><span class="p">:</span>
</span><span id="__span-0-284"><a id="__codelineno-0-284" name="__codelineno-0-284"></a>
</span><span id="__span-0-285"><a id="__codelineno-0-285" name="__codelineno-0-285"></a>        <span class="c1"># Filter the genotype matrix:</span>
</span><span id="__span-0-286"><a id="__codelineno-0-286" name="__codelineno-0-286"></a>        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">genotype</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span> <span class="ow">and</span> <span class="n">c</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">genotype</span><span class="p">:</span>
</span><span id="__span-0-287"><a id="__codelineno-0-287" name="__codelineno-0-287"></a>            <span class="bp">self</span><span class="o">.</span><span class="n">genotype</span><span class="p">[</span><span class="n">c</span><span class="p">]</span><span class="o">.</span><span class="n">filter_snps</span><span class="p">(</span><span class="n">extract_snps</span><span class="o">=</span><span class="n">extract_snps</span><span class="p">)</span>
</span><span id="__span-0-288"><a id="__codelineno-0-288" name="__codelineno-0-288"></a>
</span><span id="__span-0-289"><a id="__codelineno-0-289" name="__codelineno-0-289"></a>            <span class="c1"># If no SNPs remain in the genotype matrix for that chromosome, then remove it:</span>
</span><span id="__span-0-290"><a id="__codelineno-0-290" name="__codelineno-0-290"></a>            <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">genotype</span><span class="p">[</span><span class="n">c</span><span class="p">]</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span> <span class="o">&lt;</span> <span class="mi">1</span><span class="p">:</span>
</span><span id="__span-0-291"><a id="__codelineno-0-291" name="__codelineno-0-291"></a>                <span class="k">del</span> <span class="bp">self</span><span class="o">.</span><span class="n">genotype</span><span class="p">[</span><span class="n">c</span><span class="p">]</span>
</span><span id="__span-0-292"><a id="__codelineno-0-292" name="__codelineno-0-292"></a>
</span><span id="__span-0-293"><a id="__codelineno-0-293" name="__codelineno-0-293"></a>        <span class="c1"># Filter the summary statistics table:</span>
</span><span id="__span-0-294"><a id="__codelineno-0-294" name="__codelineno-0-294"></a>        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">sumstats_table</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span> <span class="ow">and</span> <span class="n">c</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">sumstats_table</span><span class="p">:</span>
</span><span id="__span-0-295"><a id="__codelineno-0-295" name="__codelineno-0-295"></a>            <span class="bp">self</span><span class="o">.</span><span class="n">sumstats_table</span><span class="p">[</span><span class="n">c</span><span class="p">]</span><span class="o">.</span><span class="n">filter_snps</span><span class="p">(</span><span class="n">extract_snps</span><span class="o">=</span><span class="n">extract_snps</span><span class="p">)</span>
</span><span id="__span-0-296"><a id="__codelineno-0-296" name="__codelineno-0-296"></a>
</span><span id="__span-0-297"><a id="__codelineno-0-297" name="__codelineno-0-297"></a>            <span class="c1"># If no SNPs remain in the summary statistics table for that chromosome, then remove it:</span>
</span><span id="__span-0-298"><a id="__codelineno-0-298" name="__codelineno-0-298"></a>            <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">sumstats_table</span><span class="p">[</span><span class="n">c</span><span class="p">]</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="o">&lt;</span> <span class="mi">1</span><span class="p">:</span>
</span><span id="__span-0-299"><a id="__codelineno-0-299" name="__codelineno-0-299"></a>                <span class="k">del</span> <span class="bp">self</span><span class="o">.</span><span class="n">sumstats_table</span><span class="p">[</span><span class="n">c</span><span class="p">]</span>
</span><span id="__span-0-300"><a id="__codelineno-0-300" name="__codelineno-0-300"></a>
</span><span id="__span-0-301"><a id="__codelineno-0-301" name="__codelineno-0-301"></a>        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">ld</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span> <span class="ow">and</span> <span class="n">c</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">ld</span><span class="p">:</span>
</span><span id="__span-0-302"><a id="__codelineno-0-302" name="__codelineno-0-302"></a>            <span class="bp">self</span><span class="o">.</span><span class="n">ld</span><span class="p">[</span><span class="n">c</span><span class="p">]</span><span class="o">.</span><span class="n">filter_snps</span><span class="p">(</span><span class="n">extract_snps</span><span class="o">=</span><span class="n">extract_snps</span><span class="p">)</span>
</span><span id="__span-0-303"><a id="__codelineno-0-303" name="__codelineno-0-303"></a>
</span><span id="__span-0-304"><a id="__codelineno-0-304" name="__codelineno-0-304"></a>            <span class="c1"># If no SNPs remain in the summary statistics table for that chromosome, then remove it:</span>
</span><span id="__span-0-305"><a id="__codelineno-0-305" name="__codelineno-0-305"></a>            <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">ld</span><span class="p">[</span><span class="n">c</span><span class="p">]</span><span class="o">.</span><span class="n">n_snps</span> <span class="o">&lt;</span> <span class="mi">1</span><span class="p">:</span>
</span><span id="__span-0-306"><a id="__codelineno-0-306" name="__codelineno-0-306"></a>                <span class="k">del</span> <span class="bp">self</span><span class="o">.</span><span class="n">ld</span><span class="p">[</span><span class="n">c</span><span class="p">]</span>
</span><span id="__span-0-307"><a id="__codelineno-0-307" name="__codelineno-0-307"></a>
</span><span id="__span-0-308"><a id="__codelineno-0-308" name="__codelineno-0-308"></a>        <span class="c1"># Filter the annotation matrix:</span>
</span><span id="__span-0-309"><a id="__codelineno-0-309" name="__codelineno-0-309"></a>        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">annotation</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span> <span class="ow">and</span> <span class="n">c</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">annotation</span><span class="p">:</span>
</span><span id="__span-0-310"><a id="__codelineno-0-310" name="__codelineno-0-310"></a>            <span class="bp">self</span><span class="o">.</span><span class="n">annotation</span><span class="p">[</span><span class="n">c</span><span class="p">]</span><span class="o">.</span><span class="n">filter_snps</span><span class="p">(</span><span class="n">extract_snps</span><span class="o">=</span><span class="n">extract_snps</span><span class="p">)</span>
</span><span id="__span-0-311"><a id="__codelineno-0-311" name="__codelineno-0-311"></a>
</span><span id="__span-0-312"><a id="__codelineno-0-312" name="__codelineno-0-312"></a>            <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">annotation</span><span class="p">[</span><span class="n">c</span><span class="p">]</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="o">&lt;</span> <span class="mi">1</span><span class="p">:</span>
</span><span id="__span-0-313"><a id="__codelineno-0-313" name="__codelineno-0-313"></a>                <span class="k">del</span> <span class="bp">self</span><span class="o">.</span><span class="n">annotation</span><span class="p">[</span><span class="n">c</span><span class="p">]</span>
</span></code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h2 id="magenpy.GWADataLoader.GWADataLoader.get_ld_matrices" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">get_ld_matrices</span><span class="p">()</span></code>

<a href="#magenpy.GWADataLoader.GWADataLoader.get_ld_matrices" class="headerlink" title="Permanent link">&para;</a></h2>


    <div class="doc doc-contents ">

        


    <p><span class="doc-section-title">Returns:</span></p>
    <table>
      <thead>
        <tr>
          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
            </td>
            <td>
              <div class="doc-md-description">
                <p>A dictionary containing the chromosome ID as key and corresponding LD matrices as value.</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>


            <details class="quote">
              <summary>Source code in <code>magenpy/GWADataLoader.py</code></summary>
              <div class="language-python highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal"><a href="#__codelineno-0-673">673</a></span>
<span class="normal"><a href="#__codelineno-0-674">674</a></span>
<span class="normal"><a href="#__codelineno-0-675">675</a></span>
<span class="normal"><a href="#__codelineno-0-676">676</a></span>
<span class="normal"><a href="#__codelineno-0-677">677</a></span>
<span class="normal"><a href="#__codelineno-0-678">678</a></span></pre></div></td><td class="code"><div><pre><span></span><code><span id="__span-0-673"><a id="__codelineno-0-673" name="__codelineno-0-673"></a><span class="k">def</span><span class="w"> </span><span class="nf">get_ld_matrices</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
</span><span id="__span-0-674"><a id="__codelineno-0-674" name="__codelineno-0-674"></a><span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
</span><span id="__span-0-675"><a id="__codelineno-0-675" name="__codelineno-0-675"></a><span class="sd">    :return: A dictionary containing the chromosome ID as key and corresponding LD matrices</span>
</span><span id="__span-0-676"><a id="__codelineno-0-676" name="__codelineno-0-676"></a><span class="sd">    as value.</span>
</span><span id="__span-0-677"><a id="__codelineno-0-677" name="__codelineno-0-677"></a><span class="sd">    &quot;&quot;&quot;</span>
</span><span id="__span-0-678"><a id="__codelineno-0-678" name="__codelineno-0-678"></a>    <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">ld</span>
</span></code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h2 id="magenpy.GWADataLoader.GWADataLoader.harmonize_data" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">harmonize_data</span><span class="p">()</span></code>

<a href="#magenpy.GWADataLoader.GWADataLoader.harmonize_data" class="headerlink" title="Permanent link">&para;</a></h2>


    <div class="doc doc-contents ">

        <p>This method ensures that the data sources (reference genotype,
LD matrices, summary statistics, annotations) are all aligned in terms of the
set of variants that they operate on as well as the designation of the effect allele for
each variant.</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>This method is called automatically during the initialization of the <code>GWADataLoader</code> object.
However, if you read or manipulate the data sources after initialization,
you may need to call this method again to ensure that the data sources remain aligned.</p>
</div>
<div class="admonition warning">
<p class="admonition-title">Warning</p>
<p>Harmonization for now depends on having SNP rsID be present in all the resources. Hopefully
this requirement will be relaxed in the future.</p>
</div>


            <details class="quote">
              <summary>Source code in <code>magenpy/GWADataLoader.py</code></summary>
              <div class="language-python highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal"><a href="#__codelineno-0-680">680</a></span>
<span class="normal"><a href="#__codelineno-0-681">681</a></span>
<span class="normal"><a href="#__codelineno-0-682">682</a></span>
<span class="normal"><a href="#__codelineno-0-683">683</a></span>
<span class="normal"><a href="#__codelineno-0-684">684</a></span>
<span class="normal"><a href="#__codelineno-0-685">685</a></span>
<span class="normal"><a href="#__codelineno-0-686">686</a></span>
<span class="normal"><a href="#__codelineno-0-687">687</a></span>
<span class="normal"><a href="#__codelineno-0-688">688</a></span>
<span class="normal"><a href="#__codelineno-0-689">689</a></span>
<span class="normal"><a href="#__codelineno-0-690">690</a></span>
<span class="normal"><a href="#__codelineno-0-691">691</a></span>
<span class="normal"><a href="#__codelineno-0-692">692</a></span>
<span class="normal"><a href="#__codelineno-0-693">693</a></span>
<span class="normal"><a href="#__codelineno-0-694">694</a></span>
<span class="normal"><a href="#__codelineno-0-695">695</a></span>
<span class="normal"><a href="#__codelineno-0-696">696</a></span>
<span class="normal"><a href="#__codelineno-0-697">697</a></span>
<span class="normal"><a href="#__codelineno-0-698">698</a></span>
<span class="normal"><a href="#__codelineno-0-699">699</a></span>
<span class="normal"><a href="#__codelineno-0-700">700</a></span>
<span class="normal"><a href="#__codelineno-0-701">701</a></span>
<span class="normal"><a href="#__codelineno-0-702">702</a></span>
<span class="normal"><a href="#__codelineno-0-703">703</a></span>
<span class="normal"><a href="#__codelineno-0-704">704</a></span>
<span class="normal"><a href="#__codelineno-0-705">705</a></span>
<span class="normal"><a href="#__codelineno-0-706">706</a></span>
<span class="normal"><a href="#__codelineno-0-707">707</a></span>
<span class="normal"><a href="#__codelineno-0-708">708</a></span>
<span class="normal"><a href="#__codelineno-0-709">709</a></span>
<span class="normal"><a href="#__codelineno-0-710">710</a></span>
<span class="normal"><a href="#__codelineno-0-711">711</a></span>
<span class="normal"><a href="#__codelineno-0-712">712</a></span>
<span class="normal"><a href="#__codelineno-0-713">713</a></span>
<span class="normal"><a href="#__codelineno-0-714">714</a></span>
<span class="normal"><a href="#__codelineno-0-715">715</a></span>
<span class="normal"><a href="#__codelineno-0-716">716</a></span>
<span class="normal"><a href="#__codelineno-0-717">717</a></span>
<span class="normal"><a href="#__codelineno-0-718">718</a></span>
<span class="normal"><a href="#__codelineno-0-719">719</a></span>
<span class="normal"><a href="#__codelineno-0-720">720</a></span>
<span class="normal"><a href="#__codelineno-0-721">721</a></span>
<span class="normal"><a href="#__codelineno-0-722">722</a></span>
<span class="normal"><a href="#__codelineno-0-723">723</a></span>
<span class="normal"><a href="#__codelineno-0-724">724</a></span>
<span class="normal"><a href="#__codelineno-0-725">725</a></span>
<span class="normal"><a href="#__codelineno-0-726">726</a></span>
<span class="normal"><a href="#__codelineno-0-727">727</a></span>
<span class="normal"><a href="#__codelineno-0-728">728</a></span>
<span class="normal"><a href="#__codelineno-0-729">729</a></span>
<span class="normal"><a href="#__codelineno-0-730">730</a></span>
<span class="normal"><a href="#__codelineno-0-731">731</a></span>
<span class="normal"><a href="#__codelineno-0-732">732</a></span>
<span class="normal"><a href="#__codelineno-0-733">733</a></span>
<span class="normal"><a href="#__codelineno-0-734">734</a></span>
<span class="normal"><a href="#__codelineno-0-735">735</a></span>
<span class="normal"><a href="#__codelineno-0-736">736</a></span>
<span class="normal"><a href="#__codelineno-0-737">737</a></span>
<span class="normal"><a href="#__codelineno-0-738">738</a></span>
<span class="normal"><a href="#__codelineno-0-739">739</a></span>
<span class="normal"><a href="#__codelineno-0-740">740</a></span>
<span class="normal"><a href="#__codelineno-0-741">741</a></span>
<span class="normal"><a href="#__codelineno-0-742">742</a></span>
<span class="normal"><a href="#__codelineno-0-743">743</a></span>
<span class="normal"><a href="#__codelineno-0-744">744</a></span>
<span class="normal"><a href="#__codelineno-0-745">745</a></span>
<span class="normal"><a href="#__codelineno-0-746">746</a></span>
<span class="normal"><a href="#__codelineno-0-747">747</a></span>
<span class="normal"><a href="#__codelineno-0-748">748</a></span>
<span class="normal"><a href="#__codelineno-0-749">749</a></span>
<span class="normal"><a href="#__codelineno-0-750">750</a></span>
<span class="normal"><a href="#__codelineno-0-751">751</a></span>
<span class="normal"><a href="#__codelineno-0-752">752</a></span>
<span class="normal"><a href="#__codelineno-0-753">753</a></span>
<span class="normal"><a href="#__codelineno-0-754">754</a></span>
<span class="normal"><a href="#__codelineno-0-755">755</a></span>
<span class="normal"><a href="#__codelineno-0-756">756</a></span></pre></div></td><td class="code"><div><pre><span></span><code><span id="__span-0-680"><a id="__codelineno-0-680" name="__codelineno-0-680"></a><span class="k">def</span><span class="w"> </span><span class="nf">harmonize_data</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
</span><span id="__span-0-681"><a id="__codelineno-0-681" name="__codelineno-0-681"></a><span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
</span><span id="__span-0-682"><a id="__codelineno-0-682" name="__codelineno-0-682"></a><span class="sd">    This method ensures that the data sources (reference genotype,</span>
</span><span id="__span-0-683"><a id="__codelineno-0-683" name="__codelineno-0-683"></a><span class="sd">    LD matrices, summary statistics, annotations) are all aligned in terms of the</span>
</span><span id="__span-0-684"><a id="__codelineno-0-684" name="__codelineno-0-684"></a><span class="sd">    set of variants that they operate on as well as the designation of the effect allele for</span>
</span><span id="__span-0-685"><a id="__codelineno-0-685" name="__codelineno-0-685"></a><span class="sd">    each variant.</span>
</span><span id="__span-0-686"><a id="__codelineno-0-686" name="__codelineno-0-686"></a>
</span><span id="__span-0-687"><a id="__codelineno-0-687" name="__codelineno-0-687"></a><span class="sd">    !!! note</span>
</span><span id="__span-0-688"><a id="__codelineno-0-688" name="__codelineno-0-688"></a><span class="sd">        This method is called automatically during the initialization of the `GWADataLoader` object.</span>
</span><span id="__span-0-689"><a id="__codelineno-0-689" name="__codelineno-0-689"></a><span class="sd">        However, if you read or manipulate the data sources after initialization,</span>
</span><span id="__span-0-690"><a id="__codelineno-0-690" name="__codelineno-0-690"></a><span class="sd">        you may need to call this method again to ensure that the data sources remain aligned.</span>
</span><span id="__span-0-691"><a id="__codelineno-0-691" name="__codelineno-0-691"></a>
</span><span id="__span-0-692"><a id="__codelineno-0-692" name="__codelineno-0-692"></a><span class="sd">    !!! warning</span>
</span><span id="__span-0-693"><a id="__codelineno-0-693" name="__codelineno-0-693"></a><span class="sd">        Harmonization for now depends on having SNP rsID be present in all the resources. Hopefully</span>
</span><span id="__span-0-694"><a id="__codelineno-0-694" name="__codelineno-0-694"></a><span class="sd">        this requirement will be relaxed in the future.</span>
</span><span id="__span-0-695"><a id="__codelineno-0-695" name="__codelineno-0-695"></a>
</span><span id="__span-0-696"><a id="__codelineno-0-696" name="__codelineno-0-696"></a><span class="sd">    &quot;&quot;&quot;</span>
</span><span id="__span-0-697"><a id="__codelineno-0-697" name="__codelineno-0-697"></a>
</span><span id="__span-0-698"><a id="__codelineno-0-698" name="__codelineno-0-698"></a>    <span class="n">data_sources</span> <span class="o">=</span> <span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">genotype</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">sumstats_table</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">ld</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">annotation</span><span class="p">)</span>
</span><span id="__span-0-699"><a id="__codelineno-0-699" name="__codelineno-0-699"></a>    <span class="n">initialized_data_sources</span> <span class="o">=</span> <span class="p">[</span><span class="n">ds</span> <span class="k">for</span> <span class="n">ds</span> <span class="ow">in</span> <span class="n">data_sources</span> <span class="k">if</span> <span class="n">ds</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">]</span>
</span><span id="__span-0-700"><a id="__codelineno-0-700" name="__codelineno-0-700"></a>
</span><span id="__span-0-701"><a id="__codelineno-0-701" name="__codelineno-0-701"></a>    <span class="c1"># If less than two data sources are present, skip harmonization...</span>
</span><span id="__span-0-702"><a id="__codelineno-0-702" name="__codelineno-0-702"></a>    <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">initialized_data_sources</span><span class="p">)</span> <span class="o">&lt;</span> <span class="mi">2</span><span class="p">:</span>
</span><span id="__span-0-703"><a id="__codelineno-0-703" name="__codelineno-0-703"></a>        <span class="k">return</span>
</span><span id="__span-0-704"><a id="__codelineno-0-704" name="__codelineno-0-704"></a>
</span><span id="__span-0-705"><a id="__codelineno-0-705" name="__codelineno-0-705"></a>    <span class="c1"># Get the chromosomes information from all the data sources:</span>
</span><span id="__span-0-706"><a id="__codelineno-0-706" name="__codelineno-0-706"></a>    <span class="n">chromosomes</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="nb">set</span><span class="o">.</span><span class="n">union</span><span class="p">(</span><span class="o">*</span><span class="p">[</span><span class="nb">set</span><span class="p">(</span><span class="n">ds</span><span class="o">.</span><span class="n">keys</span><span class="p">())</span> <span class="k">for</span> <span class="n">ds</span> <span class="ow">in</span> <span class="n">initialized_data_sources</span><span class="p">]))</span>
</span><span id="__span-0-707"><a id="__codelineno-0-707" name="__codelineno-0-707"></a>
</span><span id="__span-0-708"><a id="__codelineno-0-708" name="__codelineno-0-708"></a>    <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;&gt; Harmonizing data...&quot;</span><span class="p">)</span>
</span><span id="__span-0-709"><a id="__codelineno-0-709" name="__codelineno-0-709"></a>
</span><span id="__span-0-710"><a id="__codelineno-0-710" name="__codelineno-0-710"></a>    <span class="k">for</span> <span class="n">c</span> <span class="ow">in</span> <span class="n">tqdm</span><span class="p">(</span><span class="n">chromosomes</span><span class="p">,</span>
</span><span id="__span-0-711"><a id="__codelineno-0-711" name="__codelineno-0-711"></a>                  <span class="n">total</span><span class="o">=</span><span class="nb">len</span><span class="p">(</span><span class="n">chromosomes</span><span class="p">),</span>
</span><span id="__span-0-712"><a id="__codelineno-0-712" name="__codelineno-0-712"></a>                  <span class="n">desc</span><span class="o">=</span><span class="s1">&#39;Harmonizing data&#39;</span><span class="p">):</span>
</span><span id="__span-0-713"><a id="__codelineno-0-713" name="__codelineno-0-713"></a>
</span><span id="__span-0-714"><a id="__codelineno-0-714" name="__codelineno-0-714"></a>        <span class="c1"># Which initialized data sources have information for chromosome `c`</span>
</span><span id="__span-0-715"><a id="__codelineno-0-715" name="__codelineno-0-715"></a>        <span class="n">miss_chroms</span> <span class="o">=</span> <span class="p">[</span><span class="n">c</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">ds</span> <span class="k">for</span> <span class="n">ds</span> <span class="ow">in</span> <span class="n">initialized_data_sources</span><span class="p">]</span>
</span><span id="__span-0-716"><a id="__codelineno-0-716" name="__codelineno-0-716"></a>
</span><span id="__span-0-717"><a id="__codelineno-0-717" name="__codelineno-0-717"></a>        <span class="k">if</span> <span class="nb">sum</span><span class="p">(</span><span class="n">miss_chroms</span><span class="p">)</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">:</span>
</span><span id="__span-0-718"><a id="__codelineno-0-718" name="__codelineno-0-718"></a>            <span class="c1"># If the chromosome data only exists for some data sources but not others, remove the chromosome</span>
</span><span id="__span-0-719"><a id="__codelineno-0-719" name="__codelineno-0-719"></a>            <span class="c1"># from all data source.</span>
</span><span id="__span-0-720"><a id="__codelineno-0-720" name="__codelineno-0-720"></a>            <span class="c1"># Is this the best way to handle the missingness? Should we just post a warning?</span>
</span><span id="__span-0-721"><a id="__codelineno-0-721" name="__codelineno-0-721"></a>            <span class="n">logger</span><span class="o">.</span><span class="n">debug</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Chromosome </span><span class="si">{</span><span class="n">c</span><span class="si">}</span><span class="s2"> is missing in some data sources. &quot;</span>
</span><span id="__span-0-722"><a id="__codelineno-0-722" name="__codelineno-0-722"></a>                         <span class="sa">f</span><span class="s2">&quot;Removing it from all data sources.&quot;</span><span class="p">)</span>
</span><span id="__span-0-723"><a id="__codelineno-0-723" name="__codelineno-0-723"></a>            <span class="k">for</span> <span class="n">ds</span> <span class="ow">in</span> <span class="n">initialized_data_sources</span><span class="p">:</span>
</span><span id="__span-0-724"><a id="__codelineno-0-724" name="__codelineno-0-724"></a>                <span class="k">if</span> <span class="n">c</span> <span class="ow">in</span> <span class="n">ds</span><span class="p">:</span>
</span><span id="__span-0-725"><a id="__codelineno-0-725" name="__codelineno-0-725"></a>                    <span class="k">del</span> <span class="n">ds</span><span class="p">[</span><span class="n">c</span><span class="p">]</span>
</span><span id="__span-0-726"><a id="__codelineno-0-726" name="__codelineno-0-726"></a>
</span><span id="__span-0-727"><a id="__codelineno-0-727" name="__codelineno-0-727"></a>        <span class="k">else</span><span class="p">:</span>
</span><span id="__span-0-728"><a id="__codelineno-0-728" name="__codelineno-0-728"></a>
</span><span id="__span-0-729"><a id="__codelineno-0-729" name="__codelineno-0-729"></a>            <span class="c1"># Find the set of SNPs that are shared across all data sources (exclude missing values):</span>
</span><span id="__span-0-730"><a id="__codelineno-0-730" name="__codelineno-0-730"></a>            <span class="n">common_snps</span> <span class="o">=</span> <span class="n">intersect_multiple_arrays</span><span class="p">([</span><span class="n">ds</span><span class="p">[</span><span class="n">c</span><span class="p">]</span><span class="o">.</span><span class="n">snps</span> <span class="k">for</span> <span class="n">ds</span> <span class="ow">in</span> <span class="n">initialized_data_sources</span><span class="p">])</span>
</span><span id="__span-0-731"><a id="__codelineno-0-731" name="__codelineno-0-731"></a>
</span><span id="__span-0-732"><a id="__codelineno-0-732" name="__codelineno-0-732"></a>            <span class="c1"># If necessary, filter the data sources to only have the common SNPs:</span>
</span><span id="__span-0-733"><a id="__codelineno-0-733" name="__codelineno-0-733"></a>            <span class="k">for</span> <span class="n">ds</span> <span class="ow">in</span> <span class="n">initialized_data_sources</span><span class="p">:</span>
</span><span id="__span-0-734"><a id="__codelineno-0-734" name="__codelineno-0-734"></a>                <span class="k">if</span> <span class="n">ds</span><span class="p">[</span><span class="n">c</span><span class="p">]</span><span class="o">.</span><span class="n">n_snps</span> <span class="o">!=</span> <span class="nb">len</span><span class="p">(</span><span class="n">common_snps</span><span class="p">):</span>
</span><span id="__span-0-735"><a id="__codelineno-0-735" name="__codelineno-0-735"></a>                    <span class="n">ds</span><span class="p">[</span><span class="n">c</span><span class="p">]</span><span class="o">.</span><span class="n">filter_snps</span><span class="p">(</span><span class="n">extract_snps</span><span class="o">=</span><span class="n">common_snps</span><span class="p">)</span>
</span><span id="__span-0-736"><a id="__codelineno-0-736" name="__codelineno-0-736"></a>
</span><span id="__span-0-737"><a id="__codelineno-0-737" name="__codelineno-0-737"></a>            <span class="c1"># Harmonize the summary statistics data with either genotype or LD reference.</span>
</span><span id="__span-0-738"><a id="__codelineno-0-738" name="__codelineno-0-738"></a>            <span class="c1"># This procedure checks for flips in the effect allele between data sources.</span>
</span><span id="__span-0-739"><a id="__codelineno-0-739" name="__codelineno-0-739"></a>            <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">sumstats_table</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
</span><span id="__span-0-740"><a id="__codelineno-0-740" name="__codelineno-0-740"></a>
</span><span id="__span-0-741"><a id="__codelineno-0-741" name="__codelineno-0-741"></a>                <span class="n">id_cols</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">sumstats_table</span><span class="p">[</span><span class="n">c</span><span class="p">]</span><span class="o">.</span><span class="n">identifier_cols</span>
</span><span id="__span-0-742"><a id="__codelineno-0-742" name="__codelineno-0-742"></a>
</span><span id="__span-0-743"><a id="__codelineno-0-743" name="__codelineno-0-743"></a>                <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">genotype</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
</span><span id="__span-0-744"><a id="__codelineno-0-744" name="__codelineno-0-744"></a>                    <span class="bp">self</span><span class="o">.</span><span class="n">sumstats_table</span><span class="p">[</span><span class="n">c</span><span class="p">]</span><span class="o">.</span><span class="n">match</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">genotype</span><span class="p">[</span><span class="n">c</span><span class="p">]</span><span class="o">.</span><span class="n">get_snp_table</span><span class="p">(</span>
</span><span id="__span-0-745"><a id="__codelineno-0-745" name="__codelineno-0-745"></a>                        <span class="n">col_subset</span><span class="o">=</span><span class="n">id_cols</span> <span class="o">+</span> <span class="p">[</span><span class="s1">&#39;A1&#39;</span><span class="p">,</span> <span class="s1">&#39;A2&#39;</span><span class="p">]</span>
</span><span id="__span-0-746"><a id="__codelineno-0-746" name="__codelineno-0-746"></a>                    <span class="p">))</span>
</span><span id="__span-0-747"><a id="__codelineno-0-747" name="__codelineno-0-747"></a>                <span class="k">elif</span> <span class="bp">self</span><span class="o">.</span><span class="n">ld</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
</span><span id="__span-0-748"><a id="__codelineno-0-748" name="__codelineno-0-748"></a>                    <span class="bp">self</span><span class="o">.</span><span class="n">sumstats_table</span><span class="p">[</span><span class="n">c</span><span class="p">]</span><span class="o">.</span><span class="n">match</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">ld</span><span class="p">[</span><span class="n">c</span><span class="p">]</span><span class="o">.</span><span class="n">to_snp_table</span><span class="p">(</span>
</span><span id="__span-0-749"><a id="__codelineno-0-749" name="__codelineno-0-749"></a>                        <span class="n">col_subset</span><span class="o">=</span><span class="n">id_cols</span> <span class="o">+</span> <span class="p">[</span><span class="s1">&#39;A1&#39;</span><span class="p">,</span> <span class="s1">&#39;A2&#39;</span><span class="p">]</span>
</span><span id="__span-0-750"><a id="__codelineno-0-750" name="__codelineno-0-750"></a>                    <span class="p">))</span>
</span><span id="__span-0-751"><a id="__codelineno-0-751" name="__codelineno-0-751"></a>
</span><span id="__span-0-752"><a id="__codelineno-0-752" name="__codelineno-0-752"></a>                <span class="c1"># If during the allele matching process we discover incompatibilities,</span>
</span><span id="__span-0-753"><a id="__codelineno-0-753" name="__codelineno-0-753"></a>                <span class="c1"># we filter those SNPs:</span>
</span><span id="__span-0-754"><a id="__codelineno-0-754" name="__codelineno-0-754"></a>                <span class="k">for</span> <span class="n">ds</span> <span class="ow">in</span> <span class="n">initialized_data_sources</span><span class="p">:</span>
</span><span id="__span-0-755"><a id="__codelineno-0-755" name="__codelineno-0-755"></a>                    <span class="k">if</span> <span class="n">ds</span><span class="p">[</span><span class="n">c</span><span class="p">]</span><span class="o">.</span><span class="n">n_snps</span> <span class="o">!=</span> <span class="bp">self</span><span class="o">.</span><span class="n">sumstats_table</span><span class="p">[</span><span class="n">c</span><span class="p">]</span><span class="o">.</span><span class="n">n_snps</span><span class="p">:</span>
</span><span id="__span-0-756"><a id="__codelineno-0-756" name="__codelineno-0-756"></a>                        <span class="n">ds</span><span class="p">[</span><span class="n">c</span><span class="p">]</span><span class="o">.</span><span class="n">filter_snps</span><span class="p">(</span><span class="n">extract_snps</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">sumstats_table</span><span class="p">[</span><span class="n">c</span><span class="p">]</span><span class="o">.</span><span class="n">snps</span><span class="p">)</span>
</span></code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h2 id="magenpy.GWADataLoader.GWADataLoader.load_ld" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">load_ld</span><span class="p">()</span></code>

<a href="#magenpy.GWADataLoader.GWADataLoader.load_ld" class="headerlink" title="Permanent link">&para;</a></h2>


    <div class="doc doc-contents ">

        <p>A utility method to load the LD matrices to memory from on-disk storage.</p>


            <details class="quote">
              <summary>Source code in <code>magenpy/GWADataLoader.py</code></summary>
              <div class="language-python highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal"><a href="#__codelineno-0-611">611</a></span>
<span class="normal"><a href="#__codelineno-0-612">612</a></span>
<span class="normal"><a href="#__codelineno-0-613">613</a></span>
<span class="normal"><a href="#__codelineno-0-614">614</a></span>
<span class="normal"><a href="#__codelineno-0-615">615</a></span>
<span class="normal"><a href="#__codelineno-0-616">616</a></span>
<span class="normal"><a href="#__codelineno-0-617">617</a></span></pre></div></td><td class="code"><div><pre><span></span><code><span id="__span-0-611"><a id="__codelineno-0-611" name="__codelineno-0-611"></a><span class="k">def</span><span class="w"> </span><span class="nf">load_ld</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
</span><span id="__span-0-612"><a id="__codelineno-0-612" name="__codelineno-0-612"></a><span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
</span><span id="__span-0-613"><a id="__codelineno-0-613" name="__codelineno-0-613"></a><span class="sd">    A utility method to load the LD matrices to memory from on-disk storage.</span>
</span><span id="__span-0-614"><a id="__codelineno-0-614" name="__codelineno-0-614"></a><span class="sd">    &quot;&quot;&quot;</span>
</span><span id="__span-0-615"><a id="__codelineno-0-615" name="__codelineno-0-615"></a>    <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">ld</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
</span><span id="__span-0-616"><a id="__codelineno-0-616" name="__codelineno-0-616"></a>        <span class="k">for</span> <span class="n">ld</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">ld</span><span class="o">.</span><span class="n">values</span><span class="p">():</span>
</span><span id="__span-0-617"><a id="__codelineno-0-617" name="__codelineno-0-617"></a>            <span class="n">ld</span><span class="o">.</span><span class="n">load</span><span class="p">()</span>
</span></code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h2 id="magenpy.GWADataLoader.GWADataLoader.perform_gwas" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">perform_gwas</span><span class="p">(</span><span class="o">**</span><span class="n">gwa_kwargs</span><span class="p">)</span></code>

<a href="#magenpy.GWADataLoader.GWADataLoader.perform_gwas" class="headerlink" title="Permanent link">&para;</a></h2>


    <div class="doc doc-contents ">

        <p>Perform genome-wide association testing of all variants against the phenotype.
This is a utility function that calls the <code>.perform_gwas()</code> method of the
<code>GenotypeMatrix</code> objects associated with GWADataLoader.</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                <code>gwa_kwargs</code>
            </td>
            <td>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Keyword arguments to pass to the GWA functions. Consult stats.gwa.utils for relevant keyword arguments for each backend.</p>
              </div>
            </td>
            <td>
                  <code>{}</code>
            </td>
          </tr>
      </tbody>
    </table>


            <details class="quote">
              <summary>Source code in <code>magenpy/GWADataLoader.py</code></summary>
              <div class="language-python highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal"><a href="#__codelineno-0-758">758</a></span>
<span class="normal"><a href="#__codelineno-0-759">759</a></span>
<span class="normal"><a href="#__codelineno-0-760">760</a></span>
<span class="normal"><a href="#__codelineno-0-761">761</a></span>
<span class="normal"><a href="#__codelineno-0-762">762</a></span>
<span class="normal"><a href="#__codelineno-0-763">763</a></span>
<span class="normal"><a href="#__codelineno-0-764">764</a></span>
<span class="normal"><a href="#__codelineno-0-765">765</a></span>
<span class="normal"><a href="#__codelineno-0-766">766</a></span>
<span class="normal"><a href="#__codelineno-0-767">767</a></span>
<span class="normal"><a href="#__codelineno-0-768">768</a></span>
<span class="normal"><a href="#__codelineno-0-769">769</a></span>
<span class="normal"><a href="#__codelineno-0-770">770</a></span>
<span class="normal"><a href="#__codelineno-0-771">771</a></span>
<span class="normal"><a href="#__codelineno-0-772">772</a></span>
<span class="normal"><a href="#__codelineno-0-773">773</a></span>
<span class="normal"><a href="#__codelineno-0-774">774</a></span>
<span class="normal"><a href="#__codelineno-0-775">775</a></span></pre></div></td><td class="code"><div><pre><span></span><code><span id="__span-0-758"><a id="__codelineno-0-758" name="__codelineno-0-758"></a><span class="k">def</span><span class="w"> </span><span class="nf">perform_gwas</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="o">**</span><span class="n">gwa_kwargs</span><span class="p">):</span>
</span><span id="__span-0-759"><a id="__codelineno-0-759" name="__codelineno-0-759"></a><span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
</span><span id="__span-0-760"><a id="__codelineno-0-760" name="__codelineno-0-760"></a><span class="sd">    Perform genome-wide association testing of all variants against the phenotype.</span>
</span><span id="__span-0-761"><a id="__codelineno-0-761" name="__codelineno-0-761"></a><span class="sd">    This is a utility function that calls the `.perform_gwas()` method of the</span>
</span><span id="__span-0-762"><a id="__codelineno-0-762" name="__codelineno-0-762"></a><span class="sd">    `GenotypeMatrix` objects associated with GWADataLoader.</span>
</span><span id="__span-0-763"><a id="__codelineno-0-763" name="__codelineno-0-763"></a>
</span><span id="__span-0-764"><a id="__codelineno-0-764" name="__codelineno-0-764"></a><span class="sd">    :param gwa_kwargs: Keyword arguments to pass to the GWA functions. Consult stats.gwa.utils</span>
</span><span id="__span-0-765"><a id="__codelineno-0-765" name="__codelineno-0-765"></a><span class="sd">    for relevant keyword arguments for each backend.</span>
</span><span id="__span-0-766"><a id="__codelineno-0-766" name="__codelineno-0-766"></a><span class="sd">    &quot;&quot;&quot;</span>
</span><span id="__span-0-767"><a id="__codelineno-0-767" name="__codelineno-0-767"></a>
</span><span id="__span-0-768"><a id="__codelineno-0-768" name="__codelineno-0-768"></a>    <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;&gt; Performing GWAS...&quot;</span><span class="p">)</span>
</span><span id="__span-0-769"><a id="__codelineno-0-769" name="__codelineno-0-769"></a>
</span><span id="__span-0-770"><a id="__codelineno-0-770" name="__codelineno-0-770"></a>    <span class="bp">self</span><span class="o">.</span><span class="n">sumstats_table</span> <span class="o">=</span> <span class="p">{</span>
</span><span id="__span-0-771"><a id="__codelineno-0-771" name="__codelineno-0-771"></a>        <span class="n">c</span><span class="p">:</span> <span class="n">g</span><span class="o">.</span><span class="n">perform_gwas</span><span class="p">(</span><span class="o">**</span><span class="n">gwa_kwargs</span><span class="p">)</span>
</span><span id="__span-0-772"><a id="__codelineno-0-772" name="__codelineno-0-772"></a>        <span class="k">for</span> <span class="n">c</span><span class="p">,</span> <span class="n">g</span> <span class="ow">in</span> <span class="n">tqdm</span><span class="p">(</span><span class="nb">sorted</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">genotype</span><span class="o">.</span><span class="n">items</span><span class="p">(),</span> <span class="n">key</span><span class="o">=</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">x</span><span class="p">[</span><span class="mi">0</span><span class="p">]),</span>
</span><span id="__span-0-773"><a id="__codelineno-0-773" name="__codelineno-0-773"></a>                         <span class="n">total</span><span class="o">=</span><span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">genotype</span><span class="p">),</span>
</span><span id="__span-0-774"><a id="__codelineno-0-774" name="__codelineno-0-774"></a>                         <span class="n">desc</span><span class="o">=</span><span class="s1">&#39;Performing GWAS&#39;</span><span class="p">)</span>
</span><span id="__span-0-775"><a id="__codelineno-0-775" name="__codelineno-0-775"></a>    <span class="p">}</span>
</span></code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h2 id="magenpy.GWADataLoader.GWADataLoader.predict" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">predict</span><span class="p">(</span><span class="n">beta</span><span class="o">=</span><span class="kc">None</span><span class="p">)</span></code>

<a href="#magenpy.GWADataLoader.GWADataLoader.predict" class="headerlink" title="Permanent link">&para;</a></h2>


    <div class="doc doc-contents ">

        <p>Predict the phenotype for the genotyped samples using the provided effect size
estimates <code>beta</code>. For quantitative traits, this is equivalent to performing
linear scoring. For binary phenotypes, we transform the output using probit link function.</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                <code>beta</code>
            </td>
            <td>
            </td>
            <td>
              <div class="doc-md-description">
                <p>A dictionary where the keys are the chromosome numbers and the values are a vector of effect sizes for each variant on that chromosome. If the betas are not provided, we use the marginal betas by default (if those are available).</p>
              </div>
            </td>
            <td>
                  <code>None</code>
            </td>
          </tr>
      </tbody>
    </table>


            <details class="quote">
              <summary>Source code in <code>magenpy/GWADataLoader.py</code></summary>
              <div class="language-python highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal"><a href="#__codelineno-0-822">822</a></span>
<span class="normal"><a href="#__codelineno-0-823">823</a></span>
<span class="normal"><a href="#__codelineno-0-824">824</a></span>
<span class="normal"><a href="#__codelineno-0-825">825</a></span>
<span class="normal"><a href="#__codelineno-0-826">826</a></span>
<span class="normal"><a href="#__codelineno-0-827">827</a></span>
<span class="normal"><a href="#__codelineno-0-828">828</a></span>
<span class="normal"><a href="#__codelineno-0-829">829</a></span>
<span class="normal"><a href="#__codelineno-0-830">830</a></span>
<span class="normal"><a href="#__codelineno-0-831">831</a></span>
<span class="normal"><a href="#__codelineno-0-832">832</a></span>
<span class="normal"><a href="#__codelineno-0-833">833</a></span>
<span class="normal"><a href="#__codelineno-0-834">834</a></span>
<span class="normal"><a href="#__codelineno-0-835">835</a></span>
<span class="normal"><a href="#__codelineno-0-836">836</a></span>
<span class="normal"><a href="#__codelineno-0-837">837</a></span>
<span class="normal"><a href="#__codelineno-0-838">838</a></span>
<span class="normal"><a href="#__codelineno-0-839">839</a></span>
<span class="normal"><a href="#__codelineno-0-840">840</a></span>
<span class="normal"><a href="#__codelineno-0-841">841</a></span></pre></div></td><td class="code"><div><pre><span></span><code><span id="__span-0-822"><a id="__codelineno-0-822" name="__codelineno-0-822"></a><span class="k">def</span><span class="w"> </span><span class="nf">predict</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">beta</span><span class="o">=</span><span class="kc">None</span><span class="p">):</span>
</span><span id="__span-0-823"><a id="__codelineno-0-823" name="__codelineno-0-823"></a><span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
</span><span id="__span-0-824"><a id="__codelineno-0-824" name="__codelineno-0-824"></a><span class="sd">    Predict the phenotype for the genotyped samples using the provided effect size</span>
</span><span id="__span-0-825"><a id="__codelineno-0-825" name="__codelineno-0-825"></a><span class="sd">    estimates `beta`. For quantitative traits, this is equivalent to performing</span>
</span><span id="__span-0-826"><a id="__codelineno-0-826" name="__codelineno-0-826"></a><span class="sd">    linear scoring. For binary phenotypes, we transform the output using probit link function.</span>
</span><span id="__span-0-827"><a id="__codelineno-0-827" name="__codelineno-0-827"></a>
</span><span id="__span-0-828"><a id="__codelineno-0-828" name="__codelineno-0-828"></a><span class="sd">    :param beta: A dictionary where the keys are the chromosome numbers and the</span>
</span><span id="__span-0-829"><a id="__codelineno-0-829" name="__codelineno-0-829"></a><span class="sd">    values are a vector of effect sizes for each variant on that chromosome. If the</span>
</span><span id="__span-0-830"><a id="__codelineno-0-830" name="__codelineno-0-830"></a><span class="sd">    betas are not provided, we use the marginal betas by default (if those are available).</span>
</span><span id="__span-0-831"><a id="__codelineno-0-831" name="__codelineno-0-831"></a><span class="sd">    &quot;&quot;&quot;</span>
</span><span id="__span-0-832"><a id="__codelineno-0-832" name="__codelineno-0-832"></a>
</span><span id="__span-0-833"><a id="__codelineno-0-833" name="__codelineno-0-833"></a>    <span class="c1"># Perform linear scoring:</span>
</span><span id="__span-0-834"><a id="__codelineno-0-834" name="__codelineno-0-834"></a>    <span class="n">pgs</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">score</span><span class="p">(</span><span class="n">beta</span><span class="p">)</span>
</span><span id="__span-0-835"><a id="__codelineno-0-835" name="__codelineno-0-835"></a>
</span><span id="__span-0-836"><a id="__codelineno-0-836" name="__codelineno-0-836"></a>    <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">phenotype_likelihood</span> <span class="o">==</span> <span class="s1">&#39;binomial&#39;</span><span class="p">:</span>
</span><span id="__span-0-837"><a id="__codelineno-0-837" name="__codelineno-0-837"></a>        <span class="c1"># Apply probit link function:</span>
</span><span id="__span-0-838"><a id="__codelineno-0-838" name="__codelineno-0-838"></a>        <span class="kn">from</span><span class="w"> </span><span class="nn">scipy.stats</span><span class="w"> </span><span class="kn">import</span> <span class="n">norm</span>
</span><span id="__span-0-839"><a id="__codelineno-0-839" name="__codelineno-0-839"></a>        <span class="n">pgs</span> <span class="o">=</span> <span class="n">norm</span><span class="o">.</span><span class="n">cdf</span><span class="p">(</span><span class="n">pgs</span><span class="p">)</span>
</span><span id="__span-0-840"><a id="__codelineno-0-840" name="__codelineno-0-840"></a>
</span><span id="__span-0-841"><a id="__codelineno-0-841" name="__codelineno-0-841"></a>    <span class="k">return</span> <span class="n">pgs</span>
</span></code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h2 id="magenpy.GWADataLoader.GWADataLoader.read_annotations" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">read_annotations</span><span class="p">(</span><span class="n">annot_path</span><span class="p">,</span> <span class="n">annot_format</span><span class="o">=</span><span class="s1">&#39;magenpy&#39;</span><span class="p">,</span> <span class="n">parser</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="o">**</span><span class="n">parse_kwargs</span><span class="p">)</span></code>

<a href="#magenpy.GWADataLoader.GWADataLoader.read_annotations" class="headerlink" title="Permanent link">&para;</a></h2>


    <div class="doc doc-contents ">

        <p>Read the annotation matrix from file. Annotations are a set of features associated
with each SNP and are generally represented in table format.
Consult the documentation for <code>AnnotationMatrix</code> for more details.</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                <code>annot_path</code>
            </td>
            <td>
            </td>
            <td>
              <div class="doc-md-description">
                <p>The path to the annotation file(s). The path may contain a wildcard.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>annot_format</code>
            </td>
            <td>
            </td>
            <td>
              <div class="doc-md-description">
                <p>The format for the summary statistics. Currently, supports the following formats: <code>magenpy</code>, <code>ldsc</code>.</p>
              </div>
            </td>
            <td>
                  <code>&#39;magenpy&#39;</code>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>parser</code>
            </td>
            <td>
            </td>
            <td>
              <div class="doc-md-description">
                <p>If the annotation file does not follow any of the formats above, you can create your own parser by inheriting from the base <code>AnnotationMatrixParser</code> class and passing it here as an argument.</p>
              </div>
            </td>
            <td>
                  <code>None</code>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>parse_kwargs</code>
            </td>
            <td>
            </td>
            <td>
              <div class="doc-md-description">
                <p>keyword arguments for the parser. These are mainly parameters that will be passed to <code>pandas.read_csv</code> function, such as the delimiter, header information, etc.</p>
              </div>
            </td>
            <td>
                  <code>{}</code>
            </td>
          </tr>
      </tbody>
    </table>


            <details class="quote">
              <summary>Source code in <code>magenpy/GWADataLoader.py</code></summary>
              <div class="language-python highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal"><a href="#__codelineno-0-328">328</a></span>
<span class="normal"><a href="#__codelineno-0-329">329</a></span>
<span class="normal"><a href="#__codelineno-0-330">330</a></span>
<span class="normal"><a href="#__codelineno-0-331">331</a></span>
<span class="normal"><a href="#__codelineno-0-332">332</a></span>
<span class="normal"><a href="#__codelineno-0-333">333</a></span>
<span class="normal"><a href="#__codelineno-0-334">334</a></span>
<span class="normal"><a href="#__codelineno-0-335">335</a></span>
<span class="normal"><a href="#__codelineno-0-336">336</a></span>
<span class="normal"><a href="#__codelineno-0-337">337</a></span>
<span class="normal"><a href="#__codelineno-0-338">338</a></span>
<span class="normal"><a href="#__codelineno-0-339">339</a></span>
<span class="normal"><a href="#__codelineno-0-340">340</a></span>
<span class="normal"><a href="#__codelineno-0-341">341</a></span>
<span class="normal"><a href="#__codelineno-0-342">342</a></span>
<span class="normal"><a href="#__codelineno-0-343">343</a></span>
<span class="normal"><a href="#__codelineno-0-344">344</a></span>
<span class="normal"><a href="#__codelineno-0-345">345</a></span>
<span class="normal"><a href="#__codelineno-0-346">346</a></span>
<span class="normal"><a href="#__codelineno-0-347">347</a></span>
<span class="normal"><a href="#__codelineno-0-348">348</a></span>
<span class="normal"><a href="#__codelineno-0-349">349</a></span>
<span class="normal"><a href="#__codelineno-0-350">350</a></span>
<span class="normal"><a href="#__codelineno-0-351">351</a></span>
<span class="normal"><a href="#__codelineno-0-352">352</a></span>
<span class="normal"><a href="#__codelineno-0-353">353</a></span>
<span class="normal"><a href="#__codelineno-0-354">354</a></span>
<span class="normal"><a href="#__codelineno-0-355">355</a></span>
<span class="normal"><a href="#__codelineno-0-356">356</a></span>
<span class="normal"><a href="#__codelineno-0-357">357</a></span>
<span class="normal"><a href="#__codelineno-0-358">358</a></span>
<span class="normal"><a href="#__codelineno-0-359">359</a></span>
<span class="normal"><a href="#__codelineno-0-360">360</a></span>
<span class="normal"><a href="#__codelineno-0-361">361</a></span>
<span class="normal"><a href="#__codelineno-0-362">362</a></span>
<span class="normal"><a href="#__codelineno-0-363">363</a></span>
<span class="normal"><a href="#__codelineno-0-364">364</a></span>
<span class="normal"><a href="#__codelineno-0-365">365</a></span>
<span class="normal"><a href="#__codelineno-0-366">366</a></span>
<span class="normal"><a href="#__codelineno-0-367">367</a></span>
<span class="normal"><a href="#__codelineno-0-368">368</a></span>
<span class="normal"><a href="#__codelineno-0-369">369</a></span>
<span class="normal"><a href="#__codelineno-0-370">370</a></span></pre></div></td><td class="code"><div><pre><span></span><code><span id="__span-0-328"><a id="__codelineno-0-328" name="__codelineno-0-328"></a><span class="k">def</span><span class="w"> </span><span class="nf">read_annotations</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">annot_path</span><span class="p">,</span>
</span><span id="__span-0-329"><a id="__codelineno-0-329" name="__codelineno-0-329"></a>                     <span class="n">annot_format</span><span class="o">=</span><span class="s1">&#39;magenpy&#39;</span><span class="p">,</span>
</span><span id="__span-0-330"><a id="__codelineno-0-330" name="__codelineno-0-330"></a>                     <span class="n">parser</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
</span><span id="__span-0-331"><a id="__codelineno-0-331" name="__codelineno-0-331"></a>                     <span class="o">**</span><span class="n">parse_kwargs</span><span class="p">):</span>
</span><span id="__span-0-332"><a id="__codelineno-0-332" name="__codelineno-0-332"></a><span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
</span><span id="__span-0-333"><a id="__codelineno-0-333" name="__codelineno-0-333"></a><span class="sd">    Read the annotation matrix from file. Annotations are a set of features associated</span>
</span><span id="__span-0-334"><a id="__codelineno-0-334" name="__codelineno-0-334"></a><span class="sd">    with each SNP and are generally represented in table format.</span>
</span><span id="__span-0-335"><a id="__codelineno-0-335" name="__codelineno-0-335"></a><span class="sd">    Consult the documentation for `AnnotationMatrix` for more details.</span>
</span><span id="__span-0-336"><a id="__codelineno-0-336" name="__codelineno-0-336"></a>
</span><span id="__span-0-337"><a id="__codelineno-0-337" name="__codelineno-0-337"></a><span class="sd">    :param annot_path: The path to the annotation file(s). The path may contain a wildcard.</span>
</span><span id="__span-0-338"><a id="__codelineno-0-338" name="__codelineno-0-338"></a><span class="sd">    :param annot_format: The format for the summary statistics. Currently, supports the following</span>
</span><span id="__span-0-339"><a id="__codelineno-0-339" name="__codelineno-0-339"></a><span class="sd">     formats: `magenpy`, `ldsc`.</span>
</span><span id="__span-0-340"><a id="__codelineno-0-340" name="__codelineno-0-340"></a><span class="sd">    :param parser: If the annotation file does not follow any of the formats above, you can create</span>
</span><span id="__span-0-341"><a id="__codelineno-0-341" name="__codelineno-0-341"></a><span class="sd">    your own parser by inheriting from the base `AnnotationMatrixParser` class and passing it here as an argument.</span>
</span><span id="__span-0-342"><a id="__codelineno-0-342" name="__codelineno-0-342"></a><span class="sd">    :param parse_kwargs: keyword arguments for the parser. These are mainly parameters that will be passed to</span>
</span><span id="__span-0-343"><a id="__codelineno-0-343" name="__codelineno-0-343"></a><span class="sd">    `pandas.read_csv` function, such as the delimiter, header information, etc.</span>
</span><span id="__span-0-344"><a id="__codelineno-0-344" name="__codelineno-0-344"></a><span class="sd">    &quot;&quot;&quot;</span>
</span><span id="__span-0-345"><a id="__codelineno-0-345" name="__codelineno-0-345"></a>
</span><span id="__span-0-346"><a id="__codelineno-0-346" name="__codelineno-0-346"></a>    <span class="k">if</span> <span class="n">annot_path</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
</span><span id="__span-0-347"><a id="__codelineno-0-347" name="__codelineno-0-347"></a>        <span class="k">return</span>
</span><span id="__span-0-348"><a id="__codelineno-0-348" name="__codelineno-0-348"></a>
</span><span id="__span-0-349"><a id="__codelineno-0-349" name="__codelineno-0-349"></a>    <span class="c1"># Find all the relevant files in the path passed by the user:</span>
</span><span id="__span-0-350"><a id="__codelineno-0-350" name="__codelineno-0-350"></a>    <span class="k">if</span> <span class="ow">not</span> <span class="n">iterable</span><span class="p">(</span><span class="n">annot_path</span><span class="p">):</span>
</span><span id="__span-0-351"><a id="__codelineno-0-351" name="__codelineno-0-351"></a>        <span class="n">annot_files</span> <span class="o">=</span> <span class="n">get_filenames</span><span class="p">(</span><span class="n">annot_path</span><span class="p">,</span> <span class="n">extension</span><span class="o">=</span><span class="s1">&#39;.annot&#39;</span><span class="p">)</span>
</span><span id="__span-0-352"><a id="__codelineno-0-352" name="__codelineno-0-352"></a>    <span class="k">else</span><span class="p">:</span>
</span><span id="__span-0-353"><a id="__codelineno-0-353" name="__codelineno-0-353"></a>        <span class="n">annot_files</span> <span class="o">=</span> <span class="n">annot_path</span>
</span><span id="__span-0-354"><a id="__codelineno-0-354" name="__codelineno-0-354"></a>
</span><span id="__span-0-355"><a id="__codelineno-0-355" name="__codelineno-0-355"></a>    <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">annot_files</span><span class="p">)</span> <span class="o">&lt;</span> <span class="mi">1</span><span class="p">:</span>
</span><span id="__span-0-356"><a id="__codelineno-0-356" name="__codelineno-0-356"></a>        <span class="n">logger</span><span class="o">.</span><span class="n">warning</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;No annotation files were found at: </span><span class="si">{</span><span class="n">annot_path</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
</span><span id="__span-0-357"><a id="__codelineno-0-357" name="__codelineno-0-357"></a>        <span class="k">return</span>
</span><span id="__span-0-358"><a id="__codelineno-0-358" name="__codelineno-0-358"></a>
</span><span id="__span-0-359"><a id="__codelineno-0-359" name="__codelineno-0-359"></a>    <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;&gt; Reading annotation file...&quot;</span><span class="p">)</span>
</span><span id="__span-0-360"><a id="__codelineno-0-360" name="__codelineno-0-360"></a>
</span><span id="__span-0-361"><a id="__codelineno-0-361" name="__codelineno-0-361"></a>    <span class="bp">self</span><span class="o">.</span><span class="n">annotation</span> <span class="o">=</span> <span class="p">{}</span>
</span><span id="__span-0-362"><a id="__codelineno-0-362" name="__codelineno-0-362"></a>
</span><span id="__span-0-363"><a id="__codelineno-0-363" name="__codelineno-0-363"></a>    <span class="k">for</span> <span class="n">annot_file</span> <span class="ow">in</span> <span class="n">tqdm</span><span class="p">(</span><span class="n">annot_files</span><span class="p">,</span>
</span><span id="__span-0-364"><a id="__codelineno-0-364" name="__codelineno-0-364"></a>                           <span class="n">total</span><span class="o">=</span><span class="nb">len</span><span class="p">(</span><span class="n">annot_files</span><span class="p">),</span>
</span><span id="__span-0-365"><a id="__codelineno-0-365" name="__codelineno-0-365"></a>                           <span class="n">desc</span><span class="o">=</span><span class="s2">&quot;Reading annotation files&quot;</span><span class="p">):</span>
</span><span id="__span-0-366"><a id="__codelineno-0-366" name="__codelineno-0-366"></a>        <span class="n">annot_mat</span> <span class="o">=</span> <span class="n">AnnotationMatrix</span><span class="o">.</span><span class="n">from_file</span><span class="p">(</span><span class="n">annot_file</span><span class="p">,</span>
</span><span id="__span-0-367"><a id="__codelineno-0-367" name="__codelineno-0-367"></a>                                               <span class="n">annot_format</span><span class="o">=</span><span class="n">annot_format</span><span class="p">,</span>
</span><span id="__span-0-368"><a id="__codelineno-0-368" name="__codelineno-0-368"></a>                                               <span class="n">annot_parser</span><span class="o">=</span><span class="n">parser</span><span class="p">,</span>
</span><span id="__span-0-369"><a id="__codelineno-0-369" name="__codelineno-0-369"></a>                                               <span class="o">**</span><span class="n">parse_kwargs</span><span class="p">)</span>
</span><span id="__span-0-370"><a id="__codelineno-0-370" name="__codelineno-0-370"></a>        <span class="bp">self</span><span class="o">.</span><span class="n">annotation</span><span class="p">[</span><span class="n">annot_mat</span><span class="o">.</span><span class="n">chromosome</span><span class="p">]</span> <span class="o">=</span> <span class="n">annot_mat</span>
</span></code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h2 id="magenpy.GWADataLoader.GWADataLoader.read_covariates" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">read_covariates</span><span class="p">(</span><span class="n">covariates_file</span><span class="p">,</span> <span class="o">**</span><span class="n">read_csv_kwargs</span><span class="p">)</span></code>

<a href="#magenpy.GWADataLoader.GWADataLoader.read_covariates" class="headerlink" title="Permanent link">&para;</a></h2>


    <div class="doc doc-contents ">

        <p>Read the covariates file and integrate it with the sample tables and genotype matrices.</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                <code>covariates_file</code>
            </td>
            <td>
            </td>
            <td>
              <div class="doc-md-description">
                <p>The path to the covariates file (Default: tab-separated file starting with the <code>FID IID ...</code> columns and followed by the covariate columns).</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>read_csv_kwargs</code>
            </td>
            <td>
            </td>
            <td>
              <div class="doc-md-description">
                <p>keyword arguments for the <code>read_csv</code> function of <code>pandas</code>.</p>
              </div>
            </td>
            <td>
                  <code>{}</code>
            </td>
          </tr>
      </tbody>
    </table>


            <details class="quote">
              <summary>Source code in <code>magenpy/GWADataLoader.py</code></summary>
              <div class="language-python highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal"><a href="#__codelineno-0-482">482</a></span>
<span class="normal"><a href="#__codelineno-0-483">483</a></span>
<span class="normal"><a href="#__codelineno-0-484">484</a></span>
<span class="normal"><a href="#__codelineno-0-485">485</a></span>
<span class="normal"><a href="#__codelineno-0-486">486</a></span>
<span class="normal"><a href="#__codelineno-0-487">487</a></span>
<span class="normal"><a href="#__codelineno-0-488">488</a></span>
<span class="normal"><a href="#__codelineno-0-489">489</a></span>
<span class="normal"><a href="#__codelineno-0-490">490</a></span>
<span class="normal"><a href="#__codelineno-0-491">491</a></span>
<span class="normal"><a href="#__codelineno-0-492">492</a></span>
<span class="normal"><a href="#__codelineno-0-493">493</a></span>
<span class="normal"><a href="#__codelineno-0-494">494</a></span>
<span class="normal"><a href="#__codelineno-0-495">495</a></span>
<span class="normal"><a href="#__codelineno-0-496">496</a></span>
<span class="normal"><a href="#__codelineno-0-497">497</a></span>
<span class="normal"><a href="#__codelineno-0-498">498</a></span>
<span class="normal"><a href="#__codelineno-0-499">499</a></span></pre></div></td><td class="code"><div><pre><span></span><code><span id="__span-0-482"><a id="__codelineno-0-482" name="__codelineno-0-482"></a><span class="k">def</span><span class="w"> </span><span class="nf">read_covariates</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">covariates_file</span><span class="p">,</span> <span class="o">**</span><span class="n">read_csv_kwargs</span><span class="p">):</span>
</span><span id="__span-0-483"><a id="__codelineno-0-483" name="__codelineno-0-483"></a><span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
</span><span id="__span-0-484"><a id="__codelineno-0-484" name="__codelineno-0-484"></a><span class="sd">    Read the covariates file and integrate it with the sample tables and genotype matrices.</span>
</span><span id="__span-0-485"><a id="__codelineno-0-485" name="__codelineno-0-485"></a>
</span><span id="__span-0-486"><a id="__codelineno-0-486" name="__codelineno-0-486"></a><span class="sd">    :param covariates_file: The path to the covariates file</span>
</span><span id="__span-0-487"><a id="__codelineno-0-487" name="__codelineno-0-487"></a><span class="sd">    (Default: tab-separated file starting with the `FID IID ...` columns and followed by the covariate columns).</span>
</span><span id="__span-0-488"><a id="__codelineno-0-488" name="__codelineno-0-488"></a><span class="sd">    :param read_csv_kwargs: keyword arguments for the `read_csv` function of `pandas`.</span>
</span><span id="__span-0-489"><a id="__codelineno-0-489" name="__codelineno-0-489"></a><span class="sd">    &quot;&quot;&quot;</span>
</span><span id="__span-0-490"><a id="__codelineno-0-490" name="__codelineno-0-490"></a>
</span><span id="__span-0-491"><a id="__codelineno-0-491" name="__codelineno-0-491"></a>    <span class="k">if</span> <span class="n">covariates_file</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
</span><span id="__span-0-492"><a id="__codelineno-0-492" name="__codelineno-0-492"></a>        <span class="k">return</span>
</span><span id="__span-0-493"><a id="__codelineno-0-493" name="__codelineno-0-493"></a>
</span><span id="__span-0-494"><a id="__codelineno-0-494" name="__codelineno-0-494"></a>    <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;&gt; Reading covariates file...&quot;</span><span class="p">)</span>
</span><span id="__span-0-495"><a id="__codelineno-0-495" name="__codelineno-0-495"></a>
</span><span id="__span-0-496"><a id="__codelineno-0-496" name="__codelineno-0-496"></a>    <span class="k">assert</span> <span class="bp">self</span><span class="o">.</span><span class="n">sample_table</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span>
</span><span id="__span-0-497"><a id="__codelineno-0-497" name="__codelineno-0-497"></a>
</span><span id="__span-0-498"><a id="__codelineno-0-498" name="__codelineno-0-498"></a>    <span class="bp">self</span><span class="o">.</span><span class="n">sample_table</span><span class="o">.</span><span class="n">read_covariates_file</span><span class="p">(</span><span class="n">covariates_file</span><span class="p">,</span> <span class="o">**</span><span class="n">read_csv_kwargs</span><span class="p">)</span>
</span><span id="__span-0-499"><a id="__codelineno-0-499" name="__codelineno-0-499"></a>    <span class="bp">self</span><span class="o">.</span><span class="n">sync_sample_tables</span><span class="p">()</span>
</span></code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h2 id="magenpy.GWADataLoader.GWADataLoader.read_genotypes" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">read_genotypes</span><span class="p">(</span><span class="n">bed_paths</span><span class="p">,</span> <span class="n">keep_samples</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">keep_file</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">extract_snps</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">extract_file</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">min_maf</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">min_mac</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span> <span class="n">drop_duplicated</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span></code>

<a href="#magenpy.GWADataLoader.GWADataLoader.read_genotypes" class="headerlink" title="Permanent link">&para;</a></h2>


    <div class="doc doc-contents ">

        <p>Read the genotype matrix and/or associated metadata from plink's BED file format.
Consult the documentation for <code>GenotypeMatrix</code> for more details.</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                <code>bed_paths</code>
            </td>
            <td>
            </td>
            <td>
              <div class="doc-md-description">
                <p>The path to the BED file(s). You may use a wildcard here to read files for multiple chromosomes.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>keep_samples</code>
            </td>
            <td>
            </td>
            <td>
              <div class="doc-md-description">
                <p>A vector or list of sample IDs to keep when filtering the genotype matrix.</p>
              </div>
            </td>
            <td>
                  <code>None</code>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>keep_file</code>
            </td>
            <td>
            </td>
            <td>
              <div class="doc-md-description">
                <p>A path to a plink-style file containing sample IDs to keep.</p>
              </div>
            </td>
            <td>
                  <code>None</code>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>extract_snps</code>
            </td>
            <td>
            </td>
            <td>
              <div class="doc-md-description">
                <p>A vector or list of SNP IDs to keep when filtering the genotype matrix.</p>
              </div>
            </td>
            <td>
                  <code>None</code>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>extract_file</code>
            </td>
            <td>
            </td>
            <td>
              <div class="doc-md-description">
                <p>A path to a plink-style file containing SNP IDs to keep.</p>
              </div>
            </td>
            <td>
                  <code>None</code>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>min_maf</code>
            </td>
            <td>
            </td>
            <td>
              <div class="doc-md-description">
                <p>The minimum minor allele frequency cutoff.</p>
              </div>
            </td>
            <td>
                  <code>None</code>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>min_mac</code>
            </td>
            <td>
            </td>
            <td>
              <div class="doc-md-description">
                <p>The minimum minor allele count cutoff.</p>
              </div>
            </td>
            <td>
                  <code>1</code>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>drop_duplicated</code>
            </td>
            <td>
            </td>
            <td>
              <div class="doc-md-description">
                <p>If True, drop SNPs with duplicated rsID.</p>
              </div>
            </td>
            <td>
                  <code>True</code>
            </td>
          </tr>
      </tbody>
    </table>


            <details class="quote">
              <summary>Source code in <code>magenpy/GWADataLoader.py</code></summary>
              <div class="language-python highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal"><a href="#__codelineno-0-372">372</a></span>
<span class="normal"><a href="#__codelineno-0-373">373</a></span>
<span class="normal"><a href="#__codelineno-0-374">374</a></span>
<span class="normal"><a href="#__codelineno-0-375">375</a></span>
<span class="normal"><a href="#__codelineno-0-376">376</a></span>
<span class="normal"><a href="#__codelineno-0-377">377</a></span>
<span class="normal"><a href="#__codelineno-0-378">378</a></span>
<span class="normal"><a href="#__codelineno-0-379">379</a></span>
<span class="normal"><a href="#__codelineno-0-380">380</a></span>
<span class="normal"><a href="#__codelineno-0-381">381</a></span>
<span class="normal"><a href="#__codelineno-0-382">382</a></span>
<span class="normal"><a href="#__codelineno-0-383">383</a></span>
<span class="normal"><a href="#__codelineno-0-384">384</a></span>
<span class="normal"><a href="#__codelineno-0-385">385</a></span>
<span class="normal"><a href="#__codelineno-0-386">386</a></span>
<span class="normal"><a href="#__codelineno-0-387">387</a></span>
<span class="normal"><a href="#__codelineno-0-388">388</a></span>
<span class="normal"><a href="#__codelineno-0-389">389</a></span>
<span class="normal"><a href="#__codelineno-0-390">390</a></span>
<span class="normal"><a href="#__codelineno-0-391">391</a></span>
<span class="normal"><a href="#__codelineno-0-392">392</a></span>
<span class="normal"><a href="#__codelineno-0-393">393</a></span>
<span class="normal"><a href="#__codelineno-0-394">394</a></span>
<span class="normal"><a href="#__codelineno-0-395">395</a></span>
<span class="normal"><a href="#__codelineno-0-396">396</a></span>
<span class="normal"><a href="#__codelineno-0-397">397</a></span>
<span class="normal"><a href="#__codelineno-0-398">398</a></span>
<span class="normal"><a href="#__codelineno-0-399">399</a></span>
<span class="normal"><a href="#__codelineno-0-400">400</a></span>
<span class="normal"><a href="#__codelineno-0-401">401</a></span>
<span class="normal"><a href="#__codelineno-0-402">402</a></span>
<span class="normal"><a href="#__codelineno-0-403">403</a></span>
<span class="normal"><a href="#__codelineno-0-404">404</a></span>
<span class="normal"><a href="#__codelineno-0-405">405</a></span>
<span class="normal"><a href="#__codelineno-0-406">406</a></span>
<span class="normal"><a href="#__codelineno-0-407">407</a></span>
<span class="normal"><a href="#__codelineno-0-408">408</a></span>
<span class="normal"><a href="#__codelineno-0-409">409</a></span>
<span class="normal"><a href="#__codelineno-0-410">410</a></span>
<span class="normal"><a href="#__codelineno-0-411">411</a></span>
<span class="normal"><a href="#__codelineno-0-412">412</a></span>
<span class="normal"><a href="#__codelineno-0-413">413</a></span>
<span class="normal"><a href="#__codelineno-0-414">414</a></span>
<span class="normal"><a href="#__codelineno-0-415">415</a></span>
<span class="normal"><a href="#__codelineno-0-416">416</a></span>
<span class="normal"><a href="#__codelineno-0-417">417</a></span>
<span class="normal"><a href="#__codelineno-0-418">418</a></span>
<span class="normal"><a href="#__codelineno-0-419">419</a></span>
<span class="normal"><a href="#__codelineno-0-420">420</a></span>
<span class="normal"><a href="#__codelineno-0-421">421</a></span>
<span class="normal"><a href="#__codelineno-0-422">422</a></span>
<span class="normal"><a href="#__codelineno-0-423">423</a></span>
<span class="normal"><a href="#__codelineno-0-424">424</a></span>
<span class="normal"><a href="#__codelineno-0-425">425</a></span>
<span class="normal"><a href="#__codelineno-0-426">426</a></span>
<span class="normal"><a href="#__codelineno-0-427">427</a></span>
<span class="normal"><a href="#__codelineno-0-428">428</a></span>
<span class="normal"><a href="#__codelineno-0-429">429</a></span>
<span class="normal"><a href="#__codelineno-0-430">430</a></span>
<span class="normal"><a href="#__codelineno-0-431">431</a></span>
<span class="normal"><a href="#__codelineno-0-432">432</a></span>
<span class="normal"><a href="#__codelineno-0-433">433</a></span>
<span class="normal"><a href="#__codelineno-0-434">434</a></span>
<span class="normal"><a href="#__codelineno-0-435">435</a></span>
<span class="normal"><a href="#__codelineno-0-436">436</a></span>
<span class="normal"><a href="#__codelineno-0-437">437</a></span>
<span class="normal"><a href="#__codelineno-0-438">438</a></span>
<span class="normal"><a href="#__codelineno-0-439">439</a></span>
<span class="normal"><a href="#__codelineno-0-440">440</a></span>
<span class="normal"><a href="#__codelineno-0-441">441</a></span>
<span class="normal"><a href="#__codelineno-0-442">442</a></span>
<span class="normal"><a href="#__codelineno-0-443">443</a></span>
<span class="normal"><a href="#__codelineno-0-444">444</a></span>
<span class="normal"><a href="#__codelineno-0-445">445</a></span>
<span class="normal"><a href="#__codelineno-0-446">446</a></span>
<span class="normal"><a href="#__codelineno-0-447">447</a></span>
<span class="normal"><a href="#__codelineno-0-448">448</a></span></pre></div></td><td class="code"><div><pre><span></span><code><span id="__span-0-372"><a id="__codelineno-0-372" name="__codelineno-0-372"></a><span class="k">def</span><span class="w"> </span><span class="nf">read_genotypes</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span>
</span><span id="__span-0-373"><a id="__codelineno-0-373" name="__codelineno-0-373"></a>                   <span class="n">bed_paths</span><span class="p">,</span>
</span><span id="__span-0-374"><a id="__codelineno-0-374" name="__codelineno-0-374"></a>                   <span class="n">keep_samples</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
</span><span id="__span-0-375"><a id="__codelineno-0-375" name="__codelineno-0-375"></a>                   <span class="n">keep_file</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
</span><span id="__span-0-376"><a id="__codelineno-0-376" name="__codelineno-0-376"></a>                   <span class="n">extract_snps</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
</span><span id="__span-0-377"><a id="__codelineno-0-377" name="__codelineno-0-377"></a>                   <span class="n">extract_file</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
</span><span id="__span-0-378"><a id="__codelineno-0-378" name="__codelineno-0-378"></a>                   <span class="n">min_maf</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
</span><span id="__span-0-379"><a id="__codelineno-0-379" name="__codelineno-0-379"></a>                   <span class="n">min_mac</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span>
</span><span id="__span-0-380"><a id="__codelineno-0-380" name="__codelineno-0-380"></a>                   <span class="n">drop_duplicated</span><span class="o">=</span><span class="kc">True</span><span class="p">):</span>
</span><span id="__span-0-381"><a id="__codelineno-0-381" name="__codelineno-0-381"></a><span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
</span><span id="__span-0-382"><a id="__codelineno-0-382" name="__codelineno-0-382"></a><span class="sd">    Read the genotype matrix and/or associated metadata from plink&#39;s BED file format.</span>
</span><span id="__span-0-383"><a id="__codelineno-0-383" name="__codelineno-0-383"></a><span class="sd">    Consult the documentation for `GenotypeMatrix` for more details.</span>
</span><span id="__span-0-384"><a id="__codelineno-0-384" name="__codelineno-0-384"></a>
</span><span id="__span-0-385"><a id="__codelineno-0-385" name="__codelineno-0-385"></a><span class="sd">    :param bed_paths: The path to the BED file(s). You may use a wildcard here to read files for multiple</span>
</span><span id="__span-0-386"><a id="__codelineno-0-386" name="__codelineno-0-386"></a><span class="sd">    chromosomes.</span>
</span><span id="__span-0-387"><a id="__codelineno-0-387" name="__codelineno-0-387"></a><span class="sd">    :param keep_samples: A vector or list of sample IDs to keep when filtering the genotype matrix.</span>
</span><span id="__span-0-388"><a id="__codelineno-0-388" name="__codelineno-0-388"></a><span class="sd">    :param keep_file: A path to a plink-style file containing sample IDs to keep.</span>
</span><span id="__span-0-389"><a id="__codelineno-0-389" name="__codelineno-0-389"></a><span class="sd">    :param extract_snps: A vector or list of SNP IDs to keep when filtering the genotype matrix.</span>
</span><span id="__span-0-390"><a id="__codelineno-0-390" name="__codelineno-0-390"></a><span class="sd">    :param extract_file: A path to a plink-style file containing SNP IDs to keep.</span>
</span><span id="__span-0-391"><a id="__codelineno-0-391" name="__codelineno-0-391"></a><span class="sd">    :param min_maf: The minimum minor allele frequency cutoff.</span>
</span><span id="__span-0-392"><a id="__codelineno-0-392" name="__codelineno-0-392"></a><span class="sd">    :param min_mac: The minimum minor allele count cutoff.</span>
</span><span id="__span-0-393"><a id="__codelineno-0-393" name="__codelineno-0-393"></a><span class="sd">    :param drop_duplicated: If True, drop SNPs with duplicated rsID.</span>
</span><span id="__span-0-394"><a id="__codelineno-0-394" name="__codelineno-0-394"></a><span class="sd">    &quot;&quot;&quot;</span>
</span><span id="__span-0-395"><a id="__codelineno-0-395" name="__codelineno-0-395"></a>
</span><span id="__span-0-396"><a id="__codelineno-0-396" name="__codelineno-0-396"></a>    <span class="k">if</span> <span class="n">bed_paths</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
</span><span id="__span-0-397"><a id="__codelineno-0-397" name="__codelineno-0-397"></a>        <span class="k">return</span>
</span><span id="__span-0-398"><a id="__codelineno-0-398" name="__codelineno-0-398"></a>
</span><span id="__span-0-399"><a id="__codelineno-0-399" name="__codelineno-0-399"></a>    <span class="c1"># Find all the relevant files in the path passed by the user:</span>
</span><span id="__span-0-400"><a id="__codelineno-0-400" name="__codelineno-0-400"></a>    <span class="k">if</span> <span class="ow">not</span> <span class="n">iterable</span><span class="p">(</span><span class="n">bed_paths</span><span class="p">):</span>
</span><span id="__span-0-401"><a id="__codelineno-0-401" name="__codelineno-0-401"></a>        <span class="n">bed_files</span> <span class="o">=</span> <span class="n">get_filenames</span><span class="p">(</span><span class="n">bed_paths</span><span class="p">,</span> <span class="n">extension</span><span class="o">=</span><span class="s1">&#39;.bed&#39;</span><span class="p">)</span>
</span><span id="__span-0-402"><a id="__codelineno-0-402" name="__codelineno-0-402"></a>    <span class="k">else</span><span class="p">:</span>
</span><span id="__span-0-403"><a id="__codelineno-0-403" name="__codelineno-0-403"></a>        <span class="n">bed_files</span> <span class="o">=</span> <span class="n">bed_paths</span>
</span><span id="__span-0-404"><a id="__codelineno-0-404" name="__codelineno-0-404"></a>
</span><span id="__span-0-405"><a id="__codelineno-0-405" name="__codelineno-0-405"></a>    <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">bed_files</span><span class="p">)</span> <span class="o">&lt;</span> <span class="mi">1</span><span class="p">:</span>
</span><span id="__span-0-406"><a id="__codelineno-0-406" name="__codelineno-0-406"></a>        <span class="n">logger</span><span class="o">.</span><span class="n">warning</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;No BED files were found at: </span><span class="si">{</span><span class="n">bed_paths</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
</span><span id="__span-0-407"><a id="__codelineno-0-407" name="__codelineno-0-407"></a>        <span class="k">return</span>
</span><span id="__span-0-408"><a id="__codelineno-0-408" name="__codelineno-0-408"></a>
</span><span id="__span-0-409"><a id="__codelineno-0-409" name="__codelineno-0-409"></a>    <span class="c1"># Depending on the backend, select the `GenotypeMatrix` class:</span>
</span><span id="__span-0-410"><a id="__codelineno-0-410" name="__codelineno-0-410"></a>    <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">backend</span> <span class="o">==</span> <span class="s1">&#39;xarray&#39;</span><span class="p">:</span>
</span><span id="__span-0-411"><a id="__codelineno-0-411" name="__codelineno-0-411"></a>        <span class="n">gmat_class</span> <span class="o">=</span> <span class="n">xarrayGenotypeMatrix</span>
</span><span id="__span-0-412"><a id="__codelineno-0-412" name="__codelineno-0-412"></a>    <span class="k">elif</span> <span class="bp">self</span><span class="o">.</span><span class="n">backend</span> <span class="o">==</span> <span class="s1">&#39;bed-reader&#39;</span><span class="p">:</span>
</span><span id="__span-0-413"><a id="__codelineno-0-413" name="__codelineno-0-413"></a>        <span class="n">gmat_class</span> <span class="o">=</span> <span class="n">bedReaderGenotypeMatrix</span>
</span><span id="__span-0-414"><a id="__codelineno-0-414" name="__codelineno-0-414"></a>    <span class="k">else</span><span class="p">:</span>
</span><span id="__span-0-415"><a id="__codelineno-0-415" name="__codelineno-0-415"></a>        <span class="n">gmat_class</span> <span class="o">=</span> <span class="n">plinkBEDGenotypeMatrix</span>
</span><span id="__span-0-416"><a id="__codelineno-0-416" name="__codelineno-0-416"></a>
</span><span id="__span-0-417"><a id="__codelineno-0-417" name="__codelineno-0-417"></a>    <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;&gt; Reading genotype metadata...&quot;</span><span class="p">)</span>
</span><span id="__span-0-418"><a id="__codelineno-0-418" name="__codelineno-0-418"></a>
</span><span id="__span-0-419"><a id="__codelineno-0-419" name="__codelineno-0-419"></a>    <span class="bp">self</span><span class="o">.</span><span class="n">genotype</span> <span class="o">=</span> <span class="p">{}</span>
</span><span id="__span-0-420"><a id="__codelineno-0-420" name="__codelineno-0-420"></a>
</span><span id="__span-0-421"><a id="__codelineno-0-421" name="__codelineno-0-421"></a>    <span class="k">for</span> <span class="n">bfile</span> <span class="ow">in</span> <span class="n">tqdm</span><span class="p">(</span><span class="n">bed_files</span><span class="p">,</span>
</span><span id="__span-0-422"><a id="__codelineno-0-422" name="__codelineno-0-422"></a>                      <span class="n">total</span><span class="o">=</span><span class="nb">len</span><span class="p">(</span><span class="n">bed_files</span><span class="p">),</span>
</span><span id="__span-0-423"><a id="__codelineno-0-423" name="__codelineno-0-423"></a>                      <span class="n">desc</span><span class="o">=</span><span class="s2">&quot;Reading genotype metadata&quot;</span><span class="p">):</span>
</span><span id="__span-0-424"><a id="__codelineno-0-424" name="__codelineno-0-424"></a>        <span class="c1"># Read BED file and update the genotypes dictionary:</span>
</span><span id="__span-0-425"><a id="__codelineno-0-425" name="__codelineno-0-425"></a>        <span class="bp">self</span><span class="o">.</span><span class="n">genotype</span><span class="o">.</span><span class="n">update</span><span class="p">(</span><span class="n">gmat_class</span><span class="o">.</span><span class="n">from_file</span><span class="p">(</span><span class="n">bfile</span><span class="p">,</span>
</span><span id="__span-0-426"><a id="__codelineno-0-426" name="__codelineno-0-426"></a>                                                  <span class="n">temp_dir</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">temp_dir</span><span class="p">,</span>
</span><span id="__span-0-427"><a id="__codelineno-0-427" name="__codelineno-0-427"></a>                                                  <span class="n">threads</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">threads</span><span class="p">)</span><span class="o">.</span><span class="n">split_by_chromosome</span><span class="p">())</span>
</span><span id="__span-0-428"><a id="__codelineno-0-428" name="__codelineno-0-428"></a>
</span><span id="__span-0-429"><a id="__codelineno-0-429" name="__codelineno-0-429"></a>    <span class="c1"># After reading the genotype matrices, apply some standard filters:</span>
</span><span id="__span-0-430"><a id="__codelineno-0-430" name="__codelineno-0-430"></a>    <span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="p">(</span><span class="n">c</span><span class="p">,</span> <span class="n">g</span><span class="p">)</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">genotype</span><span class="o">.</span><span class="n">items</span><span class="p">()):</span>
</span><span id="__span-0-431"><a id="__codelineno-0-431" name="__codelineno-0-431"></a>
</span><span id="__span-0-432"><a id="__codelineno-0-432" name="__codelineno-0-432"></a>        <span class="c1"># Filter the genotype matrix to keep certain subsample:</span>
</span><span id="__span-0-433"><a id="__codelineno-0-433" name="__codelineno-0-433"></a>        <span class="k">if</span> <span class="n">keep_samples</span> <span class="ow">or</span> <span class="n">keep_file</span><span class="p">:</span>
</span><span id="__span-0-434"><a id="__codelineno-0-434" name="__codelineno-0-434"></a>            <span class="n">g</span><span class="o">.</span><span class="n">filter_samples</span><span class="p">(</span><span class="n">keep_samples</span><span class="o">=</span><span class="n">keep_samples</span><span class="p">,</span> <span class="n">keep_file</span><span class="o">=</span><span class="n">keep_file</span><span class="p">)</span>
</span><span id="__span-0-435"><a id="__codelineno-0-435" name="__codelineno-0-435"></a>
</span><span id="__span-0-436"><a id="__codelineno-0-436" name="__codelineno-0-436"></a>        <span class="c1"># Filter the genotype matrix to keep certain SNPs</span>
</span><span id="__span-0-437"><a id="__codelineno-0-437" name="__codelineno-0-437"></a>        <span class="k">if</span> <span class="n">extract_snps</span> <span class="ow">or</span> <span class="n">extract_file</span><span class="p">:</span>
</span><span id="__span-0-438"><a id="__codelineno-0-438" name="__codelineno-0-438"></a>            <span class="n">g</span><span class="o">.</span><span class="n">filter_snps</span><span class="p">(</span><span class="n">extract_snps</span><span class="o">=</span><span class="n">extract_snps</span><span class="p">,</span> <span class="n">extract_file</span><span class="o">=</span><span class="n">extract_file</span><span class="p">)</span>
</span><span id="__span-0-439"><a id="__codelineno-0-439" name="__codelineno-0-439"></a>
</span><span id="__span-0-440"><a id="__codelineno-0-440" name="__codelineno-0-440"></a>        <span class="c1"># Drop duplicated SNP IDs</span>
</span><span id="__span-0-441"><a id="__codelineno-0-441" name="__codelineno-0-441"></a>        <span class="k">if</span> <span class="n">drop_duplicated</span><span class="p">:</span>
</span><span id="__span-0-442"><a id="__codelineno-0-442" name="__codelineno-0-442"></a>            <span class="n">g</span><span class="o">.</span><span class="n">drop_duplicated_snps</span><span class="p">()</span>
</span><span id="__span-0-443"><a id="__codelineno-0-443" name="__codelineno-0-443"></a>
</span><span id="__span-0-444"><a id="__codelineno-0-444" name="__codelineno-0-444"></a>        <span class="c1"># Filter SNPs by minor allele frequency and/or count:</span>
</span><span id="__span-0-445"><a id="__codelineno-0-445" name="__codelineno-0-445"></a>        <span class="n">g</span><span class="o">.</span><span class="n">filter_by_allele_frequency</span><span class="p">(</span><span class="n">min_maf</span><span class="o">=</span><span class="n">min_maf</span><span class="p">,</span> <span class="n">min_mac</span><span class="o">=</span><span class="n">min_mac</span><span class="p">)</span>
</span><span id="__span-0-446"><a id="__codelineno-0-446" name="__codelineno-0-446"></a>
</span><span id="__span-0-447"><a id="__codelineno-0-447" name="__codelineno-0-447"></a>        <span class="k">if</span> <span class="n">i</span> <span class="o">==</span> <span class="mi">0</span><span class="p">:</span>
</span><span id="__span-0-448"><a id="__codelineno-0-448" name="__codelineno-0-448"></a>            <span class="bp">self</span><span class="o">.</span><span class="n">sample_table</span> <span class="o">=</span> <span class="n">g</span><span class="o">.</span><span class="n">sample_table</span>
</span></code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h2 id="magenpy.GWADataLoader.GWADataLoader.read_ld" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">read_ld</span><span class="p">(</span><span class="n">ld_store_paths</span><span class="p">)</span></code>

<a href="#magenpy.GWADataLoader.GWADataLoader.read_ld" class="headerlink" title="Permanent link">&para;</a></h2>


    <div class="doc doc-contents ">

        <p>Read the LD matrix files stored on-disk in Zarr array format.</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                <code>ld_store_paths</code>
            </td>
            <td>
            </td>
            <td>
              <div class="doc-md-description">
                <p>The path to the LD matrices. This may be a wildcard to accommodate reading data for multiple chromosomes.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
      </tbody>
    </table>


            <details class="quote">
              <summary>Source code in <code>magenpy/GWADataLoader.py</code></summary>
              <div class="language-python highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal"><a href="#__codelineno-0-578">578</a></span>
<span class="normal"><a href="#__codelineno-0-579">579</a></span>
<span class="normal"><a href="#__codelineno-0-580">580</a></span>
<span class="normal"><a href="#__codelineno-0-581">581</a></span>
<span class="normal"><a href="#__codelineno-0-582">582</a></span>
<span class="normal"><a href="#__codelineno-0-583">583</a></span>
<span class="normal"><a href="#__codelineno-0-584">584</a></span>
<span class="normal"><a href="#__codelineno-0-585">585</a></span>
<span class="normal"><a href="#__codelineno-0-586">586</a></span>
<span class="normal"><a href="#__codelineno-0-587">587</a></span>
<span class="normal"><a href="#__codelineno-0-588">588</a></span>
<span class="normal"><a href="#__codelineno-0-589">589</a></span>
<span class="normal"><a href="#__codelineno-0-590">590</a></span>
<span class="normal"><a href="#__codelineno-0-591">591</a></span>
<span class="normal"><a href="#__codelineno-0-592">592</a></span>
<span class="normal"><a href="#__codelineno-0-593">593</a></span>
<span class="normal"><a href="#__codelineno-0-594">594</a></span>
<span class="normal"><a href="#__codelineno-0-595">595</a></span>
<span class="normal"><a href="#__codelineno-0-596">596</a></span>
<span class="normal"><a href="#__codelineno-0-597">597</a></span>
<span class="normal"><a href="#__codelineno-0-598">598</a></span>
<span class="normal"><a href="#__codelineno-0-599">599</a></span>
<span class="normal"><a href="#__codelineno-0-600">600</a></span>
<span class="normal"><a href="#__codelineno-0-601">601</a></span>
<span class="normal"><a href="#__codelineno-0-602">602</a></span>
<span class="normal"><a href="#__codelineno-0-603">603</a></span>
<span class="normal"><a href="#__codelineno-0-604">604</a></span>
<span class="normal"><a href="#__codelineno-0-605">605</a></span>
<span class="normal"><a href="#__codelineno-0-606">606</a></span>
<span class="normal"><a href="#__codelineno-0-607">607</a></span>
<span class="normal"><a href="#__codelineno-0-608">608</a></span>
<span class="normal"><a href="#__codelineno-0-609">609</a></span></pre></div></td><td class="code"><div><pre><span></span><code><span id="__span-0-578"><a id="__codelineno-0-578" name="__codelineno-0-578"></a><span class="k">def</span><span class="w"> </span><span class="nf">read_ld</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">ld_store_paths</span><span class="p">):</span>
</span><span id="__span-0-579"><a id="__codelineno-0-579" name="__codelineno-0-579"></a><span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
</span><span id="__span-0-580"><a id="__codelineno-0-580" name="__codelineno-0-580"></a><span class="sd">    Read the LD matrix files stored on-disk in Zarr array format.</span>
</span><span id="__span-0-581"><a id="__codelineno-0-581" name="__codelineno-0-581"></a><span class="sd">    :param ld_store_paths: The path to the LD matrices. This may be a wildcard to accommodate reading data</span>
</span><span id="__span-0-582"><a id="__codelineno-0-582" name="__codelineno-0-582"></a><span class="sd">    for multiple chromosomes.</span>
</span><span id="__span-0-583"><a id="__codelineno-0-583" name="__codelineno-0-583"></a><span class="sd">    &quot;&quot;&quot;</span>
</span><span id="__span-0-584"><a id="__codelineno-0-584" name="__codelineno-0-584"></a>
</span><span id="__span-0-585"><a id="__codelineno-0-585" name="__codelineno-0-585"></a>    <span class="k">if</span> <span class="n">ld_store_paths</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
</span><span id="__span-0-586"><a id="__codelineno-0-586" name="__codelineno-0-586"></a>        <span class="k">return</span>
</span><span id="__span-0-587"><a id="__codelineno-0-587" name="__codelineno-0-587"></a>
</span><span id="__span-0-588"><a id="__codelineno-0-588" name="__codelineno-0-588"></a>    <span class="k">if</span> <span class="ow">not</span> <span class="n">iterable</span><span class="p">(</span><span class="n">ld_store_paths</span><span class="p">):</span>
</span><span id="__span-0-589"><a id="__codelineno-0-589" name="__codelineno-0-589"></a>        <span class="k">if</span> <span class="s1">&#39;s3://&#39;</span> <span class="ow">in</span> <span class="n">ld_store_paths</span><span class="p">:</span>
</span><span id="__span-0-590"><a id="__codelineno-0-590" name="__codelineno-0-590"></a>            <span class="kn">from</span><span class="w"> </span><span class="nn">.utils.system_utils</span><span class="w"> </span><span class="kn">import</span> <span class="n">glob_s3_path</span>
</span><span id="__span-0-591"><a id="__codelineno-0-591" name="__codelineno-0-591"></a>            <span class="n">ld_store_files</span> <span class="o">=</span> <span class="n">glob_s3_path</span><span class="p">(</span><span class="n">ld_store_paths</span><span class="p">)</span>
</span><span id="__span-0-592"><a id="__codelineno-0-592" name="__codelineno-0-592"></a>        <span class="k">else</span><span class="p">:</span>
</span><span id="__span-0-593"><a id="__codelineno-0-593" name="__codelineno-0-593"></a>            <span class="n">ld_store_files</span> <span class="o">=</span> <span class="n">get_filenames</span><span class="p">(</span><span class="n">ld_store_paths</span><span class="p">,</span> <span class="n">extension</span><span class="o">=</span><span class="s1">&#39;.zgroup&#39;</span><span class="p">)</span>
</span><span id="__span-0-594"><a id="__codelineno-0-594" name="__codelineno-0-594"></a>    <span class="k">else</span><span class="p">:</span>
</span><span id="__span-0-595"><a id="__codelineno-0-595" name="__codelineno-0-595"></a>        <span class="n">ld_store_files</span> <span class="o">=</span> <span class="n">ld_store_paths</span>
</span><span id="__span-0-596"><a id="__codelineno-0-596" name="__codelineno-0-596"></a>
</span><span id="__span-0-597"><a id="__codelineno-0-597" name="__codelineno-0-597"></a>    <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">ld_store_files</span><span class="p">)</span> <span class="o">&lt;</span> <span class="mi">1</span><span class="p">:</span>
</span><span id="__span-0-598"><a id="__codelineno-0-598" name="__codelineno-0-598"></a>        <span class="n">logger</span><span class="o">.</span><span class="n">warning</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;No LD matrix files were found at: </span><span class="si">{</span><span class="n">ld_store_paths</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
</span><span id="__span-0-599"><a id="__codelineno-0-599" name="__codelineno-0-599"></a>        <span class="k">return</span>
</span><span id="__span-0-600"><a id="__codelineno-0-600" name="__codelineno-0-600"></a>
</span><span id="__span-0-601"><a id="__codelineno-0-601" name="__codelineno-0-601"></a>    <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;&gt; Reading LD metadata...&quot;</span><span class="p">)</span>
</span><span id="__span-0-602"><a id="__codelineno-0-602" name="__codelineno-0-602"></a>
</span><span id="__span-0-603"><a id="__codelineno-0-603" name="__codelineno-0-603"></a>    <span class="bp">self</span><span class="o">.</span><span class="n">ld</span> <span class="o">=</span> <span class="p">{}</span>
</span><span id="__span-0-604"><a id="__codelineno-0-604" name="__codelineno-0-604"></a>
</span><span id="__span-0-605"><a id="__codelineno-0-605" name="__codelineno-0-605"></a>    <span class="k">for</span> <span class="n">f</span> <span class="ow">in</span> <span class="n">tqdm</span><span class="p">(</span><span class="n">ld_store_files</span><span class="p">,</span>
</span><span id="__span-0-606"><a id="__codelineno-0-606" name="__codelineno-0-606"></a>                  <span class="n">total</span><span class="o">=</span><span class="nb">len</span><span class="p">(</span><span class="n">ld_store_files</span><span class="p">),</span>
</span><span id="__span-0-607"><a id="__codelineno-0-607" name="__codelineno-0-607"></a>                  <span class="n">desc</span><span class="o">=</span><span class="s2">&quot;Reading LD metadata&quot;</span><span class="p">):</span>
</span><span id="__span-0-608"><a id="__codelineno-0-608" name="__codelineno-0-608"></a>        <span class="n">z</span> <span class="o">=</span> <span class="n">LDMatrix</span><span class="o">.</span><span class="n">from_path</span><span class="p">(</span><span class="n">f</span><span class="p">)</span>
</span><span id="__span-0-609"><a id="__codelineno-0-609" name="__codelineno-0-609"></a>        <span class="bp">self</span><span class="o">.</span><span class="n">ld</span><span class="p">[</span><span class="n">z</span><span class="o">.</span><span class="n">chromosome</span><span class="p">]</span> <span class="o">=</span> <span class="n">z</span>
</span></code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h2 id="magenpy.GWADataLoader.GWADataLoader.read_phenotype" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">read_phenotype</span><span class="p">(</span><span class="n">phenotype_file</span><span class="p">,</span> <span class="n">drop_na</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="o">**</span><span class="n">read_csv_kwargs</span><span class="p">)</span></code>

<a href="#magenpy.GWADataLoader.GWADataLoader.read_phenotype" class="headerlink" title="Permanent link">&para;</a></h2>


    <div class="doc doc-contents ">

        <p>Read the phenotype file and integrate it with the sample tables and genotype matrices.</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                <code>phenotype_file</code>
            </td>
            <td>
            </td>
            <td>
              <div class="doc-md-description">
                <p>The path to the phenotype file (Default: tab-separated file with <code>FID IID phenotype</code> columns). If different, supply details as additional arguments to this function.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>drop_na</code>
            </td>
            <td>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Drop samples with missing phenotype information.</p>
              </div>
            </td>
            <td>
                  <code>True</code>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>read_csv_kwargs</code>
            </td>
            <td>
            </td>
            <td>
              <div class="doc-md-description">
                <p>keyword arguments for the <code>read_csv</code> function of <code>pandas</code>.</p>
              </div>
            </td>
            <td>
                  <code>{}</code>
            </td>
          </tr>
      </tbody>
    </table>


            <details class="quote">
              <summary>Source code in <code>magenpy/GWADataLoader.py</code></summary>
              <div class="language-python highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal"><a href="#__codelineno-0-450">450</a></span>
<span class="normal"><a href="#__codelineno-0-451">451</a></span>
<span class="normal"><a href="#__codelineno-0-452">452</a></span>
<span class="normal"><a href="#__codelineno-0-453">453</a></span>
<span class="normal"><a href="#__codelineno-0-454">454</a></span>
<span class="normal"><a href="#__codelineno-0-455">455</a></span>
<span class="normal"><a href="#__codelineno-0-456">456</a></span>
<span class="normal"><a href="#__codelineno-0-457">457</a></span>
<span class="normal"><a href="#__codelineno-0-458">458</a></span>
<span class="normal"><a href="#__codelineno-0-459">459</a></span>
<span class="normal"><a href="#__codelineno-0-460">460</a></span>
<span class="normal"><a href="#__codelineno-0-461">461</a></span>
<span class="normal"><a href="#__codelineno-0-462">462</a></span>
<span class="normal"><a href="#__codelineno-0-463">463</a></span>
<span class="normal"><a href="#__codelineno-0-464">464</a></span>
<span class="normal"><a href="#__codelineno-0-465">465</a></span>
<span class="normal"><a href="#__codelineno-0-466">466</a></span>
<span class="normal"><a href="#__codelineno-0-467">467</a></span>
<span class="normal"><a href="#__codelineno-0-468">468</a></span>
<span class="normal"><a href="#__codelineno-0-469">469</a></span></pre></div></td><td class="code"><div><pre><span></span><code><span id="__span-0-450"><a id="__codelineno-0-450" name="__codelineno-0-450"></a><span class="k">def</span><span class="w"> </span><span class="nf">read_phenotype</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">phenotype_file</span><span class="p">,</span> <span class="n">drop_na</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="o">**</span><span class="n">read_csv_kwargs</span><span class="p">):</span>
</span><span id="__span-0-451"><a id="__codelineno-0-451" name="__codelineno-0-451"></a><span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
</span><span id="__span-0-452"><a id="__codelineno-0-452" name="__codelineno-0-452"></a><span class="sd">    Read the phenotype file and integrate it with the sample tables and genotype matrices.</span>
</span><span id="__span-0-453"><a id="__codelineno-0-453" name="__codelineno-0-453"></a>
</span><span id="__span-0-454"><a id="__codelineno-0-454" name="__codelineno-0-454"></a><span class="sd">    :param phenotype_file: The path to the phenotype file</span>
</span><span id="__span-0-455"><a id="__codelineno-0-455" name="__codelineno-0-455"></a><span class="sd">    (Default: tab-separated file with `FID IID phenotype` columns). If different, supply</span>
</span><span id="__span-0-456"><a id="__codelineno-0-456" name="__codelineno-0-456"></a><span class="sd">    details as additional arguments to this function.</span>
</span><span id="__span-0-457"><a id="__codelineno-0-457" name="__codelineno-0-457"></a><span class="sd">    :param drop_na: Drop samples with missing phenotype information.</span>
</span><span id="__span-0-458"><a id="__codelineno-0-458" name="__codelineno-0-458"></a><span class="sd">    :param read_csv_kwargs: keyword arguments for the `read_csv` function of `pandas`.</span>
</span><span id="__span-0-459"><a id="__codelineno-0-459" name="__codelineno-0-459"></a><span class="sd">    &quot;&quot;&quot;</span>
</span><span id="__span-0-460"><a id="__codelineno-0-460" name="__codelineno-0-460"></a>
</span><span id="__span-0-461"><a id="__codelineno-0-461" name="__codelineno-0-461"></a>    <span class="k">if</span> <span class="n">phenotype_file</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
</span><span id="__span-0-462"><a id="__codelineno-0-462" name="__codelineno-0-462"></a>        <span class="k">return</span>
</span><span id="__span-0-463"><a id="__codelineno-0-463" name="__codelineno-0-463"></a>
</span><span id="__span-0-464"><a id="__codelineno-0-464" name="__codelineno-0-464"></a>    <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;&gt; Reading phenotype file...&quot;</span><span class="p">)</span>
</span><span id="__span-0-465"><a id="__codelineno-0-465" name="__codelineno-0-465"></a>
</span><span id="__span-0-466"><a id="__codelineno-0-466" name="__codelineno-0-466"></a>    <span class="k">assert</span> <span class="bp">self</span><span class="o">.</span><span class="n">sample_table</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span>
</span><span id="__span-0-467"><a id="__codelineno-0-467" name="__codelineno-0-467"></a>
</span><span id="__span-0-468"><a id="__codelineno-0-468" name="__codelineno-0-468"></a>    <span class="bp">self</span><span class="o">.</span><span class="n">sample_table</span><span class="o">.</span><span class="n">read_phenotype_file</span><span class="p">(</span><span class="n">phenotype_file</span><span class="p">,</span> <span class="n">drop_na</span><span class="o">=</span><span class="n">drop_na</span><span class="p">,</span> <span class="o">**</span><span class="n">read_csv_kwargs</span><span class="p">)</span>
</span><span id="__span-0-469"><a id="__codelineno-0-469" name="__codelineno-0-469"></a>    <span class="bp">self</span><span class="o">.</span><span class="n">sync_sample_tables</span><span class="p">()</span>
</span></code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h2 id="magenpy.GWADataLoader.GWADataLoader.read_summary_statistics" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">read_summary_statistics</span><span class="p">(</span><span class="n">sumstats_path</span><span class="p">,</span> <span class="n">sumstats_format</span><span class="o">=</span><span class="s1">&#39;magenpy&#39;</span><span class="p">,</span> <span class="n">parser</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">drop_duplicated</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="o">**</span><span class="n">parse_kwargs</span><span class="p">)</span></code>

<a href="#magenpy.GWADataLoader.GWADataLoader.read_summary_statistics" class="headerlink" title="Permanent link">&para;</a></h2>


    <div class="doc doc-contents ">

        <p>Read GWAS summary statistics file(s) and parse them to <code>SumstatsTable</code> objects.</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                <code>sumstats_path</code>
            </td>
            <td>
            </td>
            <td>
              <div class="doc-md-description">
                <p>The path to the summary statistics file(s). The path may be a wildcard.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>sumstats_format</code>
            </td>
            <td>
            </td>
            <td>
              <div class="doc-md-description">
                <p>The format for the summary statistics. Currently supports the following formats: <code>plink1.9</code>, <code>plink2</code>, <code>magenpy</code>, <code>fastGWA</code>, <code>COJO</code>, <code>SAIGE</code>, or <code>GWASCatalog</code> for the standard summary statistics format (also known as <code>ssf</code> or <code>gwas-ssf</code>).</p>
              </div>
            </td>
            <td>
                  <code>&#39;magenpy&#39;</code>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>parser</code>
            </td>
            <td>
            </td>
            <td>
              <div class="doc-md-description">
                <p>If the summary statistics file does not follow any of the formats above, you can create your own parser by inheriting from the base <code>SumstatsParser</code> class and passing it here as an argument.</p>
              </div>
            </td>
            <td>
                  <code>None</code>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>drop_duplicated</code>
            </td>
            <td>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Drop SNPs with duplicated rsIDs.</p>
              </div>
            </td>
            <td>
                  <code>True</code>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>parse_kwargs</code>
            </td>
            <td>
            </td>
            <td>
              <div class="doc-md-description">
                <p>keyword arguments for the parser. These are mainly parameters that will be passed to <code>pandas.read_csv</code> function, such as the delimiter, header information, etc.</p>
              </div>
            </td>
            <td>
                  <code>{}</code>
            </td>
          </tr>
      </tbody>
    </table>


            <details class="quote">
              <summary>Source code in <code>magenpy/GWADataLoader.py</code></summary>
              <div class="language-python highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal"><a href="#__codelineno-0-501">501</a></span>
<span class="normal"><a href="#__codelineno-0-502">502</a></span>
<span class="normal"><a href="#__codelineno-0-503">503</a></span>
<span class="normal"><a href="#__codelineno-0-504">504</a></span>
<span class="normal"><a href="#__codelineno-0-505">505</a></span>
<span class="normal"><a href="#__codelineno-0-506">506</a></span>
<span class="normal"><a href="#__codelineno-0-507">507</a></span>
<span class="normal"><a href="#__codelineno-0-508">508</a></span>
<span class="normal"><a href="#__codelineno-0-509">509</a></span>
<span class="normal"><a href="#__codelineno-0-510">510</a></span>
<span class="normal"><a href="#__codelineno-0-511">511</a></span>
<span class="normal"><a href="#__codelineno-0-512">512</a></span>
<span class="normal"><a href="#__codelineno-0-513">513</a></span>
<span class="normal"><a href="#__codelineno-0-514">514</a></span>
<span class="normal"><a href="#__codelineno-0-515">515</a></span>
<span class="normal"><a href="#__codelineno-0-516">516</a></span>
<span class="normal"><a href="#__codelineno-0-517">517</a></span>
<span class="normal"><a href="#__codelineno-0-518">518</a></span>
<span class="normal"><a href="#__codelineno-0-519">519</a></span>
<span class="normal"><a href="#__codelineno-0-520">520</a></span>
<span class="normal"><a href="#__codelineno-0-521">521</a></span>
<span class="normal"><a href="#__codelineno-0-522">522</a></span>
<span class="normal"><a href="#__codelineno-0-523">523</a></span>
<span class="normal"><a href="#__codelineno-0-524">524</a></span>
<span class="normal"><a href="#__codelineno-0-525">525</a></span>
<span class="normal"><a href="#__codelineno-0-526">526</a></span>
<span class="normal"><a href="#__codelineno-0-527">527</a></span>
<span class="normal"><a href="#__codelineno-0-528">528</a></span>
<span class="normal"><a href="#__codelineno-0-529">529</a></span>
<span class="normal"><a href="#__codelineno-0-530">530</a></span>
<span class="normal"><a href="#__codelineno-0-531">531</a></span>
<span class="normal"><a href="#__codelineno-0-532">532</a></span>
<span class="normal"><a href="#__codelineno-0-533">533</a></span>
<span class="normal"><a href="#__codelineno-0-534">534</a></span>
<span class="normal"><a href="#__codelineno-0-535">535</a></span>
<span class="normal"><a href="#__codelineno-0-536">536</a></span>
<span class="normal"><a href="#__codelineno-0-537">537</a></span>
<span class="normal"><a href="#__codelineno-0-538">538</a></span>
<span class="normal"><a href="#__codelineno-0-539">539</a></span>
<span class="normal"><a href="#__codelineno-0-540">540</a></span>
<span class="normal"><a href="#__codelineno-0-541">541</a></span>
<span class="normal"><a href="#__codelineno-0-542">542</a></span>
<span class="normal"><a href="#__codelineno-0-543">543</a></span>
<span class="normal"><a href="#__codelineno-0-544">544</a></span>
<span class="normal"><a href="#__codelineno-0-545">545</a></span>
<span class="normal"><a href="#__codelineno-0-546">546</a></span>
<span class="normal"><a href="#__codelineno-0-547">547</a></span>
<span class="normal"><a href="#__codelineno-0-548">548</a></span>
<span class="normal"><a href="#__codelineno-0-549">549</a></span>
<span class="normal"><a href="#__codelineno-0-550">550</a></span>
<span class="normal"><a href="#__codelineno-0-551">551</a></span>
<span class="normal"><a href="#__codelineno-0-552">552</a></span>
<span class="normal"><a href="#__codelineno-0-553">553</a></span>
<span class="normal"><a href="#__codelineno-0-554">554</a></span>
<span class="normal"><a href="#__codelineno-0-555">555</a></span>
<span class="normal"><a href="#__codelineno-0-556">556</a></span>
<span class="normal"><a href="#__codelineno-0-557">557</a></span>
<span class="normal"><a href="#__codelineno-0-558">558</a></span>
<span class="normal"><a href="#__codelineno-0-559">559</a></span>
<span class="normal"><a href="#__codelineno-0-560">560</a></span>
<span class="normal"><a href="#__codelineno-0-561">561</a></span>
<span class="normal"><a href="#__codelineno-0-562">562</a></span>
<span class="normal"><a href="#__codelineno-0-563">563</a></span>
<span class="normal"><a href="#__codelineno-0-564">564</a></span>
<span class="normal"><a href="#__codelineno-0-565">565</a></span>
<span class="normal"><a href="#__codelineno-0-566">566</a></span>
<span class="normal"><a href="#__codelineno-0-567">567</a></span>
<span class="normal"><a href="#__codelineno-0-568">568</a></span>
<span class="normal"><a href="#__codelineno-0-569">569</a></span>
<span class="normal"><a href="#__codelineno-0-570">570</a></span>
<span class="normal"><a href="#__codelineno-0-571">571</a></span>
<span class="normal"><a href="#__codelineno-0-572">572</a></span>
<span class="normal"><a href="#__codelineno-0-573">573</a></span>
<span class="normal"><a href="#__codelineno-0-574">574</a></span>
<span class="normal"><a href="#__codelineno-0-575">575</a></span>
<span class="normal"><a href="#__codelineno-0-576">576</a></span></pre></div></td><td class="code"><div><pre><span></span><code><span id="__span-0-501"><a id="__codelineno-0-501" name="__codelineno-0-501"></a><span class="k">def</span><span class="w"> </span><span class="nf">read_summary_statistics</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span>
</span><span id="__span-0-502"><a id="__codelineno-0-502" name="__codelineno-0-502"></a>                            <span class="n">sumstats_path</span><span class="p">,</span>
</span><span id="__span-0-503"><a id="__codelineno-0-503" name="__codelineno-0-503"></a>                            <span class="n">sumstats_format</span><span class="o">=</span><span class="s1">&#39;magenpy&#39;</span><span class="p">,</span>
</span><span id="__span-0-504"><a id="__codelineno-0-504" name="__codelineno-0-504"></a>                            <span class="n">parser</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
</span><span id="__span-0-505"><a id="__codelineno-0-505" name="__codelineno-0-505"></a>                            <span class="n">drop_duplicated</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
</span><span id="__span-0-506"><a id="__codelineno-0-506" name="__codelineno-0-506"></a>                            <span class="o">**</span><span class="n">parse_kwargs</span><span class="p">):</span>
</span><span id="__span-0-507"><a id="__codelineno-0-507" name="__codelineno-0-507"></a><span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
</span><span id="__span-0-508"><a id="__codelineno-0-508" name="__codelineno-0-508"></a><span class="sd">    Read GWAS summary statistics file(s) and parse them to `SumstatsTable` objects.</span>
</span><span id="__span-0-509"><a id="__codelineno-0-509" name="__codelineno-0-509"></a>
</span><span id="__span-0-510"><a id="__codelineno-0-510" name="__codelineno-0-510"></a><span class="sd">    :param sumstats_path: The path to the summary statistics file(s). The path may be a wildcard.</span>
</span><span id="__span-0-511"><a id="__codelineno-0-511" name="__codelineno-0-511"></a><span class="sd">    :param sumstats_format: The format for the summary statistics. Currently supports the following</span>
</span><span id="__span-0-512"><a id="__codelineno-0-512" name="__codelineno-0-512"></a><span class="sd">     formats: `plink1.9`, `plink2`, `magenpy`, `fastGWA`, `COJO`, `SAIGE`, or `GWASCatalog` for the standard</span>
</span><span id="__span-0-513"><a id="__codelineno-0-513" name="__codelineno-0-513"></a><span class="sd">     summary statistics format (also known as `ssf` or `gwas-ssf`).</span>
</span><span id="__span-0-514"><a id="__codelineno-0-514" name="__codelineno-0-514"></a><span class="sd">    :param parser: If the summary statistics file does not follow any of the formats above, you can create</span>
</span><span id="__span-0-515"><a id="__codelineno-0-515" name="__codelineno-0-515"></a><span class="sd">    your own parser by inheriting from the base `SumstatsParser` class and passing it here as an argument.</span>
</span><span id="__span-0-516"><a id="__codelineno-0-516" name="__codelineno-0-516"></a><span class="sd">    :param drop_duplicated: Drop SNPs with duplicated rsIDs.</span>
</span><span id="__span-0-517"><a id="__codelineno-0-517" name="__codelineno-0-517"></a><span class="sd">    :param parse_kwargs: keyword arguments for the parser. These are mainly parameters that will be passed to</span>
</span><span id="__span-0-518"><a id="__codelineno-0-518" name="__codelineno-0-518"></a><span class="sd">    `pandas.read_csv` function, such as the delimiter, header information, etc.</span>
</span><span id="__span-0-519"><a id="__codelineno-0-519" name="__codelineno-0-519"></a><span class="sd">    &quot;&quot;&quot;</span>
</span><span id="__span-0-520"><a id="__codelineno-0-520" name="__codelineno-0-520"></a>
</span><span id="__span-0-521"><a id="__codelineno-0-521" name="__codelineno-0-521"></a>    <span class="k">if</span> <span class="n">sumstats_path</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
</span><span id="__span-0-522"><a id="__codelineno-0-522" name="__codelineno-0-522"></a>        <span class="k">return</span>
</span><span id="__span-0-523"><a id="__codelineno-0-523" name="__codelineno-0-523"></a>
</span><span id="__span-0-524"><a id="__codelineno-0-524" name="__codelineno-0-524"></a>    <span class="k">if</span> <span class="ow">not</span> <span class="n">iterable</span><span class="p">(</span><span class="n">sumstats_path</span><span class="p">):</span>
</span><span id="__span-0-525"><a id="__codelineno-0-525" name="__codelineno-0-525"></a>        <span class="n">sumstats_files</span> <span class="o">=</span> <span class="n">get_filenames</span><span class="p">(</span><span class="n">sumstats_path</span><span class="p">)</span>
</span><span id="__span-0-526"><a id="__codelineno-0-526" name="__codelineno-0-526"></a>
</span><span id="__span-0-527"><a id="__codelineno-0-527" name="__codelineno-0-527"></a>        <span class="kn">from</span><span class="w"> </span><span class="nn">.utils.system_utils</span><span class="w"> </span><span class="kn">import</span> <span class="n">valid_url</span>
</span><span id="__span-0-528"><a id="__codelineno-0-528" name="__codelineno-0-528"></a>        <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">sumstats_files</span><span class="p">)</span> <span class="o">&lt;</span> <span class="mi">1</span> <span class="ow">and</span> <span class="n">valid_url</span><span class="p">(</span><span class="n">sumstats_path</span><span class="p">):</span>
</span><span id="__span-0-529"><a id="__codelineno-0-529" name="__codelineno-0-529"></a>            <span class="n">sumstats_files</span> <span class="o">=</span> <span class="p">[</span><span class="n">sumstats_path</span><span class="p">]</span>
</span><span id="__span-0-530"><a id="__codelineno-0-530" name="__codelineno-0-530"></a>    <span class="k">else</span><span class="p">:</span>
</span><span id="__span-0-531"><a id="__codelineno-0-531" name="__codelineno-0-531"></a>        <span class="n">sumstats_files</span> <span class="o">=</span> <span class="n">sumstats_path</span>
</span><span id="__span-0-532"><a id="__codelineno-0-532" name="__codelineno-0-532"></a>
</span><span id="__span-0-533"><a id="__codelineno-0-533" name="__codelineno-0-533"></a>    <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">sumstats_files</span><span class="p">)</span> <span class="o">&lt;</span> <span class="mi">1</span><span class="p">:</span>
</span><span id="__span-0-534"><a id="__codelineno-0-534" name="__codelineno-0-534"></a>        <span class="n">logger</span><span class="o">.</span><span class="n">warning</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;No summary statistics files were found at: </span><span class="si">{</span><span class="n">sumstats_path</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
</span><span id="__span-0-535"><a id="__codelineno-0-535" name="__codelineno-0-535"></a>        <span class="k">return</span>
</span><span id="__span-0-536"><a id="__codelineno-0-536" name="__codelineno-0-536"></a>
</span><span id="__span-0-537"><a id="__codelineno-0-537" name="__codelineno-0-537"></a>    <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;&gt; Reading summary statistics...&quot;</span><span class="p">)</span>
</span><span id="__span-0-538"><a id="__codelineno-0-538" name="__codelineno-0-538"></a>
</span><span id="__span-0-539"><a id="__codelineno-0-539" name="__codelineno-0-539"></a>    <span class="bp">self</span><span class="o">.</span><span class="n">sumstats_table</span> <span class="o">=</span> <span class="p">{}</span>
</span><span id="__span-0-540"><a id="__codelineno-0-540" name="__codelineno-0-540"></a>
</span><span id="__span-0-541"><a id="__codelineno-0-541" name="__codelineno-0-541"></a>    <span class="k">for</span> <span class="n">f</span> <span class="ow">in</span> <span class="n">tqdm</span><span class="p">(</span><span class="n">sumstats_files</span><span class="p">,</span>
</span><span id="__span-0-542"><a id="__codelineno-0-542" name="__codelineno-0-542"></a>                  <span class="n">total</span><span class="o">=</span><span class="nb">len</span><span class="p">(</span><span class="n">sumstats_files</span><span class="p">),</span>
</span><span id="__span-0-543"><a id="__codelineno-0-543" name="__codelineno-0-543"></a>                  <span class="n">desc</span><span class="o">=</span><span class="s2">&quot;Reading summary statistics&quot;</span><span class="p">):</span>
</span><span id="__span-0-544"><a id="__codelineno-0-544" name="__codelineno-0-544"></a>
</span><span id="__span-0-545"><a id="__codelineno-0-545" name="__codelineno-0-545"></a>        <span class="n">ss_tab</span> <span class="o">=</span> <span class="n">SumstatsTable</span><span class="o">.</span><span class="n">from_file</span><span class="p">(</span><span class="n">f</span><span class="p">,</span>
</span><span id="__span-0-546"><a id="__codelineno-0-546" name="__codelineno-0-546"></a>                                         <span class="n">sumstats_format</span><span class="o">=</span><span class="n">sumstats_format</span><span class="p">,</span>
</span><span id="__span-0-547"><a id="__codelineno-0-547" name="__codelineno-0-547"></a>                                         <span class="n">parser</span><span class="o">=</span><span class="n">parser</span><span class="p">,</span>
</span><span id="__span-0-548"><a id="__codelineno-0-548" name="__codelineno-0-548"></a>                                         <span class="o">**</span><span class="n">parse_kwargs</span><span class="p">)</span>
</span><span id="__span-0-549"><a id="__codelineno-0-549" name="__codelineno-0-549"></a>
</span><span id="__span-0-550"><a id="__codelineno-0-550" name="__codelineno-0-550"></a>        <span class="k">if</span> <span class="n">drop_duplicated</span><span class="p">:</span>
</span><span id="__span-0-551"><a id="__codelineno-0-551" name="__codelineno-0-551"></a>            <span class="n">ss_tab</span><span class="o">.</span><span class="n">drop_duplicates</span><span class="p">()</span>
</span><span id="__span-0-552"><a id="__codelineno-0-552" name="__codelineno-0-552"></a>
</span><span id="__span-0-553"><a id="__codelineno-0-553" name="__codelineno-0-553"></a>        <span class="k">if</span> <span class="s1">&#39;CHR&#39;</span> <span class="ow">in</span> <span class="n">ss_tab</span><span class="o">.</span><span class="n">table</span><span class="o">.</span><span class="n">columns</span><span class="p">:</span>
</span><span id="__span-0-554"><a id="__codelineno-0-554" name="__codelineno-0-554"></a>            <span class="bp">self</span><span class="o">.</span><span class="n">sumstats_table</span><span class="o">.</span><span class="n">update</span><span class="p">(</span><span class="n">ss_tab</span><span class="o">.</span><span class="n">split_by_chromosome</span><span class="p">())</span>
</span><span id="__span-0-555"><a id="__codelineno-0-555" name="__codelineno-0-555"></a>        <span class="k">else</span><span class="p">:</span>
</span><span id="__span-0-556"><a id="__codelineno-0-556" name="__codelineno-0-556"></a>            <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">genotype</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
</span><span id="__span-0-557"><a id="__codelineno-0-557" name="__codelineno-0-557"></a>                <span class="n">ref_table</span> <span class="o">=</span> <span class="p">{</span><span class="n">c</span><span class="p">:</span> <span class="n">g</span><span class="o">.</span><span class="n">snps</span> <span class="k">for</span> <span class="n">c</span><span class="p">,</span> <span class="n">g</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">genotype</span><span class="o">.</span><span class="n">items</span><span class="p">()}</span>
</span><span id="__span-0-558"><a id="__codelineno-0-558" name="__codelineno-0-558"></a>            <span class="k">elif</span> <span class="bp">self</span><span class="o">.</span><span class="n">ld</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
</span><span id="__span-0-559"><a id="__codelineno-0-559" name="__codelineno-0-559"></a>                <span class="n">ref_table</span> <span class="o">=</span> <span class="p">{</span><span class="n">c</span><span class="p">:</span> <span class="n">ld</span><span class="o">.</span><span class="n">snps</span> <span class="k">for</span> <span class="n">c</span><span class="p">,</span> <span class="n">ld</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">ld</span><span class="o">.</span><span class="n">items</span><span class="p">()}</span>
</span><span id="__span-0-560"><a id="__codelineno-0-560" name="__codelineno-0-560"></a>            <span class="k">else</span><span class="p">:</span>
</span><span id="__span-0-561"><a id="__codelineno-0-561" name="__codelineno-0-561"></a>                <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;Cannot index summary statistics tables without chromosome information!&quot;</span><span class="p">)</span>
</span><span id="__span-0-562"><a id="__codelineno-0-562" name="__codelineno-0-562"></a>
</span><span id="__span-0-563"><a id="__codelineno-0-563" name="__codelineno-0-563"></a>            <span class="bp">self</span><span class="o">.</span><span class="n">sumstats_table</span><span class="o">.</span><span class="n">update</span><span class="p">(</span><span class="n">ss_tab</span><span class="o">.</span><span class="n">split_by_chromosome</span><span class="p">(</span><span class="n">snps_per_chrom</span><span class="o">=</span><span class="n">ref_table</span><span class="p">))</span>
</span><span id="__span-0-564"><a id="__codelineno-0-564" name="__codelineno-0-564"></a>
</span><span id="__span-0-565"><a id="__codelineno-0-565" name="__codelineno-0-565"></a>    <span class="c1"># If SNP information is not present in the sumstats tables, try to impute it</span>
</span><span id="__span-0-566"><a id="__codelineno-0-566" name="__codelineno-0-566"></a>    <span class="c1"># using other reference tables:</span>
</span><span id="__span-0-567"><a id="__codelineno-0-567" name="__codelineno-0-567"></a>
</span><span id="__span-0-568"><a id="__codelineno-0-568" name="__codelineno-0-568"></a>    <span class="n">missing_snp</span> <span class="o">=</span> <span class="nb">any</span><span class="p">(</span><span class="s1">&#39;SNP&#39;</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">ss</span><span class="o">.</span><span class="n">table</span><span class="o">.</span><span class="n">columns</span> <span class="k">for</span> <span class="n">ss</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">sumstats_table</span><span class="o">.</span><span class="n">values</span><span class="p">())</span>
</span><span id="__span-0-569"><a id="__codelineno-0-569" name="__codelineno-0-569"></a>
</span><span id="__span-0-570"><a id="__codelineno-0-570" name="__codelineno-0-570"></a>    <span class="k">if</span> <span class="n">missing_snp</span> <span class="ow">and</span> <span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">genotype</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span> <span class="ow">or</span> <span class="bp">self</span><span class="o">.</span><span class="n">ld</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">):</span>
</span><span id="__span-0-571"><a id="__codelineno-0-571" name="__codelineno-0-571"></a>
</span><span id="__span-0-572"><a id="__codelineno-0-572" name="__codelineno-0-572"></a>        <span class="n">ref_table</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">to_snp_table</span><span class="p">(</span><span class="n">col_subset</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;CHR&#39;</span><span class="p">,</span> <span class="s1">&#39;POS&#39;</span><span class="p">,</span> <span class="s1">&#39;SNP&#39;</span><span class="p">],</span> <span class="n">per_chromosome</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
</span><span id="__span-0-573"><a id="__codelineno-0-573" name="__codelineno-0-573"></a>
</span><span id="__span-0-574"><a id="__codelineno-0-574" name="__codelineno-0-574"></a>        <span class="k">for</span> <span class="n">c</span><span class="p">,</span> <span class="n">ss</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">sumstats_table</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
</span><span id="__span-0-575"><a id="__codelineno-0-575" name="__codelineno-0-575"></a>            <span class="k">if</span> <span class="s1">&#39;SNP&#39;</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">ss</span><span class="o">.</span><span class="n">table</span><span class="o">.</span><span class="n">columns</span> <span class="ow">and</span> <span class="n">c</span> <span class="ow">in</span> <span class="n">ref_table</span><span class="p">:</span>
</span><span id="__span-0-576"><a id="__codelineno-0-576" name="__codelineno-0-576"></a>                <span class="n">ss</span><span class="o">.</span><span class="n">infer_snp_id</span><span class="p">(</span><span class="n">ref_table</span><span class="p">[</span><span class="n">c</span><span class="p">],</span> <span class="n">allow_na</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
</span></code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h2 id="magenpy.GWADataLoader.GWADataLoader.release_ld" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">release_ld</span><span class="p">()</span></code>

<a href="#magenpy.GWADataLoader.GWADataLoader.release_ld" class="headerlink" title="Permanent link">&para;</a></h2>


    <div class="doc doc-contents ">

        <p>A utility function to release the LD matrices from memory.</p>


            <details class="quote">
              <summary>Source code in <code>magenpy/GWADataLoader.py</code></summary>
              <div class="language-python highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal"><a href="#__codelineno-0-619">619</a></span>
<span class="normal"><a href="#__codelineno-0-620">620</a></span>
<span class="normal"><a href="#__codelineno-0-621">621</a></span>
<span class="normal"><a href="#__codelineno-0-622">622</a></span>
<span class="normal"><a href="#__codelineno-0-623">623</a></span>
<span class="normal"><a href="#__codelineno-0-624">624</a></span>
<span class="normal"><a href="#__codelineno-0-625">625</a></span></pre></div></td><td class="code"><div><pre><span></span><code><span id="__span-0-619"><a id="__codelineno-0-619" name="__codelineno-0-619"></a><span class="k">def</span><span class="w"> </span><span class="nf">release_ld</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
</span><span id="__span-0-620"><a id="__codelineno-0-620" name="__codelineno-0-620"></a><span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
</span><span id="__span-0-621"><a id="__codelineno-0-621" name="__codelineno-0-621"></a><span class="sd">    A utility function to release the LD matrices from memory.</span>
</span><span id="__span-0-622"><a id="__codelineno-0-622" name="__codelineno-0-622"></a><span class="sd">    &quot;&quot;&quot;</span>
</span><span id="__span-0-623"><a id="__codelineno-0-623" name="__codelineno-0-623"></a>    <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">ld</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
</span><span id="__span-0-624"><a id="__codelineno-0-624" name="__codelineno-0-624"></a>        <span class="k">for</span> <span class="n">ld</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">ld</span><span class="o">.</span><span class="n">values</span><span class="p">():</span>
</span><span id="__span-0-625"><a id="__codelineno-0-625" name="__codelineno-0-625"></a>            <span class="n">ld</span><span class="o">.</span><span class="n">release</span><span class="p">()</span>
</span></code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h2 id="magenpy.GWADataLoader.GWADataLoader.score" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">score</span><span class="p">(</span><span class="n">beta</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">standardize_genotype</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span></code>

<a href="#magenpy.GWADataLoader.GWADataLoader.score" class="headerlink" title="Permanent link">&para;</a></h2>


    <div class="doc doc-contents ">

        <p>Perform linear scoring, i.e. multiply the genotype matrix by the vector of effect sizes, <code>beta</code>.</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                <code>beta</code>
            </td>
            <td>
            </td>
            <td>
              <div class="doc-md-description">
                <p>A dictionary where the keys are the chromosome numbers and the values are a vector of effect sizes for each variant on that chromosome. If the betas are not provided, we use the marginal betas by default (if those are available).</p>
              </div>
            </td>
            <td>
                  <code>None</code>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>standardize_genotype</code>
            </td>
            <td>
            </td>
            <td>
              <div class="doc-md-description">
                <p>If True, standardize the genotype matrix before scoring.</p>
              </div>
            </td>
            <td>
                  <code>False</code>
            </td>
          </tr>
      </tbody>
    </table>


            <details class="quote">
              <summary>Source code in <code>magenpy/GWADataLoader.py</code></summary>
              <div class="language-python highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal"><a href="#__codelineno-0-777">777</a></span>
<span class="normal"><a href="#__codelineno-0-778">778</a></span>
<span class="normal"><a href="#__codelineno-0-779">779</a></span>
<span class="normal"><a href="#__codelineno-0-780">780</a></span>
<span class="normal"><a href="#__codelineno-0-781">781</a></span>
<span class="normal"><a href="#__codelineno-0-782">782</a></span>
<span class="normal"><a href="#__codelineno-0-783">783</a></span>
<span class="normal"><a href="#__codelineno-0-784">784</a></span>
<span class="normal"><a href="#__codelineno-0-785">785</a></span>
<span class="normal"><a href="#__codelineno-0-786">786</a></span>
<span class="normal"><a href="#__codelineno-0-787">787</a></span>
<span class="normal"><a href="#__codelineno-0-788">788</a></span>
<span class="normal"><a href="#__codelineno-0-789">789</a></span>
<span class="normal"><a href="#__codelineno-0-790">790</a></span>
<span class="normal"><a href="#__codelineno-0-791">791</a></span>
<span class="normal"><a href="#__codelineno-0-792">792</a></span>
<span class="normal"><a href="#__codelineno-0-793">793</a></span>
<span class="normal"><a href="#__codelineno-0-794">794</a></span>
<span class="normal"><a href="#__codelineno-0-795">795</a></span>
<span class="normal"><a href="#__codelineno-0-796">796</a></span>
<span class="normal"><a href="#__codelineno-0-797">797</a></span>
<span class="normal"><a href="#__codelineno-0-798">798</a></span>
<span class="normal"><a href="#__codelineno-0-799">799</a></span>
<span class="normal"><a href="#__codelineno-0-800">800</a></span>
<span class="normal"><a href="#__codelineno-0-801">801</a></span>
<span class="normal"><a href="#__codelineno-0-802">802</a></span>
<span class="normal"><a href="#__codelineno-0-803">803</a></span>
<span class="normal"><a href="#__codelineno-0-804">804</a></span>
<span class="normal"><a href="#__codelineno-0-805">805</a></span>
<span class="normal"><a href="#__codelineno-0-806">806</a></span>
<span class="normal"><a href="#__codelineno-0-807">807</a></span>
<span class="normal"><a href="#__codelineno-0-808">808</a></span>
<span class="normal"><a href="#__codelineno-0-809">809</a></span>
<span class="normal"><a href="#__codelineno-0-810">810</a></span>
<span class="normal"><a href="#__codelineno-0-811">811</a></span>
<span class="normal"><a href="#__codelineno-0-812">812</a></span>
<span class="normal"><a href="#__codelineno-0-813">813</a></span>
<span class="normal"><a href="#__codelineno-0-814">814</a></span>
<span class="normal"><a href="#__codelineno-0-815">815</a></span>
<span class="normal"><a href="#__codelineno-0-816">816</a></span>
<span class="normal"><a href="#__codelineno-0-817">817</a></span>
<span class="normal"><a href="#__codelineno-0-818">818</a></span>
<span class="normal"><a href="#__codelineno-0-819">819</a></span>
<span class="normal"><a href="#__codelineno-0-820">820</a></span></pre></div></td><td class="code"><div><pre><span></span><code><span id="__span-0-777"><a id="__codelineno-0-777" name="__codelineno-0-777"></a><span class="k">def</span><span class="w"> </span><span class="nf">score</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">beta</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">standardize_genotype</span><span class="o">=</span><span class="kc">False</span><span class="p">):</span>
</span><span id="__span-0-778"><a id="__codelineno-0-778" name="__codelineno-0-778"></a><span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
</span><span id="__span-0-779"><a id="__codelineno-0-779" name="__codelineno-0-779"></a><span class="sd">    Perform linear scoring, i.e. multiply the genotype matrix by the vector of effect sizes, `beta`.</span>
</span><span id="__span-0-780"><a id="__codelineno-0-780" name="__codelineno-0-780"></a>
</span><span id="__span-0-781"><a id="__codelineno-0-781" name="__codelineno-0-781"></a><span class="sd">    :param beta: A dictionary where the keys are the chromosome numbers and the</span>
</span><span id="__span-0-782"><a id="__codelineno-0-782" name="__codelineno-0-782"></a><span class="sd">    values are a vector of effect sizes for each variant on that chromosome. If the</span>
</span><span id="__span-0-783"><a id="__codelineno-0-783" name="__codelineno-0-783"></a><span class="sd">    betas are not provided, we use the marginal betas by default (if those are available).</span>
</span><span id="__span-0-784"><a id="__codelineno-0-784" name="__codelineno-0-784"></a><span class="sd">    :param standardize_genotype: If True, standardize the genotype matrix before scoring.</span>
</span><span id="__span-0-785"><a id="__codelineno-0-785" name="__codelineno-0-785"></a><span class="sd">    &quot;&quot;&quot;</span>
</span><span id="__span-0-786"><a id="__codelineno-0-786" name="__codelineno-0-786"></a>
</span><span id="__span-0-787"><a id="__codelineno-0-787" name="__codelineno-0-787"></a>    <span class="k">if</span> <span class="n">beta</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
</span><span id="__span-0-788"><a id="__codelineno-0-788" name="__codelineno-0-788"></a>        <span class="k">try</span><span class="p">:</span>
</span><span id="__span-0-789"><a id="__codelineno-0-789" name="__codelineno-0-789"></a>            <span class="n">beta</span> <span class="o">=</span> <span class="p">{</span><span class="n">c</span><span class="p">:</span> <span class="n">s</span><span class="o">.</span><span class="n">marginal_beta</span> <span class="ow">or</span> <span class="n">s</span><span class="o">.</span><span class="n">get_snp_pseudo_corr</span><span class="p">()</span> <span class="k">for</span> <span class="n">c</span><span class="p">,</span> <span class="n">s</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">sumstats_table</span><span class="o">.</span><span class="n">items</span><span class="p">()}</span>
</span><span id="__span-0-790"><a id="__codelineno-0-790" name="__codelineno-0-790"></a>        <span class="k">except</span> <span class="ne">Exception</span><span class="p">:</span>
</span><span id="__span-0-791"><a id="__codelineno-0-791" name="__codelineno-0-791"></a>            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;To perform linear scoring, you must &quot;</span>
</span><span id="__span-0-792"><a id="__codelineno-0-792" name="__codelineno-0-792"></a>                             <span class="s2">&quot;provide effect size estimates (BETA)!&quot;</span><span class="p">)</span>
</span><span id="__span-0-793"><a id="__codelineno-0-793" name="__codelineno-0-793"></a>
</span><span id="__span-0-794"><a id="__codelineno-0-794" name="__codelineno-0-794"></a>    <span class="c1"># Here, we have a very ugly way of accounting for</span>
</span><span id="__span-0-795"><a id="__codelineno-0-795" name="__codelineno-0-795"></a>    <span class="c1"># the fact that the chromosomes may be coded differently between the genotype</span>
</span><span id="__span-0-796"><a id="__codelineno-0-796" name="__codelineno-0-796"></a>    <span class="c1"># and the beta dictionary. Maybe we can find a better solution in the future.</span>
</span><span id="__span-0-797"><a id="__codelineno-0-797" name="__codelineno-0-797"></a>    <span class="n">common_chr_g</span><span class="p">,</span> <span class="n">common_chr_b</span> <span class="o">=</span> <span class="n">match_chromosomes</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">genotype</span><span class="o">.</span><span class="n">keys</span><span class="p">(),</span> <span class="n">beta</span><span class="o">.</span><span class="n">keys</span><span class="p">(),</span> <span class="n">return_both</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
</span><span id="__span-0-798"><a id="__codelineno-0-798" name="__codelineno-0-798"></a>
</span><span id="__span-0-799"><a id="__codelineno-0-799" name="__codelineno-0-799"></a>    <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">common_chr_g</span><span class="p">)</span> <span class="o">&lt;</span> <span class="mi">1</span><span class="p">:</span>
</span><span id="__span-0-800"><a id="__codelineno-0-800" name="__codelineno-0-800"></a>        <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;No common chromosomes found between &quot;</span>
</span><span id="__span-0-801"><a id="__codelineno-0-801" name="__codelineno-0-801"></a>                         <span class="s2">&quot;the genotype and the effect size estimates!&quot;</span><span class="p">)</span>
</span><span id="__span-0-802"><a id="__codelineno-0-802" name="__codelineno-0-802"></a>
</span><span id="__span-0-803"><a id="__codelineno-0-803" name="__codelineno-0-803"></a>    <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;&gt; Generating polygenic scores...&quot;</span><span class="p">)</span>
</span><span id="__span-0-804"><a id="__codelineno-0-804" name="__codelineno-0-804"></a>
</span><span id="__span-0-805"><a id="__codelineno-0-805" name="__codelineno-0-805"></a>    <span class="n">pgs</span> <span class="o">=</span> <span class="kc">None</span>
</span><span id="__span-0-806"><a id="__codelineno-0-806" name="__codelineno-0-806"></a>
</span><span id="__span-0-807"><a id="__codelineno-0-807" name="__codelineno-0-807"></a>    <span class="k">for</span> <span class="n">c_g</span><span class="p">,</span> <span class="n">c_b</span> <span class="ow">in</span> <span class="n">tqdm</span><span class="p">(</span><span class="nb">zip</span><span class="p">(</span><span class="n">common_chr_g</span><span class="p">,</span> <span class="n">common_chr_b</span><span class="p">),</span>
</span><span id="__span-0-808"><a id="__codelineno-0-808" name="__codelineno-0-808"></a>                         <span class="n">total</span><span class="o">=</span><span class="nb">len</span><span class="p">(</span><span class="n">common_chr_g</span><span class="p">),</span>
</span><span id="__span-0-809"><a id="__codelineno-0-809" name="__codelineno-0-809"></a>                         <span class="n">desc</span><span class="o">=</span><span class="s1">&#39;Generating polygenic scores&#39;</span><span class="p">):</span>
</span><span id="__span-0-810"><a id="__codelineno-0-810" name="__codelineno-0-810"></a>
</span><span id="__span-0-811"><a id="__codelineno-0-811" name="__codelineno-0-811"></a>        <span class="k">if</span> <span class="n">pgs</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
</span><span id="__span-0-812"><a id="__codelineno-0-812" name="__codelineno-0-812"></a>            <span class="n">pgs</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">genotype</span><span class="p">[</span><span class="n">c_g</span><span class="p">]</span><span class="o">.</span><span class="n">score</span><span class="p">(</span><span class="n">beta</span><span class="p">[</span><span class="n">c_b</span><span class="p">],</span> <span class="n">standardize_genotype</span><span class="o">=</span><span class="n">standardize_genotype</span><span class="p">)</span>
</span><span id="__span-0-813"><a id="__codelineno-0-813" name="__codelineno-0-813"></a>        <span class="k">else</span><span class="p">:</span>
</span><span id="__span-0-814"><a id="__codelineno-0-814" name="__codelineno-0-814"></a>            <span class="n">pgs</span> <span class="o">+=</span> <span class="bp">self</span><span class="o">.</span><span class="n">genotype</span><span class="p">[</span><span class="n">c_g</span><span class="p">]</span><span class="o">.</span><span class="n">score</span><span class="p">(</span><span class="n">beta</span><span class="p">[</span><span class="n">c_b</span><span class="p">],</span> <span class="n">standardize_genotype</span><span class="o">=</span><span class="n">standardize_genotype</span><span class="p">)</span>
</span><span id="__span-0-815"><a id="__codelineno-0-815" name="__codelineno-0-815"></a>
</span><span id="__span-0-816"><a id="__codelineno-0-816" name="__codelineno-0-816"></a>    <span class="c1"># If we only have a single set of betas, flatten the PGS vector:</span>
</span><span id="__span-0-817"><a id="__codelineno-0-817" name="__codelineno-0-817"></a>    <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">pgs</span><span class="o">.</span><span class="n">shape</span><span class="p">)</span> <span class="o">&gt;</span> <span class="mi">1</span> <span class="ow">and</span> <span class="n">pgs</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span> <span class="o">==</span> <span class="mi">1</span><span class="p">:</span>
</span><span id="__span-0-818"><a id="__codelineno-0-818" name="__codelineno-0-818"></a>        <span class="n">pgs</span> <span class="o">=</span> <span class="n">pgs</span><span class="o">.</span><span class="n">flatten</span><span class="p">()</span>
</span><span id="__span-0-819"><a id="__codelineno-0-819" name="__codelineno-0-819"></a>
</span><span id="__span-0-820"><a id="__codelineno-0-820" name="__codelineno-0-820"></a>    <span class="k">return</span> <span class="n">pgs</span>
</span></code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h2 id="magenpy.GWADataLoader.GWADataLoader.set_phenotype" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">set_phenotype</span><span class="p">(</span><span class="n">new_phenotype</span><span class="p">,</span> <span class="n">phenotype_likelihood</span><span class="o">=</span><span class="kc">None</span><span class="p">)</span></code>

<a href="#magenpy.GWADataLoader.GWADataLoader.set_phenotype" class="headerlink" title="Permanent link">&para;</a></h2>


    <div class="doc doc-contents ">

        <p>A convenience method to update the phenotype column for the samples.</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                <code>new_phenotype</code>
            </td>
            <td>
            </td>
            <td>
              <div class="doc-md-description">
                <p>A vector or list of phenotype values.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>phenotype_likelihood</code>
            </td>
            <td>
            </td>
            <td>
              <div class="doc-md-description">
                <p>The phenotype likelihood (e.g. <code>binomial</code>, <code>gaussian</code>). Optional.</p>
              </div>
            </td>
            <td>
                  <code>None</code>
            </td>
          </tr>
      </tbody>
    </table>


            <details class="quote">
              <summary>Source code in <code>magenpy/GWADataLoader.py</code></summary>
              <div class="language-python highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal"><a href="#__codelineno-0-471">471</a></span>
<span class="normal"><a href="#__codelineno-0-472">472</a></span>
<span class="normal"><a href="#__codelineno-0-473">473</a></span>
<span class="normal"><a href="#__codelineno-0-474">474</a></span>
<span class="normal"><a href="#__codelineno-0-475">475</a></span>
<span class="normal"><a href="#__codelineno-0-476">476</a></span>
<span class="normal"><a href="#__codelineno-0-477">477</a></span>
<span class="normal"><a href="#__codelineno-0-478">478</a></span>
<span class="normal"><a href="#__codelineno-0-479">479</a></span>
<span class="normal"><a href="#__codelineno-0-480">480</a></span></pre></div></td><td class="code"><div><pre><span></span><code><span id="__span-0-471"><a id="__codelineno-0-471" name="__codelineno-0-471"></a><span class="k">def</span><span class="w"> </span><span class="nf">set_phenotype</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">new_phenotype</span><span class="p">,</span> <span class="n">phenotype_likelihood</span><span class="o">=</span><span class="kc">None</span><span class="p">):</span>
</span><span id="__span-0-472"><a id="__codelineno-0-472" name="__codelineno-0-472"></a><span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
</span><span id="__span-0-473"><a id="__codelineno-0-473" name="__codelineno-0-473"></a><span class="sd">    A convenience method to update the phenotype column for the samples.</span>
</span><span id="__span-0-474"><a id="__codelineno-0-474" name="__codelineno-0-474"></a><span class="sd">    :param new_phenotype: A vector or list of phenotype values.</span>
</span><span id="__span-0-475"><a id="__codelineno-0-475" name="__codelineno-0-475"></a><span class="sd">    :param phenotype_likelihood: The phenotype likelihood (e.g. `binomial`, `gaussian`). Optional.</span>
</span><span id="__span-0-476"><a id="__codelineno-0-476" name="__codelineno-0-476"></a><span class="sd">    &quot;&quot;&quot;</span>
</span><span id="__span-0-477"><a id="__codelineno-0-477" name="__codelineno-0-477"></a>
</span><span id="__span-0-478"><a id="__codelineno-0-478" name="__codelineno-0-478"></a>    <span class="bp">self</span><span class="o">.</span><span class="n">sample_table</span><span class="o">.</span><span class="n">set_phenotype</span><span class="p">(</span><span class="n">new_phenotype</span><span class="p">,</span>
</span><span id="__span-0-479"><a id="__codelineno-0-479" name="__codelineno-0-479"></a>                                    <span class="n">phenotype_likelihood</span><span class="o">=</span><span class="n">phenotype_likelihood</span> <span class="ow">or</span> <span class="bp">self</span><span class="o">.</span><span class="n">phenotype_likelihood</span><span class="p">)</span>
</span><span id="__span-0-480"><a id="__codelineno-0-480" name="__codelineno-0-480"></a>    <span class="bp">self</span><span class="o">.</span><span class="n">sync_sample_tables</span><span class="p">()</span>
</span></code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h2 id="magenpy.GWADataLoader.GWADataLoader.split_by_chromosome" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">split_by_chromosome</span><span class="p">()</span></code>

<a href="#magenpy.GWADataLoader.GWADataLoader.split_by_chromosome" class="headerlink" title="Permanent link">&para;</a></h2>


    <div class="doc doc-contents ">

        <p>A utility method to split a GWADataLoader object by chromosome ID, such that
we would have one <code>GWADataLoader</code> object per chromosome. The method returns a dictionary
where the key is the chromosome number and the value is the <code>GWADataLoader</code> object corresponding
to that chromosome only.</p>


            <details class="quote">
              <summary>Source code in <code>magenpy/GWADataLoader.py</code></summary>
              <div class="language-python highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal"><a href="#__codelineno-0-949">949</a></span>
<span class="normal"><a href="#__codelineno-0-950">950</a></span>
<span class="normal"><a href="#__codelineno-0-951">951</a></span>
<span class="normal"><a href="#__codelineno-0-952">952</a></span>
<span class="normal"><a href="#__codelineno-0-953">953</a></span>
<span class="normal"><a href="#__codelineno-0-954">954</a></span>
<span class="normal"><a href="#__codelineno-0-955">955</a></span>
<span class="normal"><a href="#__codelineno-0-956">956</a></span>
<span class="normal"><a href="#__codelineno-0-957">957</a></span>
<span class="normal"><a href="#__codelineno-0-958">958</a></span>
<span class="normal"><a href="#__codelineno-0-959">959</a></span>
<span class="normal"><a href="#__codelineno-0-960">960</a></span>
<span class="normal"><a href="#__codelineno-0-961">961</a></span>
<span class="normal"><a href="#__codelineno-0-962">962</a></span>
<span class="normal"><a href="#__codelineno-0-963">963</a></span>
<span class="normal"><a href="#__codelineno-0-964">964</a></span>
<span class="normal"><a href="#__codelineno-0-965">965</a></span>
<span class="normal"><a href="#__codelineno-0-966">966</a></span>
<span class="normal"><a href="#__codelineno-0-967">967</a></span>
<span class="normal"><a href="#__codelineno-0-968">968</a></span>
<span class="normal"><a href="#__codelineno-0-969">969</a></span>
<span class="normal"><a href="#__codelineno-0-970">970</a></span>
<span class="normal"><a href="#__codelineno-0-971">971</a></span>
<span class="normal"><a href="#__codelineno-0-972">972</a></span>
<span class="normal"><a href="#__codelineno-0-973">973</a></span>
<span class="normal"><a href="#__codelineno-0-974">974</a></span>
<span class="normal"><a href="#__codelineno-0-975">975</a></span></pre></div></td><td class="code"><div><pre><span></span><code><span id="__span-0-949"><a id="__codelineno-0-949" name="__codelineno-0-949"></a><span class="k">def</span><span class="w"> </span><span class="nf">split_by_chromosome</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
</span><span id="__span-0-950"><a id="__codelineno-0-950" name="__codelineno-0-950"></a><span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
</span><span id="__span-0-951"><a id="__codelineno-0-951" name="__codelineno-0-951"></a><span class="sd">    A utility method to split a GWADataLoader object by chromosome ID, such that</span>
</span><span id="__span-0-952"><a id="__codelineno-0-952" name="__codelineno-0-952"></a><span class="sd">    we would have one `GWADataLoader` object per chromosome. The method returns a dictionary</span>
</span><span id="__span-0-953"><a id="__codelineno-0-953" name="__codelineno-0-953"></a><span class="sd">    where the key is the chromosome number and the value is the `GWADataLoader` object corresponding</span>
</span><span id="__span-0-954"><a id="__codelineno-0-954" name="__codelineno-0-954"></a><span class="sd">    to that chromosome only.</span>
</span><span id="__span-0-955"><a id="__codelineno-0-955" name="__codelineno-0-955"></a><span class="sd">    &quot;&quot;&quot;</span>
</span><span id="__span-0-956"><a id="__codelineno-0-956" name="__codelineno-0-956"></a>
</span><span id="__span-0-957"><a id="__codelineno-0-957" name="__codelineno-0-957"></a>    <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">chromosomes</span><span class="p">)</span> <span class="o">==</span> <span class="mi">1</span><span class="p">:</span>
</span><span id="__span-0-958"><a id="__codelineno-0-958" name="__codelineno-0-958"></a>        <span class="k">return</span> <span class="p">{</span><span class="bp">self</span><span class="o">.</span><span class="n">chromosomes</span><span class="p">[</span><span class="mi">0</span><span class="p">]:</span> <span class="bp">self</span><span class="p">}</span>
</span><span id="__span-0-959"><a id="__codelineno-0-959" name="__codelineno-0-959"></a>
</span><span id="__span-0-960"><a id="__codelineno-0-960" name="__codelineno-0-960"></a>    <span class="k">else</span><span class="p">:</span>
</span><span id="__span-0-961"><a id="__codelineno-0-961" name="__codelineno-0-961"></a>        <span class="n">split_dict</span> <span class="o">=</span> <span class="p">{}</span>
</span><span id="__span-0-962"><a id="__codelineno-0-962" name="__codelineno-0-962"></a>
</span><span id="__span-0-963"><a id="__codelineno-0-963" name="__codelineno-0-963"></a>        <span class="k">for</span> <span class="n">c</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">chromosomes</span><span class="p">:</span>
</span><span id="__span-0-964"><a id="__codelineno-0-964" name="__codelineno-0-964"></a>            <span class="n">split_dict</span><span class="p">[</span><span class="n">c</span><span class="p">]</span> <span class="o">=</span> <span class="n">copy</span><span class="o">.</span><span class="n">copy</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span>
</span><span id="__span-0-965"><a id="__codelineno-0-965" name="__codelineno-0-965"></a>
</span><span id="__span-0-966"><a id="__codelineno-0-966" name="__codelineno-0-966"></a>            <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">genotype</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span> <span class="ow">and</span> <span class="n">c</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">genotype</span><span class="p">:</span>
</span><span id="__span-0-967"><a id="__codelineno-0-967" name="__codelineno-0-967"></a>                <span class="n">split_dict</span><span class="p">[</span><span class="n">c</span><span class="p">]</span><span class="o">.</span><span class="n">genotype</span> <span class="o">=</span> <span class="p">{</span><span class="n">c</span><span class="p">:</span> <span class="bp">self</span><span class="o">.</span><span class="n">genotype</span><span class="p">[</span><span class="n">c</span><span class="p">]}</span>
</span><span id="__span-0-968"><a id="__codelineno-0-968" name="__codelineno-0-968"></a>            <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">sumstats_table</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span> <span class="ow">and</span> <span class="n">c</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">sumstats_table</span><span class="p">:</span>
</span><span id="__span-0-969"><a id="__codelineno-0-969" name="__codelineno-0-969"></a>                <span class="n">split_dict</span><span class="p">[</span><span class="n">c</span><span class="p">]</span><span class="o">.</span><span class="n">sumstats_table</span> <span class="o">=</span> <span class="p">{</span><span class="n">c</span><span class="p">:</span> <span class="bp">self</span><span class="o">.</span><span class="n">sumstats_table</span><span class="p">[</span><span class="n">c</span><span class="p">]}</span>
</span><span id="__span-0-970"><a id="__codelineno-0-970" name="__codelineno-0-970"></a>            <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">ld</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span> <span class="ow">and</span> <span class="n">c</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">ld</span><span class="p">:</span>
</span><span id="__span-0-971"><a id="__codelineno-0-971" name="__codelineno-0-971"></a>                <span class="n">split_dict</span><span class="p">[</span><span class="n">c</span><span class="p">]</span><span class="o">.</span><span class="n">ld</span> <span class="o">=</span> <span class="p">{</span><span class="n">c</span><span class="p">:</span> <span class="bp">self</span><span class="o">.</span><span class="n">ld</span><span class="p">[</span><span class="n">c</span><span class="p">]}</span>
</span><span id="__span-0-972"><a id="__codelineno-0-972" name="__codelineno-0-972"></a>            <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">annotation</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span> <span class="ow">and</span> <span class="n">c</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">annotation</span><span class="p">:</span>
</span><span id="__span-0-973"><a id="__codelineno-0-973" name="__codelineno-0-973"></a>                <span class="n">split_dict</span><span class="p">[</span><span class="n">c</span><span class="p">]</span><span class="o">.</span><span class="n">annotation</span> <span class="o">=</span> <span class="p">{</span><span class="n">c</span><span class="p">:</span> <span class="bp">self</span><span class="o">.</span><span class="n">annotation</span><span class="p">[</span><span class="n">c</span><span class="p">]}</span>
</span><span id="__span-0-974"><a id="__codelineno-0-974" name="__codelineno-0-974"></a>
</span><span id="__span-0-975"><a id="__codelineno-0-975" name="__codelineno-0-975"></a>        <span class="k">return</span> <span class="n">split_dict</span>
</span></code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h2 id="magenpy.GWADataLoader.GWADataLoader.split_by_samples" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">split_by_samples</span><span class="p">(</span><span class="n">proportions</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">groups</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">keep_original</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span></code>

<a href="#magenpy.GWADataLoader.GWADataLoader.split_by_samples" class="headerlink" title="Permanent link">&para;</a></h2>


    <div class="doc doc-contents ">

        <p>Split the <code>GWADataLoader</code> object by samples, if genotype or sample data
is available. The user must provide a list or proportion of samples in each split,
and the method will return a list of <code>GWADataLoader</code> objects with only the samples
designated for each split. This may be a useful utility for training/testing split or some
other downstream tasks.</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                <code>proportions</code>
            </td>
            <td>
            </td>
            <td>
              <div class="doc-md-description">
                <p>A list with the proportion of samples in each split. Must add to 1.</p>
              </div>
            </td>
            <td>
                  <code>None</code>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>groups</code>
            </td>
            <td>
            </td>
            <td>
              <div class="doc-md-description">
                <p>A list of lists containing the sample IDs in each split.</p>
              </div>
            </td>
            <td>
                  <code>None</code>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>keep_original</code>
            </td>
            <td>
            </td>
            <td>
              <div class="doc-md-description">
                <p>If True, keep the original <code>GWADataLoader</code> object and do not transform it in the splitting process.</p>
              </div>
            </td>
            <td>
                  <code>True</code>
            </td>
          </tr>
      </tbody>
    </table>


            <details class="quote">
              <summary>Source code in <code>magenpy/GWADataLoader.py</code></summary>
              <div class="language-python highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal"><a href="#__codelineno-0-977"> 977</a></span>
<span class="normal"><a href="#__codelineno-0-978"> 978</a></span>
<span class="normal"><a href="#__codelineno-0-979"> 979</a></span>
<span class="normal"><a href="#__codelineno-0-980"> 980</a></span>
<span class="normal"><a href="#__codelineno-0-981"> 981</a></span>
<span class="normal"><a href="#__codelineno-0-982"> 982</a></span>
<span class="normal"><a href="#__codelineno-0-983"> 983</a></span>
<span class="normal"><a href="#__codelineno-0-984"> 984</a></span>
<span class="normal"><a href="#__codelineno-0-985"> 985</a></span>
<span class="normal"><a href="#__codelineno-0-986"> 986</a></span>
<span class="normal"><a href="#__codelineno-0-987"> 987</a></span>
<span class="normal"><a href="#__codelineno-0-988"> 988</a></span>
<span class="normal"><a href="#__codelineno-0-989"> 989</a></span>
<span class="normal"><a href="#__codelineno-0-990"> 990</a></span>
<span class="normal"><a href="#__codelineno-0-991"> 991</a></span>
<span class="normal"><a href="#__codelineno-0-992"> 992</a></span>
<span class="normal"><a href="#__codelineno-0-993"> 993</a></span>
<span class="normal"><a href="#__codelineno-0-994"> 994</a></span>
<span class="normal"><a href="#__codelineno-0-995"> 995</a></span>
<span class="normal"><a href="#__codelineno-0-996"> 996</a></span>
<span class="normal"><a href="#__codelineno-0-997"> 997</a></span>
<span class="normal"><a href="#__codelineno-0-998"> 998</a></span>
<span class="normal"><a href="#__codelineno-0-999"> 999</a></span>
<span class="normal"><a href="#__codelineno-0-1000">1000</a></span>
<span class="normal"><a href="#__codelineno-0-1001">1001</a></span>
<span class="normal"><a href="#__codelineno-0-1002">1002</a></span>
<span class="normal"><a href="#__codelineno-0-1003">1003</a></span>
<span class="normal"><a href="#__codelineno-0-1004">1004</a></span>
<span class="normal"><a href="#__codelineno-0-1005">1005</a></span>
<span class="normal"><a href="#__codelineno-0-1006">1006</a></span>
<span class="normal"><a href="#__codelineno-0-1007">1007</a></span>
<span class="normal"><a href="#__codelineno-0-1008">1008</a></span>
<span class="normal"><a href="#__codelineno-0-1009">1009</a></span>
<span class="normal"><a href="#__codelineno-0-1010">1010</a></span>
<span class="normal"><a href="#__codelineno-0-1011">1011</a></span>
<span class="normal"><a href="#__codelineno-0-1012">1012</a></span>
<span class="normal"><a href="#__codelineno-0-1013">1013</a></span>
<span class="normal"><a href="#__codelineno-0-1014">1014</a></span>
<span class="normal"><a href="#__codelineno-0-1015">1015</a></span>
<span class="normal"><a href="#__codelineno-0-1016">1016</a></span>
<span class="normal"><a href="#__codelineno-0-1017">1017</a></span>
<span class="normal"><a href="#__codelineno-0-1018">1018</a></span>
<span class="normal"><a href="#__codelineno-0-1019">1019</a></span>
<span class="normal"><a href="#__codelineno-0-1020">1020</a></span></pre></div></td><td class="code"><div><pre><span></span><code><span id="__span-0-977"><a id="__codelineno-0-977" name="__codelineno-0-977"></a><span class="k">def</span><span class="w"> </span><span class="nf">split_by_samples</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">proportions</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">groups</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">keep_original</span><span class="o">=</span><span class="kc">True</span><span class="p">):</span>
</span><span id="__span-0-978"><a id="__codelineno-0-978" name="__codelineno-0-978"></a><span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
</span><span id="__span-0-979"><a id="__codelineno-0-979" name="__codelineno-0-979"></a><span class="sd">    Split the `GWADataLoader` object by samples, if genotype or sample data</span>
</span><span id="__span-0-980"><a id="__codelineno-0-980" name="__codelineno-0-980"></a><span class="sd">    is available. The user must provide a list or proportion of samples in each split,</span>
</span><span id="__span-0-981"><a id="__codelineno-0-981" name="__codelineno-0-981"></a><span class="sd">    and the method will return a list of `GWADataLoader` objects with only the samples</span>
</span><span id="__span-0-982"><a id="__codelineno-0-982" name="__codelineno-0-982"></a><span class="sd">    designated for each split. This may be a useful utility for training/testing split or some</span>
</span><span id="__span-0-983"><a id="__codelineno-0-983" name="__codelineno-0-983"></a><span class="sd">    other downstream tasks.</span>
</span><span id="__span-0-984"><a id="__codelineno-0-984" name="__codelineno-0-984"></a>
</span><span id="__span-0-985"><a id="__codelineno-0-985" name="__codelineno-0-985"></a><span class="sd">    :param proportions: A list with the proportion of samples in each split. Must add to 1.</span>
</span><span id="__span-0-986"><a id="__codelineno-0-986" name="__codelineno-0-986"></a><span class="sd">    :param groups: A list of lists containing the sample IDs in each split.</span>
</span><span id="__span-0-987"><a id="__codelineno-0-987" name="__codelineno-0-987"></a><span class="sd">    :param keep_original: If True, keep the original `GWADataLoader` object and do not</span>
</span><span id="__span-0-988"><a id="__codelineno-0-988" name="__codelineno-0-988"></a><span class="sd">    transform it in the splitting process.</span>
</span><span id="__span-0-989"><a id="__codelineno-0-989" name="__codelineno-0-989"></a><span class="sd">    &quot;&quot;&quot;</span>
</span><span id="__span-0-990"><a id="__codelineno-0-990" name="__codelineno-0-990"></a>
</span><span id="__span-0-991"><a id="__codelineno-0-991" name="__codelineno-0-991"></a>    <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">sample_table</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
</span><span id="__span-0-992"><a id="__codelineno-0-992" name="__codelineno-0-992"></a>        <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;The sample table is not set!&quot;</span><span class="p">)</span>
</span><span id="__span-0-993"><a id="__codelineno-0-993" name="__codelineno-0-993"></a>
</span><span id="__span-0-994"><a id="__codelineno-0-994" name="__codelineno-0-994"></a>    <span class="k">if</span> <span class="n">groups</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
</span><span id="__span-0-995"><a id="__codelineno-0-995" name="__codelineno-0-995"></a>        <span class="k">if</span> <span class="n">proportions</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
</span><span id="__span-0-996"><a id="__codelineno-0-996" name="__codelineno-0-996"></a>            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;To split a `GWADataloader` object by samples, the user must provide either the list &quot;</span>
</span><span id="__span-0-997"><a id="__codelineno-0-997" name="__codelineno-0-997"></a>                             <span class="s2">&quot;or proportion of individuals in each split.&quot;</span><span class="p">)</span>
</span><span id="__span-0-998"><a id="__codelineno-0-998" name="__codelineno-0-998"></a>        <span class="k">else</span><span class="p">:</span>
</span><span id="__span-0-999"><a id="__codelineno-0-999" name="__codelineno-0-999"></a>
</span><span id="__span-0-1000"><a id="__codelineno-0-1000" name="__codelineno-0-1000"></a>            <span class="c1"># Assign each sample to a different split randomly by drawing from a multinomial:</span>
</span><span id="__span-0-1001"><a id="__codelineno-0-1001" name="__codelineno-0-1001"></a>            <span class="n">random_split</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">random</span><span class="o">.</span><span class="n">multinomial</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="n">proportions</span><span class="p">,</span> <span class="n">size</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">sample_size</span><span class="p">)</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="nb">bool</span><span class="p">)</span>
</span><span id="__span-0-1002"><a id="__codelineno-0-1002" name="__codelineno-0-1002"></a>            <span class="c1"># Extract the individuals in each group from the multinomial sample:</span>
</span><span id="__span-0-1003"><a id="__codelineno-0-1003" name="__codelineno-0-1003"></a>            <span class="n">groups</span> <span class="o">=</span> <span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">samples</span><span class="p">[</span><span class="n">random_split</span><span class="p">[:,</span> <span class="n">i</span><span class="p">]]</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">random_split</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">1</span><span class="p">])]</span>
</span><span id="__span-0-1004"><a id="__codelineno-0-1004" name="__codelineno-0-1004"></a>
</span><span id="__span-0-1005"><a id="__codelineno-0-1005" name="__codelineno-0-1005"></a>    <span class="n">gdls</span> <span class="o">=</span> <span class="p">[]</span>
</span><span id="__span-0-1006"><a id="__codelineno-0-1006" name="__codelineno-0-1006"></a>    <span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="n">g</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">groups</span><span class="p">):</span>
</span><span id="__span-0-1007"><a id="__codelineno-0-1007" name="__codelineno-0-1007"></a>
</span><span id="__span-0-1008"><a id="__codelineno-0-1008" name="__codelineno-0-1008"></a>        <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">g</span><span class="p">)</span> <span class="o">&lt;</span> <span class="mi">1</span><span class="p">:</span>
</span><span id="__span-0-1009"><a id="__codelineno-0-1009" name="__codelineno-0-1009"></a>            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Group </span><span class="si">{</span><span class="n">i</span><span class="si">}</span><span class="s2"> is empty! Please ensure that all splits have at least one sample.&quot;</span><span class="p">)</span>
</span><span id="__span-0-1010"><a id="__codelineno-0-1010" name="__codelineno-0-1010"></a>
</span><span id="__span-0-1011"><a id="__codelineno-0-1011" name="__codelineno-0-1011"></a>        <span class="k">if</span> <span class="p">(</span><span class="n">i</span> <span class="o">+</span> <span class="mi">1</span><span class="p">)</span> <span class="o">==</span> <span class="nb">len</span><span class="p">(</span><span class="n">groups</span><span class="p">)</span> <span class="ow">and</span> <span class="ow">not</span> <span class="n">keep_original</span><span class="p">:</span>
</span><span id="__span-0-1012"><a id="__codelineno-0-1012" name="__codelineno-0-1012"></a>            <span class="n">new_gdl</span> <span class="o">=</span> <span class="bp">self</span>
</span><span id="__span-0-1013"><a id="__codelineno-0-1013" name="__codelineno-0-1013"></a>        <span class="k">else</span><span class="p">:</span>
</span><span id="__span-0-1014"><a id="__codelineno-0-1014" name="__codelineno-0-1014"></a>            <span class="n">new_gdl</span> <span class="o">=</span> <span class="n">copy</span><span class="o">.</span><span class="n">deepcopy</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span>
</span><span id="__span-0-1015"><a id="__codelineno-0-1015" name="__codelineno-0-1015"></a>
</span><span id="__span-0-1016"><a id="__codelineno-0-1016" name="__codelineno-0-1016"></a>        <span class="n">new_gdl</span><span class="o">.</span><span class="n">filter_samples</span><span class="p">(</span><span class="n">keep_samples</span><span class="o">=</span><span class="n">g</span><span class="p">)</span>
</span><span id="__span-0-1017"><a id="__codelineno-0-1017" name="__codelineno-0-1017"></a>
</span><span id="__span-0-1018"><a id="__codelineno-0-1018" name="__codelineno-0-1018"></a>        <span class="n">gdls</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">new_gdl</span><span class="p">)</span>
</span><span id="__span-0-1019"><a id="__codelineno-0-1019" name="__codelineno-0-1019"></a>
</span><span id="__span-0-1020"><a id="__codelineno-0-1020" name="__codelineno-0-1020"></a>    <span class="k">return</span> <span class="n">gdls</span>
</span></code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h2 id="magenpy.GWADataLoader.GWADataLoader.sync_sample_tables" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">sync_sample_tables</span><span class="p">()</span></code>

<a href="#magenpy.GWADataLoader.GWADataLoader.sync_sample_tables" class="headerlink" title="Permanent link">&para;</a></h2>


    <div class="doc doc-contents ">

        <p>A utility method to sync the sample tables of the
<code>GenotypeMatrix</code> objects with the sample table under
the <code>GWADataLoader</code> object. This is especially important
when setting new phenotypes (from the simulators) or reading
covariates files, etc.</p>


            <details class="quote">
              <summary>Source code in <code>magenpy/GWADataLoader.py</code></summary>
              <div class="language-python highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal"><a href="#__codelineno-0-937">937</a></span>
<span class="normal"><a href="#__codelineno-0-938">938</a></span>
<span class="normal"><a href="#__codelineno-0-939">939</a></span>
<span class="normal"><a href="#__codelineno-0-940">940</a></span>
<span class="normal"><a href="#__codelineno-0-941">941</a></span>
<span class="normal"><a href="#__codelineno-0-942">942</a></span>
<span class="normal"><a href="#__codelineno-0-943">943</a></span>
<span class="normal"><a href="#__codelineno-0-944">944</a></span>
<span class="normal"><a href="#__codelineno-0-945">945</a></span>
<span class="normal"><a href="#__codelineno-0-946">946</a></span>
<span class="normal"><a href="#__codelineno-0-947">947</a></span></pre></div></td><td class="code"><div><pre><span></span><code><span id="__span-0-937"><a id="__codelineno-0-937" name="__codelineno-0-937"></a><span class="k">def</span><span class="w"> </span><span class="nf">sync_sample_tables</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
</span><span id="__span-0-938"><a id="__codelineno-0-938" name="__codelineno-0-938"></a><span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
</span><span id="__span-0-939"><a id="__codelineno-0-939" name="__codelineno-0-939"></a><span class="sd">    A utility method to sync the sample tables of the</span>
</span><span id="__span-0-940"><a id="__codelineno-0-940" name="__codelineno-0-940"></a><span class="sd">    `GenotypeMatrix` objects with the sample table under</span>
</span><span id="__span-0-941"><a id="__codelineno-0-941" name="__codelineno-0-941"></a><span class="sd">    the `GWADataLoader` object. This is especially important</span>
</span><span id="__span-0-942"><a id="__codelineno-0-942" name="__codelineno-0-942"></a><span class="sd">    when setting new phenotypes (from the simulators) or reading</span>
</span><span id="__span-0-943"><a id="__codelineno-0-943" name="__codelineno-0-943"></a><span class="sd">    covariates files, etc.</span>
</span><span id="__span-0-944"><a id="__codelineno-0-944" name="__codelineno-0-944"></a><span class="sd">    &quot;&quot;&quot;</span>
</span><span id="__span-0-945"><a id="__codelineno-0-945" name="__codelineno-0-945"></a>
</span><span id="__span-0-946"><a id="__codelineno-0-946" name="__codelineno-0-946"></a>    <span class="k">for</span> <span class="n">c</span><span class="p">,</span> <span class="n">g</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">genotype</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
</span><span id="__span-0-947"><a id="__codelineno-0-947" name="__codelineno-0-947"></a>        <span class="n">g</span><span class="o">.</span><span class="n">set_sample_table</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">sample_table</span><span class="p">)</span>
</span></code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h2 id="magenpy.GWADataLoader.GWADataLoader.to_individual_table" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">to_individual_table</span><span class="p">()</span></code>

<a href="#magenpy.GWADataLoader.GWADataLoader.to_individual_table" class="headerlink" title="Permanent link">&para;</a></h2>


    <div class="doc doc-contents ">

        


    <p><span class="doc-section-title">Returns:</span></p>
    <table>
      <thead>
        <tr>
          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
            </td>
            <td>
              <div class="doc-md-description">
                <p>A plink-style dataframe of individual IDs, in the form of Family ID (FID) and Individual ID (IID).</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>


            <details class="quote">
              <summary>Source code in <code>magenpy/GWADataLoader.py</code></summary>
              <div class="language-python highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal"><a href="#__codelineno-0-843">843</a></span>
<span class="normal"><a href="#__codelineno-0-844">844</a></span>
<span class="normal"><a href="#__codelineno-0-845">845</a></span>
<span class="normal"><a href="#__codelineno-0-846">846</a></span>
<span class="normal"><a href="#__codelineno-0-847">847</a></span>
<span class="normal"><a href="#__codelineno-0-848">848</a></span>
<span class="normal"><a href="#__codelineno-0-849">849</a></span>
<span class="normal"><a href="#__codelineno-0-850">850</a></span></pre></div></td><td class="code"><div><pre><span></span><code><span id="__span-0-843"><a id="__codelineno-0-843" name="__codelineno-0-843"></a><span class="k">def</span><span class="w"> </span><span class="nf">to_individual_table</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
</span><span id="__span-0-844"><a id="__codelineno-0-844" name="__codelineno-0-844"></a><span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
</span><span id="__span-0-845"><a id="__codelineno-0-845" name="__codelineno-0-845"></a><span class="sd">    :return: A plink-style dataframe of individual IDs, in the form of</span>
</span><span id="__span-0-846"><a id="__codelineno-0-846" name="__codelineno-0-846"></a><span class="sd">    Family ID (FID) and Individual ID (IID).</span>
</span><span id="__span-0-847"><a id="__codelineno-0-847" name="__codelineno-0-847"></a><span class="sd">    &quot;&quot;&quot;</span>
</span><span id="__span-0-848"><a id="__codelineno-0-848" name="__codelineno-0-848"></a>    <span class="k">assert</span> <span class="bp">self</span><span class="o">.</span><span class="n">sample_table</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span>
</span><span id="__span-0-849"><a id="__codelineno-0-849" name="__codelineno-0-849"></a>
</span><span id="__span-0-850"><a id="__codelineno-0-850" name="__codelineno-0-850"></a>    <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">sample_table</span><span class="o">.</span><span class="n">get_individual_table</span><span class="p">()</span>
</span></code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h2 id="magenpy.GWADataLoader.GWADataLoader.to_phenotype_table" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">to_phenotype_table</span><span class="p">()</span></code>

<a href="#magenpy.GWADataLoader.GWADataLoader.to_phenotype_table" class="headerlink" title="Permanent link">&para;</a></h2>


    <div class="doc doc-contents ">

        


    <p><span class="doc-section-title">Returns:</span></p>
    <table>
      <thead>
        <tr>
          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
            </td>
            <td>
              <div class="doc-md-description">
                <p>A plink-style dataframe with each individual's Family ID (FID), Individual ID (IID), and phenotype value.</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>


            <details class="quote">
              <summary>Source code in <code>magenpy/GWADataLoader.py</code></summary>
              <div class="language-python highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal"><a href="#__codelineno-0-852">852</a></span>
<span class="normal"><a href="#__codelineno-0-853">853</a></span>
<span class="normal"><a href="#__codelineno-0-854">854</a></span>
<span class="normal"><a href="#__codelineno-0-855">855</a></span>
<span class="normal"><a href="#__codelineno-0-856">856</a></span>
<span class="normal"><a href="#__codelineno-0-857">857</a></span>
<span class="normal"><a href="#__codelineno-0-858">858</a></span>
<span class="normal"><a href="#__codelineno-0-859">859</a></span>
<span class="normal"><a href="#__codelineno-0-860">860</a></span></pre></div></td><td class="code"><div><pre><span></span><code><span id="__span-0-852"><a id="__codelineno-0-852" name="__codelineno-0-852"></a><span class="k">def</span><span class="w"> </span><span class="nf">to_phenotype_table</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
</span><span id="__span-0-853"><a id="__codelineno-0-853" name="__codelineno-0-853"></a><span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
</span><span id="__span-0-854"><a id="__codelineno-0-854" name="__codelineno-0-854"></a><span class="sd">    :return: A plink-style dataframe with each individual&#39;s Family ID (FID),</span>
</span><span id="__span-0-855"><a id="__codelineno-0-855" name="__codelineno-0-855"></a><span class="sd">    Individual ID (IID), and phenotype value.</span>
</span><span id="__span-0-856"><a id="__codelineno-0-856" name="__codelineno-0-856"></a><span class="sd">    &quot;&quot;&quot;</span>
</span><span id="__span-0-857"><a id="__codelineno-0-857" name="__codelineno-0-857"></a>
</span><span id="__span-0-858"><a id="__codelineno-0-858" name="__codelineno-0-858"></a>    <span class="k">assert</span> <span class="bp">self</span><span class="o">.</span><span class="n">sample_table</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span>
</span><span id="__span-0-859"><a id="__codelineno-0-859" name="__codelineno-0-859"></a>
</span><span id="__span-0-860"><a id="__codelineno-0-860" name="__codelineno-0-860"></a>    <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">sample_table</span><span class="o">.</span><span class="n">get_phenotype_table</span><span class="p">()</span>
</span></code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h2 id="magenpy.GWADataLoader.GWADataLoader.to_snp_table" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">to_snp_table</span><span class="p">(</span><span class="n">col_subset</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">per_chromosome</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">resource</span><span class="o">=</span><span class="s1">&#39;auto&#39;</span><span class="p">)</span></code>

<a href="#magenpy.GWADataLoader.GWADataLoader.to_snp_table" class="headerlink" title="Permanent link">&para;</a></h2>


    <div class="doc doc-contents ">

        <p>Get a dataframe of SNP data for all variants
across different chromosomes.</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                <code>col_subset</code>
            </td>
            <td>
            </td>
            <td>
              <div class="doc-md-description">
                <p>The subset of columns to obtain.</p>
              </div>
            </td>
            <td>
                  <code>None</code>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>per_chromosome</code>
            </td>
            <td>
            </td>
            <td>
              <div class="doc-md-description">
                <p>If True, returns a dictionary where the key is the chromosome number and the value is the SNP table per chromosome.</p>
              </div>
            </td>
            <td>
                  <code>False</code>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>resource</code>
            </td>
            <td>
            </td>
            <td>
              <div class="doc-md-description">
                <p>The data source to extract the SNP table from. By default, the method will try to extract the SNP table from the genotype matrix. If the genotype matrix is not available, then it will try to extract the SNP information from the LD matrix or the summary statistics table. Possible values: <code>auto</code>, <code>genotype</code>, <code>ld</code>, <code>sumstats</code>.</p>
              </div>
            </td>
            <td>
                  <code>&#39;auto&#39;</code>
            </td>
          </tr>
      </tbody>
    </table>


    <p><span class="doc-section-title">Returns:</span></p>
    <table>
      <thead>
        <tr>
          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
            </td>
            <td>
              <div class="doc-md-description">
                <p>A dataframe (or dictionary of dataframes) of SNP data.</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>


            <details class="quote">
              <summary>Source code in <code>magenpy/GWADataLoader.py</code></summary>
              <div class="language-python highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal"><a href="#__codelineno-0-862">862</a></span>
<span class="normal"><a href="#__codelineno-0-863">863</a></span>
<span class="normal"><a href="#__codelineno-0-864">864</a></span>
<span class="normal"><a href="#__codelineno-0-865">865</a></span>
<span class="normal"><a href="#__codelineno-0-866">866</a></span>
<span class="normal"><a href="#__codelineno-0-867">867</a></span>
<span class="normal"><a href="#__codelineno-0-868">868</a></span>
<span class="normal"><a href="#__codelineno-0-869">869</a></span>
<span class="normal"><a href="#__codelineno-0-870">870</a></span>
<span class="normal"><a href="#__codelineno-0-871">871</a></span>
<span class="normal"><a href="#__codelineno-0-872">872</a></span>
<span class="normal"><a href="#__codelineno-0-873">873</a></span>
<span class="normal"><a href="#__codelineno-0-874">874</a></span>
<span class="normal"><a href="#__codelineno-0-875">875</a></span>
<span class="normal"><a href="#__codelineno-0-876">876</a></span>
<span class="normal"><a href="#__codelineno-0-877">877</a></span>
<span class="normal"><a href="#__codelineno-0-878">878</a></span>
<span class="normal"><a href="#__codelineno-0-879">879</a></span>
<span class="normal"><a href="#__codelineno-0-880">880</a></span>
<span class="normal"><a href="#__codelineno-0-881">881</a></span>
<span class="normal"><a href="#__codelineno-0-882">882</a></span>
<span class="normal"><a href="#__codelineno-0-883">883</a></span>
<span class="normal"><a href="#__codelineno-0-884">884</a></span>
<span class="normal"><a href="#__codelineno-0-885">885</a></span>
<span class="normal"><a href="#__codelineno-0-886">886</a></span>
<span class="normal"><a href="#__codelineno-0-887">887</a></span>
<span class="normal"><a href="#__codelineno-0-888">888</a></span>
<span class="normal"><a href="#__codelineno-0-889">889</a></span>
<span class="normal"><a href="#__codelineno-0-890">890</a></span>
<span class="normal"><a href="#__codelineno-0-891">891</a></span>
<span class="normal"><a href="#__codelineno-0-892">892</a></span>
<span class="normal"><a href="#__codelineno-0-893">893</a></span>
<span class="normal"><a href="#__codelineno-0-894">894</a></span>
<span class="normal"><a href="#__codelineno-0-895">895</a></span>
<span class="normal"><a href="#__codelineno-0-896">896</a></span>
<span class="normal"><a href="#__codelineno-0-897">897</a></span>
<span class="normal"><a href="#__codelineno-0-898">898</a></span>
<span class="normal"><a href="#__codelineno-0-899">899</a></span>
<span class="normal"><a href="#__codelineno-0-900">900</a></span>
<span class="normal"><a href="#__codelineno-0-901">901</a></span>
<span class="normal"><a href="#__codelineno-0-902">902</a></span>
<span class="normal"><a href="#__codelineno-0-903">903</a></span>
<span class="normal"><a href="#__codelineno-0-904">904</a></span>
<span class="normal"><a href="#__codelineno-0-905">905</a></span>
<span class="normal"><a href="#__codelineno-0-906">906</a></span>
<span class="normal"><a href="#__codelineno-0-907">907</a></span>
<span class="normal"><a href="#__codelineno-0-908">908</a></span>
<span class="normal"><a href="#__codelineno-0-909">909</a></span>
<span class="normal"><a href="#__codelineno-0-910">910</a></span></pre></div></td><td class="code"><div><pre><span></span><code><span id="__span-0-862"><a id="__codelineno-0-862" name="__codelineno-0-862"></a><span class="k">def</span><span class="w"> </span><span class="nf">to_snp_table</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">col_subset</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">per_chromosome</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">resource</span><span class="o">=</span><span class="s1">&#39;auto&#39;</span><span class="p">):</span>
</span><span id="__span-0-863"><a id="__codelineno-0-863" name="__codelineno-0-863"></a><span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
</span><span id="__span-0-864"><a id="__codelineno-0-864" name="__codelineno-0-864"></a><span class="sd">    Get a dataframe of SNP data for all variants</span>
</span><span id="__span-0-865"><a id="__codelineno-0-865" name="__codelineno-0-865"></a><span class="sd">    across different chromosomes.</span>
</span><span id="__span-0-866"><a id="__codelineno-0-866" name="__codelineno-0-866"></a>
</span><span id="__span-0-867"><a id="__codelineno-0-867" name="__codelineno-0-867"></a><span class="sd">    :param col_subset: The subset of columns to obtain.</span>
</span><span id="__span-0-868"><a id="__codelineno-0-868" name="__codelineno-0-868"></a><span class="sd">    :param per_chromosome: If True, returns a dictionary where the key</span>
</span><span id="__span-0-869"><a id="__codelineno-0-869" name="__codelineno-0-869"></a><span class="sd">    is the chromosome number and the value is the SNP table per</span>
</span><span id="__span-0-870"><a id="__codelineno-0-870" name="__codelineno-0-870"></a><span class="sd">    chromosome.</span>
</span><span id="__span-0-871"><a id="__codelineno-0-871" name="__codelineno-0-871"></a><span class="sd">    :param resource: The data source to extract the SNP table from. By default, the method</span>
</span><span id="__span-0-872"><a id="__codelineno-0-872" name="__codelineno-0-872"></a><span class="sd">    will try to extract the SNP table from the genotype matrix. If the genotype matrix is not</span>
</span><span id="__span-0-873"><a id="__codelineno-0-873" name="__codelineno-0-873"></a><span class="sd">    available, then it will try to extract the SNP information from the LD matrix or the summary</span>
</span><span id="__span-0-874"><a id="__codelineno-0-874" name="__codelineno-0-874"></a><span class="sd">    statistics table. Possible values: `auto`, `genotype`, `ld`, `sumstats`.</span>
</span><span id="__span-0-875"><a id="__codelineno-0-875" name="__codelineno-0-875"></a>
</span><span id="__span-0-876"><a id="__codelineno-0-876" name="__codelineno-0-876"></a><span class="sd">    :return: A dataframe (or dictionary of dataframes) of SNP data.</span>
</span><span id="__span-0-877"><a id="__codelineno-0-877" name="__codelineno-0-877"></a><span class="sd">    &quot;&quot;&quot;</span>
</span><span id="__span-0-878"><a id="__codelineno-0-878" name="__codelineno-0-878"></a>
</span><span id="__span-0-879"><a id="__codelineno-0-879" name="__codelineno-0-879"></a>    <span class="c1"># Sanity checks:</span>
</span><span id="__span-0-880"><a id="__codelineno-0-880" name="__codelineno-0-880"></a>    <span class="k">assert</span> <span class="n">resource</span> <span class="ow">in</span> <span class="p">(</span><span class="s1">&#39;auto&#39;</span><span class="p">,</span> <span class="s1">&#39;genotype&#39;</span><span class="p">,</span> <span class="s1">&#39;ld&#39;</span><span class="p">,</span> <span class="s1">&#39;sumstats&#39;</span><span class="p">)</span>
</span><span id="__span-0-881"><a id="__codelineno-0-881" name="__codelineno-0-881"></a>
</span><span id="__span-0-882"><a id="__codelineno-0-882" name="__codelineno-0-882"></a>    <span class="k">if</span> <span class="n">resource</span> <span class="o">!=</span> <span class="s1">&#39;auto&#39;</span><span class="p">:</span>
</span><span id="__span-0-883"><a id="__codelineno-0-883" name="__codelineno-0-883"></a>        <span class="k">if</span> <span class="n">resource</span> <span class="o">==</span> <span class="s1">&#39;genotype&#39;</span> <span class="ow">and</span> <span class="bp">self</span><span class="o">.</span><span class="n">genotype</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
</span><span id="__span-0-884"><a id="__codelineno-0-884" name="__codelineno-0-884"></a>            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;Genotype matrix is not available!&quot;</span><span class="p">)</span>
</span><span id="__span-0-885"><a id="__codelineno-0-885" name="__codelineno-0-885"></a>        <span class="k">if</span> <span class="n">resource</span> <span class="o">==</span> <span class="s1">&#39;ld&#39;</span> <span class="ow">and</span> <span class="bp">self</span><span class="o">.</span><span class="n">ld</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
</span><span id="__span-0-886"><a id="__codelineno-0-886" name="__codelineno-0-886"></a>            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;LD matrix is not available!&quot;</span><span class="p">)</span>
</span><span id="__span-0-887"><a id="__codelineno-0-887" name="__codelineno-0-887"></a>        <span class="k">if</span> <span class="n">resource</span> <span class="o">==</span> <span class="s1">&#39;sumstats&#39;</span> <span class="ow">and</span> <span class="bp">self</span><span class="o">.</span><span class="n">sumstats_table</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
</span><span id="__span-0-888"><a id="__codelineno-0-888" name="__codelineno-0-888"></a>            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;Summary statistics table is not available!&quot;</span><span class="p">)</span>
</span><span id="__span-0-889"><a id="__codelineno-0-889" name="__codelineno-0-889"></a>    <span class="k">else</span><span class="p">:</span>
</span><span id="__span-0-890"><a id="__codelineno-0-890" name="__codelineno-0-890"></a>        <span class="k">if</span> <span class="nb">all</span><span class="p">(</span><span class="n">ds</span> <span class="ow">is</span> <span class="kc">None</span> <span class="k">for</span> <span class="n">ds</span> <span class="ow">in</span> <span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">genotype</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">ld</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">sumstats_table</span><span class="p">)):</span>
</span><span id="__span-0-891"><a id="__codelineno-0-891" name="__codelineno-0-891"></a>            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;No data sources available to extract SNP data from!&quot;</span><span class="p">)</span>
</span><span id="__span-0-892"><a id="__codelineno-0-892" name="__codelineno-0-892"></a>
</span><span id="__span-0-893"><a id="__codelineno-0-893" name="__codelineno-0-893"></a>    <span class="c1"># Extract the SNP data:</span>
</span><span id="__span-0-894"><a id="__codelineno-0-894" name="__codelineno-0-894"></a>
</span><span id="__span-0-895"><a id="__codelineno-0-895" name="__codelineno-0-895"></a>    <span class="n">snp_tables</span> <span class="o">=</span> <span class="p">{}</span>
</span><span id="__span-0-896"><a id="__codelineno-0-896" name="__codelineno-0-896"></a>
</span><span id="__span-0-897"><a id="__codelineno-0-897" name="__codelineno-0-897"></a>    <span class="k">if</span> <span class="n">resource</span> <span class="ow">in</span> <span class="p">(</span><span class="s1">&#39;auto&#39;</span><span class="p">,</span> <span class="s1">&#39;genotype&#39;</span><span class="p">)</span> <span class="ow">and</span> <span class="bp">self</span><span class="o">.</span><span class="n">genotype</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
</span><span id="__span-0-898"><a id="__codelineno-0-898" name="__codelineno-0-898"></a>        <span class="k">for</span> <span class="n">c</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">chromosomes</span><span class="p">:</span>
</span><span id="__span-0-899"><a id="__codelineno-0-899" name="__codelineno-0-899"></a>            <span class="n">snp_tables</span><span class="p">[</span><span class="n">c</span><span class="p">]</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">genotype</span><span class="p">[</span><span class="n">c</span><span class="p">]</span><span class="o">.</span><span class="n">get_snp_table</span><span class="p">(</span><span class="n">col_subset</span><span class="o">=</span><span class="n">col_subset</span><span class="p">)</span>
</span><span id="__span-0-900"><a id="__codelineno-0-900" name="__codelineno-0-900"></a>    <span class="k">elif</span> <span class="n">resource</span> <span class="ow">in</span> <span class="p">(</span><span class="s1">&#39;auto&#39;</span><span class="p">,</span> <span class="s1">&#39;ld&#39;</span><span class="p">)</span> <span class="ow">and</span> <span class="bp">self</span><span class="o">.</span><span class="n">ld</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
</span><span id="__span-0-901"><a id="__codelineno-0-901" name="__codelineno-0-901"></a>        <span class="k">for</span> <span class="n">c</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">chromosomes</span><span class="p">:</span>
</span><span id="__span-0-902"><a id="__codelineno-0-902" name="__codelineno-0-902"></a>            <span class="n">snp_tables</span><span class="p">[</span><span class="n">c</span><span class="p">]</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">ld</span><span class="p">[</span><span class="n">c</span><span class="p">]</span><span class="o">.</span><span class="n">to_snp_table</span><span class="p">(</span><span class="n">col_subset</span><span class="o">=</span><span class="n">col_subset</span><span class="p">)</span>
</span><span id="__span-0-903"><a id="__codelineno-0-903" name="__codelineno-0-903"></a>    <span class="k">else</span><span class="p">:</span>
</span><span id="__span-0-904"><a id="__codelineno-0-904" name="__codelineno-0-904"></a>        <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">to_summary_statistics_table</span><span class="p">(</span><span class="n">col_subset</span><span class="o">=</span><span class="n">col_subset</span><span class="p">,</span>
</span><span id="__span-0-905"><a id="__codelineno-0-905" name="__codelineno-0-905"></a>                                                <span class="n">per_chromosome</span><span class="o">=</span><span class="n">per_chromosome</span><span class="p">)</span>
</span><span id="__span-0-906"><a id="__codelineno-0-906" name="__codelineno-0-906"></a>
</span><span id="__span-0-907"><a id="__codelineno-0-907" name="__codelineno-0-907"></a>    <span class="k">if</span> <span class="n">per_chromosome</span><span class="p">:</span>
</span><span id="__span-0-908"><a id="__codelineno-0-908" name="__codelineno-0-908"></a>        <span class="k">return</span> <span class="n">snp_tables</span>
</span><span id="__span-0-909"><a id="__codelineno-0-909" name="__codelineno-0-909"></a>    <span class="k">else</span><span class="p">:</span>
</span><span id="__span-0-910"><a id="__codelineno-0-910" name="__codelineno-0-910"></a>        <span class="k">return</span> <span class="n">pd</span><span class="o">.</span><span class="n">concat</span><span class="p">(</span><span class="nb">list</span><span class="p">(</span><span class="n">snp_tables</span><span class="o">.</span><span class="n">values</span><span class="p">()))</span>
</span></code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h2 id="magenpy.GWADataLoader.GWADataLoader.to_summary_statistics_table" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">to_summary_statistics_table</span><span class="p">(</span><span class="n">col_subset</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">per_chromosome</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span></code>

<a href="#magenpy.GWADataLoader.GWADataLoader.to_summary_statistics_table" class="headerlink" title="Permanent link">&para;</a></h2>


    <div class="doc doc-contents ">

        <p>Get a dataframe of the GWAS summary statistics for all variants
across different chromosomes.</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                <code>col_subset</code>
            </td>
            <td>
            </td>
            <td>
              <div class="doc-md-description">
                <p>The subset of columns (or summary statistics) to obtain.</p>
              </div>
            </td>
            <td>
                  <code>None</code>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>per_chromosome</code>
            </td>
            <td>
            </td>
            <td>
              <div class="doc-md-description">
                <p>If True, returns a dictionary where the key is the chromosome number and the value is the summary statistics table per chromosome.</p>
              </div>
            </td>
            <td>
                  <code>False</code>
            </td>
          </tr>
      </tbody>
    </table>


    <p><span class="doc-section-title">Returns:</span></p>
    <table>
      <thead>
        <tr>
          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
            </td>
            <td>
              <div class="doc-md-description">
                <p>A dataframe (or dictionary of dataframes) of summary statistics.</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>


            <details class="quote">
              <summary>Source code in <code>magenpy/GWADataLoader.py</code></summary>
              <div class="language-python highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal"><a href="#__codelineno-0-912">912</a></span>
<span class="normal"><a href="#__codelineno-0-913">913</a></span>
<span class="normal"><a href="#__codelineno-0-914">914</a></span>
<span class="normal"><a href="#__codelineno-0-915">915</a></span>
<span class="normal"><a href="#__codelineno-0-916">916</a></span>
<span class="normal"><a href="#__codelineno-0-917">917</a></span>
<span class="normal"><a href="#__codelineno-0-918">918</a></span>
<span class="normal"><a href="#__codelineno-0-919">919</a></span>
<span class="normal"><a href="#__codelineno-0-920">920</a></span>
<span class="normal"><a href="#__codelineno-0-921">921</a></span>
<span class="normal"><a href="#__codelineno-0-922">922</a></span>
<span class="normal"><a href="#__codelineno-0-923">923</a></span>
<span class="normal"><a href="#__codelineno-0-924">924</a></span>
<span class="normal"><a href="#__codelineno-0-925">925</a></span>
<span class="normal"><a href="#__codelineno-0-926">926</a></span>
<span class="normal"><a href="#__codelineno-0-927">927</a></span>
<span class="normal"><a href="#__codelineno-0-928">928</a></span>
<span class="normal"><a href="#__codelineno-0-929">929</a></span>
<span class="normal"><a href="#__codelineno-0-930">930</a></span>
<span class="normal"><a href="#__codelineno-0-931">931</a></span>
<span class="normal"><a href="#__codelineno-0-932">932</a></span>
<span class="normal"><a href="#__codelineno-0-933">933</a></span>
<span class="normal"><a href="#__codelineno-0-934">934</a></span>
<span class="normal"><a href="#__codelineno-0-935">935</a></span></pre></div></td><td class="code"><div><pre><span></span><code><span id="__span-0-912"><a id="__codelineno-0-912" name="__codelineno-0-912"></a><span class="k">def</span><span class="w"> </span><span class="nf">to_summary_statistics_table</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">col_subset</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">per_chromosome</span><span class="o">=</span><span class="kc">False</span><span class="p">):</span>
</span><span id="__span-0-913"><a id="__codelineno-0-913" name="__codelineno-0-913"></a><span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
</span><span id="__span-0-914"><a id="__codelineno-0-914" name="__codelineno-0-914"></a><span class="sd">    Get a dataframe of the GWAS summary statistics for all variants</span>
</span><span id="__span-0-915"><a id="__codelineno-0-915" name="__codelineno-0-915"></a><span class="sd">    across different chromosomes.</span>
</span><span id="__span-0-916"><a id="__codelineno-0-916" name="__codelineno-0-916"></a>
</span><span id="__span-0-917"><a id="__codelineno-0-917" name="__codelineno-0-917"></a><span class="sd">    :param col_subset: The subset of columns (or summary statistics) to obtain.</span>
</span><span id="__span-0-918"><a id="__codelineno-0-918" name="__codelineno-0-918"></a><span class="sd">    :param per_chromosome: If True, returns a dictionary where the key</span>
</span><span id="__span-0-919"><a id="__codelineno-0-919" name="__codelineno-0-919"></a><span class="sd">    is the chromosome number and the value is the summary statistics table per</span>
</span><span id="__span-0-920"><a id="__codelineno-0-920" name="__codelineno-0-920"></a><span class="sd">    chromosome.</span>
</span><span id="__span-0-921"><a id="__codelineno-0-921" name="__codelineno-0-921"></a>
</span><span id="__span-0-922"><a id="__codelineno-0-922" name="__codelineno-0-922"></a><span class="sd">    :return: A dataframe (or dictionary of dataframes) of summary statistics.</span>
</span><span id="__span-0-923"><a id="__codelineno-0-923" name="__codelineno-0-923"></a><span class="sd">    &quot;&quot;&quot;</span>
</span><span id="__span-0-924"><a id="__codelineno-0-924" name="__codelineno-0-924"></a>
</span><span id="__span-0-925"><a id="__codelineno-0-925" name="__codelineno-0-925"></a>    <span class="k">assert</span> <span class="bp">self</span><span class="o">.</span><span class="n">sumstats_table</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span>
</span><span id="__span-0-926"><a id="__codelineno-0-926" name="__codelineno-0-926"></a>
</span><span id="__span-0-927"><a id="__codelineno-0-927" name="__codelineno-0-927"></a>    <span class="n">snp_tables</span> <span class="o">=</span> <span class="p">{}</span>
</span><span id="__span-0-928"><a id="__codelineno-0-928" name="__codelineno-0-928"></a>
</span><span id="__span-0-929"><a id="__codelineno-0-929" name="__codelineno-0-929"></a>    <span class="k">for</span> <span class="n">c</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">chromosomes</span><span class="p">:</span>
</span><span id="__span-0-930"><a id="__codelineno-0-930" name="__codelineno-0-930"></a>        <span class="n">snp_tables</span><span class="p">[</span><span class="n">c</span><span class="p">]</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">sumstats_table</span><span class="p">[</span><span class="n">c</span><span class="p">]</span><span class="o">.</span><span class="n">to_table</span><span class="p">(</span><span class="n">col_subset</span><span class="o">=</span><span class="n">col_subset</span><span class="p">)</span>
</span><span id="__span-0-931"><a id="__codelineno-0-931" name="__codelineno-0-931"></a>
</span><span id="__span-0-932"><a id="__codelineno-0-932" name="__codelineno-0-932"></a>    <span class="k">if</span> <span class="n">per_chromosome</span><span class="p">:</span>
</span><span id="__span-0-933"><a id="__codelineno-0-933" name="__codelineno-0-933"></a>        <span class="k">return</span> <span class="n">snp_tables</span>
</span><span id="__span-0-934"><a id="__codelineno-0-934" name="__codelineno-0-934"></a>    <span class="k">else</span><span class="p">:</span>
</span><span id="__span-0-935"><a id="__codelineno-0-935" name="__codelineno-0-935"></a>        <span class="k">return</span> <span class="n">pd</span><span class="o">.</span><span class="n">concat</span><span class="p">(</span><span class="nb">list</span><span class="p">(</span><span class="n">snp_tables</span><span class="o">.</span><span class="n">values</span><span class="p">()))</span>
</span></code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>



  </div>

    </div>

</div>












                
              </article>
            </div>
          
          
<script>var target=document.getElementById(location.hash.slice(1));target&&target.name&&(target.checked=target.name.startsWith("__tabbed_"))</script>
        </div>
        
      </main>
      
        <footer class="md-footer">
  
    
  
  <div class="md-footer-meta md-typeset">
    <div class="md-footer-meta__inner md-grid">
      <div class="md-copyright">
  
  
    Made with
    <a href="https://squidfunk.github.io/mkdocs-material/" target="_blank" rel="noopener">
      Material for MkDocs
    </a>
  
</div>
      
    </div>
  </div>
</footer>
      
    </div>
    <div class="md-dialog" data-md-component="dialog">
      <div class="md-dialog__inner md-typeset"></div>
    </div>
    
    
    
      
      <script id="__config" type="application/json">{"base": "../..", "features": ["announce.dismiss", "content.action.edit", "content.action.view", "content.code.annotate", "content.code.copy", "content.tooltips", "navigation.expand", "navigation.footer", "navigation.indexes", "navigation.tracking", "search.highlight", "search.share", "search.suggest", "toc.follow"], "search": "../../assets/javascripts/workers/search.d50fe291.min.js", "tags": null, "translations": {"clipboard.copied": "Copied to clipboard", "clipboard.copy": "Copy to clipboard", "search.result.more.one": "1 more on this page", "search.result.more.other": "# more on this page", "search.result.none": "No matching documents", "search.result.one": "1 matching document", "search.result.other": "# matching documents", "search.result.placeholder": "Type to start searching", "search.result.term.missing": "Missing", "select.version": "Select version"}, "version": null}</script>
    
    
      <script src="../../assets/javascripts/bundle.13a4f30d.min.js"></script>
      
    
  </body>
</html>